{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pronto\n",
    "import os.path as osp\n",
    "import re\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tcre.env import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download pr.owl from http://purl.obolibrary.org/obo/pr.owl\n",
    "onto = pronto.Ontology(osp.join(DATA_DIR, 'meta', 'pro', 'pr.owl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(316242, collections.OrderedDict)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(onto.terms), type(onto.terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein               216442\n",
       "gene                   98908\n",
       "ncbi_taxonomy            580\n",
       "PSI-MOD                  120\n",
       "cellular_component       115\n",
       "chebi_ontology            36\n",
       "molecular_function        25\n",
       "sequence                   7\n",
       "bfo                        5\n",
       "biological_process         3\n",
       "obi                        1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show namespace frequency\n",
    "pd.Series([ns for k, t in onto.terms.items() for ns in t.other['namespace']]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "213413"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Top-level \"Protein\" term (see https://www.ebi.ac.uk/ols/ontologies/pr/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FPR_000000001)\n",
    "parent = onto['PR:000000001']\n",
    "\n",
    "# Fetch children\n",
    "terms = [parent] + list(parent.rchildren())\n",
    "len(terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein    213413\n",
       "dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show namespace frequency (for selected subtree)\n",
    "pd.Series([ns for t in terms for ns in t.other['namespace']]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(False    213413\n",
       " dtype: int64, 1)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that ids are always present and always unique\n",
    "pd.Series([t.id is None for t in terms]).value_counts(), pd.Series([t.id or t in terms]).value_counts().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('PR:000001836',\n",
       " 'CD7 molecule',\n",
       " Description('A protein that is a translation product of the human CD7 gene or a 1:1 ortholog thereof.', ['PRO:WCB']),\n",
       " {'hasExactSynonym': ['CD7',\n",
       "   'GP40',\n",
       "   'T-cell antigen CD7',\n",
       "   'T-cell leukemia antigen',\n",
       "   'T-cell surface antigen Leu-9',\n",
       "   'TP41'],\n",
       "  'comment': ['Category=gene. Requested by=CL.'],\n",
       "  'xref': ['PIRSF:PIRSF038791'],\n",
       "  'namespace': ['protein']},\n",
       " {'hasExactSynonym': ['CD7',\n",
       "   'GP40',\n",
       "   'T-cell antigen CD7',\n",
       "   'T-cell leukemia antigen',\n",
       "   'T-cell surface antigen Leu-9',\n",
       "   'TP41'],\n",
       "  'comment': ['Category=gene. Requested by=CL.'],\n",
       "  'xref': ['PIRSF:PIRSF038791'],\n",
       "  'namespace': ['protein']},\n",
       " {<Synonym: \"CD7\" EXACT []>,\n",
       "  <Synonym: \"GP40\" EXACT []>,\n",
       "  <Synonym: \"T-cell antigen CD7\" EXACT []>,\n",
       "  <Synonym: \"T-cell leukemia antigen\" EXACT []>,\n",
       "  <Synonym: \"T-cell surface antigen Leu-9\" EXACT []>,\n",
       "  <Synonym: \"TP41\" EXACT []>},\n",
       " 'T-cell antigen CD7')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show fields for an example term\n",
    "t = terms[230]\n",
    "t.id, t.name, t.desc, t.other, t.other, t.synonyms, list(t.synonyms)[0].desc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EXACT      436863\n",
       "RELATED    149152\n",
       "BROAD        4763\n",
       "NARROW       1316\n",
       "dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Synonym type frequencies\n",
    "syn_types = pd.Series([s.scope for t in terms for s in t.synonyms]).value_counts()\n",
    "syn_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>nm</th>\n",
       "      <th>sc</th>\n",
       "      <th>desc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>BROAD</td>\n",
       "      <td>bHLH clade D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>BROAD</td>\n",
       "      <td>bHLH class V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PR:000000020</td>\n",
       "      <td>Myc protein</td>\n",
       "      <td>BROAD</td>\n",
       "      <td>bHLH class III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PR:000000020</td>\n",
       "      <td>Myc protein</td>\n",
       "      <td>BROAD</td>\n",
       "      <td>bHLH clade B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>PR:000000735</td>\n",
       "      <td>staphylococcal enterotoxin A</td>\n",
       "      <td>BROAD</td>\n",
       "      <td>ET-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>EXACT</td>\n",
       "      <td>DNA-binding protein inhibitor ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PR:000000005</td>\n",
       "      <td>TGF-beta receptor type-2</td>\n",
       "      <td>EXACT</td>\n",
       "      <td>TGF-beta receptor type II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PR:000000005</td>\n",
       "      <td>TGF-beta receptor type-2</td>\n",
       "      <td>EXACT</td>\n",
       "      <td>TbetaR-II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PR:000000005</td>\n",
       "      <td>TGF-beta receptor type-2</td>\n",
       "      <td>EXACT</td>\n",
       "      <td>TGFR-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PR:000000005</td>\n",
       "      <td>TGF-beta receptor type-2</td>\n",
       "      <td>EXACT</td>\n",
       "      <td>transforming growth factor-beta receptor type II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4993</th>\n",
       "      <td>PR:000004066</td>\n",
       "      <td>acidic leucine-rich nuclear phosphoprotein 32 ...</td>\n",
       "      <td>NARROW</td>\n",
       "      <td>putative HLA-DR-associated protein I-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4418</th>\n",
       "      <td>PR:000003927</td>\n",
       "      <td>fatty aldehyde dehydrogenase</td>\n",
       "      <td>NARROW</td>\n",
       "      <td>aldehyde dehydrogenase 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4979</th>\n",
       "      <td>PR:000004065</td>\n",
       "      <td>acidic leucine-rich nuclear phosphoprotein 32 ...</td>\n",
       "      <td>NARROW</td>\n",
       "      <td>putative HLA-DR-associated protein I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3906</th>\n",
       "      <td>PR:000003840</td>\n",
       "      <td>lysophosphatidylcholine acyltransferase</td>\n",
       "      <td>NARROW</td>\n",
       "      <td>lysophosphatidylcholine acyltransferase 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053</th>\n",
       "      <td>PR:000001822</td>\n",
       "      <td>MHC class II histocompatibility antigen gamma ...</td>\n",
       "      <td>NARROW</td>\n",
       "      <td>HLA-DR antigens-associated invariant chain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>RELATED</td>\n",
       "      <td>ID protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PR:000000019</td>\n",
       "      <td>mitogen-activated protein kinase</td>\n",
       "      <td>RELATED</td>\n",
       "      <td>kinase-related transforming protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PR:000000008</td>\n",
       "      <td>TGF-beta-like cystine-knot cytokine</td>\n",
       "      <td>RELATED</td>\n",
       "      <td>TGF-beta superfamily</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PR:000000672</td>\n",
       "      <td>cation channel sperm-associated protein subuni...</td>\n",
       "      <td>RELATED</td>\n",
       "      <td>C14orf161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PR:000000009</td>\n",
       "      <td>anti-Muellerian hormone type-2 receptor</td>\n",
       "      <td>RELATED</td>\n",
       "      <td>AMHR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id                                                 nm  \\\n",
       "1     PR:000000003                  HLH DNA-binding protein inhibitor   \n",
       "3     PR:000000003                  HLH DNA-binding protein inhibitor   \n",
       "25    PR:000000020                                        Myc protein   \n",
       "26    PR:000000020                                        Myc protein   \n",
       "37    PR:000000735                       staphylococcal enterotoxin A   \n",
       "2     PR:000000003                  HLH DNA-binding protein inhibitor   \n",
       "4     PR:000000005                           TGF-beta receptor type-2   \n",
       "5     PR:000000005                           TGF-beta receptor type-2   \n",
       "6     PR:000000005                           TGF-beta receptor type-2   \n",
       "7     PR:000000005                           TGF-beta receptor type-2   \n",
       "4993  PR:000004066  acidic leucine-rich nuclear phosphoprotein 32 ...   \n",
       "4418  PR:000003927                       fatty aldehyde dehydrogenase   \n",
       "4979  PR:000004065  acidic leucine-rich nuclear phosphoprotein 32 ...   \n",
       "3906  PR:000003840            lysophosphatidylcholine acyltransferase   \n",
       "1053  PR:000001822  MHC class II histocompatibility antigen gamma ...   \n",
       "0     PR:000000003                  HLH DNA-binding protein inhibitor   \n",
       "23    PR:000000019                   mitogen-activated protein kinase   \n",
       "11    PR:000000008                TGF-beta-like cystine-knot cytokine   \n",
       "31    PR:000000672  cation channel sperm-associated protein subuni...   \n",
       "15    PR:000000009            anti-Muellerian hormone type-2 receptor   \n",
       "\n",
       "           sc                                              desc  \n",
       "1       BROAD                                      bHLH clade D  \n",
       "3       BROAD                                      bHLH class V  \n",
       "25      BROAD                                    bHLH class III  \n",
       "26      BROAD                                      bHLH clade B  \n",
       "37      BROAD                                              ET-A  \n",
       "2       EXACT                  DNA-binding protein inhibitor ID  \n",
       "4       EXACT                         TGF-beta receptor type II  \n",
       "5       EXACT                                         TbetaR-II  \n",
       "6       EXACT                                            TGFR-2  \n",
       "7       EXACT  transforming growth factor-beta receptor type II  \n",
       "4993   NARROW            putative HLA-DR-associated protein I-2  \n",
       "4418   NARROW                          aldehyde dehydrogenase 3  \n",
       "4979   NARROW              putative HLA-DR-associated protein I  \n",
       "3906   NARROW         lysophosphatidylcholine acyltransferase 4  \n",
       "1053   NARROW        HLA-DR antigens-associated invariant chain  \n",
       "0     RELATED                                        ID protein  \n",
       "23    RELATED               kinase-related transforming protein  \n",
       "11    RELATED                              TGF-beta superfamily  \n",
       "31    RELATED                                         C14orf161  \n",
       "15    RELATED                                              AMHR  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([(t.id, t.name, s.scope, s.desc) for t in terms for s in t.synonyms], columns=['id', 'nm', 'sc', 'desc'])\\\n",
    "    .groupby('sc').head(5).sort_values('sc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"Category=gene. Requested by=CL\" --> gene\n",
    "category_regex = re.compile(r'(?<=Category=)[a-z-]+(?=\\.)')\n",
    "\n",
    "def get_category(t):\n",
    "    if 'comment' in t.other and t.other['comment']:\n",
    "        cats = category_regex.findall(t.other['comment'][0])\n",
    "        return cats[0] if cats else None\n",
    "    return None\n",
    "\n",
    "# Map each term to its category\n",
    "category_map = {t.id: get_category(t) for t in terms}\n",
    "assert len(category_map) == len(terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PR:000000001']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[t.id for t in t.parents]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 805507 entries, 0 to 213412\n",
      "Data columns (total 8 columns):\n",
      "ancestors    805507 non-null object\n",
      "category     805409 non-null object\n",
      "id           805507 non-null object\n",
      "label        805507 non-null object\n",
      "namespace    805507 non-null object\n",
      "parent       331586 non-null object\n",
      "syn          805507 non-null object\n",
      "syn_typ      805507 non-null object\n",
      "dtypes: object(8)\n",
      "memory usage: 55.3+ MB\n"
     ]
    }
   ],
   "source": [
    "# Synonym type priorities\n",
    "PRIORITY = {'label': 5, 'exact': 4, 'narrow': 3, 'related': 2, 'broad': 1}\n",
    "SPECIES = {'any': 3, 'human': 2, 'mouse': 1}\n",
    "\n",
    "def get_row(t, syn, syn_typ):\n",
    "    # Get ancestor ID list\n",
    "    ancestors = [ta.id for ta in t.parents]\n",
    "    \n",
    "    # Assign \"parent\" as first ancestor in \"gene\" category \n",
    "    # (e.g. \"interferon gamma isoform 1 (human)\" gets parent \"interferon gamma\")\n",
    "    parent = [aid for aid in ancestors if category_map.get(aid) == 'gene']\n",
    "    parent = parent[0] if parent else None\n",
    "    \n",
    "    return dict(\n",
    "        id=t.id, namespace=t.other['namespace'][0], \n",
    "        label=t.name, syn=syn.desc,\n",
    "        category=category_map.get(t.id), \n",
    "        parent=parent, ancestors=ancestors,\n",
    "        syn_typ=syn_typ if syn_typ else syn.scope.lower()\n",
    "    )\n",
    "\n",
    "def get_frame(syn_fn, syn_typ=None):    \n",
    "    return pd.DataFrame([\n",
    "        get_row(t, s, syn_typ)\n",
    "        for t in terms\n",
    "        for s in syn_fn(t)\n",
    "    ])\n",
    "            \n",
    "dfa = pd.concat([\n",
    "    get_frame(lambda t: t.synonyms),\n",
    "    get_frame(lambda t: [pronto.Synonym(t.name)], 'label')\n",
    "])\n",
    "dfa.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Validate that all fields are not null except where missing values are expected:\n",
    "# parent - can be null if entry has no ancestor in \"gene\" category\n",
    "# category - can be null if category regex extraction results in nothing \n",
    "#    (not sure why such an important field is jammed into a string description rather than made into a formal property within the ontology)\n",
    "for c in dfa.columns.difference(['parent', 'category']):\n",
    "    assert dfa[c].notnull().all(), f'Found null values for field {c}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ancestors</th>\n",
       "      <th>category</th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>namespace</th>\n",
       "      <th>parent</th>\n",
       "      <th>syn</th>\n",
       "      <th>syn_typ</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[PR:000000001]</td>\n",
       "      <td>family</td>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>ID protein</td>\n",
       "      <td>related</td>\n",
       "      <td>any</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[PR:000000001]</td>\n",
       "      <td>family</td>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>bHLH clade D</td>\n",
       "      <td>broad</td>\n",
       "      <td>any</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[PR:000000001]</td>\n",
       "      <td>family</td>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>DNA-binding protein inhibitor ID</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[PR:000000001]</td>\n",
       "      <td>family</td>\n",
       "      <td>PR:000000003</td>\n",
       "      <td>HLH DNA-binding protein inhibitor</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>bHLH class V</td>\n",
       "      <td>broad</td>\n",
       "      <td>any</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[PR:000000001]</td>\n",
       "      <td>gene</td>\n",
       "      <td>PR:000000005</td>\n",
       "      <td>TGF-beta receptor type-2</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>TGF-beta receptor type II</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ancestors category            id                              label  \\\n",
       "0  [PR:000000001]   family  PR:000000003  HLH DNA-binding protein inhibitor   \n",
       "1  [PR:000000001]   family  PR:000000003  HLH DNA-binding protein inhibitor   \n",
       "2  [PR:000000001]   family  PR:000000003  HLH DNA-binding protein inhibitor   \n",
       "3  [PR:000000001]   family  PR:000000003  HLH DNA-binding protein inhibitor   \n",
       "4  [PR:000000001]     gene  PR:000000005           TGF-beta receptor type-2   \n",
       "\n",
       "  namespace parent                               syn  syn_typ species  \n",
       "0   protein   None                        ID protein  related     any  \n",
       "1   protein   None                      bHLH clade D    broad     any  \n",
       "2   protein   None  DNA-binding protein inhibitor ID    exact     any  \n",
       "3   protein   None                      bHLH class V    broad     any  \n",
       "4   protein   None         TGF-beta receptor type II    exact     any  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfa.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exact      436863\n",
       "label      213413\n",
       "related    149152\n",
       "broad        4763\n",
       "narrow       1316\n",
       "Name: syn_typ, dtype: int64"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfa['syn_typ'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "organism-gene            476217\n",
       "organism-sequence        156723\n",
       "gene                     119254\n",
       "organism-modification     23980\n",
       "sequence                  16958\n",
       "modification              10510\n",
       "family                      963\n",
       "organism-seqgroup           561\n",
       "organism-family              85\n",
       "union                        65\n",
       "organism-genegroup           54\n",
       "complex                      20\n",
       "organism-complex             19\n",
       "Name: category, dtype: int64"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfa['category'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "any                    324251\n",
       "human                  206231\n",
       "mouse                  122978\n",
       "rat                     52025\n",
       "yeast                   29749\n",
       "fruit fly               21994\n",
       "worm                    20540\n",
       "zebrafish               15163\n",
       "chicken                 10671\n",
       "cow                       797\n",
       "pig                       272\n",
       "rabbit                    238\n",
       "dog                       125\n",
       "frog                      115\n",
       "guinea pig                110\n",
       "spinach                    45\n",
       "sheep                      40\n",
       "rice                       31\n",
       "opium poppy                25\n",
       "maize                      25\n",
       "wheat                      20\n",
       "turkey                     12\n",
       "green monkey               12\n",
       "horse                       9\n",
       "honeybee                    9\n",
       "cat                         6\n",
       "horseradish                 5\n",
       "kidney bean                 5\n",
       "great scarlet poppy         2\n",
       "cobra                       2\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract species from labels (i.e. names): \"suppressor of kinetochore protein 1 (yeast)\" --> yeast\n",
    "dfa['species'] = np.where(\n",
    "    dfa['category'].str.contains('organism'), \n",
    "    dfa['label'].str.strip().str.extract(r'(?<=\\()([a-z ]+?)(?=\\)$)', expand=False),\n",
    "    None\n",
    ")\n",
    "# Replace unknown species classifications with \"any\", as these are more generic \n",
    "# ontology elements that could apply to any organism (but not necessarily all of them)\n",
    "dfa['species'] = dfa['species'].fillna('any')\n",
    "dfa['species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "174347"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a set of all proteins with either IDs explicitly associated with target \n",
    "# organisms are that are ancestors of such proteins\n",
    "\n",
    "def get_target_ids(df):\n",
    "    return list(set(df['id'].tolist() + [v for a in df['ancestors'] for v in a]))\n",
    "    \n",
    "target_ids = get_target_ids(dfa[dfa['species'].isin(list(SPECIES.keys()))])\n",
    "len(target_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "' \\\\((any|yeast|human|chicken|rat|mouse|zebrafish|worm|fruit fly|green monkey|frog|pig|cow|dog|sheep|maize|wheat|horse|rabbit|guinea pig|kidney bean|horseradish|turkey|cobra|great scarlet poppy|opium poppy|spinach|honeybee|rice|cat)\\\\)$'"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build a regular expression for removing species in synonyms;\n",
    "# e.g. \"IL4-induced protein 1 (mouse)\" --> \"IL4-induced protein 1\"\n",
    "# * Note that the species strings are matched explicitly as there are often parenthetical elements in labels\n",
    "species_regex = ' \\(({})\\)$'.format('|'.join(dfa['species'].unique()))\n",
    "species_regex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 653228 entries, 0 to 213412\n",
      "Data columns (total 8 columns):\n",
      "category     653130 non-null object\n",
      "id           653228 non-null object\n",
      "label        653228 non-null object\n",
      "namespace    653228 non-null object\n",
      "parent       255340 non-null object\n",
      "syn          653228 non-null object\n",
      "syn_typ      653228 non-null object\n",
      "species      653228 non-null object\n",
      "dtypes: object(8)\n",
      "memory usage: 44.9+ MB\n"
     ]
    }
   ],
   "source": [
    "# Filter to records and subtrees relevant to target species\n",
    "# species_regex = r'\\(({})\\)'.format('|'.join(list(SPECIES.values())))\n",
    "# .assign(syn=lambda df: df['syn'].str.replace(species_regex, '').str.strip())\n",
    "dff = (\n",
    "    dfa\n",
    "    # Restrict to species-specific sub-trees noting that a couple\n",
    "    # ancestors end up being for off-target species (hence the \n",
    "    # second filter on species again)\n",
    "    .pipe(lambda df: df[df['id'].isin(target_ids)])\n",
    "    .pipe(lambda df: df[df['species'].isin(list(SPECIES.keys()))])\n",
    "    # Remove species in synonym strings and drop any resulting duplicates within the same term id\n",
    "    .assign(syn=lambda df: df['syn'].str.strip().str.replace(species_regex, '').str.strip())\n",
    "    .drop_duplicates(['id', 'label', 'syn', 'syn_typ', 'species'])\n",
    "    .drop('ancestors', axis=1)\n",
    ")\n",
    "dff.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "any      324251\n",
       "human    206072\n",
       "mouse    122905\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dff['species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_ids</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>17</th>\n",
       "      <th>21</th>\n",
       "      <th>23</th>\n",
       "      <th>35</th>\n",
       "      <th>39</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>species</th>\n",
       "      <th>syn_typ</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">any</th>\n",
       "      <th>broad</th>\n",
       "      <td>599</td>\n",
       "      <td>622</td>\n",
       "      <td>94</td>\n",
       "      <td>49</td>\n",
       "      <td>26</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exact</th>\n",
       "      <td>164335</td>\n",
       "      <td>260</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <td>78101</td>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>narrow</th>\n",
       "      <td>250</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>related</th>\n",
       "      <td>68215</td>\n",
       "      <td>4111</td>\n",
       "      <td>357</td>\n",
       "      <td>81</td>\n",
       "      <td>25</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">human</th>\n",
       "      <th>broad</th>\n",
       "      <td>573</td>\n",
       "      <td>94</td>\n",
       "      <td>20</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exact</th>\n",
       "      <td>118967</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <td>60728</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>narrow</th>\n",
       "      <td>961</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>related</th>\n",
       "      <td>21790</td>\n",
       "      <td>1063</td>\n",
       "      <td>111</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">mouse</th>\n",
       "      <th>broad</th>\n",
       "      <td>102</td>\n",
       "      <td>71</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exact</th>\n",
       "      <td>72094</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>label</th>\n",
       "      <td>35455</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>narrow</th>\n",
       "      <td>66</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>related</th>\n",
       "      <td>13632</td>\n",
       "      <td>554</td>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "n_ids                1     2    3   4   5   6   7   8   9   10  11  12  13  \\\n",
       "species syn_typ                                                              \n",
       "any     broad       599   622   94  49  26  13   3   0   5   3   2   2   1   \n",
       "        exact    164335   260    8   0   0   0   0   0   0   0   0   0   0   \n",
       "        label     78101    23    3   0   0   0   0   0   0   0   0   0   0   \n",
       "        narrow      250     6    1   0   0   0   0   0   0   0   0   0   0   \n",
       "        related   68215  4111  357  81  25  12   4   1   2   1   0   0   5   \n",
       "human   broad       573    94   20  13   6   3   3   1   1   1   0   0   1   \n",
       "        exact    118967     1    0   0   0   0   0   0   0   0   0   0   0   \n",
       "        label     60728     0    0   0   0   0   0   0   0   0   0   0   0   \n",
       "        narrow      961     0    0   0   0   0   0   0   0   0   0   0   0   \n",
       "        related   21790  1063  111  18   3   0   1   0   1   1   0   0   0   \n",
       "mouse   broad       102    71    6   9   7   3   1   0   0   0   0   1   0   \n",
       "        exact     72094    18    0   0   0   0   0   0   0   0   0   0   0   \n",
       "        label     35455     3    0   0   0   0   0   0   0   0   0   0   0   \n",
       "        narrow       66     0    0   0   0   0   0   0   0   0   0   0   0   \n",
       "        related   13632   554   37   3   0   1   0   0   1   0   0   0   0   \n",
       "\n",
       "n_ids            14  15  17  21  23  35  39  \n",
       "species syn_typ                              \n",
       "any     broad     0   0   1   0   1   0   1  \n",
       "        exact     0   0   0   0   0   0   0  \n",
       "        label     0   0   0   0   0   0   0  \n",
       "        narrow    0   0   0   0   0   0   0  \n",
       "        related   0   2   0   0   0   0   0  \n",
       "human   broad     1   0   0   1   0   1   0  \n",
       "        exact     0   0   0   0   0   0   0  \n",
       "        label     0   0   0   0   0   0   0  \n",
       "        narrow    0   0   0   0   0   0   0  \n",
       "        related   0   0   0   0   0   0   0  \n",
       "mouse   broad     0   0   0   0   0   0   0  \n",
       "        exact     0   0   0   0   0   0   0  \n",
       "        label     0   0   0   0   0   0   0  \n",
       "        narrow    0   0   0   0   0   0   0  \n",
       "        related   0   0   0   0   0   0   0  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get frequencies with which synonyms match for different labels (i.e. terms).\n",
    "# This should be useful for deciding which synonyms are too broad to be helpful\n",
    "dff.groupby(['syn', 'syn_typ', 'species'])['id'].nunique().rename('n_ids').reset_index()\\\n",
    "    .groupby(['species', 'syn_typ', 'n_ids']).size().unstack().fillna(0).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 640587 entries, 444955 to 182118\n",
      "Data columns (total 9 columns):\n",
      "category     640489 non-null object\n",
      "id           640587 non-null object\n",
      "label        640587 non-null object\n",
      "namespace    640587 non-null object\n",
      "parent       251235 non-null object\n",
      "syn          640587 non-null object\n",
      "syn_typ      640587 non-null object\n",
      "species      640587 non-null object\n",
      "priority     640587 non-null int64\n",
      "dtypes: int64(1), object(8)\n",
      "memory usage: 48.9+ MB\n"
     ]
    }
   ],
   "source": [
    "# When a synonym exists for multiple records within same species, resolve by using\n",
    "# synonym source priority (e.g. label = highest, exact = next highest, broad = lowest)\n",
    "# * when a synonym has multiple records for the same species and synonym type, use the highest id\n",
    "dfd = (\n",
    "    dff\n",
    "    .assign(priority=lambda df: df['syn_typ'].map(PRIORITY))\n",
    "    # Note that NA values will are last by default (i.e. lowest)\n",
    "    .sort_values(['species', 'syn', 'priority', 'id'], ascending=False)\n",
    "    .groupby(['species', 'syn'], group_keys=False).head(1)\n",
    ")\n",
    "assert dfd.groupby(['species', 'syn']).size().max() == 1\n",
    "dfd.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "any      314794\n",
       "human    204020\n",
       "mouse    121773\n",
       "Name: species, dtype: int64"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfd['species'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4    355667\n",
       "5    174313\n",
       "2    107475\n",
       "1      1884\n",
       "3      1248\n",
       "Name: priority, dtype: int64"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfd['priority'].fillna(-1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 503426 entries, 56926 to 182118\n",
      "Data columns (total 10 columns):\n",
      "category      503328 non-null object\n",
      "id            503426 non-null object\n",
      "label         503426 non-null object\n",
      "namespace     503426 non-null object\n",
      "parent        131835 non-null object\n",
      "syn           503426 non-null object\n",
      "syn_typ       503426 non-null object\n",
      "species       503426 non-null object\n",
      "priority      503426 non-null int64\n",
      "species_id    503426 non-null int64\n",
      "dtypes: int64(2), object(8)\n",
      "memory usage: 42.2+ MB\n"
     ]
    }
   ],
   "source": [
    "# Finally, make synonyms unique by choosing based on species preference\n",
    "# * This may not be desirable in the future as it will do something like make \"IL12A\"\n",
    "# always point the IL12A record for humans rather than humans or mice\n",
    "\n",
    "def assign_species_id(df):\n",
    "    df['species_id'] = df['species'].map(SPECIES)\n",
    "    assert df['species_id'].notnull().all()\n",
    "    return df\n",
    "\n",
    "dfe = (\n",
    "    dfd\n",
    "    .pipe(assign_species_id)\n",
    "    # Note that NA values will are last by default (i.e. lowest)\n",
    "    .sort_values(['syn', 'species_id', 'id'], ascending=False)\n",
    "    .groupby(['syn'], group_keys=False).head(1)\n",
    ")\n",
    "assert dfe.groupby('syn').size().max() == 1\n",
    "assert dfe['species_id'].notnull().all()\n",
    "dfe.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>id</th>\n",
       "      <th>label</th>\n",
       "      <th>namespace</th>\n",
       "      <th>parent</th>\n",
       "      <th>syn</th>\n",
       "      <th>syn_typ</th>\n",
       "      <th>species</th>\n",
       "      <th>priority</th>\n",
       "      <th>species_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1240</th>\n",
       "      <td>gene</td>\n",
       "      <td>PR:000001866</td>\n",
       "      <td>interleukin-4 receptor subunit alpha</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>IL4RA</td>\n",
       "      <td>related</td>\n",
       "      <td>any</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99846</th>\n",
       "      <td>modification</td>\n",
       "      <td>PR:000018552</td>\n",
       "      <td>interleukin-4 receptor subunit alpha, signal p...</td>\n",
       "      <td>protein</td>\n",
       "      <td>PR:000001866</td>\n",
       "      <td>IL4R/SigPep-</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99847</th>\n",
       "      <td>modification</td>\n",
       "      <td>PR:000018553</td>\n",
       "      <td>interleukin-4 receptor subunit alpha proteolyt...</td>\n",
       "      <td>protein</td>\n",
       "      <td>PR:000001866</td>\n",
       "      <td>IL4R/ClvPrd</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1237</th>\n",
       "      <td>gene</td>\n",
       "      <td>PR:000001866</td>\n",
       "      <td>interleukin-4 receptor subunit alpha</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>IL4R</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204387</th>\n",
       "      <td>modification</td>\n",
       "      <td>PR:000020307</td>\n",
       "      <td>L-amino-acid oxidase, signal peptide removed form</td>\n",
       "      <td>protein</td>\n",
       "      <td>PR:000009005</td>\n",
       "      <td>IL4I1/SigPep-</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25385</th>\n",
       "      <td>gene</td>\n",
       "      <td>PR:000009005</td>\n",
       "      <td>L-amino-acid oxidase</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>IL4I1</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571099</th>\n",
       "      <td>organism-gene</td>\n",
       "      <td>PR:Q6UX52</td>\n",
       "      <td>protein IL-40 (human)</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>IL40</td>\n",
       "      <td>related</td>\n",
       "      <td>human</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93962</th>\n",
       "      <td>modification</td>\n",
       "      <td>PR:000018452</td>\n",
       "      <td>interleukin-4, signal peptide removed form</td>\n",
       "      <td>protein</td>\n",
       "      <td>PR:000001391</td>\n",
       "      <td>IL4/SigPep-</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25386</th>\n",
       "      <td>gene</td>\n",
       "      <td>PR:000009005</td>\n",
       "      <td>L-amino-acid oxidase</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>IL4-induced protein 1</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>856</th>\n",
       "      <td>gene</td>\n",
       "      <td>PR:000001391</td>\n",
       "      <td>interleukin-4</td>\n",
       "      <td>protein</td>\n",
       "      <td>None</td>\n",
       "      <td>IL4</td>\n",
       "      <td>exact</td>\n",
       "      <td>any</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             category            id  \\\n",
       "1240             gene  PR:000001866   \n",
       "99846    modification  PR:000018552   \n",
       "99847    modification  PR:000018553   \n",
       "1237             gene  PR:000001866   \n",
       "204387   modification  PR:000020307   \n",
       "25385            gene  PR:000009005   \n",
       "571099  organism-gene     PR:Q6UX52   \n",
       "93962    modification  PR:000018452   \n",
       "25386            gene  PR:000009005   \n",
       "856              gene  PR:000001391   \n",
       "\n",
       "                                                    label namespace  \\\n",
       "1240                 interleukin-4 receptor subunit alpha   protein   \n",
       "99846   interleukin-4 receptor subunit alpha, signal p...   protein   \n",
       "99847   interleukin-4 receptor subunit alpha proteolyt...   protein   \n",
       "1237                 interleukin-4 receptor subunit alpha   protein   \n",
       "204387  L-amino-acid oxidase, signal peptide removed form   protein   \n",
       "25385                                L-amino-acid oxidase   protein   \n",
       "571099                              protein IL-40 (human)   protein   \n",
       "93962          interleukin-4, signal peptide removed form   protein   \n",
       "25386                                L-amino-acid oxidase   protein   \n",
       "856                                         interleukin-4   protein   \n",
       "\n",
       "              parent                    syn  syn_typ species  priority  \\\n",
       "1240            None                  IL4RA  related     any         2   \n",
       "99846   PR:000001866           IL4R/SigPep-    exact     any         4   \n",
       "99847   PR:000001866            IL4R/ClvPrd    exact     any         4   \n",
       "1237            None                   IL4R    exact     any         4   \n",
       "204387  PR:000009005          IL4I1/SigPep-    exact     any         4   \n",
       "25385           None                  IL4I1    exact     any         4   \n",
       "571099          None                   IL40  related   human         2   \n",
       "93962   PR:000001391            IL4/SigPep-    exact     any         4   \n",
       "25386           None  IL4-induced protein 1    exact     any         4   \n",
       "856             None                    IL4    exact     any         4   \n",
       "\n",
       "        species_id  \n",
       "1240             3  \n",
       "99846            3  \n",
       "99847            3  \n",
       "1237             3  \n",
       "204387           3  \n",
       "25385            3  \n",
       "571099           2  \n",
       "93962            3  \n",
       "25386            3  \n",
       "856              3  "
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfe[dfe['syn'].str.contains('^IL4')]\n",
    "#dfe[dfe['syn'] == 'IL12A']\n",
    "#dfe[dfe['syn'] == 'TGFB']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lab/repos/t-cell-relation-extraction/data/meta/raw/pro.raw.csv.gz'"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = osp.join(META_DATA_DIR, 'raw', 'pro.raw.csv.gz')\n",
    "dfe.to_csv(path, index=False)\n",
    "path"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
