{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import os.path as osp\n",
    "import glob\n",
    "import tqdm\n",
    "import tarfile\n",
    "import pandas as pd\n",
    "import re\n",
    "%run env.py\n",
    "%run src/integration.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lab/data/pmc_oa/bulk/files'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oadir = osp.join(DATA_DIR, 'pmc_oa', 'bulk', 'files')\n",
    "oadir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/comm_use.C-H.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/comm_use.I-N.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/comm_use.O-Z.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.A-B.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.C-H.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.I-N.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.O-Z.xml.tar.gz']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "archives = glob.glob(osp.join(oadir, '*.xml.tar.gz'))\n",
    "archives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_path(p, archive):\n",
    "    parts = p.split('/')\n",
    "    # Curr_HIV/AIDS_Rep/Curr_HIV_AIDS_Rep_2014_Dec_4_11_487-495.txt\n",
    "    venue, name = '/'.join(parts[:-1]), parts[-1]\n",
    "    aid = name.replace('.txt', '').replace('.nxml', '') if name.startswith('PMC') else None\n",
    "    return dict(id=aid, path=p, venue=venue, name=name, archive=archive)\n",
    "\n",
    "def file_meta_stream(archives):\n",
    "    for a in tqdm.tqdm(archives):\n",
    "        tar = tarfile.open(a, \"r:gz\")\n",
    "        files = [f for f in tar.getmembers() if f.isfile()]\n",
    "        for f in tqdm.tqdm(files, mininterval=10):\n",
    "            yield tar, f, parse_path(f.path, a)\n",
    "                \n",
    "BODY_REGEX = r'<body>.*</body>'\n",
    "ABSTRACT_REGEX = r'<abstract>.*</abstract>'\n",
    "\n",
    "def doc_stream(file_meta):\n",
    "    for tar, file, meta in file_meta:\n",
    "        try:\n",
    "            xml = tar.extractfile(file).read().decode('utf-8', errors='ignore')\n",
    "            text = (re.findall(BODY_REGEX, xml) or [''])[0] + (re.findall(ABSTRACT_REGEX, xml) or [''])[0] \n",
    "            yield meta, text\n",
    "        except Exception as e:\n",
    "            raise type(e)(e.message + f'; (archive = {tar.name}, file = {file.path}')\n",
    "\n",
    "SEARCH_TERMS = [\n",
    "    'human', 'mouse', 'murine', \n",
    "    ' t cell', ' t-cell', ' t lymphocyte', ' t-lymphocyte',\n",
    "    'cd3', 'cd4', 'cd8', \n",
    "    'expression', 'cytokine', 'phenotype', 'surface',\n",
    "    'differentiate', 'differentiation', 'differentiated',\n",
    "    'polarization', 'polarize', 'induce', 'induction'\n",
    "]\n",
    "\n",
    "def row_stream(docs, terms=SEARCH_TERMS):\n",
    "    for meta, text in docs:\n",
    "        row = dict(meta)\n",
    "        ltext = text.lower()\n",
    "        for t in SEARCH_TERMS:\n",
    "            row[f'term:{t}'] = t in ltext\n",
    "        yield row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/8 [00:00<?, ?it/s]\n",
      "  0%|          | 0/291368 [00:00<?, ?it/s]\u001b[A\n",
      "  4%|▍         | 11888/291368 [00:10<03:55, 1188.75it/s]\u001b[A\n",
      "  8%|▊         | 23776/291368 [00:25<04:19, 1032.43it/s]\u001b[A\n",
      " 12%|█▏        | 33816/291368 [00:35<04:11, 1023.72it/s]\u001b[A\n",
      " 16%|█▌        | 46395/291368 [00:45<03:45, 1084.26it/s]\u001b[A\n",
      " 20%|██        | 58974/291368 [00:57<03:39, 1059.29it/s]\u001b[A\n",
      " 24%|██▎       | 69027/291368 [01:07<03:34, 1038.16it/s]\u001b[A\n",
      " 24%|██▎       | 69027/291368 [01:18<03:34, 1038.16it/s]\u001b[A\n",
      " 27%|██▋       | 78202/291368 [01:18<03:37, 979.87it/s] \u001b[A\n",
      " 31%|███       | 89800/291368 [01:28<03:16, 1027.68it/s]\u001b[A\n",
      " 34%|███▍      | 99186/291368 [01:38<03:12, 999.11it/s] \u001b[A\n",
      " 34%|███▍      | 99186/291368 [01:38<03:12, 999.11it/s]\u001b[A\n",
      " 38%|███▊      | 110402/291368 [01:48<02:55, 1032.94it/s]\u001b[A\n",
      " 42%|████▏     | 122115/291368 [01:58<02:38, 1070.82it/s]\u001b[A\n",
      " 45%|████▌     | 131651/291368 [02:08<02:34, 1032.58it/s]\u001b[A\n",
      " 45%|████▌     | 131651/291368 [02:08<02:34, 1032.58it/s]\u001b[A\n",
      " 45%|████▌     | 131651/291368 [02:18<02:34, 1032.58it/s]\u001b[A\n",
      " 48%|████▊     | 139461/291368 [02:18<02:41, 941.33it/s] \u001b[A\n",
      " 52%|█████▏    | 152267/291368 [02:28<02:16, 1022.60it/s]\u001b[A\n",
      " 52%|█████▏    | 152267/291368 [02:38<02:16, 1022.60it/s]\u001b[A\n",
      " 56%|█████▋    | 164129/291368 [02:38<01:59, 1066.59it/s]\u001b[A\n",
      " 64%|██████▎   | 185201/291368 [02:48<01:24, 1252.07it/s]\u001b[A\n",
      " 64%|██████▎   | 185201/291368 [02:58<01:24, 1252.07it/s]\u001b[A\n",
      " 67%|██████▋   | 195422/291368 [02:58<01:21, 1172.59it/s]\u001b[A\n",
      " 71%|███████   | 207009/291368 [03:08<01:12, 1168.33it/s]\u001b[A\n",
      " 75%|███████▌  | 218845/291368 [03:18<01:01, 1172.86it/s]\u001b[A\n",
      " 75%|███████▌  | 218845/291368 [03:28<01:01, 1172.86it/s]\u001b[A\n",
      " 79%|███████▊  | 228788/291368 [03:28<00:56, 1112.57it/s]\u001b[A\n",
      " 83%|████████▎ | 240842/291368 [03:38<00:44, 1138.87it/s]\u001b[A\n",
      " 88%|████████▊ | 255560/291368 [03:48<00:29, 1221.76it/s]\u001b[A\n",
      " 88%|████████▊ | 255560/291368 [03:58<00:29, 1221.76it/s]\u001b[A\n",
      " 92%|█████████▏| 267352/291368 [03:58<00:19, 1208.28it/s]\u001b[A\n",
      " 97%|█████████▋| 283375/291368 [04:08<00:06, 1304.49it/s]\u001b[A\n",
      " 12%|█▎        | 1/8 [06:41<46:50, 401.43s/it]166.85it/s]\u001b[A\n",
      "  0%|          | 0/336804 [00:00<?, ?it/s]\u001b[A\n",
      "  4%|▎         | 11888/336804 [00:10<04:33, 1188.76it/s]\u001b[A\n",
      "  4%|▎         | 11888/336804 [00:21<04:33, 1188.76it/s]\u001b[A\n",
      "  6%|▌         | 19385/336804 [00:21<05:31, 957.59it/s] \u001b[A\n",
      "  9%|▊         | 29206/336804 [00:31<05:18, 964.81it/s]\u001b[A\n",
      " 12%|█▏        | 40814/336804 [00:41<04:51, 1016.27it/s]\u001b[A\n",
      " 12%|█▏        | 40814/336804 [00:51<04:51, 1016.27it/s]\u001b[A\n",
      " 15%|█▌        | 51105/336804 [00:51<04:40, 1019.91it/s]\u001b[A\n",
      " 18%|█▊        | 61924/336804 [01:01<04:24, 1037.75it/s]\u001b[A\n",
      " 18%|█▊        | 61924/336804 [01:11<04:24, 1037.75it/s]\u001b[A\n",
      " 21%|██        | 69219/336804 [01:11<04:50, 920.77it/s] \u001b[A\n",
      " 23%|██▎       | 76510/336804 [01:21<05:07, 847.42it/s]\u001b[A\n",
      " 25%|██▍       | 83656/336804 [01:32<05:22, 785.43it/s]\u001b[A\n",
      " 27%|██▋       | 90501/336804 [01:42<05:27, 752.11it/s]\u001b[A\n",
      " 29%|██▉       | 97345/336804 [01:53<05:40, 703.85it/s]\u001b[A\n",
      " 31%|███       | 104846/336804 [02:03<05:23, 717.11it/s]\u001b[A\n",
      " 33%|███▎      | 112347/336804 [02:13<05:10, 721.79it/s]\u001b[A\n",
      " 36%|███▌      | 119677/336804 [02:24<05:05, 710.00it/s]\u001b[A\n",
      " 38%|███▊      | 126517/336804 [02:34<05:00, 698.84it/s]\u001b[A\n",
      " 40%|███▉      | 134370/336804 [02:44<04:40, 722.69it/s]\u001b[A\n",
      " 42%|████▏     | 142223/336804 [02:55<04:28, 724.06it/s]\u001b[A\n",
      " 44%|████▍     | 149624/336804 [03:05<04:16, 728.80it/s]\u001b[A\n",
      " 47%|████▋     | 157164/336804 [03:15<04:04, 736.14it/s]\u001b[A\n",
      " 50%|████▉     | 167448/336804 [03:25<03:30, 804.73it/s]\u001b[A\n",
      " 53%|█████▎    | 179817/336804 [03:35<02:54, 898.95it/s]\u001b[A\n",
      " 57%|█████▋    | 192186/336804 [03:47<02:36, 923.80it/s]\u001b[A\n",
      " 60%|█████▉    | 202066/336804 [03:57<02:23, 942.16it/s]\u001b[A\n",
      " 65%|██████▌   | 219331/336804 [04:07<01:47, 1090.82it/s]\u001b[A\n",
      " 65%|██████▌   | 219331/336804 [04:21<01:47, 1090.82it/s]\u001b[A\n",
      " 68%|██████▊   | 230323/336804 [04:21<01:47, 988.53it/s] \u001b[A\n",
      " 72%|███████▏  | 243035/336804 [04:31<01:28, 1059.18it/s]\u001b[A\n",
      " 72%|███████▏  | 243035/336804 [04:41<01:28, 1059.18it/s]\u001b[A\n",
      " 75%|███████▌  | 252890/336804 [04:41<01:21, 1035.64it/s]\u001b[A\n",
      " 79%|███████▉  | 266181/336804 [04:51<01:03, 1109.09it/s]\u001b[A\n",
      " 84%|████████▎ | 281502/336804 [05:01<00:45, 1209.24it/s]\u001b[A\n",
      " 84%|████████▎ | 281502/336804 [05:11<00:45, 1209.24it/s]\u001b[A\n",
      " 88%|████████▊ | 296540/336804 [05:11<00:31, 1284.48it/s]\u001b[A\n",
      " 25%|██▌       | 2/8 [15:05<43:14, 432.34s/it]061.82it/s]\u001b[A\n",
      "  0%|          | 0/343696 [00:00<?, ?it/s]\u001b[A\n",
      "  3%|▎         | 9512/343696 [00:10<05:51, 951.14it/s]\u001b[A\n",
      "  6%|▋         | 21732/343696 [00:20<05:15, 1018.88it/s]\u001b[A\n",
      " 10%|▉         | 34127/343696 [00:30<04:47, 1076.35it/s]\u001b[A\n",
      " 13%|█▎        | 45909/343696 [00:43<04:54, 1010.19it/s]\u001b[A\n",
      " 13%|█▎        | 45909/343696 [00:43<04:54, 1010.19it/s]\u001b[A\n",
      " 17%|█▋        | 56874/343696 [00:53<04:37, 1034.62it/s]\u001b[A\n",
      " 20%|█▉        | 68656/343696 [01:03<04:16, 1073.86it/s]\u001b[A\n",
      " 23%|██▎       | 77943/343696 [01:13<04:19, 1025.58it/s]\u001b[A\n",
      " 23%|██▎       | 77943/343696 [01:13<04:19, 1025.58it/s]\u001b[A\n",
      " 27%|██▋       | 91431/343696 [01:23<03:48, 1105.01it/s]\u001b[A\n",
      " 27%|██▋       | 91431/343696 [01:33<03:48, 1105.01it/s]\u001b[A\n",
      " 31%|███       | 104844/343696 [01:33<03:24, 1166.50it/s]\u001b[A\n",
      " 31%|███       | 104844/343696 [01:43<03:24, 1166.50it/s]\u001b[A\n",
      " 34%|███▎      | 115160/343696 [01:43<03:23, 1122.32it/s]\u001b[A\n",
      " 36%|███▌      | 123460/343696 [01:53<03:36, 1014.94it/s]\u001b[A\n",
      " 36%|███▌      | 123460/343696 [01:53<03:36, 1014.94it/s]\u001b[A\n",
      " 38%|███▊      | 131884/343696 [02:03<03:41, 956.15it/s] \u001b[A\n",
      " 41%|████      | 139826/343696 [02:13<03:46, 900.87it/s]\u001b[A\n",
      " 41%|████      | 139826/343696 [02:13<03:46, 900.87it/s]\u001b[A\n",
      " 44%|████▍     | 150565/343696 [02:23<03:24, 946.59it/s]\u001b[A\n",
      " 47%|████▋     | 161623/343696 [02:33<03:04, 989.31it/s]\u001b[A\n",
      " 47%|████▋     | 161623/343696 [02:43<03:04, 989.31it/s]\u001b[A\n",
      " 50%|████▉     | 170945/343696 [02:43<02:57, 971.07it/s]\u001b[A\n",
      " 52%|█████▏    | 180256/343696 [02:56<03:06, 875.02it/s]\u001b[A\n",
      " 55%|█████▍    | 187366/343696 [03:07<03:17, 791.98it/s]\u001b[A\n",
      " 58%|█████▊    | 200013/343696 [03:17<02:41, 892.00it/s]\u001b[A\n",
      " 62%|██████▏   | 212660/343696 [03:28<02:15, 965.22it/s]\u001b[A\n",
      " 65%|██████▌   | 224600/343696 [03:39<02:01, 977.74it/s]\u001b[A\n",
      " 68%|██████▊   | 234683/343696 [03:51<01:54, 953.74it/s]\u001b[A\n",
      " 71%|███████   | 244206/343696 [04:01<01:44, 953.30it/s]\u001b[A\n",
      " 75%|███████▍  | 257445/343696 [04:11<01:22, 1040.69it/s]\u001b[A\n",
      " 79%|███████▉  | 270684/343696 [04:22<01:08, 1071.34it/s]\u001b[A\n",
      " 82%|████████▏ | 283317/343696 [04:32<00:53, 1122.49it/s]\u001b[A\n",
      " 86%|████████▌ | 295950/343696 [04:42<00:41, 1155.20it/s]\u001b[A\n",
      " 90%|████████▉ | 308345/343696 [04:53<00:30, 1162.64it/s]\u001b[A\n",
      " 93%|█████████▎| 321200/343696 [05:03<00:18, 1196.95it/s]\u001b[A\n",
      "100%|█████████▉| 343307/343696 [05:13<00:00, 1387.87it/s]\u001b[A\n",
      " 38%|███▊      | 3/8 [23:30<37:50, 454.02s/it]095.66it/s]\u001b[A\n",
      "  0%|          | 0/542020 [00:00<?, ?it/s]\u001b[A\n",
      "  3%|▎         | 14579/542020 [00:10<06:01, 1457.87it/s]\u001b[A\n",
      "  5%|▌         | 29158/542020 [00:22<06:17, 1357.35it/s]\u001b[A\n",
      "  8%|▊         | 40851/542020 [00:32<06:30, 1281.97it/s]\u001b[A\n",
      " 10%|▉         | 52364/542020 [00:42<06:34, 1239.68it/s]\u001b[A\n",
      " 10%|▉         | 52364/542020 [00:54<06:34, 1239.68it/s]\u001b[A\n",
      " 12%|█▏        | 63326/542020 [00:54<07:03, 1130.10it/s]\u001b[A\n",
      " 14%|█▍        | 76532/542020 [01:04<06:34, 1181.20it/s]\u001b[A\n",
      " 17%|█▋        | 89811/542020 [01:14<06:10, 1221.68it/s]\u001b[A\n",
      " 17%|█▋        | 89811/542020 [01:24<06:10, 1221.68it/s]\u001b[A\n",
      " 18%|█▊        | 99155/542020 [01:24<06:36, 1118.24it/s]\u001b[A\n",
      " 20%|██        | 110284/542020 [01:34<06:26, 1116.62it/s]\u001b[A\n",
      " 23%|██▎       | 122424/542020 [01:44<06:06, 1144.14it/s]\u001b[A\n",
      " 25%|██▍       | 134665/542020 [01:54<05:49, 1166.99it/s]\u001b[A\n",
      " 25%|██▍       | 134665/542020 [02:04<05:49, 1166.99it/s]\u001b[A\n",
      " 26%|██▋       | 143609/542020 [02:04<06:12, 1068.93it/s]\u001b[A\n",
      " 26%|██▋       | 143609/542020 [02:14<06:12, 1068.93it/s]\u001b[A\n",
      " 28%|██▊       | 151163/542020 [02:14<06:51, 950.40it/s] \u001b[A\n",
      " 28%|██▊       | 151163/542020 [02:24<06:51, 950.40it/s]\u001b[A\n",
      " 29%|██▉       | 158448/542020 [02:24<07:20, 870.68it/s]\u001b[A\n",
      " 31%|███       | 165476/542020 [02:34<07:43, 812.34it/s]\u001b[A\n",
      " 31%|███       | 165476/542020 [02:34<07:43, 812.34it/s]\u001b[A\n",
      " 31%|███       | 165476/542020 [02:44<07:43, 812.34it/s]\u001b[A\n",
      " 32%|███▏      | 172277/542020 [02:44<08:01, 767.38it/s]\u001b[A\n",
      " 33%|███▎      | 180969/542020 [02:54<07:33, 795.32it/s]\u001b[A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 36%|███▌      | 193033/542020 [03:04<06:33, 885.88it/s]\u001b[A\n",
      " 38%|███▊      | 206218/542020 [03:14<05:41, 982.59it/s]\u001b[A\n",
      " 38%|███▊      | 206218/542020 [03:24<05:41, 982.59it/s]\u001b[A\n",
      " 40%|████      | 219274/542020 [03:24<05:04, 1061.19it/s]\u001b[A\n",
      " 43%|████▎     | 232325/542020 [03:34<04:35, 1124.21it/s]\u001b[A\n",
      " 43%|████▎     | 232325/542020 [03:44<04:35, 1124.21it/s]\u001b[A\n",
      " 45%|████▌     | 245302/542020 [03:44<04:13, 1170.96it/s]\u001b[A\n",
      " 48%|████▊     | 258541/542020 [03:54<03:53, 1212.99it/s]\u001b[A\n",
      " 48%|████▊     | 258541/542020 [04:04<03:53, 1212.99it/s]\u001b[A\n",
      " 49%|████▉     | 267795/542020 [04:04<04:07, 1109.21it/s]\u001b[A\n",
      " 51%|█████     | 274946/542020 [04:14<04:40, 951.66it/s] \u001b[A\n",
      " 51%|█████     | 274946/542020 [04:14<04:40, 951.66it/s]\u001b[A\n",
      " 51%|█████     | 274946/542020 [04:24<04:40, 951.66it/s]\u001b[A\n",
      " 52%|█████▏    | 282068/542020 [04:24<05:00, 864.26it/s]\u001b[A\n",
      " 54%|█████▎    | 290185/542020 [04:34<04:57, 847.77it/s]\u001b[A\n",
      " 55%|█████▌    | 298865/542020 [04:44<04:44, 853.73it/s]\u001b[A\n",
      " 57%|█████▋    | 308249/542020 [04:54<04:26, 877.46it/s]\u001b[A\n",
      " 57%|█████▋    | 308249/542020 [05:04<04:26, 877.46it/s]\u001b[A\n",
      " 58%|█████▊    | 315446/542020 [05:04<04:35, 823.11it/s]\u001b[A\n",
      " 58%|█████▊    | 315446/542020 [05:14<04:35, 823.11it/s]\u001b[A\n",
      " 59%|█████▉    | 322295/542020 [05:14<04:43, 776.01it/s]\u001b[A\n",
      " 59%|█████▉    | 322295/542020 [05:24<04:43, 776.01it/s]\u001b[A\n",
      " 61%|██████    | 329112/542020 [05:24<04:45, 744.94it/s]\u001b[A\n",
      " 63%|██████▎   | 339316/542020 [05:34<04:10, 810.59it/s]\u001b[A\n",
      " 65%|██████▌   | 352759/542020 [05:44<03:25, 920.18it/s]\u001b[A\n",
      " 65%|██████▌   | 352759/542020 [05:54<03:25, 920.18it/s]\u001b[A\n",
      " 68%|██████▊   | 365868/542020 [05:54<02:54, 1010.38it/s]\u001b[A\n",
      " 70%|███████   | 380130/542020 [06:04<02:26, 1107.22it/s]\u001b[A\n",
      " 73%|███████▎  | 394483/542020 [06:14<02:04, 1188.73it/s]\u001b[A\n",
      " 73%|███████▎  | 394483/542020 [06:24<02:04, 1188.73it/s]\u001b[A\n",
      " 75%|███████▌  | 408045/542020 [06:24<01:48, 1234.26it/s]\u001b[A\n",
      " 75%|███████▌  | 408045/542020 [06:34<01:48, 1234.26it/s]\u001b[A\n",
      " 78%|███████▊  | 420141/542020 [06:34<01:39, 1226.58it/s]\u001b[A\n",
      " 80%|███████▉  | 431357/542020 [06:44<01:32, 1192.88it/s]\u001b[A\n",
      " 80%|███████▉  | 431357/542020 [06:44<01:32, 1192.88it/s]\u001b[A\n",
      " 80%|███████▉  | 431357/542020 [06:54<01:32, 1192.88it/s]\u001b[A\n",
      " 81%|████████▏ | 441303/542020 [06:54<01:29, 1125.35it/s]\u001b[A\n",
      " 83%|████████▎ | 451660/542020 [07:04<01:22, 1096.87it/s]\u001b[A\n",
      " 85%|████████▌ | 462419/542020 [07:14<01:12, 1090.44it/s]\u001b[A\n",
      " 88%|████████▊ | 474369/542020 [07:24<01:00, 1119.81it/s]\u001b[A\n",
      " 88%|████████▊ | 474369/542020 [07:34<01:00, 1119.81it/s]\u001b[A\n",
      " 90%|████████▉ | 485536/542020 [07:34<00:50, 1118.65it/s]\u001b[A\n",
      " 90%|████████▉ | 485536/542020 [07:44<00:50, 1118.65it/s]\u001b[A\n",
      " 91%|█████████▏| 495831/542020 [07:44<00:42, 1090.06it/s]\u001b[A\n",
      " 93%|█████████▎| 506105/542020 [07:54<00:33, 1070.39it/s]\u001b[A\n",
      " 93%|█████████▎| 506105/542020 [07:54<00:33, 1070.39it/s]\u001b[A\n",
      " 96%|█████████▌| 517868/542020 [08:04<00:21, 1100.09it/s]\u001b[A\n",
      " 98%|█████████▊| 529654/542020 [08:14<00:11, 1122.52it/s]\u001b[A\n",
      " 50%|█████     | 4/8 [37:28<37:57, 569.33s/it]077.23it/s]\u001b[A\n",
      "  0%|          | 0/148449 [00:00<?, ?it/s]\u001b[A\n",
      " 11%|█         | 16141/148449 [00:10<01:21, 1614.07it/s]\u001b[A\n",
      " 22%|██▏       | 32282/148449 [00:20<01:13, 1574.24it/s]\u001b[A\n",
      " 36%|███▌      | 52904/148449 [00:30<00:56, 1694.43it/s]\u001b[A\n",
      " 36%|███▌      | 52904/148449 [00:42<00:56, 1694.43it/s]\u001b[A\n",
      " 46%|████▋     | 68847/148449 [00:42<00:49, 1602.95it/s]\u001b[A\n",
      " 56%|█████▌    | 83486/148449 [00:52<00:41, 1558.48it/s]\u001b[A\n",
      " 65%|██████▌   | 96701/148449 [01:02<00:34, 1478.67it/s]\u001b[A\n",
      " 65%|██████▌   | 96701/148449 [01:02<00:34, 1478.67it/s]\u001b[A\n",
      " 74%|███████▍  | 110422/148449 [01:12<00:26, 1444.95it/s]\u001b[A\n",
      " 85%|████████▌ | 126278/148449 [01:22<00:14, 1484.32it/s]\u001b[A\n",
      " 94%|█████████▍| 140177/148449 [01:32<00:05, 1454.42it/s]\u001b[A\n",
      " 94%|█████████▍| 140177/148449 [01:32<00:05, 1454.42it/s]\u001b[A\n",
      " 62%|██████▎   | 5/8 [40:03<22:14, 444.98s/it]555.93it/s]\u001b[A\n",
      "  0%|          | 0/215696 [00:00<?, ?it/s]\u001b[A\n",
      "  9%|▉         | 19721/215696 [00:10<01:39, 1971.93it/s]\u001b[A\n",
      " 18%|█▊        | 39441/215696 [00:22<01:35, 1849.42it/s]\u001b[A\n",
      " 25%|██▌       | 54522/215696 [00:32<01:34, 1714.58it/s]\u001b[A\n",
      " 25%|██▌       | 54522/215696 [00:32<01:34, 1714.58it/s]\u001b[A\n",
      " 25%|██▌       | 54522/215696 [00:42<01:34, 1714.58it/s]\u001b[A\n",
      " 31%|███       | 66065/215696 [00:42<01:39, 1496.34it/s]\u001b[A\n",
      " 31%|███       | 66065/215696 [00:52<01:39, 1496.34it/s]\u001b[A\n",
      " 36%|███▌      | 77326/215696 [00:52<01:41, 1361.78it/s]\u001b[A\n",
      " 36%|███▌      | 77326/215696 [01:02<01:41, 1361.78it/s]\u001b[A\n",
      " 41%|████      | 88227/215696 [01:02<01:40, 1266.85it/s]\u001b[A\n",
      " 45%|████▍     | 96984/215696 [01:12<01:46, 1116.98it/s]\u001b[A\n",
      " 45%|████▍     | 96984/215696 [01:12<01:46, 1116.98it/s]\u001b[A\n",
      " 50%|████▉     | 107709/215696 [01:22<01:37, 1103.24it/s]\u001b[A\n",
      " 56%|█████▌    | 121011/215696 [01:32<01:21, 1162.76it/s]\u001b[A\n",
      " 56%|█████▌    | 121011/215696 [01:42<01:21, 1162.76it/s]\u001b[A\n",
      " 62%|██████▏   | 133402/215696 [01:42<01:09, 1184.38it/s]\u001b[A\n",
      " 67%|██████▋   | 144902/215696 [01:52<01:00, 1173.79it/s]\u001b[A\n",
      " 67%|██████▋   | 144902/215696 [01:52<01:00, 1173.79it/s]\u001b[A\n",
      " 73%|███████▎  | 157479/215696 [02:02<00:48, 1197.69it/s]\u001b[A\n",
      " 79%|███████▉  | 170428/215696 [02:12<00:36, 1225.27it/s]\u001b[A\n",
      " 75%|███████▌  | 6/8 [43:48<12:37, 378.93s/it]533.46it/s]\u001b[A\n",
      "  0%|          | 0/426134 [00:00<?, ?it/s]\u001b[A\n",
      "  3%|▎         | 11866/426134 [00:10<05:49, 1186.55it/s]\u001b[A\n",
      "  6%|▌         | 24255/426134 [00:20<05:34, 1201.76it/s]\u001b[A\n",
      "  9%|▉         | 38091/426134 [00:30<05:10, 1251.08it/s]\u001b[A\n",
      " 12%|█▏        | 51927/426134 [00:40<04:54, 1272.74it/s]\u001b[A\n",
      " 15%|█▌        | 65191/426134 [00:52<04:54, 1225.39it/s]\u001b[A\n",
      " 18%|█▊        | 76467/426134 [01:02<04:53, 1193.25it/s]\u001b[A\n",
      " 21%|██        | 87712/426134 [01:12<04:55, 1145.58it/s]\u001b[A\n",
      " 25%|██▍       | 106012/426134 [01:22<04:08, 1290.35it/s]\u001b[A\n",
      " 29%|██▉       | 125120/426134 [01:32<03:30, 1429.55it/s]\u001b[A\n",
      " 34%|███▍      | 144225/426134 [01:44<03:11, 1474.75it/s]\u001b[A\n",
      " 40%|███▉      | 169065/426134 [01:54<02:33, 1679.41it/s]\u001b[A\n",
      " 44%|████▍     | 188185/426134 [02:05<02:17, 1731.95it/s]\u001b[A\n",
      " 44%|████▍     | 188185/426134 [02:05<02:17, 1731.95it/s]\u001b[A\n",
      " 47%|████▋     | 200372/426134 [02:15<02:26, 1537.40it/s]\u001b[A\n",
      " 47%|████▋     | 200372/426134 [02:15<02:26, 1537.40it/s]\u001b[A\n",
      " 50%|█████     | 213767/426134 [02:25<02:24, 1472.02it/s]\u001b[A\n",
      " 53%|█████▎    | 225167/426134 [02:35<02:28, 1353.55it/s]\u001b[A\n",
      " 53%|█████▎    | 225167/426134 [02:35<02:28, 1353.55it/s]\u001b[A\n",
      " 53%|█████▎    | 225167/426134 [02:45<02:28, 1353.55it/s]\u001b[A\n",
      " 55%|█████▌    | 234895/426134 [02:45<02:37, 1211.00it/s]\u001b[A\n",
      " 58%|█████▊    | 247942/426134 [02:55<02:23, 1237.65it/s]\u001b[A\n",
      " 62%|██████▏   | 262332/426134 [03:05<02:06, 1291.87it/s]\u001b[A\n",
      " 62%|██████▏   | 262332/426134 [03:15<02:06, 1291.87it/s]\u001b[A\n",
      " 64%|██████▍   | 272519/426134 [03:15<02:08, 1195.26it/s]\u001b[A\n",
      " 68%|██████▊   | 288838/426134 [03:25<01:45, 1299.56it/s]\u001b[A\n",
      " 72%|███████▏  | 307191/426134 [03:35<01:23, 1424.24it/s]\u001b[A\n",
      " 76%|███████▋  | 325542/426134 [03:46<01:07, 1494.82it/s]\u001b[A\n",
      " 80%|████████  | 342838/426134 [03:56<00:53, 1558.24it/s]\u001b[A\n",
      " 85%|████████▍ | 360133/426134 [04:08<00:44, 1487.46it/s]\u001b[A\n",
      " 88%|████████▊ | 375259/426134 [04:18<00:34, 1494.90it/s]\u001b[A\n",
      " 99%|█████████▊| 420191/426134 [04:28<00:03, 1869.06it/s]\u001b[A\n",
      " 88%|████████▊ | 7/8 [51:07<06:36, 396.95s/it]565.05it/s]\u001b[A\n",
      "  0%|          | 0/154980 [00:00<?, ?it/s]\u001b[A\n",
      " 16%|█▌        | 24446/154980 [00:10<00:53, 2444.48it/s]\u001b[A\n",
      " 30%|███       | 46950/154980 [00:28<00:58, 1858.07it/s]\u001b[A\n",
      " 30%|███       | 46950/154980 [00:28<00:58, 1858.07it/s]\u001b[A\n",
      " 41%|████      | 63584/154980 [00:38<00:50, 1795.02it/s]\u001b[A\n",
      " 41%|████      | 63584/154980 [00:48<00:50, 1795.02it/s]\u001b[A\n",
      " 49%|████▊     | 75269/154980 [00:48<00:51, 1545.89it/s]\u001b[A\n",
      " 49%|████▊     | 75269/154980 [00:58<00:51, 1545.89it/s]\u001b[A\n",
      " 55%|█████▍    | 84933/154980 [00:58<00:53, 1309.82it/s]\u001b[A\n",
      " 65%|██████▍   | 100242/154980 [01:08<00:39, 1369.12it/s]\u001b[A\n",
      " 72%|███████▏  | 111249/154980 [01:18<00:34, 1275.58it/s]\u001b[A\n",
      " 72%|███████▏  | 111249/154980 [01:18<00:34, 1275.58it/s]\u001b[A\n",
      " 82%|████████▏ | 126326/154980 [01:28<00:21, 1337.34it/s]\u001b[A\n",
      " 89%|████████▉ | 138295/154980 [01:38<00:12, 1291.60it/s]\u001b[A\n",
      " 89%|████████▉ | 138295/154980 [01:38<00:12, 1291.60it/s]\u001b[A\n",
      " 89%|████████▉ | 138295/154980 [01:48<00:12, 1291.60it/s]\u001b[A\n",
      " 95%|█████████▌| 147553/154980 [01:48<00:06, 1154.40it/s]\u001b[A\n",
      "100%|██████████| 8/8 [54:04<00:00, 330.95s/it]372.78it/s]\u001b[A\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2459147"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stream = row_stream(doc_stream(file_meta_stream(archives)))\n",
    "rows = list(stream)\n",
    "len(rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2459147 entries, 0 to 2459146\n",
      "Data columns (total 26 columns):\n",
      "archive                 2459147 non-null object\n",
      "id                      2459146 non-null object\n",
      "name                    2459147 non-null object\n",
      "path                    2459147 non-null object\n",
      "term: t cell            2459147 non-null bool\n",
      "term: t lymphocyte      2459147 non-null bool\n",
      "term: t-cell            2459147 non-null bool\n",
      "term: t-lymphocyte      2459147 non-null bool\n",
      "term:cd3                2459147 non-null bool\n",
      "term:cd4                2459147 non-null bool\n",
      "term:cd8                2459147 non-null bool\n",
      "term:cytokine           2459147 non-null bool\n",
      "term:differentiate      2459147 non-null bool\n",
      "term:differentiated     2459147 non-null bool\n",
      "term:differentiation    2459147 non-null bool\n",
      "term:expression         2459147 non-null bool\n",
      "term:human              2459147 non-null bool\n",
      "term:induce             2459147 non-null bool\n",
      "term:induction          2459147 non-null bool\n",
      "term:mouse              2459147 non-null bool\n",
      "term:murine             2459147 non-null bool\n",
      "term:phenotype          2459147 non-null bool\n",
      "term:polarization       2459147 non-null bool\n",
      "term:polarize           2459147 non-null bool\n",
      "term:surface            2459147 non-null bool\n",
      "venue                   2459147 non-null object\n",
      "dtypes: bool(21), object(5)\n",
      "memory usage: 143.1+ MB\n"
     ]
    }
   ],
   "source": [
    "pd.set_option('display.max_info_rows', 10000000)\n",
    "df = pd.DataFrame(rows)\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export and Reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lab/data/pmc_oa/bulk/file_meta.feather'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "export_file = osp.join(DATA_DIR, 'pmc_oa', 'bulk', 'file_meta.feather')\n",
    "df.to_feather(export_file)\n",
    "export_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2459147 entries, 0 to 2459146\n",
      "Data columns (total 26 columns):\n",
      "archive                 2459147 non-null object\n",
      "id                      2459146 non-null object\n",
      "name                    2459147 non-null object\n",
      "path                    2459147 non-null object\n",
      "term: t cell            2459147 non-null bool\n",
      "term: t lymphocyte      2459147 non-null bool\n",
      "term: t-cell            2459147 non-null bool\n",
      "term: t-lymphocyte      2459147 non-null bool\n",
      "term:cd3                2459147 non-null bool\n",
      "term:cd4                2459147 non-null bool\n",
      "term:cd8                2459147 non-null bool\n",
      "term:cytokine           2459147 non-null bool\n",
      "term:differentiate      2459147 non-null bool\n",
      "term:differentiated     2459147 non-null bool\n",
      "term:differentiation    2459147 non-null bool\n",
      "term:expression         2459147 non-null bool\n",
      "term:human              2459147 non-null bool\n",
      "term:induce             2459147 non-null bool\n",
      "term:induction          2459147 non-null bool\n",
      "term:mouse              2459147 non-null bool\n",
      "term:murine             2459147 non-null bool\n",
      "term:phenotype          2459147 non-null bool\n",
      "term:polarization       2459147 non-null bool\n",
      "term:polarize           2459147 non-null bool\n",
      "term:surface            2459147 non-null bool\n",
      "venue                   2459147 non-null object\n",
      "dtypes: bool(21), object(5)\n",
      "memory usage: 143.1+ MB\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_feather(export_file)\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2459147 entries, 0 to 2459146\n",
      "Data columns (total 16 columns):\n",
      "archive               2459147 non-null object\n",
      "id                    2459146 non-null object\n",
      "name                  2459147 non-null object\n",
      "path                  2459147 non-null object\n",
      "term:cytokine         2459147 non-null bool\n",
      "term:differentiate    2459147 non-null bool\n",
      "term:expression       2459147 non-null bool\n",
      "term:human            2459147 non-null bool\n",
      "term:induce           2459147 non-null bool\n",
      "term:mouse            2459147 non-null bool\n",
      "term:phenotype        2459147 non-null bool\n",
      "term:polarize         2459147 non-null bool\n",
      "term:surface          2459147 non-null bool\n",
      "venue                 2459147 non-null object\n",
      "term:cd               2459147 non-null bool\n",
      "term:t cell           2459147 non-null bool\n",
      "dtypes: bool(11), object(5)\n",
      "memory usage: 119.6+ MB\n"
     ]
    }
   ],
   "source": [
    "grps = {\n",
    "    'cd': ['cd3', 'cd4', 'cd8'],\n",
    "    'differentiate': ['differentiate', 'differentiated', 'differentiation'],\n",
    "    'induce': ['induce', 'induction'],\n",
    "    'mouse': ['mouse', 'murine'],\n",
    "    'polarize': ['polarization', 'polarize'],\n",
    "    't cell': [' t cell', ' t lymphocyte', ' t-cell', ' t-lymphocyte']\n",
    "}\n",
    "dfg = df.copy()\n",
    "for k, g in grps.items():\n",
    "    k = 'term:' + k\n",
    "    g = ['term:' + v for v in g]\n",
    "    dfg[k] = dfg[g].any(axis=1)\n",
    "    dfg = dfg.drop([c for c in g if c != k], axis=1)\n",
    "dfg.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dfg.groupby(dfg.filter(regex='term:').columns.tolist()).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "term:t cell  term:human  term:mouse  term:cd\n",
       "False        False       False       False      1712822\n",
       "                                     True         14809\n",
       "                         True        False        47241\n",
       "                                     True          2368\n",
       "             True        False       False       400022\n",
       "                                     True         13460\n",
       "                         True        False       138904\n",
       "                                     True         18836\n",
       "True         False       False       False        12775\n",
       "                                     True         11734\n",
       "                         True        False         2911\n",
       "                                     True          3923\n",
       "             True        False       False        14353\n",
       "                                     True         16028\n",
       "                         True        False        16295\n",
       "                                     True         32666\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg.groupby(['term:' + c for c in ['t cell', 'human', 'mouse', 'cd']]).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "term:human  term:t cell  term:cd  term:phenotype  term:surface\n",
       "False       False        False    False           False           1576049\n",
       "                                                  True             128187\n",
       "                                  True            False             46113\n",
       "                                                  True               9714\n",
       "                         True     False           False             11337\n",
       "                                                  True               3062\n",
       "                                  True            False              1673\n",
       "                                                  True               1105\n",
       "            True         False    False           False             11288\n",
       "                                                  True               2391\n",
       "                                  True            False              1541\n",
       "                                                  True                466\n",
       "                         True     False           False              9274\n",
       "                                                  True               2476\n",
       "                                  True            False              2337\n",
       "                                                  True               1570\n",
       "True        False        False    False           False            344306\n",
       "                                                  True             101401\n",
       "                                  True            False             65172\n",
       "                                                  True              28047\n",
       "                         True     False           False             11104\n",
       "                                                  True               7964\n",
       "                                  True            False              5499\n",
       "                                                  True               7729\n",
       "            True         False    False           False             13975\n",
       "                                                  True               7538\n",
       "                                  True            False              5247\n",
       "                                                  True               3888\n",
       "                         True     False           False             12710\n",
       "                                                  True              12759\n",
       "                                  True            False              7508\n",
       "                                                  True              15717\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg.groupby(['term:' + c for c in ['human', 't cell', 'cd', 'phenotype', 'surface']]).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "term:differentiate  term:polarize  term:cytokine  term:induce\n",
       "False               False          False          False           3695\n",
       "                                                  True            5803\n",
       "                                   True           False           1258\n",
       "                                                  True            9301\n",
       "                    True           False          False             62\n",
       "                                                  True             276\n",
       "                                   True           False             52\n",
       "                                                  True            1194\n",
       "True                False          False          False           1169\n",
       "                                                  True            4189\n",
       "                                   True           False            845\n",
       "                                                  True           15273\n",
       "                    True           False          False             34\n",
       "                                                  True             365\n",
       "                                   True           False             65\n",
       "                                                  True            5113\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg.groupby(['term:' + c for c in [\n",
    "    'human', 't cell', 'cd', 'differentiate', 'polarize', 'cytokine', 'induce'\n",
    "]]).size().loc[(True, True, True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    2443119\n",
       "True       16028\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_mask(df):\n",
    "    return df['term:t cell'] & df['term:human'] & (~df['term:mouse']) & df['term:cd']\n",
    "    #return df['term:t cell'] & df['term:human'] & df['term:cd']\n",
    "mask = get_mask(dfg)\n",
    "dfe = dfg[mask].drop(dfg.filter(regex='term:').columns.tolist(), axis=1)\n",
    "mask.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 16028 entries, 583 to 2459096\n",
      "Data columns (total 5 columns):\n",
      "archive    16028 non-null object\n",
      "id         16028 non-null object\n",
      "name       16028 non-null object\n",
      "path       16028 non-null object\n",
      "venue      16028 non-null object\n",
      "dtypes: object(5)\n",
      "memory usage: 751.3+ KB\n"
     ]
    }
   ],
   "source": [
    "dfe.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>archive</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "      <th>venue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...</td>\n",
       "      <td>PMC1636025</td>\n",
       "      <td>PMC1636025.nxml</td>\n",
       "      <td>Biol_Direct/PMC1636025.nxml</td>\n",
       "      <td>Biol_Direct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...</td>\n",
       "      <td>PMC2390528</td>\n",
       "      <td>PMC2390528.nxml</td>\n",
       "      <td>Biol_Direct/PMC2390528.nxml</td>\n",
       "      <td>Biol_Direct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...</td>\n",
       "      <td>PMC3203257</td>\n",
       "      <td>PMC3203257.nxml</td>\n",
       "      <td>Biol_Direct/PMC3203257.nxml</td>\n",
       "      <td>Biol_Direct</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               archive          id  \\\n",
       "583  /lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...  PMC1636025   \n",
       "618  /lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...  PMC2390528   \n",
       "773  /lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...  PMC3203257   \n",
       "\n",
       "                name                         path        venue  \n",
       "583  PMC1636025.nxml  Biol_Direct/PMC1636025.nxml  Biol_Direct  \n",
       "618  PMC2390528.nxml  Biol_Direct/PMC2390528.nxml  Biol_Direct  \n",
       "773  PMC3203257.nxml  Biol_Direct/PMC3203257.nxml  Biol_Direct  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfe.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1600, 8)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfe['venue'].nunique(), dfe['archive'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>archive</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "      <th>venue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...</td>\n",
       "      <td>PMC1636025</td>\n",
       "      <td>PMC1636025.nxml</td>\n",
       "      <td>Biol_Direct/PMC1636025.nxml</td>\n",
       "      <td>Biol_Direct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...</td>\n",
       "      <td>PMC2390528</td>\n",
       "      <td>PMC2390528.nxml</td>\n",
       "      <td>Biol_Direct/PMC2390528.nxml</td>\n",
       "      <td>Biol_Direct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...</td>\n",
       "      <td>PMC3203257</td>\n",
       "      <td>PMC3203257.nxml</td>\n",
       "      <td>Biol_Direct/PMC3203257.nxml</td>\n",
       "      <td>Biol_Direct</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               archive          id  \\\n",
       "583  /lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...  PMC1636025   \n",
       "618  /lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...  PMC2390528   \n",
       "773  /lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.t...  PMC3203257   \n",
       "\n",
       "                name                         path        venue  \n",
       "583  PMC1636025.nxml  Biol_Direct/PMC1636025.nxml  Biol_Direct  \n",
       "618  PMC2390528.nxml  Biol_Direct/PMC2390528.nxml  Biol_Direct  \n",
       "773  PMC3203257.nxml  Biol_Direct/PMC3203257.nxml  Biol_Direct  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfe[dfe['id'].notnull()].head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export Corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyarrow.parquet as pq\n",
    "import pyarrow as pa\n",
    "\n",
    "def filtered_file_meta_stream(df, meta_stream):\n",
    "    paths = df['path'].unique()\n",
    "    for tar, file, meta in meta_stream:\n",
    "        if meta['path'] in paths:\n",
    "            yield tar, file, meta\n",
    "\n",
    "def full_text_stream(file_meta):\n",
    "    for tar, file, meta in file_meta:\n",
    "        try:\n",
    "            text = tar.extractfile(file).read().decode('utf-8', errors='ignore')\n",
    "            row = {'arch_' + k:v for k, v in meta.items()}\n",
    "            yield row, text\n",
    "        except Exception as e:\n",
    "            raise type(e)(e.message + f'; (archive = {tar.name}, file = {file.path}')\n",
    "\n",
    "def extract_corpus(stream, output_file, batch_size=1000):\n",
    "    dfs = []\n",
    "    writer = None\n",
    "    \n",
    "    def flush(dfs, writer):\n",
    "        dfs = pd.concat(dfs)\n",
    "        table = pa.Table.from_pandas(dfs, preserve_index=False)\n",
    "        if writer is None:\n",
    "            writer = pq.ParquetWriter(output_file, table.schema)\n",
    "        writer.write_table(table)\n",
    "        return writer\n",
    "    \n",
    "    for row, text in stream:\n",
    "        df = parse_nxml(text)\n",
    "        df = df.assign(**{k:v for k, v in row.items() if k.startswith('arch_')})\n",
    "        \n",
    "        # Convert to string to avoid issue with all null vs datetime type fields\n",
    "        for c in df.filter(regex='date_'):\n",
    "            df[c] = df[c].astype(str)\n",
    "\n",
    "        if len(df) > 0:\n",
    "            dfs.append(df)\n",
    "        if len(dfs) >= batch_size:\n",
    "            writer = flush(dfs, writer)\n",
    "            dfs = []\n",
    "    if len(dfs) > 0:\n",
    "        writer = flush(dfs, writer)\n",
    "    if writer is not None:\n",
    "        writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lab/data/articles/import/20190621/corpus.parquet'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_file = osp.join(DATA_DIR, 'articles', 'import', '20190621', 'corpus.parquet')\n",
    "output_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/lab/data/pmc_oa/bulk/files/comm_use.A-B.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/comm_use.C-H.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/comm_use.I-N.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/comm_use.O-Z.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.A-B.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.C-H.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.I-N.xml.tar.gz',\n",
       " '/lab/data/pmc_oa/bulk/files/non_comm_use.O-Z.xml.tar.gz']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_archives = list(dfe['archive'].unique())\n",
    "target_archives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "  0%|          | 0/8 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/291368 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 5864/291368 [00:10<08:19, 571.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 5864/291368 [00:24<08:19, 571.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▎         | 10351/291368 [00:24<10:11, 459.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  5%|▌         | 15598/291368 [00:34<09:38, 476.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  7%|▋         | 20810/291368 [00:48<10:21, 435.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 27706/291368 [00:58<08:58, 489.63it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 12%|█▏        | 35841/291368 [01:10<07:51, 542.19it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▌        | 45028/291368 [01:20<06:38, 618.19it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▌        | 45028/291368 [01:34<06:38, 618.19it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 18%|█▊        | 52534/291368 [01:34<06:50, 581.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|█▉        | 57637/291368 [01:45<07:01, 554.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 21%|██▏       | 62638/291368 [02:03<09:05, 418.95it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 24%|██▍       | 69350/291368 [02:13<07:50, 472.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 27%|██▋       | 77821/291368 [02:24<06:33, 542.13it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|██▉       | 87282/291368 [02:34<05:33, 611.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 96789/291368 [02:44<04:43, 685.19it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 96789/291368 [02:56<04:43, 685.19it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 36%|███▋      | 105823/291368 [02:57<04:29, 689.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|███▉      | 115320/291368 [03:09<04:01, 728.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|███▉      | 115320/291368 [03:27<04:01, 728.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 42%|████▏     | 123251/291368 [03:28<04:44, 591.40it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 44%|████▍     | 128984/291368 [03:38<04:38, 583.03it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 46%|████▌     | 134657/291368 [03:48<04:32, 575.76it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 49%|████▊     | 141443/291368 [03:58<04:08, 603.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████     | 148229/291368 [04:09<03:51, 617.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 54%|█████▎    | 155966/291368 [04:19<03:26, 656.49it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 58%|█████▊    | 167710/291368 [04:29<02:43, 756.59it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 64%|██████▎   | 185280/291368 [04:39<01:56, 912.44it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 66%|██████▋   | 193162/291368 [04:59<02:30, 653.35it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 66%|██████▋   | 193162/291368 [04:59<02:30, 653.35it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 66%|██████▋   | 193162/291368 [05:09<02:30, 653.35it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 69%|██████▉   | 201602/291368 [05:09<02:08, 698.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 71%|███████   | 207195/291368 [05:19<02:09, 649.85it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 71%|███████   | 207195/291368 [05:19<02:09, 649.85it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 73%|███████▎  | 214063/291368 [05:29<01:57, 660.50it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 219331/291368 [05:39<01:57, 611.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 219331/291368 [05:39<01:57, 611.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 219331/291368 [05:49<01:57, 611.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 77%|███████▋  | 225010/291368 [05:50<01:53, 583.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 79%|███████▉  | 230286/291368 [06:01<01:53, 539.54it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 82%|████████▏ | 238586/291368 [06:11<01:27, 602.83it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 85%|████████▍ | 246886/291368 [06:22<01:09, 640.26it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 88%|████████▊ | 255429/291368 [06:33<00:53, 674.68it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 90%|█████████ | 263144/291368 [06:48<00:45, 619.78it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 92%|█████████▏| 268353/291368 [06:59<00:39, 575.75it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 95%|█████████▍| 276480/291368 [07:09<00:23, 630.93it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 95%|█████████▍| 276480/291368 [07:21<00:23, 630.93it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 96%|█████████▋| 280697/291368 [07:21<00:21, 496.13it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 291368/291368 [07:29<00:00, 647.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 12%|█▎        | 1/8 [10:05<1:10:35, 605.08s/it]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/336804 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 8056/336804 [00:10<06:50, 800.93it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  5%|▍         | 16066/336804 [00:23<07:17, 733.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  5%|▍         | 16066/336804 [00:35<07:17, 733.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  6%|▌         | 20037/336804 [00:35<10:07, 521.70it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  8%|▊         | 26427/336804 [00:45<09:22, 552.10it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|█         | 34694/336804 [00:55<08:12, 613.20it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|█         | 34694/336804 [01:06<08:12, 613.20it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 12%|█▏        | 41583/336804 [01:06<08:00, 615.02it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▍        | 49597/336804 [01:16<07:14, 661.14it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 17%|█▋        | 57611/336804 [01:28<06:54, 673.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|█▉        | 66037/336804 [01:38<06:17, 716.66it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|█▉        | 66037/336804 [01:49<06:17, 716.66it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 21%|██        | 70536/336804 [01:50<07:51, 564.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 22%|██▏       | 74316/336804 [02:03<09:53, 442.20it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 23%|██▎       | 78211/336804 [02:13<10:19, 417.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 24%|██▍       | 81905/336804 [02:24<11:00, 386.14it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 27%|██▋       | 90664/336804 [02:34<08:50, 463.97it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|██▉       | 99423/336804 [02:49<07:55, 499.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|██▉       | 99423/336804 [03:00<07:55, 499.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|███       | 102384/336804 [03:00<10:01, 389.66it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|███       | 102384/336804 [03:20<10:01, 389.66it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███       | 103851/336804 [03:20<22:50, 170.00it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███       | 104586/336804 [03:36<40:15, 96.14it/s] \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███       | 105198/336804 [03:46<47:10, 81.84it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███▏      | 105968/336804 [03:56<48:07, 79.95it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 106732/336804 [04:06<48:41, 78.75it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 107493/336804 [04:16<49:49, 76.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 108217/336804 [04:27<51:17, 74.28it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 108968/336804 [04:37<51:12, 74.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 109710/336804 [04:47<51:10, 73.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 110445/336804 [05:00<54:52, 68.74it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 111098/336804 [05:10<55:46, 67.45it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 35%|███▌      | 118571/336804 [05:20<39:12, 92.76it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 35%|███▌      | 118571/336804 [05:34<39:12, 92.76it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 125808/336804 [05:34<28:35, 123.00it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 125808/336804 [05:44<28:35, 123.00it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 38%|███▊      | 127865/336804 [05:44<25:11, 138.21it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 39%|███▉      | 130608/336804 [05:55<21:17, 161.38it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|███▉      | 133267/336804 [06:05<18:47, 180.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|████      | 136032/336804 [06:15<16:37, 201.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 41%|████▏     | 139460/336804 [06:26<14:22, 228.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 42%|████▏     | 142811/336804 [06:37<13:17, 243.26it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 44%|████▎     | 147212/336804 [06:48<11:15, 280.52it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▌     | 152117/336804 [06:58<09:33, 321.86it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 47%|████▋     | 157022/336804 [07:08<08:28, 353.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 48%|████▊     | 162829/336804 [07:19<07:17, 397.38it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████     | 170987/336804 [07:29<05:53, 469.62it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 179144/336804 [07:39<04:55, 533.68it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 56%|█████▌    | 186973/336804 [07:55<04:47, 520.52it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 56%|█████▌    | 186973/336804 [08:06<04:47, 520.52it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 58%|█████▊    | 195822/336804 [08:06<04:02, 580.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 61%|██████    | 203773/336804 [08:18<03:38, 608.27it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 63%|██████▎   | 210615/336804 [08:32<03:44, 561.64it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▌   | 219849/336804 [08:43<03:06, 626.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▌   | 219849/336804 [08:56<03:06, 626.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 66%|██████▌   | 222344/336804 [08:56<05:07, 372.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 66%|██████▌   | 222344/336804 [09:06<05:07, 372.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 67%|██████▋   | 225366/336804 [09:06<05:25, 341.84it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 68%|██████▊   | 228238/336804 [09:19<06:05, 296.76it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 70%|██████▉   | 235298/336804 [09:29<04:42, 358.78it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 72%|███████▏  | 242598/336804 [09:39<03:42, 423.37it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 74%|███████▍  | 249898/336804 [09:53<03:15, 445.61it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 76%|███████▌  | 254977/336804 [10:05<03:03, 446.99it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 78%|███████▊  | 261814/336804 [10:15<02:30, 498.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 80%|████████  | 270684/336804 [10:25<01:55, 573.83it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 83%|████████▎ | 279526/336804 [10:37<01:34, 607.37it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 85%|████████▌ | 287570/336804 [10:47<01:15, 655.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 293629/336804 [10:59<01:10, 610.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 293629/336804 [10:59<01:10, 610.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 90%|█████████ | 303658/336804 [11:09<00:47, 692.06it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 95%|█████████▌| 320671/336804 [11:19<00:19, 841.88it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 336804/336804 [11:30<00:00, 487.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 25%|██▌       | 2/8 [24:41<1:08:39, 686.61s/it]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/343696 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 6050/343696 [00:10<09:21, 601.06it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▎         | 12061/343696 [00:20<09:18, 593.61it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▎         | 12061/343696 [00:31<09:18, 593.61it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▍         | 14362/343696 [00:31<14:13, 385.73it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  5%|▍         | 16485/343696 [00:42<18:22, 296.87it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  6%|▌         | 20671/343696 [00:52<16:33, 325.24it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  8%|▊         | 28953/343696 [01:02<13:11, 397.70it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 11%|█         | 37235/343696 [01:12<10:57, 466.31it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 13%|█▎        | 45041/343696 [01:25<09:55, 501.20it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▍        | 51114/343696 [01:36<09:26, 516.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 17%|█▋        | 59514/343696 [01:46<08:06, 583.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|█▉        | 67914/343696 [01:58<07:31, 610.79it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 22%|██▏       | 74758/343696 [02:10<07:25, 604.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 22%|██▏       | 74758/343696 [02:22<07:25, 604.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 23%|██▎       | 80520/343696 [02:23<07:56, 552.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 25%|██▍       | 85130/343696 [02:33<08:26, 510.13it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 27%|██▋       | 92448/343696 [02:44<07:30, 557.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 29%|██▉       | 99573/343696 [02:57<07:23, 550.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███       | 104910/343696 [03:11<08:19, 477.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 112408/343696 [03:22<07:12, 535.02it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 112408/343696 [03:32<07:12, 535.02it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 35%|███▍      | 119311/343696 [03:33<06:41, 559.43it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 125967/343696 [03:43<06:15, 579.79it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 38%|███▊      | 132304/343696 [03:57<06:34, 535.99it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|███▉      | 136861/343696 [04:13<08:05, 426.19it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 41%|████      | 139745/343696 [04:23<09:18, 365.35it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 43%|████▎     | 146339/343696 [04:33<07:47, 421.78it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▍     | 153565/343696 [04:44<06:36, 479.26it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▍     | 153565/343696 [04:54<06:36, 479.26it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 46%|████▌     | 158598/343696 [04:55<06:32, 472.06it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 48%|████▊     | 163793/343696 [05:05<06:10, 484.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 48%|████▊     | 163793/343696 [05:25<06:10, 484.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 49%|████▉     | 168665/343696 [05:25<07:50, 371.62it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 49%|████▉     | 168665/343696 [05:36<07:50, 371.62it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 50%|█████     | 173105/343696 [05:36<07:24, 383.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 52%|█████▏    | 177268/343696 [05:50<07:58, 347.93it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 52%|█████▏    | 180123/343696 [06:03<09:03, 301.21it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 182417/343696 [06:14<10:04, 266.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 54%|█████▍    | 184752/343696 [06:24<10:40, 248.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 55%|█████▌    | 189939/343696 [06:35<08:43, 293.87it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 57%|█████▋    | 197115/343696 [06:45<06:53, 354.11it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 60%|█████▉    | 205530/343696 [06:55<05:22, 427.77it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 60%|█████▉    | 205530/343696 [07:06<05:22, 427.77it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 62%|██████▏   | 211921/343696 [07:06<04:40, 469.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 64%|██████▍   | 221263/343696 [07:17<03:45, 541.88it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 67%|██████▋   | 229716/343696 [07:32<03:27, 548.12it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 69%|██████▉   | 236925/343696 [07:42<03:00, 590.58it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 71%|███████   | 244134/343696 [07:54<02:48, 590.30it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 73%|███████▎  | 251162/343696 [08:04<02:30, 615.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 258664/343696 [08:14<02:10, 650.47it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 258664/343696 [08:26<02:10, 650.47it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 77%|███████▋  | 265532/343696 [08:26<02:02, 635.90it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 79%|███████▉  | 271576/343696 [08:36<01:56, 619.15it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 81%|████████▏ | 279365/343696 [08:46<01:37, 658.63it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 84%|████████▍ | 288633/343696 [08:56<01:16, 721.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 84%|████████▍ | 288633/343696 [09:06<01:16, 721.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 86%|████████▌ | 295005/343696 [09:06<01:10, 690.34it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 86%|████████▌ | 295005/343696 [09:26<01:10, 690.34it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 298427/343696 [09:27<02:06, 357.46it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 88%|████████▊ | 303863/343696 [09:37<01:39, 398.39it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 90%|████████▉ | 309299/343696 [09:47<01:20, 428.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 92%|█████████▏| 316056/343696 [09:57<00:57, 481.59it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 94%|█████████▍| 322813/343696 [10:07<00:40, 521.77it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 97%|█████████▋| 332303/343696 [10:17<00:18, 603.24it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|█████████▉| 343481/343696 [10:28<00:00, 698.74it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 343696/343696 [10:28<00:00, 547.14it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 38%|███▊      | 3/8 [38:27<1:00:41, 728.34s/it]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/542020 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 9206/542020 [00:10<09:38, 920.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 9206/542020 [00:21<09:38, 920.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  3%|▎         | 16839/542020 [00:21<10:34, 827.82it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▍         | 23542/542020 [00:40<14:51, 581.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▍         | 23542/542020 [00:51<14:51, 581.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  6%|▌         | 32647/542020 [00:51<13:09, 645.56it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  8%|▊         | 41337/542020 [01:03<12:31, 666.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  9%|▉         | 48540/542020 [01:15<12:54, 637.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  9%|▉         | 48540/542020 [01:31<12:54, 637.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 52026/542020 [01:31<19:40, 415.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 52026/542020 [01:51<19:40, 415.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 53116/542020 [01:51<58:58, 138.17it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 11%|█         | 59085/542020 [02:01<44:49, 179.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 12%|█▏        | 65639/542020 [02:11<34:38, 229.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 13%|█▎        | 72206/542020 [02:21<27:30, 284.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▍        | 78723/542020 [02:32<22:43, 339.82it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 16%|█▌        | 85562/542020 [02:42<19:16, 394.69it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 17%|█▋        | 91896/542020 [02:53<17:03, 439.74it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 18%|█▊        | 97890/542020 [03:05<16:24, 451.07it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 19%|█▉        | 102690/542020 [03:18<16:55, 432.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|██        | 108642/542020 [03:28<15:20, 470.59it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 21%|██▏       | 115501/542020 [03:38<13:52, 512.25it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 23%|██▎       | 121958/542020 [03:49<13:02, 536.90it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 24%|██▎       | 128007/542020 [03:59<12:29, 552.04it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 25%|██▍       | 134971/542020 [04:09<11:31, 588.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 25%|██▍       | 134971/542020 [04:24<11:31, 588.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 26%|██▌       | 141869/542020 [04:24<12:12, 546.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 27%|██▋       | 146542/542020 [04:36<13:44, 479.45it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 28%|██▊       | 150513/542020 [04:46<14:27, 451.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 29%|██▊       | 154484/542020 [04:59<16:03, 402.12it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 29%|██▉       | 158057/542020 [05:09<16:34, 386.26it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|██▉       | 161774/542020 [05:19<16:36, 381.77it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███       | 165491/542020 [05:31<17:42, 354.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 31%|███       | 168742/542020 [05:41<18:04, 344.07it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 172476/542020 [05:51<17:32, 351.26it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 176169/542020 [06:02<17:28, 349.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 180392/542020 [06:12<16:28, 365.94it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 34%|███▍      | 184517/542020 [06:23<15:56, 373.94it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 35%|███▍      | 189185/542020 [06:33<14:47, 397.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 36%|███▌      | 193853/542020 [06:44<14:20, 404.59it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 199423/542020 [06:54<12:57, 440.77it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 38%|███▊      | 204993/542020 [07:04<12:07, 462.98it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 38%|███▊      | 204993/542020 [07:15<12:07, 462.98it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 39%|███▊      | 209830/542020 [07:15<11:55, 464.59it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|███▉      | 215609/542020 [07:25<11:01, 493.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|███▉      | 215609/542020 [07:35<11:01, 493.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 41%|████      | 220903/542020 [07:35<10:40, 501.09it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 42%|████▏     | 227252/542020 [07:45<09:48, 534.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 42%|████▏     | 227252/542020 [07:55<09:48, 534.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 43%|████▎     | 232942/542020 [07:55<09:32, 540.27it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 44%|████▍     | 239215/542020 [08:05<08:58, 562.52it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▌     | 245439/542020 [08:17<08:55, 554.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 46%|████▋     | 251451/542020 [08:27<08:32, 567.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 48%|████▊     | 257837/542020 [08:37<08:04, 587.04it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 49%|████▊     | 264223/542020 [08:49<08:04, 573.88it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 50%|████▉     | 269677/542020 [09:02<08:45, 518.52it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████     | 273910/542020 [09:12<09:13, 484.05it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████▏    | 278397/542020 [09:22<09:27, 464.75it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 52%|█████▏    | 283090/542020 [09:32<09:15, 466.10it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 52%|█████▏    | 283090/542020 [09:45<09:15, 466.10it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 285865/542020 [09:45<12:23, 344.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 285865/542020 [09:55<12:23, 344.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 286772/542020 [09:55<23:03, 184.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 287857/542020 [10:06<27:52, 152.00it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 54%|█████▍    | 291571/542020 [10:16<22:36, 184.60it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 54%|█████▍    | 295266/542020 [10:32<21:01, 195.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 54%|█████▍    | 295266/542020 [10:45<21:01, 195.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 55%|█████▍    | 296791/542020 [10:45<25:31, 160.15it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 55%|█████▌    | 298138/542020 [10:55<26:49, 151.50it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 55%|█████▌    | 299811/542020 [11:05<25:56, 155.63it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 56%|█████▌    | 303930/542020 [11:15<20:44, 191.35it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 57%|█████▋    | 309198/542020 [11:26<16:25, 236.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 57%|█████▋    | 309198/542020 [11:36<16:25, 236.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 58%|█████▊    | 312519/542020 [11:36<14:49, 257.87it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 58%|█████▊    | 315794/542020 [11:46<13:47, 273.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 59%|█████▉    | 318980/542020 [11:57<13:32, 274.38it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 59%|█████▉    | 321745/542020 [12:09<14:08, 259.57it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 60%|█████▉    | 324545/542020 [12:19<13:39, 265.37it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 60%|██████    | 327350/542020 [12:30<13:17, 269.32it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 61%|██████    | 330141/542020 [12:40<13:19, 265.13it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 62%|██████▏   | 333798/542020 [12:50<12:00, 288.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 63%|██████▎   | 341900/542020 [13:00<09:19, 357.85it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▍   | 350725/542020 [13:10<07:19, 435.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 66%|██████▋   | 359549/542020 [13:21<05:58, 509.66it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 68%|██████▊   | 368005/542020 [13:31<05:00, 578.47it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 70%|███████   | 381297/542020 [13:41<03:50, 696.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 73%|███████▎  | 394589/542020 [13:59<03:28, 705.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▍  | 405036/542020 [14:11<03:00, 759.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 76%|███████▋  | 414265/542020 [14:21<02:39, 799.70it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 76%|███████▋  | 414265/542020 [14:31<02:39, 799.70it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 78%|███████▊  | 420888/542020 [14:31<02:44, 737.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 79%|███████▉  | 428771/542020 [14:41<02:30, 751.83it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 81%|████████▏ | 441385/542020 [14:51<01:57, 855.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 81%|████████▏ | 441385/542020 [15:03<01:57, 855.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 83%|████████▎ | 448016/542020 [15:04<02:11, 715.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 84%|████████▍ | 454670/542020 [15:14<02:05, 696.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 85%|████████▌ | 461235/542020 [15:24<01:58, 680.44it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 86%|████████▋ | 468219/542020 [15:34<01:47, 685.31it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 88%|████████▊ | 475189/542020 [15:47<01:45, 635.50it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 88%|████████▊ | 475189/542020 [16:04<01:45, 635.50it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 89%|████████▊ | 480071/542020 [16:04<02:13, 463.25it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 90%|████████▉ | 487521/542020 [16:14<01:44, 522.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 90%|████████▉ | 487521/542020 [16:24<01:44, 522.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 91%|█████████ | 494038/542020 [16:25<01:26, 551.98it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 92%|█████████▏| 500394/542020 [16:35<01:13, 562.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 94%|█████████▎| 506830/542020 [16:46<01:00, 581.01it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 95%|█████████▍| 513119/542020 [16:58<00:52, 551.85it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 96%|█████████▌| 519257/542020 [17:09<00:40, 567.20it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 97%|█████████▋| 525323/542020 [17:20<00:29, 560.71it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 99%|█████████▊| 535026/542020 [17:30<00:10, 641.24it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 542020/542020 [17:44<00:00, 509.12it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 50%|█████     | 4/8 [1:01:51<1:02:04, 931.10s/it]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/148449 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  6%|▌         | 9047/148449 [00:10<02:34, 904.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▍        | 21834/148449 [00:20<02:07, 991.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▍        | 21834/148449 [00:34<02:07, 991.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 21%|██        | 31157/148449 [00:34<02:16, 860.49it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 29%|██▉       | 42881/148449 [00:44<01:52, 935.13it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 54605/148449 [01:03<01:56, 805.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 44%|████▎     | 64899/148449 [01:13<01:37, 860.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 44%|████▎     | 64899/148449 [01:24<01:37, 860.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 48%|████▊     | 70662/148449 [01:24<01:49, 711.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████     | 75726/148449 [01:35<01:55, 628.73it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 56%|█████▌    | 82445/148449 [01:45<01:44, 628.70it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 60%|█████▉    | 88732/148449 [01:55<01:35, 627.86it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▌   | 96953/148449 [02:06<01:16, 670.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 71%|███████   | 104943/148449 [02:17<01:04, 674.23it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 76%|███████▌  | 112139/148449 [02:27<00:52, 687.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 81%|████████  | 120570/148449 [02:37<00:38, 726.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 128938/148449 [02:49<00:26, 727.37it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 92%|█████████▏| 137142/148449 [02:59<00:15, 751.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 99%|█████████▉| 147356/148449 [03:09<00:01, 815.82it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 148449/148449 [03:12<00:00, 772.45it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 62%|██████▎   | 5/8 [1:06:01<36:20, 726.76s/it]  \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/215696 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  6%|▌         | 13381/215696 [00:10<02:31, 1338.04it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 20646/215696 [00:25<03:46, 861.92it/s] \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 20646/215696 [00:25<03:46, 861.92it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 10%|▉         | 20646/215696 [00:35<03:46, 861.92it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 14%|█▍        | 30586/215696 [00:35<03:26, 896.46it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 14%|█▍        | 30586/215696 [00:45<03:26, 896.46it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 17%|█▋        | 36413/215696 [00:45<03:54, 763.60it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|██        | 43652/215696 [00:55<03:49, 751.24it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 25%|██▍       | 52949/215696 [01:05<03:24, 797.13it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 29%|██▉       | 62246/215696 [01:21<03:32, 721.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 33%|███▎      | 70699/215696 [01:31<03:12, 754.96it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 78879/215696 [01:45<03:18, 687.73it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 78879/215696 [01:45<03:18, 687.73it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 39%|███▉      | 83779/215696 [01:55<03:35, 613.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 39%|███▉      | 83779/215696 [01:55<03:35, 613.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 39%|███▉      | 83779/215696 [02:05<03:35, 613.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 41%|████      | 88764/215696 [02:05<03:41, 572.40it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 43%|████▎     | 93720/215696 [02:17<03:52, 525.50it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▌     | 98132/215696 [02:27<04:03, 483.34it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 47%|████▋     | 102204/215696 [02:37<04:08, 457.58it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 49%|████▉     | 106616/215696 [02:48<04:01, 452.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████▏    | 111018/215696 [02:58<03:57, 440.17it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 55%|█████▍    | 118466/215696 [03:08<03:13, 501.36it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 59%|█████▊    | 126451/215696 [03:18<02:38, 563.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 59%|█████▊    | 126451/215696 [03:36<02:38, 563.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 62%|██████▏   | 133731/215696 [03:36<02:40, 510.17it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▍   | 139279/215696 [03:46<02:26, 522.78it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 68%|██████▊   | 147299/215696 [03:56<01:57, 583.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 68%|██████▊   | 147299/215696 [04:06<01:57, 583.72it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 72%|███████▏  | 155227/215696 [04:07<01:37, 619.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 162441/215696 [04:19<01:28, 601.02it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 78%|███████▊  | 168859/215696 [04:30<01:17, 605.53it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 81%|████████  | 175023/215696 [04:40<01:06, 607.46it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 90%|████████▉ | 193265/215696 [04:50<00:29, 759.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 98%|█████████▊| 211507/215696 [05:02<00:04, 894.44it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 215696/215696 [05:05<00:00, 706.71it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 75%|███████▌  | 6/8 [1:12:28<20:49, 624.87s/it]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/426134 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  1%|          | 4430/426134 [00:10<15:57, 440.35it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  2%|▏         | 10340/426134 [00:20<14:35, 474.83it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  4%|▍         | 16744/426134 [00:30<13:20, 511.39it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  5%|▌         | 22978/426134 [00:42<13:00, 516.46it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  7%|▋         | 31024/426134 [00:52<11:24, 577.51it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  9%|▉         | 40020/426134 [01:02<10:01, 642.25it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 11%|█▏        | 48718/426134 [01:20<10:36, 592.55it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 13%|█▎        | 56416/426134 [01:30<09:41, 635.34it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 15%|█▌        | 64313/426134 [01:40<08:56, 674.91it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 17%|█▋        | 72210/426134 [01:50<08:30, 693.01it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 19%|█▊        | 79756/426134 [02:00<08:07, 710.41it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 19%|█▊        | 79756/426134 [02:11<08:07, 710.41it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 20%|██        | 86773/426134 [02:11<08:07, 695.86it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 22%|██▏       | 93415/426134 [02:28<09:57, 557.18it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 23%|██▎       | 98772/426134 [02:38<09:54, 550.37it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 24%|██▍       | 104123/426134 [02:49<09:55, 540.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 26%|██▌       | 110731/426134 [02:59<09:11, 571.65it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 28%|██▊       | 117203/426134 [03:11<09:15, 556.28it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 28%|██▊       | 117203/426134 [03:11<09:15, 556.28it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 28%|██▊       | 117203/426134 [03:21<09:15, 556.28it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 28%|██▊       | 120285/426134 [03:21<11:24, 446.71it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 30%|██▉       | 127436/426134 [03:31<09:53, 503.39it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 137891/426134 [03:42<08:10, 588.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 32%|███▏      | 137891/426134 [03:53<08:10, 588.22it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 34%|███▍      | 147007/426134 [03:53<07:18, 636.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 37%|███▋      | 158906/426134 [04:03<06:01, 739.50it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|████      | 171948/426134 [04:13<04:59, 849.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 43%|████▎     | 184990/426134 [04:26<04:28, 899.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 43%|████▎     | 184990/426134 [04:38<04:28, 899.42it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▌     | 192636/426134 [04:38<04:55, 790.34it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 47%|████▋     | 201007/426134 [04:48<04:40, 803.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 47%|████▋     | 201007/426134 [04:59<04:40, 803.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 49%|████▊     | 207026/426134 [04:59<05:09, 708.20it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████     | 215605/426134 [05:10<04:45, 736.75it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 51%|█████     | 215605/426134 [05:29<04:45, 736.75it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 52%|█████▏    | 223379/426134 [05:29<05:42, 592.24it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 227694/426134 [05:39<06:12, 532.71it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 54%|█████▍    | 232009/426134 [05:49<06:31, 495.75it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 56%|█████▌    | 237492/426134 [05:59<06:12, 506.32it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 58%|█████▊    | 245072/426134 [06:09<05:22, 561.98it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 60%|█████▉    | 254247/426134 [06:20<04:31, 632.62it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 62%|██████▏   | 263199/426134 [06:34<04:18, 630.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 63%|██████▎   | 269528/426134 [06:44<04:08, 629.90it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▌   | 277734/426134 [06:54<03:40, 674.24it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 68%|██████▊   | 288827/426134 [07:04<03:00, 761.48it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 70%|███████   | 299735/426134 [07:18<02:44, 766.84it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 73%|███████▎  | 310644/426134 [07:28<02:17, 841.86it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 73%|███████▎  | 310644/426134 [07:40<02:17, 841.86it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 321398/426134 [07:40<02:00, 869.28it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 78%|███████▊  | 330806/426134 [07:52<01:53, 840.87it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 80%|███████▉  | 340792/426134 [08:02<01:36, 880.70it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 82%|████████▏ | 350694/426134 [08:12<01:23, 901.58it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 85%|████████▍ | 360239/426134 [08:26<01:19, 826.78it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 86%|████████▌ | 367166/426134 [08:37<01:17, 758.64it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 88%|████████▊ | 374890/426134 [08:47<01:07, 762.71it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 90%|█████████ | 384873/426134 [08:57<00:50, 820.82it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 94%|█████████▍| 402672/426134 [09:07<00:23, 979.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 94%|█████████▍| 402672/426134 [09:22<00:23, 979.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 99%|█████████▊| 420152/426134 [09:22<00:05, 1021.69it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 426134/426134 [09:30<00:00, 747.07it/s] \u001b[A\u001b[A\u001b[A\n",
      "\n",
      " 88%|████████▊ | 7/8 [1:24:39<10:56, 656.57s/it]\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  0%|          | 0/154980 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "  5%|▌         | 8355/154980 [00:10<03:00, 812.52it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 13%|█▎        | 20488/154980 [00:20<02:29, 901.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 13%|█▎        | 20488/154980 [00:39<02:29, 901.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 16%|█▌        | 25008/154980 [00:39<04:29, 481.87it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 21%|██        | 32869/154980 [00:49<03:43, 545.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 21%|██        | 32869/154980 [00:59<03:43, 545.16it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 23%|██▎       | 36144/154980 [01:00<04:25, 446.77it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 26%|██▌       | 39721/154980 [01:10<04:37, 415.68it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 29%|██▉       | 44596/154980 [01:20<04:14, 433.99it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 35%|███▍      | 53692/154980 [01:31<03:19, 508.69it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 40%|████      | 62194/154980 [01:41<02:43, 568.29it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 45%|████▌     | 70016/154980 [01:52<02:19, 610.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 50%|████▉     | 77416/154980 [02:06<02:13, 581.38it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 53%|█████▎    | 82643/154980 [02:17<02:10, 556.07it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 59%|█████▉    | 91570/154980 [02:27<01:41, 626.15it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 65%|██████▍   | 100440/154980 [02:39<01:24, 645.90it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 70%|██████▉   | 108471/154980 [02:49<01:07, 685.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 75%|███████▌  | 116550/154980 [02:59<00:53, 718.37it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 81%|████████  | 125902/154980 [03:10<00:37, 765.80it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 134566/154980 [03:21<00:26, 774.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 134566/154980 [03:21<00:26, 774.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 87%|████████▋ | 134566/154980 [03:31<00:26, 774.89it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 89%|████████▉ | 138475/154980 [03:31<00:28, 581.67it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 92%|█████████▏| 142153/154980 [03:47<00:32, 397.32it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      " 96%|█████████▌| 148111/154980 [03:57<00:15, 441.41it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "100%|██████████| 154980/154980 [04:05<00:00, 632.08it/s]\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "100%|██████████| 8/8 [1:29:47<00:00, 551.90s/it]\u001b[A\u001b[A"
     ]
    }
   ],
   "source": [
    "stream = full_text_stream(filtered_file_meta_stream(dfe, file_meta_stream(target_archives)))\n",
    "extract_corpus(stream, output_file)     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "ArrowIOError",
     "evalue": "Invalid parquet file. Corrupt footer.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mArrowIOError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-9c09ff948ae4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mParquetFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;31m#pf = pq.read_pandas(output_file)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.6/site-packages/pyarrow/parquet.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, metadata, common_metadata, memory_map)\u001b[0m\n\u001b[1;32m    128\u001b[0m                  memory_map=True):\n\u001b[1;32m    129\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mParquetReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 130\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muse_memory_map\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmemory_map\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmetadata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    131\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommon_metadata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcommon_metadata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_nested_paths_by_prefix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_build_nested_paths\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.6/site-packages/pyarrow/_parquet.pyx\u001b[0m in \u001b[0;36mpyarrow._parquet.ParquetReader.open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.6/site-packages/pyarrow/error.pxi\u001b[0m in \u001b[0;36mpyarrow.lib.check_status\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mArrowIOError\u001b[0m: Invalid parquet file. Corrupt footer."
     ]
    }
   ],
   "source": [
    "pf = pq.ParquetFile(output_file)\n",
    "#pf = pq.read_pandas(output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 899 entries, 0 to 898\n",
      "Data columns (total 16 columns):\n",
      "abstract          856 non-null object\n",
      "body              899 non-null object\n",
      "date_accepted     804 non-null datetime64[ns]\n",
      "date_pub          899 non-null datetime64[ns]\n",
      "date_received     794 non-null datetime64[ns]\n",
      "id_doi            862 non-null object\n",
      "id_pmc            899 non-null object\n",
      "id_pmid           881 non-null object\n",
      "journal_ids       899 non-null object\n",
      "journal_titles    899 non-null object\n",
      "title             899 non-null object\n",
      "arch_id           899 non-null object\n",
      "arch_path         899 non-null object\n",
      "arch_venue        899 non-null object\n",
      "arch_name         899 non-null object\n",
      "arch_archive      899 non-null object\n",
      "dtypes: datetime64[ns](3), object(13)\n",
      "memory usage: 112.5+ KB\n"
     ]
    }
   ],
   "source": [
    "dftmp = pf.read_row_group(1).to_pandas()\n",
    "dftmp.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"BackgroundGlobal gene expression has been used successfully to elicit biological behavior in different soft tissue tumors. Pterygium as a human disease, noted to be more prevalent than 20% of some populations, is of immense biological interest for a few reasons.First, the pathogenesis of this condition is hotly debated. Hypothesis driven approaches have not resolved the relative importance of competing mechanisms for this disease. Theories that have been proposed include inflammatory influence, degeneration of connective tissue, genetic instability, angiogenesis, redox-related toxicity, cellular proliferation, aberration of apoptosis, exuberant wound healing, altered lipid metabolism, mast cell infiltration. and stem cell dysfunction. Conventional approaches to disease mechanism, by virtue of their narrow focus, were not helpful to assess relative contribution of widely heterogenous processes. Furthermore, a fundamental issue about the diseased tissue remains un-resolved in this context: the origin of the epithelium overlying pterygial lesions, though suspected to be conjunctival in origin, is not entirely certain.Second, the spectrum of tumor size and behavior is tremendous, ranging from the inconspicuous lesion barely encroaching on the peripheral cornea, to the rapidly growing, menacing tumors that obscure the visual axes and threaten vision. In some contexts, additional molecular events. may drive the original tumor to behave differently from the original lesion. It is also intriguing that unlike frank malignancies, these lesions, however fast growing, do not erode through the full thickness cornea, highlighting the presence of distinct processes from those manifested in malignant tumors.Third, pterygium is a mixed soft tissue tumor that is strongly associated with non-ionising ultraviolet radiation. Unlike the case of cutaneous melanomas, the cell type that responded to the environmental trigger may not be epithelial but rather, the fibro-vascular component – a truly unusual phenomenon in human tumor biology because ultraviolet radiation-induced tumors encountered in humans are epithelial in origin and ultraviolet radiation effects on keloids, if any, are generally inhibitory.Lastly, the inability of researchers to discover an analogous tumor in animals or reconstitute the disease in organ cultures imply that pterygium involves specie-specific mechanisms mediated by in-vivo cell-cell or cell-matrix interactions. The molecules involved in this disease may have therefore arisen from divergent evolution in the human ocular surface.In view of the controversies on the multiple mechanisms of pterygia formation, we advocate an unbiased, global gene expression approach to decipher the dominant gene expression patterns that may underlie specific molecular events of interest. Recently, the genome-wide microarray data in whole tissue pterygium as well as microarray data limited to cultured pterygial fibroblasts, have been published. We have performed a study using gene microarray on recurrent pterygium, primary pterygium and un-involved conjunctiva, examined the gene subsets which were differentially regulated, compared our findings with these previous microarray studies, and discussed some of the findings in the light of a complementary proteomics approach.In clinical practice, the treatment of this condition is surgical excision. However, some cases aggressively recur after surgery. Conjunctival auto-grafting as an adjunctive procedure may reduce recurrence, though the explanation for this is not entirely clear. For these reasons, we speculate that a non-hypothesis driven approach may also be useful to discover gene expression signatures that predispose lesions to a more aggressive phenotype. Such information will benefit clinicians, who can appropriately anticipate otherwise 'un-expected' biological behavior in their treatment of this disorder.MethodsMaterialsAntibodies against MUC5AC, keratin 6 and CD24 were from Labvision, Neomarkers, Fremont, CA, USA, anti-MSMB antibodies were from US Biologicals, Massachusetts, MA, USA, anti-CEACAM5 and anti-NR4A2 antibodies were from Abcam, Cambridge, UK.Samples used for the StudyThe procurement and use of both human tissues in this study was in compliance with the tenets of the Declaration of Helsinki. The study was approved by the Institutional Review Board of Singapore Eye Research Institute. Written informed consent was obtained from donors after explanation of the nature and possible consequences of the study. Human tissues samples were obtained from patients diagnosed with primary pterygium and came from different races, Chinese, Malay and Indian. All patients underwent pterygium excision in conjunction with the use of an upper bulbar conjunctival free autograft placed over the site of the original lesion. All pterygia specimens used were nasal pterygia. The whole pterygium tissue and a small portion of the conjunctival patch (approx. 1 × 3 mm) from the superotemporal conjunctiva were collected and were rapidly frozen in liquid nitrogen after removal and stored at -150°C. For immunohistology purpose, the samples were collected on ice and embedded in Optimal Cutting Temperature (OCT, Sakura, USA) in the laboratory.Microarray ExperimentThe procedure for the microarray experiment has previously been published. All microarray chips and related protocols and equipment for the processing of these chips were from Affymetrix Inc., Santa Clara, CA. The human genome GeneChip U133A consisting of more than 22000 probe sets was used for this study.A group of 8 primary pterygium samples, harvested from 4 males (aged 40 to 50 years old) and 4 females (aged 50 to 60 years old) was used in this experiment. Another group of 4 control conjunctiva tissues, each pooled from 4 individual samples, was used as controls to the diseased tissues. Considering that only a tiny piece of the conjunctiva tissue could be obtained from each patient during the surgery, pooling of 4 conjunctiva tissues was necessary to obtain enough starting material.Total RNA was extracted using TRIzol Reagent (Invitrogen, CA) and purified with RNeasy Mini Kit (Qiagen, Valencia, CA) according to the manufacturer's instructions. Five micrograms of each purified RNA sample were prepared according to the Affymetrix standard protocol. Fifteen micrograms of biotin-labelled cRNA using BioArray RNA Transcript Labelling Kit (ENZO Life Sciences, NY) were fragmented and the appropriate volume injected separately into the probe array chips.The transcripts were hybridized onto the immobilized oligonucleotide sequence on array for 16 hours at 45°C under 60 rpm rotation using GeneChip Hybridization Oven 640. Washing and Strepavidin-staining steps were performed using Affymetrix Fluidics Station 450. The chips were scanned using GeneChip Scanner3000 and the image data were further analyzed using Microarray Suite v.5.0. Data analysis included pre-processing and normalization before identification of differentially expressed genes. Pre-processing adjusted for non-specific binding and background noise, whereas normalization removed systematic variation in the data due to effects other than biological differences. We used the Robust Multi-array Average (RMA) model  to extract the gene expression signals from probe intensities without taking the mismatch probe signals into consideration. Cross-array normalization was performed using the intensity-based log ratio median method  with the first array as the reference. Gene-level normalization was performed by normalizing all samples to the median of the expression level of the control (un-involved conjunctiva) samples. The data were annotated using gene annotation headings, bioprocesses and molecular functions from the NetsAffx database.Analysis of Microarray DataThe data reported in this study have been deposited in NCBI's Gene Expression Omnibus (GEO,  with GEO series accession number GSE2513). Data were visualised and explored using the GeneSpring GX 7.3 platform (Agilent Technology, Redwood City, CA).For selection of differentially expressed genes, the modified t-statistic (SAM)  with 100% of the standard deviation percentile as the fudge constant was used. The threshold for significantly changed genes was set at a false discovery rate (FDR) of 5%. After delineating a list of differentially expressed genes, we performed various types of analysis to search for a pattern of global gene expression. The methods ranged from K-means clustering, to the construction of hierarchical dendrograms on a subset of genes, as well as the visualisation of gene expression pattern summarised by principal component analysis. In additional to analysis of differentially expressed genes, we employed the Gene Set Enrichment Analysis (GSEA) method for identification of those pathways that were more affected in pterygium tissues. The goal of GSEA is to determine whether members of a gene set tend to occur toward the top or bottom of the list, in which case the gene set is correlated with the phenotypic class distinction. In contrast to the methods for extraction of differentially expressed genes, GSEA considers the collective up- or down-regulation of a gene set rather than individual genes. The GSEA software package downloaded from the Board Institute's website was used for the identification of activated or deactivated pathways in primary pterygial and conjunctival tissues. Microarray data for 8 primary pterygial samples and 4 samples of conjunctival tissue were used. Various datasets from publicly accessible databases (Biocarta, STKE, PubMed and KEGG) were dissected into over 300 gene sets. The two phenotypic classes used were pterygium and uninvolved conjunctiva tissues.As an independent method of analyzing signaling pathways in pterygium, we exported the list of genes significantly up-regulated or down-regulated by 2 fold into the Pathway Studio 5.0 software (Ariadne Genomics Inc, Rockville, MD). The option 'Find all shortest paths between selected entities' was used on the up-regulated and down-regulated genes sequentially. The number of connectivities was limited to 2 and the genes mapped into known pathways displayed on a chart, and the relevant relationships listed in a table. The purpose of performing this analysis was to identify important upstream regulators of the differentially expressed genes as well as downstream effectors, in an unbiased fashion, based on known biological knowledge.Methods used to extract and analyse tear proteins from patients who had undergone surgery for pterygium have already been described. Wherever possible, the fold change data for genes corresponding to detected tear proteins were tabulated.Real Time Reverse Transcription Polymerase Chain ReactionSix pairs of independent pterygium and uninvolved conjunctiva tissues were used in the real time reverse transcription polymerase chain reaction (qPCR) experiment. One microgram of the each total RNA preparation was reverse-transcribed to single stranded cDNAs using an oligo-dT primer with Superscript II Rnase H Reverse Transcriptase (Invitrogen, USA). Primer pairs specific to each gene used in the qPCR are listed in Additional file. qPCR was performed using SYBR Green PCR Master Mix (Applied Biosystems) in an ABI Prism 7700 Sequence Detection System (Applied Biosystems). The thermal cycling conditions were as follows: 95°C for 10 min, 45 cycles at 95°C for 30 s and 60°C for 1 min. Data obtained from qPCR were analyzed using the comparative CT method as previously described by Livak. Paired t-statistics with p-value < 0.05 was used to determine whether the qPCR results in pterygial tissue were significantly different from those in un-involved conjunctiva.ImmunohistochemistryA separate group of paired pterygial and conjunctival tissue samples from three patients were used for immunohistochemistry. Sections were cut at 5 μm thicknesses from blocks of freshly frozen pterygium and matched un-involved conjunctiva embedded in the OCT. The post-fixed sections were incubated with the primary antibody at 4°C overnight. Specific antibodies against the following were used: MUC5AC, MSMB, CD24, CEACAM5, keratin 6 and NR4A2. After 3 washes with PBS, secondary antibody used was either fluorescein isothiocyanate (FITC)- or Rhodamine- conjugated anti IgG (Santa Cruz, USA), incubated with the sections for 40 minutes at room temperature. After 3 final washes each section was mounted with in a fluorescence mounting medium (DAKO, Denmark). Images were captured using a 40× Achrostigmat lens on an Axioplan2 microscope equipped with an AxioCam MR camera (Carl Zeiss, Germany).ResultsGlobal Gene Expression Profiling and analysis of pterygium versus conjunctivaGene expression analysisWith gene level normalization using un-involved conjunctival samples as reference, gene expression analysis showed that a total of 105 probe sets were significantly changed (p < 0.05) by at least 2 fold in primary pterygial tissue from uninvolved conjunctiva. Among these, 60 probe sets were up-regulated and 45 probe sets were down-regulated (Table ). Without performing gene level normalization, 114 unique genes out of 156 probe sets were significantly changed in pterygium. (Figure )Table 1Genes significantly changed by at least 2-fold difference in pterygium (p < 0.05)Up-regulated in pterygiumProbe IDGenesymbolGenbank accessionFold change209125_atKRT6AJ002696.399205009_atTFF1NM_0032255.671210297_s_atMSMBU221784.175211719_x_atFN1BC0058584.112212464_s_atFN1X027613.939216442_x_atFN1AK0267373.827207811_atKRT12NM_0002233.781210495_x_atFN1AF1300953.738207430_s_atMSMBNM_0024433.732202310_s_atCOL1A1K012283.635216379_x_atAK0001683.475209771_x_atCD24AA7611813.42218990_s_atSPRR3NM_0054163.401201884_atCEACAM5NM_0043633.287214385_s_atMUC5BAI5216463.244202404_s_atCOL1A2NM_0000893.149205064_atSPRR1BNM_0031253.053204777_s_atMALNM_0023713.043211161_s_atAF1300822.937219087_atASPNNM_0176802.903214303_x_atMUC5ACAW1927952.897209047_atAQP1AL5183912.773213796_atSPRR1AAI9239842.772205694_atTYRP1NM_0005502.736214580_x_atKRT6AAL5695112.645217272_s_atSERPINB13AJ0016982.545218963_s_atKRT23NM_0155152.506218353_atRGS5NM_0252262.476205547_s_atTAGLNNM_0031862.467217388_s_atKYNUD556392.441201438_atCOL6A3NM_0043692.423203477_atCOL15A1NM_0018552.374203980_atFABP4NM_0014422.364202311_s_atCOL1A1AI7436212.334215076_s_atCOL3A1AU1441672.331213975_s_atLYZAV7119042.322209071_s_atRGS5AF1595702.275210809_s_atPOSTND136652.261202878_s_atC1QR1NM_0120722.242217528_atCLCA2BF0031342.226204682_atLTBP2NM_0004282.21212190_atSERPINE2AL5413022.201219410_atFLJ10134NM_0180042.171209848_s_atSILVU018742.15204351_atS100PNM_0059802.114204971_atCSTANM_0052132.092206427_s_atMLANAU066542.088202917_s_atS100A8NM_0029642.084203382_s_atAPOENM_0000412.081211980_atCOL4A1AI9226052.077200665_s_atSPARCNM_0031182.071201667_atGJA1NM_0001652.06218723_s_atRGC32NM_0140592.056203691_atPI3NM_0026382.05206276_atLY6DNM_0036952.04219412_atRAB38NM_0223372.04208982_atPECAM1AW5745042.029203504_s_atABCA1NM_0055022.026201852_x_atCOL3A1AI8137582.01203570_atLOXL1NM_0055762.01Down-regulated in pterygiumProbe IDGeneGenbank accessionFold change217232_x_atHBBAF0591800.5208078_s_atTCF8NM_0307510.499201842_s_atEFEMP1AI8267990.496201008_s_atTXNIPAA8122320.495215176_x_atAW4048940.494206391_atRARRES1NM_0028880.494212225_atSUI1AL5168540.492201505_atLAMB1NM_0022910.488205979_atSCGB2A1NM_0024070.487209278_s_atTFPI2L276240.482211003_x_atTGM2BC0035510.479211573_x_atTGM2M984780.479202431_s_atMYCNM_0024670.475213831_atHLA-DQA1X004520.47210095_s_atIGFBP3M311590.465206392_s_atRARRES1NM_0028880.46211651_s_atLAMB1M202060.455201843_s_atEFEMP1NM_0041050.443202081_atIER2NM_0049070.439202340_x_atNR4A1NM_0021350.435200664_s_atDNAJB1BG5372550.435209116_x_atHBBM250790.431215723_s_atPLD1AJ2762300.43205910_s_atCELNM_0018070.417219759_atLRAPNM_0223500.416204286_s_atPMAIP1NM_0211270.406206424_atCYP26A1NM_0007830.403206393_atTNNI2NM_0032820.396201466_s_atJUNNM_0022280.388201464_x_atJUNBG4918440.368212143_s_atIGFBP3BF3402280.359201473_atJUNBNM_0022290.356217022_s_atMGC27165S557350.355201236_s_atBTG2NM_0067630.342204285_s_atPMAIP1AI8576390.301201041_s_atDUSP1NM_0044170.296201044_x_atDUSP1AA5308920.265204622_x_atNR4A2NM_0061860.22201694_s_atEGR1NM_0019640.215204621_s_atNR4A2AI9350960.192213674_x_atIGHMAI8580040.188216248_s_atNR4A2S771540.17202768_atFOSBNM_0067320.167202672_s_atATF3NM_0016740.159209189_atFOSBC0044900.072Figure 1Differentially expressed transcripts. Heat map showing expression of 156 probe sets in conjunctiva and pterygium. Red represents high level of expression, green represents lower level expression and black represents no change.Gene Ontology AnalysisAll significantly changed genes were categorized based on biological functions using categories from the GeneOntology database (Figure ). Among the up-regulated genes, genes coding for cell adhesion (8), extracellular matrix (ECM) (7) and structural proteins (8) accounted for 38% of up-regulated genes. Examples of up-regulated cell adhesion molecules include CADH11, MCAM, FN1, CEACAM5, ECM component include COL1A2, COL1A1, COL4A2, COL6A3 and structural proteins, ACTA2, KRT12, KRT6, KRT23, SPRR1B and SPRR3. Several of these genes also encode for proteins involved in wound healing, including COL6A3, CD24, FN1, KRT6a, TFF1, SPRR1b, and MUC5AC. Another group of up-regulated genes is the 'mitogenic proteins' group exemplified by MSMB, CEACAM5 and CD24.Figure 2Biological function of dysfunctional genes. Pie Charts showing the percentage of genes with different biological functions that were dysregulated in pterygium.Some immune response and transcription factors were down-regulated in primary pterygium. Examples of the former include immunoglobulin heavy chain genes, whereas examples of the latter include ATF3, BTG2, ERG1, FOS, FOSB, JUN, NR4A1 and NR4A2. Other down-regulated genes include those encoding for transport proteins, ie., HBB and HBD, and those involved in stress-response, including DUSP1 and GADD45B.When the composition of the list of significantly changed genes was studied in terms of certain gene ontology categories, we observed that there is an over-representation of such genes compared to their expected frequency, considering the proportion of the probes representing these genes present on the chip (Table ). This suggests that structural molecules and molecules with transporter activity play a role in pterygium.Table 2Proportion of significantly changed genes by functional category.Gene Ontology (GO)Observed %Expected %p valueExtracellular region (5576)16.35.30.00698Extracellular matrix (31012)13.952.40.000503Intermediate filament cytoskeleton (45111)6.9770.530.00197Intermediate filament (5882)6.9770.580.00197Muscle development (7517)6.9771.160.0135Transporter activity (5215)23.2612.060.0291Structural molecule activity (5198)20.935.4240.000414Extracellular matrix structural constituents (5210)9.3020.6750.000198Oxidoreductase activity acting on paired donors (16705)6.9770.6310.0025Note: This refers to primary pterygium versus control. Only functional categories which represent at least 5% of the significantly changed genes are shown. The numbers in the brackets represent the GO category ID.Analysis of enriched pathwaysGene set enrichment analysis (GSEA) was performed using gene expression data from the conjunctival and primary pterygium groups (Figure ). Significantly enriched pathways in pterygium were related to ECM formation, tissue invasion, angiogenesis and lipid metabolism. Candidate genes found to be involved in the ECM pathway were cadherins (CDH5 and CDH3), integrins (eg., ITGb1), cystatin C (CST3) and intracellular adhesion molecule 1(ICAM1). Angiogenesis related genes include akt1, akt2, tyrosine kinase 2 (Tyk2), phosphinositide-3-kinases (PIK3C2A, PIK3CD, PIK3R1, PIK3R2, PIK3R4) and janus kinases (JAK1, JAK2). The genes involved in 'tissue invasion' pathways include plasminogen activator and receptor (PLAU, PLAUR), matrix metalloproteinase (MMP1,2,9), tissue inhibitor of metalloproteinase 1 (TIMP1, TIMP3), serine (or cysteine) proteinase inhibitor clade B member 5 (SERPINB5), s100 calcium binding protein (S100A4), secreted phosphoprotein1 (spp1) and mucin (MUC1). The lipid formation pathway was also significantly enriched in pterygium tissues, for example, the sterol regulatory element binding transcription factor1 (SREBF1), dual specific phosphatase 14 (DUSP14), hexokinase2 (HK2) and the solute carrier family 2 member 4 (SLC2A4).Figure 3Gene Set Expression analysis. Diagram showing the affected pathways involved in pathogenesis of pterygium by Gene Set Expression analysis.The only significantly enriched pathway in un-involved conjunctiva was related to apoptosis. The genes in this pathway include the calpains (CAPN1, CAPN2, CAPN3, CAPNS1) and bcl-antagonist of cell death (BAD2).Pathway analysis to identify upstream and downstream signalsUp-regulated genes and down-regulated genes were analysed in pathway studio. Figure  summarised signaling network that resulted from this analysis. Primary pterygium may be characterized by stress-induced down-regulation of transcription factors (Egr1, Jun and Fos), with defective wound healing as the major process responsible for the disease phenotype. Scrutiny of the networks in Figure  illustrates a few other relevant processes: up-regulation of PECAM1 and down-regulation of TFPI-2 may contribute to aberrant vascularisation, whereas down-regulation of DUSP1 and up-regulation of GP75 may contribute to an abnormal response to oxidative stress. The identities and biological/molecular functions of these relationships are shown in Additional files  and.Figure 4Signaling in pterygium. Schematic representation of potential signaling pathways involved in primary pterygium. These pathways could be affected in conjunctival epithelial cells, fibroblasts or vascular endothelial cells. Pathways were identified by incorporating the microarray results (genes which were differentially expressed between normal conjunctival and pterygium tissue) into Pathway Studio. Green symbols represent down-regulation, whereas red symbols represent up-regulation of genes. Solid lines represent positive regulation, thin arrow with a cross line represent inhibition. Purple symbols represent intermediate molecules in the potential pathways that may have altered function, for example, CDKN1A or p21 (Cip) may have reduced function due to down-regulated egr1 transcription factor. GP75: Tyrosinase-related protein 1, CSTA: cystatin A, TAGLN: transgelin, GJA1: Gap junction protein α-1 or Connexin-43, ABCA1: ATP binding cassette subfamily A1 protein, CEL: carboxyl ester lipase, DUSP1: dual specificity phosphatase I, PMAIP1: phorbol-12-myristate-13-acetate-induced protein 1, BTG2: B cell translocation gene 2, CYP26A1: cytochrome P450 family 26 A1 isoform, NR4A1 and 4A2: nuclear receptors 4A1 and 2, SPARC: secreted protein acidic rich in cysteine, TGM-2: transglutaminase 2, TFPI-2: Tissue factor pathway inhibitor 2 and IGFBP3: insulin-like growth factor binding protein 3. Green symbols within the nucleus represent transcription factors genes that were depressed, probably due to the upstream stress signaling such as those related to ultraviolet light (not shown). Due to quenching effect, there may be an increased in the transcriptional promoter activity of other transcription factors such as SP1, CEBP and SP3 (open symbols).Clustering analysisTo extract some useful knowledge or patterns from global gene expression data, we performed a few cluster analyses. The K-means cluster method was computed using 5 clusters on the 156 gene probes differentially expressed in pterygium compared to controls, over 1000 iterations. Table  shows the final characteristics of the clusters. Two clusters contained genes that were down-regulated in pterygium, whereas three other clusters showed genes that were up-regulated. Visual examination of the role of these genes revealed biological functions that are consistent with the above-mentioned pathway and gene ontology studies.Table 3K-means cluster analysis supplemented with GO browser results.ClusterNo of GenesAverage RadiusStatus in pterygiumMajor function by GO1582.463Down-regulatedResponse to stimulus2131.803Up-regulatedMetabolism3371.507Up-regulatedCell adhesion, response to wounding4222.357Up-regulatedLipid related5261.827Down-regulatedCell proliferationTo identify co-expressed gene sets or the similarity of samples, we further performed hierarchical clustering analysis on the data for the 156 gene probes. Two closely related clusters of genes up-regulated in conjunctiva was detected (Figure ), and similarly, 2 clusters of up-regulated genes in pterygium were detected (Figure ). Scrutiny of the composition of these clusters yields a number of cytoskeletal proteins, immunoglobulins, cancer markers and transcription factors.Figure 5Hierarchical clustering analysis. Dendrograms were constructed using hierarchical clustering algorithms, performed on the subset of significantly changed genes. The shorter the length of the branches, the more co-expressed the members of the genes are. Magnified portions of the dendrograms showing the clusters of co-expression for genes down-regulated (A) and up-regulated (B) in primary pterygium relative to control. Note that the 2 clusters in A involved transcription factors and immunoglobulins, whereas 2 clusters in B were involved in structural proteins and extracellular matrix.We further performed principal component analysis on the entire set of genes in the chip, as well as several selected categories of Gene Ontology previously implicated in the pathogenesis of pterygium (Figure ) for the following two purposes: 1. To show the compactness of the samples (represented by the distance between circles) in each condition and 2. To reveal the possible existence of a plane which can separate the 2 conditions. The 3D scatter plots of the first three prinicipal components show that the expression of genes coding for ECM component (Figure ) was able to differentiate the conjunctiva and pterygium samples better than those in the antioxidant (Figure ) and apoptosis (Figure ) categories. When all the genes were included in the analysis (Figure ), a clear plane between the pterygial and conjunctival samples was not so evident. This may be due to the inclusion of many genes that were not significantly regulated but whose expression data contained extensive noise. For this analysis, we deliberately included processes like 'reproduction' (Figure ) being unlikely to play a significant role in pterygium pathogenesis, acted as a negative control.Figure 6Principal component analysis. Three-D scatter plots constructed from principal component analyses of gene expression data. In the scatter plots the first, second and third principal components were plotted on the x, y and z axis respectively. Red circles represent conjunctival samples (control) whereas yellow circles denote primary pterygial samples. A. Performed on all genes, B-L. Using subsets of Gene Ontology genes corresponding to a biological process or molecular function and J-L. Using subsets of Gene Ontology genes corresponding to specific cellular components. Brackets represent the number of genes in the overall analysis (A) or in subsets (B-L)Differences between primary and recurrent pterygiumRecurrent pterygium [see Additional file ] differs from primary cases as it has a different morphology and often worse prognosis. Table  shows the genes that were differentially regulated in recurrent compared to primary pterygium. Additional file  shows that a large number of genes in recurrent pterygium were down-regulated or up-regulated relative to primary pterygium and conjunctiva tissues. Since there were a rather large number of differentially regulated genes, only those with at least 2 fold change between primary and recurrent pterygia are shown in table. It is interesting that the up-regulated genes are not the same as those up-regulated in primary pterygium compared to conjunctiva. Examples include stearoyl-CoA desaturase 5 that converts fatty acids into monounsaturated forms in the endoplasmic reticulum and involved in dyslipidemia, the ubiquinol-cytochrome c reductase involved in the mitochondrial respiratory chain, the DNA recombination repair protein RAD51, the neuronal thread protein (AD7C-NTP), which is a extracellular protein involved in apoptosis, and the gene for Mediterranean fever (MEFV) which is involved in regulation of transcription and inflammatory response. Involvement of the gene for sialophorin (SPN or CD43) and NEK5 (Never in mitosis gene a) are consistent with increased cell migration because SPN is a negative regulator of cell adhesion and NEK5 is involved in microtubule function and motility. Additional file  shows possible mechanisms of pterygium recurrence involving these mediators.Table 4Distinguishing recurrent from primary pterygia (p < 0.05), list of genes down-regulated or up-regulated by at least 2-fold.A. Name of down-regulated genesNo of folds down-regulatedB. Name of up-regulated genesNo of folds up-regulatedPSCA-4.34SCD54.05COL1A1-4.21C12orf383.12COL6A2-3.24MEFV3.10COL3A1-3.09UQCRQ2.68HDDC2-2.84RAD51L32.43ASPN-2.77SPN2.36YWHAE-2.66AD7C-NTP2.19KRT7-2.65NEK52.18THRAP5-2.62LOC3892862.00HNRPH2-2.60UBQLN2-2.43FLJ20273-2.38MUC7-2.38UNC84A-2.35MSMB-2.34LTBP3-2.33NDFIP1-2.31BAT1-2.30AKAP9-2.28SMC1A-2.24BNIP3L-2.23COPE-2.22CAPZA2-2.20RBM3-2.18PKM2-2.18SLC34A2-2.15C9orf95-2.15ZDHHC11-2.14SPARC-2.12DICER1-2.07LOXL1-2.07SPINT1-2.06LSM14A-2.06PSMC6-2.06UBE1-2.06MARCKS-2.05PFN1-2.04H1FX-2.03DDX41-2.03MGP-2.02NDUFS8-2.01PSCA-4.34Validation of gene microarray dataRelative quantitative real time polymerase chain reaction (qPCR) was performed as an independent laboratory approach to validate the transcriptional level changes in the microarray experiment. These results (Figure ) show that for all genes selected for validation with qPCR, the direction and magnitude of changes were consistent with the results obtained from the microarray analysis.Figure 7Validation of microarray data. Bar graph showing the correlation of microarray data with real time PCR transcript levels. Black bars represent number of folds of change by GeneChip experiment, gray bars represent the number of folds of change using real time PCR. A normalised ratio (Y-axis) of more than 1 indicates up-regulation in pterygium, whereas a ratio of less than 1 indicates down-regulation in pterygium. The X axis shows an arbitrarily selected panel of genes.We then addressed the question whether protein levels were also affected in pterygium. Several proteins were chosen arbitrarily to address the protein expression levels as well as the localization in pterygium tissues. Our results (Figure ) show that protein expression levels for keratin 6, CEACAM5, CD24, MSMB and MUC5AC were increased, whereas NR4A2 and IGFBP3 were reduced in pterygium, consistent with the transcript changes detected by the microarray analysis. Furthermore, the immunofluorescent staining yielded interesting information about the localization of these proteins in pterygium. Immunofluorescent staining (Figure ) shows that keratin 6a stained strongly in the superficial layer in pterygial epithelium but not in un-involved conjunctiva. Similarly, CEACAM5 was detected in the squamous layer of pterygial epithelium but was not detectable in conjunctiva. There was up-regulation and nuclear accummulation of the cell adhesion molecule CD24 in the pterygium epithelium, but CD24 was not detectable in conjunctiva. MSMB, on the other hand, was detected prominently in the basal epithelial layer of pterygium, with some staining also in the superficial stromal cells adjacent to the basal epithelia. Lastly, MUC5AC protein was detected in both pterygial and conjunctival tissue sections predominantly in goblet cells. However, the intensity of MUC5AC staining in pterygium was stronger than that in conjunctiva. IGFBP3, present in un-involved conjunctival stroma and epithelium, and NR4A2, present in conjunctival epithelium, were reduced in pterygium (Figure ).Figure 8Tissue localisation of important molecules in pterygium. Immunohistochemical staining images of uninvolved conjunctiva and pterygium tissue. Nuclear position (and indirectly, tissue architecture) was shown using counter-staining with DAPI, in blue. The same magnification was used in all images: Scale Bar = 100 micrometers. A. Proteins with increased expression in pterygium. Staining for keratin 6A, CECAM5, CD24 and MUC5AC was shown in green and MSMB, in red. B. Proteins with decreased expression in pterygium. Staining for IGFBP3 and NR4A2 was shown in green and red respectively.The up-regulation of MSMB transcripts (Figure ) and proteins (Figure ) was also verified by semi-quantitative reverse transcription polymerase chain reaction and Western blot respectively.Figure 9Important molecules in pterygium pathogenesis. A and C. Ethidium bromide stained gel images from semi-quantitative reverse transcription polymerase chain reaction. A. Using specific primers against PSP57 (amplicon length of 452 bp) and PSP94 (amplicon length of 350), PCR products corresponding to PSP94 could be visualised at 30 cycles, whereas PSP57 products could only be just detected at 40 cycles. The sequences were confirmed by sequencing the specific band after cutting the band and extraction of DNA. Note that in patients 1–3, PSP94 transcripts were up-regulated in pterygium relative to conjunctival controls. B. Western blot images, using specific antibodies against MSMB and GADPH (loading control). Note that the PSP proteins were up-regulated in pterygium relative to controls. C. The alternative spliced forms of fibronectin transcripts ED-A and ED-B were detected by visualisation of PCR products. Note that all forms of the transcripts were up-regulated in pterygium compared to conjunctival controls.Comparison of microarray data with tear proteinsAnother report from our center has previously documented the analysis of tear proteins in tears of patients after surgery for pterygium. Briefly, this study employed reverse-phase high-pressure liquid chromatography followed by tryptic digestion and characterisation of proteins using nanoLC-nano-ESI-MS/MS. Eleven genes corresponding to the detected tear proteins were significantly up-regulated in pterygium compared to un-involved conjunctival tissue (Table ). Examples of the up-regulated genes were those encoding for prolactin induced protein I and the S100 A8 proteins. Another eleven genes corresponding to the detected tear proteins were found to be down-regulated in pterygium. These include lipophilin C, ribonuclease 4, complement C3 and histone 1b. Twenty-nine genes corresponding to detected tear proteins were not significantly up or down-regulated in pterygium. Since there were no controls in this study, it is difficult to interpret some of these findings. Furthermore it was not possible to compare the microarray results with every instance of the protein data because more than one type of probe sets may contribute to the synthesis of the protein and not all gene probe sets have signal-noise ratio high enough for this analysis.Table 5Gene microarray data corresponding to detected tear proteins.Name of gene corresponding to tear proteinGenbank AccessionNo of fold changep valueA. Genes significantly regulated in pterygiumProlactin-induced protein INM_0026522.1490.000213S100A8 (Calgranulin A)NM_0029641.740.0000082Defensin alpha A1U739451.5060.0149S100 A9NM_0029651.4420.00204Hepatocellular carcinoma associated protein (PIGR)NM_0026441.3710.0299VimentinAI9225991.2780.0324Beta-2 microglobulinNM_0040481.1880.0378Cathepsin GNM_0019111.1410.00984S100 A12NM_0056211.1220.0268Clusterin (Apolipoprotein J)AI9827541.080.0287Vitamin D binding proteinNM_0005831.0520.0437Cystatin SNNM_0018980.9460.0121Histone 1bNM_0053220.9410.00456Orosomucoid 1 (alpha 1 acid glycoprotein)NM_0006070.940.0027Anti-thrombin IIID298320.9370.000451Beta actinX003510.8880.0263Proline rich PeptideNM_0252630.8640.00355Ribonuclease 4NM_0029370.7250.000675Alpha-1-antitrypsin (SERPINA1)NM_0002950.720.0399Ceruloplasmin (ferroxidase)NM_0000960.6370.000375Complement C3NM_0000640.6070.00443Lipophilin C (SCGB2A1)NM_0024070.4020.000045B. Genes not significantly regulated in pterygiumIGHM proteinL235170.890.0511Lipophilin ANM_0065520.9550.073Zn-alpha2-glycoproteinD904271.2230.0788Apolipoprotein A-INM_0000390.9340.0874Immunoglobulin kappa lightM634380.5590.0878Mucin 7L132831.0230.0895Lipocalin 1NM_0022970.9490.135Alpha-fibrinogenNM_0218711.0380.261TransferrinA10734070.970.319Proline rich 4 (lacrimal)NM_0072440.9590.32Nasopharyngeal carcinoma associated proline rich protein 4NM_0072440.9590.32LactoferrinNM_0023431.0680.366Complement Factor-BAF3496791.0480.386Immunoglobulin J chainAV7332660.7070.398cystatin SANM_0013220.9770.415Apolipoprotein A-IVNM_0004820.9770.442TransthyretinAH626901.0240.493LysozymeNM_0204260.9820.527Complement factor HX046970.9560.542Annexin INM_0007000.9590.583Secretary leukocyte protease inhibitorNM_0030640.9440.592Proline rich protein 5NM_0123901.1450.597Coagulation factor H (thrombin)NM_0005060.9790.6Neutrophil elastaseNM_0019720.990.68Apolipoprotein HNM_0000421.0130.691HemopexinBC0053951.0130.7AlbuminAF1166451.0070.8HaptoglobinNM_0051431.0050.836Basic proline rich proteinX078821.0040.905Note: the study did not have expression data for the genes corresponding to the following tear proteins: defensins HNP-2 and HNP-3, lacritin and DMBT1.DiscussionMajor findingsA global gene expression analysis of pterygium showed distinct differences between primary pterygium and uninvolved conjunctiva. Several pathways were significantly affected in pterygium. These were: increase in the production of extracellular matrix, structural proteins, mitotic proteins, and protein involved in tissue invasion.Recurrent pterygia demonstrated a different signature composed of other perturbed genes. For example, the COL4A6 (AL031177) and the RAB6B (BC002510) were down- and up-regulated respectively in recurrent pterygia compared to primary pterygia. The former encodes for one of the 6 subunits of collagen IV, a major component of the basement membrane, whereas the latter is a RAS family oncogene. The finding suggests that recurrence is a distinct biological phenomenon from the formation of primary pterygium, even though in general, we did not detect obvious microscopic changes between primary and recurrent pterygia.Microarray analysis of gene expression is a useful approach for understanding the molecular mechanism of disease. We used both qPCR and immunohistochemistry to show that both RNA level and protein levels of specific mediators were dysregulated, validating the results from the gene microarray approach. Some 22 genes corresponding to tear proteins detected in pterygium subjects were significantly up or down-regulated in pterygium relative to conjunctival tissue. Many of the processes discovered in this study are highly novel and were not previously associated with pterygium, for example, immunohistochemistry show that cell adhesion molecules (ie., CD24) may be increased but abnormally localized in the nuclei in pterygium epithelium and therefore cell adhesion properties may be disturbed. Another example is the evidence for Goblet cell dysfunction. In pterygium, there was elevation of transcript and protein expression of MUC5AC and MSMB. MUC5AC is a well-known marker of conjunctival mucous-secreting Goblet cells, and MSMB is a cutaneous squamous cell carcinoma related protein that is associated with respiratory tract Goblet cells.Comparison with previous studiesThe data from a previous study utilized only 2 primary and 1 recurrent pterygia, with no stratification of data between the clinical sub-types. This study highlighted the up-regulation of 29 genes common to primary and recurrent pterygium. Our results supported the dysregulation of 19 of these genes in pterygium (Table ), for example, TRAP100, MIP-4, RBP-1, MAP-17 and PECAM1. One apparent discrepancy was that the CLIC2 was up-regulated in our study, but significantly down-regulated in John-Aryankalayil et al. Discrepancies reflect differences in study methodology, for example, this study used an older microarray chip HG_U95Av2, containing considerably less probes than the U133A; and the earlier study may have differences in probe, chip or gene-level normalizations which were not reported. Differences in normalization may account for the differences in the folds of change in specific genes between different studies. In John-Aryankalayil et al's study, although a few up-regulated genes from different functional categories were tabulated, there was no attempt to evaluate the patterns of expression or relative contribution of broad biological processes. Unlike this study, we did not restrict to listing differentially expressed genes by fold change.Table 6Comparison of microarray analysis between two studiesGeneFold change in John-Aryankalayil et alFold change in our studySignificantly regulated in current study (p < 0.05)SILV (silver homolog)2.22.3CD31 (PECAM1)2.12.2C5orf132.52.2RBP13.62.0Von Willebrand factor2.52.0RAB312.01.7Clathrin (CLTB)2.41.7DAB22.11.5S100 A9 (calgranulin B)2.21.4MAP 172.41.4MIP-44.91.4ECM13.21.3COMP2.01.3versican3.31.2TRAP1005.01.2CD32 receptor2.91.2mRNA, clone AL0501542.81.1CLIC20.51.1Prostaglandin F2alpha receptor3.21.1Not significantly regulated in current study (p > 0.05)NDRG43.11.1Carbonic anhydrase CA10.51.1ABCG12.31.1ADPRTL22.31.1Casein kinase II2.51.0Collagen III4.11.0Fibronectin8.51.0LZTR12.21.0Collagen VI2.11.0Myosin heavy chain II3.10.9FUBP0.50.9Ephrin-A10.50.9Per10.50.9Lipocalin 22.40.9Previous studies have shown that gene expression profiles in tumor and wound response were similar, lending support to our opinion that wound response is the major theme in global pterygial gene expression.Possible mechanisms of pterygia formationWe believe that our data support a predominantly wound healing pattern of gene expression in pterygium. Genes encoding for extracellular matrix, structural and adhesion molecules including wound healing related proteins, collagen subtypes, keratin 6A and fibronectin were significantly up-regulated in pterygium. Other up-regulated proteins like small proline rich protein 1B (SPRR1B), CD24, S100 calcium binding protein, SPARC, TFF1, SPRR1B and SERPINB13 also govern the wound healing process and cornification of epithelium, again, supporting the hypothesis of aberrant wound response in pterygium. Fibronectin alternatively splice transcript with EDA, up-regulated in our pterygial specimens, was shown to be up-regulated in wound healing. In order to further validate this finding, semi-quantitative PCR was performed on paired pterygial and conjunctival tissue specimens from 3 patients to examine the level of fibronectin transcript with EDA. The results show that pterygium expressed higher levels of fibronectin transcript with EDA compared to uninvolved conjunctiva (Figure ).Our data also suggest that in pterygia formation, there may be increased cellular proliferation and motility on one hand, and reduced cell death on the other. Genes up-regulated in pterygium include fibronectin (FN1), CEACAM5 (CEA), CD24, SPARC, MSMB and TFF1. CEACAM5 (CEA), a common cancer marker, was up-regulated 3.8-fold in pterygium. SPARC was known to have anti-adhesive property, and induced cancer cell motility. The transcripts coding for microseminoprotein (MSMB) or PSP94 was up-regulated 3.6-fold, whereas transcripts for calcium binding protein S100A8 was up-regulated 5.7-fold in pterygium. Interestingly, in our analysis of tears from patients with pterygium, S100 proteins were also detected. Trefoil factors promote restitution of epithelial cells and are abundantly secreted onto the mucosal surface rapidly after mucosal injury. In contrast to the above, genes coding for apoptosis (TGM2, IGFBP3 and DUSP1) were down-regulated in pterygium, reinforcing the over-proliferative tendency in pterygium.The stress-inducible transcription regulator genes including ATF3, BTG2, EGR1, ERG2, FOS, JUN, NR4A1 and NR4A2 were surprisingly, down-regulated in pterygium relative to uninvolved conjunctiva. One has to bear in mind however, that our tissue specimens may represent chronic UV stimulation. The expression of these transcription factors may very well be elevated shortly after the commencement of the initial stimulus.The transcript level of immunoglobulin subunit in pterygium was relatively lower as compared to uninvolved conjunctiva. This does not support a role for B cell mediated immunity in pterygium formation. However, the co-expression of some immunoglobulin genes (Figure ) may have some biological significance, such as an immunological basis for the inappropriate wounding response.Possible mechanisms of pterygia recurrenceOne hypothesis is that a low level of inflammation may stimulate epithelial and fibroblast cell migration due to factors such as SPN [see Additional file ]. This allows continuation of inflammatory factors and cytokines to be produced, potentiating a cycle [see Additional file ]. Factors such as SPN are unique to recurrence only, since they were up-regulated compared to primary pterygium cases.In an alternative hypothesis, the formation of an exuberant scar or a 'go' signal, may or may not be coupled with a reduction of 'stop' or inhibitory signals. One such 'stop signal' may be the MSMB protein. Any up-regulation of this anti-metastatic gene in pterygium could give an inhibiting signal to slow the growth rate in pterygium. When pterygium is surgically removed, this inhibiting factor may also be reduced, encouraging the remnants of fibrous tissue to proliferate more aggressively, attaining the phenotype of recurrent pterygium. This may also explain the success of conjunctival autographs as adjunctive procedure after excision of the lesion in the prevention of recurrence. The graft tissue may 'replenish' any surgically-induced loss of 'stop' signals. In support of this hypothesis, the level of MSMB was significantly depressed in recurrent compared to primary pterygia.Strengths and LimitationsThe strengths of this study include the use of a variety of analytical approaches to interpret the microarray data. The tissue specimens have been harvested in a center which manages a high volume of such disease. Samples have been processed in a very standardized fashion, obtained from patients with accurate clinical diagnoses.One limitation of the study is that we did not evaluate p53 and related genes such as the Vasoendothelial growth factor (VEGF). Pterygia specimens contained a mixture of cell types whereas un-involved conjunctiva consisted largely of epithelial tissue. Therefore, any observations concerning differential gene expression could be related to differences in composition of cells as well as to pathological processes.We did not notice regional differences in the intensity of CD24 immunofluorescent staining along the length of the pterygium epithelium on longitudinal section [see Additional file ]. However, we cannot conclusively state that there is no difference in the pattern of CD24 expression between the head and the body of the pterygium. Since the limits of the pterygium body was defined clinically by the surgeon, we were uncertain if the bulk of the pterygium body was available for immunohistology.Since we did not obtain temporally located pterygia in this study, our findings may not be applicable to this less frequently occurring type of pterygia.Specimens analysed were at the later stage of pterygium formation when surgical removal was needed. No data at the early stage of disease were available for investigation.In addition to these limitations, there are inherent limitations with hieuristic and even more deterministic methods of analyzing global gene expression. For example, the results of the k-means clustering depend arbitrarily on the initial number of clusters and the number of iterations. However, we attempt to reduce the effects of these shortcomings by employing a variety of different data-mining techniques and interpreting results as a whole and not in isolation. We drew conclusions based on findings that are consistent between different procedures and refrained from over-interpretation of isolated anomalies.If changes in global gene expression in cases of recurrent pterygium do not occur until a very late time point, it may not be possible to predict recurrence of the condition by analysing the tissue obtained during the time of first excision. However, the gene expression data may still be useful for understanding the biology of recurrence. The tear protein analysis also has limitations. It is difficult to determine whether the detected proteins were physiological, related to the surgical trauma, or arising from the disease. Nevertheless the 22 significantly up-regulated genes that have products in human tears may be potential biomarkers for the disease.Potential applicationsThis study illustrates that an unbiased global gene expression approach is useful to address the disease mechanisms in a controversial condition. Further studies on more specimens may allow the use of a sub-set of genes to prognosticate lesions. Non-surgical treatment for pterygium currently does not exist. Since wound healing and matrix dysregulation is a major theme, known modulators of wound healing may be explored in pterygium in a targeted way, using topical application. Other novel processes require further evaluation. Studies in the regulation of cell adhesion by CD24 and mucous processing pathways are required to understand pterygium formation. Studies are also required to elicit the origin of the overlying epithelium in pterygium tissue.Nevertheless, our study suggests that after resection of the initial tumor, clinicians can modify treatment in selected patients based on objective criteria after analysing tumor gene expression or tumor proteins.ConclusionBased on differential gene profiling in pterygium, it can be concluded that an aberrant wound healing process is the major pathogenetic process in pterygial formation. The other secondary processes such as cornification and attempted barrier formation may be compensatory in nature. This contrasts sharply with frank malignant neoplasms, where gene expression signatures are dominanted by increased proliferation, reduced apoptosis, cell cycling anomalies and genomic instability. The existence of an aggressive phenotype to account for post excision recurrence may be related to an imbalance of growth signals rather than due to mere prolongation of the stimulus that initiated primary pterygial formation.AbbreviationsCT: threshold cycle; (c)DNA: (complementary) deoxyribonucleic acid; ECM: extracellular matrix; FDR: false discovery rate; GEO: Gene expression omnibus; GSEA: Gene set enrichment analysis; OCT: optimal cutting temperature; (q) PCR: (quantitative) polymerase chain reaction; RMA: Robust Multi-array Average; (c)RNA: (complementary) ribonucleic acid.Competing interestsThe authors declare that they have no competing interests.Authors' contributionsLT drafted and revised the manuscript, performed the analysis and coordinated study. JC carried out the molecular studies and microarray experiments, HY carried out the analysis on GSEA and provided help with bioinformatics, LPKA conceived of the study, coordinated study, revised the manuscript. DTHT critically reviewed manuscript, provided patient materials and RB initiated the study and formulated the design, revised the manuscript and approved the submission.Pre-publication historyThe pre-publication history for this paper can be accessed here:Supplementary MaterialAdditional File 1Clinical appearance of pterygium on color photographs. A. Primary pterygium and B. Recurrent pterygium. Note the excessive scarring that is typical of recurrent pterygium.Click here for fileAdditional File 2Primary and recurrent pterygium specific genes. Heat map showing pterygium specific and recurrence specific genes. Note that in recurrent pterygium a larger number of genes were relatively down-regulated (green) compared to primary pterygium and un-involved conjunctiva, and a smaller but still large number of genes were up-regulated (red) compared to primary pterygium and un-involved conjunctiva.Click here for fileAdditional File 3Postulated pterygium recurrence pathways. These pathways A and B were obtained from pathway analysis. Red symbols: genes up-regulated in recurrent pterygia compared to primary pterygia.Click here for fileAdditional File 4Immunofluorescence image showing staining for CD24 in the pterygium epithelium.Click here for fileAdditional File 5Relationships in the pathway studio 5.0 analysis for up-regulated genes.Click here for fileAdditional File 6Relationships in the pathway studio 5.0 analysis for down-regulated genes.Click here for fileAdditional File 7Primers used for PCR.Click here for file\""
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dftmp.iloc[0]['body']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df[df['id'] == 'PMC5052263']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<!DOCTYPE article PUBLIC \"-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD v1.0 20120330//EN\" \"JATS-archivearticle1.dtd\">\\n<article xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns:mml=\"http://www.w3.org/1998/Math/MathML\" article-type=\"research-article\"><?properties open_access?><front><journal-meta><journal-id journal-id-type=\"nlm-ta\">Front Immunol</journal-id><journal-id journal-id-type=\"iso-abbrev\">Front Immunol</journal-id><journal-id journal-id-type=\"publisher-id\">Front. Immunol.</journal-id><journal-title-group><journal-title>Frontiers in Immunology</journal-title></journal-title-group><issn pub-type=\"epub\">1664-3224</issn><publisher><publisher-name>Frontiers Media S.A.</publisher-name></publisher></journal-meta><article-meta><article-id pub-id-type=\"pmid\">31191530</article-id><article-id pub-id-type=\"pmc\">6546853</article-id><article-id pub-id-type=\"doi\">10.3389/fimmu.2019.01148</article-id><article-categories><subj-group subj-group-type=\"heading\"><subject>Immunology</subject><subj-group><subject>Original Research</subject></subj-group></subj-group></article-categories><title-group><article-title>Comprehensive Characterization of a Next-Generation Antiviral T-Cell Product and Feasibility for Application in Immunosuppressed Transplant Patients</article-title></title-group><contrib-group><contrib contrib-type=\"author\"><name><surname>Amini</surname><given-names>Leila</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff2\"><sup>2</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff4\"><sup>4</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/642379/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Vollmer</surname><given-names>Tino</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/667138/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Wendering</surname><given-names>Desiree J.</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff4\"><sup>4</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/637208/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Jurisch</surname><given-names>Anke</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/701575/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Landwehr-Kenzel</surname><given-names>Sybille</given-names></name><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref><xref ref-type=\"aff\" rid=\"aff6\"><sup>6</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/165903/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Otto</surname><given-names>Natalie Maureen</given-names></name><xref ref-type=\"aff\" rid=\"aff2\"><sup>2</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref></contrib><contrib contrib-type=\"author\"><name><surname>J&#x000fc;rchott</surname><given-names>Karsten</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/723708/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Volk</surname><given-names>Hans-Dieter</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/23718/overview\"/></contrib><contrib contrib-type=\"author\"><name><surname>Reinke</surname><given-names>Petra</given-names></name><xref ref-type=\"aff\" rid=\"aff2\"><sup>2</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref></contrib><contrib contrib-type=\"author\"><name><surname>Schmueck-Henneresse</surname><given-names>Michael</given-names></name><xref ref-type=\"aff\" rid=\"aff1\"><sup>1</sup></xref><xref ref-type=\"aff\" rid=\"aff3\"><sup>3</sup></xref><xref ref-type=\"aff\" rid=\"aff5\"><sup>5</sup></xref><xref ref-type=\"corresp\" rid=\"c001\"><sup>*</sup></xref><uri xlink:type=\"simple\" xlink:href=\"http://loop.frontiersin.org/people/236711/overview\"/></contrib></contrib-group><aff id=\"aff1\"><sup>1</sup><institution>Institute for Medical Immunology, Charit&#x000e9; University Medicine Berlin</institution>, <addr-line>Berlin</addr-line>, <country>Germany</country></aff><aff id=\"aff2\"><sup>2</sup><institution>Renal and Transplant Research Unit, Department of Nephrology and Internal Intensive Care, Charit&#x000e9; University Medicine Berlin</institution>, <addr-line>Berlin</addr-line>, <country>Germany</country></aff><aff id=\"aff3\"><sup>3</sup><institution>Berlin Institute of Health Center for Regenerative Therapies (BCRT), Charit&#x000e9; University Medicine Berlin</institution>, <addr-line>Berlin</addr-line>, <country>Germany</country></aff><aff id=\"aff4\"><sup>4</sup><institution>Berlin-Brandenburg School for Regenerative Therapies, Charit&#x000e9; University Medicine Berlin</institution>, <addr-line>Berlin</addr-line>, <country>Germany</country></aff><aff id=\"aff5\"><sup>5</sup><institution>Berlin Center for Advanced Therapies, Charit&#x000e9; University Medicine Berlin</institution>, <addr-line>Berlin</addr-line>, <country>Germany</country></aff><aff id=\"aff6\"><sup>6</sup><institution>Department for Pediatric Pulmonology, Immunology and Intensive Care Medicine, Charit&#x000e9; University Medicine Berlin</institution>, <addr-line>Berlin</addr-line>, <country>Germany</country></aff><author-notes><fn fn-type=\"edited-by\"><p>Edited by: Hermann Einsele, University of Wuerzburg, Germany</p></fn><fn fn-type=\"edited-by\"><p>Reviewed by: Lucas Coelho Marli&#x000e8;re Arruda, Karolinska Institute (KI), Sweden; Jianing Fu, Columbia University, United States</p></fn><corresp id=\"c001\">*Correspondence: Michael Schmueck-Henneresse <email>michael.schmueck-henneresse@charite.de</email></corresp><fn fn-type=\"other\" id=\"fn001\"><p>This article was submitted to Alloimmunity and Transplantation, a section of the journal Frontiers in Immunology</p></fn></author-notes><pub-date pub-type=\"epub\"><day>28</day><month>5</month><year>2019</year></pub-date><pub-date pub-type=\"collection\"><year>2019</year></pub-date><volume>10</volume><elocation-id>1148</elocation-id><history><date date-type=\"received\"><day>08</day><month>1</month><year>2019</year></date><date date-type=\"accepted\"><day>07</day><month>5</month><year>2019</year></date></history><permissions><copyright-statement>Copyright &#x000a9; 2019 Amini, Vollmer, Wendering, Jurisch, Landwehr-Kenzel, Otto, J&#x000fc;rchott, Volk, Reinke and Schmueck-Henneresse.</copyright-statement><copyright-year>2019</copyright-year><copyright-holder>Amini, Vollmer, Wendering, Jurisch, Landwehr-Kenzel, Otto, J&#x000fc;rchott, Volk, Reinke and Schmueck-Henneresse</copyright-holder><license xlink:href=\"http://creativecommons.org/licenses/by/4.0/\"><license-p>This is an open-access article distributed under the terms of the Creative Commons Attribution License (CC BY). The use, distribution or reproduction in other forums is permitted, provided the original author(s) and the copyright owner(s) are credited and that the original publication in this journal is cited, in accordance with accepted academic practice. No use, distribution or reproduction is permitted which does not comply with these terms.</license-p></license></permissions><abstract><p>Viral infections have a major impact on morbidity and mortality of immunosuppressed solid organ transplant (SOT) patients because of missing or failure of adequate pharmacologic antiviral treatment. Adoptive antiviral T-cell therapy (AVTT), regenerating disturbed endogenous T-cell immunity, emerged as an attractive alternative approach to combat severe viral complications in immunocompromised patients. AVTT is successful in patients after hematopoietic stem cell transplantation where T-cell products (TCPs) are manufactured from healthy donors. In contrast, in the SOT setting TCPs are derived from/applied back to immunosuppressed patients. We and others demonstrated feasibility of TCP generation from SOT patients and first clinical proof-of-concept trials revealing promising data. However, the initial efficacy is frequently lost long-term, because of limited survival of transferred short-lived T-cells indicating a need for next-generation TCPs. Our recent data suggest that Rapamycin treatment during TCP manufacture, conferring partial inhibition of mTOR, might improve its composition. The aim of this study was to confirm these promising observations in a setting closer to clinical challenges and to deeply characterize the next-generation TCPs. Using cytomegalovirus (CMV) as model, our next-generation Rapamycin-treated (Rapa-)TCP showed consistently increased proportions of CD4<sup>+</sup> T-cells as well as CD4<sup>+</sup> and CD8<sup>+</sup> central-memory T-cells (T<sub>CM</sub>). In addition, Rapamycin sustained T-cell function despite withdrawal of Rapamycin, showed superior T-cell viability and resistance to apoptosis, stable metabolism upon activation, preferential expansion of T<sub>CM</sub>, partial conversion of other memory T-cell subsets to T<sub>CM</sub> and increased clonal diversity. On transcriptome level, we observed a gene expression profile denoting long-lived early memory T-cells with potent effector functions. Furthermore, we successfully applied the novel protocol for the generation of Rapa-TCPs to 19/19 SOT patients in a comparative study, irrespective of their history of CMV reactivation. Moreover, comparison of paired TCPs generated before/after transplantation did not reveal inferiority of the latter despite exposition to maintenance immunosuppression <italic>post</italic>-SOT. Our data imply that the Rapa-TCPs, exhibiting longevity and sustained T-cell memory, are a reasonable treatment option for SOT patients. Based on our success to manufacture Rapa-TCPs from SOT patients under maintenance immunosuppression, now, we seek ultimate clinical proof of efficacy in a clinical study.</p></abstract><kwd-group><kwd>cytomegalovirus</kwd><kwd>adoptive T-cell therapy</kwd><kwd>solid organ transplantation</kwd><kwd>Rapamycin</kwd><kwd>mTOR</kwd><kwd>immune regeneration</kwd></kwd-group><counts><fig-count count=\"7\"/><table-count count=\"0\"/><equation-count count=\"0\"/><ref-count count=\"100\"/><page-count count=\"20\"/><word-count count=\"14103\"/></counts></article-meta></front><body><sec sec-type=\"intro\" id=\"s1\"><title>Introduction</title><p>Severe viral infections have a major impact on the clinical course of immunocompromised patients. Despite availability of powerful antiviral medication, cytomegalovirus (CMV) still accounts for significant morbidity and mortality in solid organ transplant (SOT) recipients (<xref rid=\"B1\" ref-type=\"bibr\">1</xref>). CMV can trigger direct and indirect morbidities such as chronic allograft rejection or in the case of kidney transplantation (KTx) chronic nephropathy (<xref rid=\"B2\" ref-type=\"bibr\">2</xref>, <xref rid=\"B3\" ref-type=\"bibr\">3</xref>). Therapeutic control of CMV may be hampered by the development of anti-viral drug resistance (<xref rid=\"B4\" ref-type=\"bibr\">4</xref>). Moreover, after discontinuation of anti-viral prophylaxis, late-onset CMV disease frequently occurs and overall mortality is significantly higher in CMV-infected compared to uninfected KTx patients (<xref rid=\"B1\" ref-type=\"bibr\">1</xref>). Of note, T-cell-mediated anti-CMV immunity was reported to be predictive for the development of late-onset disease (<xref rid=\"B5\" ref-type=\"bibr\">5</xref>) and anti-CMV<sub>IE&#x02212;1</sub>-specific CD8<sup>+</sup> T-cell responses stratify risk of CMV disease in heart and lung transplant as well as KTx patients (<xref rid=\"B6\" ref-type=\"bibr\">6</xref>, <xref rid=\"B7\" ref-type=\"bibr\">7</xref>). In addition, the magnitude of the CMV<sub>pp65</sub>- and CMV<sub>IE&#x02212;1</sub>-specific T-cell responses turned out to be protective against complications with CMV in hematopoietic stem cell transplantation (HSCT) (<xref rid=\"B8\" ref-type=\"bibr\">8</xref>, <xref rid=\"B9\" ref-type=\"bibr\">9</xref>). Consequently, regeneration of the endogenous T-cell response against these antigens, as aspired by AVTT, may prevent and reduce virus-associated morbidities/mortality in the SOT setting. Other viruses with impact on SOT outcomes are Epstein-Barr-virus and BK-virus, for which less efficient or no antiviral drugs are available. T-cells play a key role in protection from severe viral infections (<xref rid=\"B7\" ref-type=\"bibr\">7</xref>, <xref rid=\"B10\" ref-type=\"bibr\">10</xref>, <xref rid=\"B11\" ref-type=\"bibr\">11</xref>). Thus, adoptive T-cell therapy (AVTT) is a potent novel treatment strategy to tackle fatal viral complications in immunosuppressed transplant patients. Mechanisms of success or failure of new AVTT approaches need to be thoroughly understood and specific characteristics of patient cohorts have to be considered for successful translation of AVTT.</p><p>For clinical application of AVTT, <italic>ex vivo</italic> enrichment and expansion of virus-specific T-cells under GMP conditions are crucial and thus various protocols have been developed for CMV-specific AVTT after HSCT (<xref rid=\"B12\" ref-type=\"bibr\">12</xref>&#x02013;<xref rid=\"B17\" ref-type=\"bibr\">17</xref>). However, the success of these approaches is limited in SOT patients due to the T-cell products (TCPs) being derived from patients instead of healthy HSCT donors, the lack of lymphodepletive preconditioning and the need for concomitant immunosuppression. Nevertheless, we and other groups demonstrated not only safety of AVTT, but also significant reduction of viral load and control of clinical symptoms of CMV disease in SOT recipients under maintenance immunosuppression in proof-of-concept studies (<xref rid=\"B18\" ref-type=\"bibr\">18</xref>&#x02013;<xref rid=\"B21\" ref-type=\"bibr\">21</xref>). These observations are in line with positive results of AVTT for treatment of patients with EBV-related <italic>post</italic>-transplant lymphoproliferative disease (<xref rid=\"B22\" ref-type=\"bibr\">22</xref>&#x02013;<xref rid=\"B24\" ref-type=\"bibr\">24</xref>). Yet, long-term efficacy failed in some patients, who experienced recurrence of CMV or EBV load and symptoms (<xref rid=\"B18\" ref-type=\"bibr\">18</xref>&#x02013;<xref rid=\"B25\" ref-type=\"bibr\">25</xref>). To adapt AVTT to combat these clinical challenges, it is crucial to consider the respective patient cohort and the TCPs\\' characteristics. Specifically, our aim was to increase longevity of transferred T-cells to improve sustainability of clinical efficacy of AVTT in SOT patients. Failure of long-term control of CMV/EBV infections may be due to limited persistence of adoptively transferred T-cells <italic>in vivo</italic>, which might occur due to the late differentiation state implying limited longevity of infused T-cells. Therefore, advancing the quality of adoptively transferred TCPs with defined compositions by the enrichment for distinct T-cell memory subsets may improve therapeutic outcome. In particular, central-memory T-cells (T<sub>CM</sub>; CCR7<sup>+</sup> CD62L<sup>+</sup> CD45RO<sup>+</sup> CD45RA<sup>&#x02212;</sup>) and memory-stem T-cells (T<sub>SCM</sub>; CCR7<sup>+</sup> CD62L<sup>+</sup> CD45RO<sup>&#x02212;</sup> CD45RA<sup>+</sup> CD95<sup>+</sup>) have high proliferative potential, self-renewal capacity and are reported to show superior engraftment, persistence, and survival compared to more differentiated memory T-cells (<xref rid=\"B26\" ref-type=\"bibr\">26</xref>&#x02013;<xref rid=\"B33\" ref-type=\"bibr\">33</xref>). Conversely, late-differentiated short-lived effector-memory T-cells (T<sub>EM</sub>; CCR7<sup>&#x02212;</sup> CD62L<sup>&#x02212;</sup> CD45RA<sup>&#x02212;</sup>CD45RO<sup>+</sup>) and terminally-differentiated effector T-cells (T<sub>EMRA</sub>; CCR7<sup>&#x02212;</sup> CD62L<sup>&#x02212;</sup> CD45RA<sup>+</sup>CD45RO<sup>&#x02212;</sup>) exert immediate effector function, but fail to establish long-lasting protective memory, because of poor proliferative potential and limited survival following antigenic rechallenge (<xref rid=\"B26\" ref-type=\"bibr\">26</xref>, <xref rid=\"B27\" ref-type=\"bibr\">27</xref>, <xref rid=\"B34\" ref-type=\"bibr\">34</xref>). Remarkably, these observations match clinical data demonstrating T-cell reconstitution after HSCT and prevention of CMV disease related to T<sub>CM</sub> proportions in peripheral blood (<xref rid=\"B35\" ref-type=\"bibr\">35</xref>). Direct sorting strategies to isolate only CMV-specific long-lived T-cells are barely feasible under GMP conditions and would yield very small cell numbers likely not sufficient for successful AVTT in immunosuppressed SOT recipients. To ensure applicability in a clinical setting, we recently optimized our GMP-conform manufacturing process for autologous virus-specific TCPs and succeeded in attenuating T-cell differentiation by treatment with low doses of Rapamycin (inhibits the <italic>mechanistic-target-of-rapamycin-complex-1</italic>: mTOR-C1, favorable results with 20 nM) during expansion cultures (<xref rid=\"B18\" ref-type=\"bibr\">18</xref>, <xref rid=\"B36\" ref-type=\"bibr\">36</xref>, <xref rid=\"B37\" ref-type=\"bibr\">37</xref>). This next-generation antiviral TCP comprises enriched proportions of early-differentiated T<sub>CM</sub> being superior for AVTT (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>&#x02013;<xref rid=\"B39\" ref-type=\"bibr\">39</xref>). Furthermore, next-generation Rapamycin-treated (Rapa-)TCPs contain higher proportions of CD4<sup>+</sup> T-cells (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>) reported to improve clinical efficacy (<xref rid=\"B29\" ref-type=\"bibr\">29</xref>, <xref rid=\"B40\" ref-type=\"bibr\">40</xref>, <xref rid=\"B41\" ref-type=\"bibr\">41</xref>).</p><p>Detailed knowledge regarding the characteristics of Rapa-TCPs is a prerequisite for realization of clinical translation. Thus, we closely investigated the molecular properties of this Rapa-TCP regarding dependence on cytokine supplementation regiments during <italic>in vitro</italic> expansion, long-term stability, survival/sensitivity to apoptosis, metabolism, transcriptome, clonal composition, the role of the different memory T-cell subsets and applicability to SOT patient samples. Our data reveal a beneficial early differentiated phenotype, profound function, elevated clonal diversity, and superior survival of Rapa-TCPs compared to first-generation TCPs, which is further underlined and confirmed by a distinct gene expression signature revealed by mRNA sequencing.</p><p>We used <italic>in vitro</italic> models to mimic the situation of TCPs once injected into a patient coping with CMV disease, <italic>i.e</italic>., massive antigen exposure and withdrawal of Rapamycin. Here, we observed a preserved capacity for CMV-specific production of effector cytokines. Moreover, we tested manufacture of CMV-specific TCPs from material of patients with chronic end-stage renal disease (ESRD) before and after KTx to investigate the impact of chronic immunosuppression, showing no benefits of TCP generation before KTx. We further investigated the influence of CMV-reactivation history after KTx on the differentiation of virus-reactive memory T-cells and the resulting composition of untreated and Rapa-TCPs, implying feasibility of TCP generation from all groups investigated. This next-generation AVTT approach may also be applied to other viral specificities, such as EBV and BKV or even cancer-specific T-cells. Prospectively, implementation of next-generation AVTT may allow for reduction or complete ablation of toxic anti-viral medication and minimize the risk for virus-associated complications in the SOT setting.</p></sec><sec sec-type=\"methods\" id=\"s2\"><title>Methods</title><sec><title>Patients\\' and Healthy Donors\\' Blood Samples</title><p>Venous blood samples were collected from 19 healthy donors (HDs) (10 m/9 f; 25&#x02013;81 years) and 19 KTx patients (11 m/8 f; 34&#x02013;78 years; <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S2</xref>) of the Kidney Transplant Ambulance, Charit&#x000e9; Virchow Klinikum, Berlin. We worked with buffy coats from 3 of the 19 different HDs to have sufficient cells for different cell sorting steps. Peripheral blood mononuclear cells (PBMC) were isolated by Biocoll Separating Solution density gradient centrifugation (Biochrom). The Charit&#x000e9; Ethics Committee (IRB) approved the study protocol and all blood donors provided written informed consent. Detailed characteristics of each patient are presented in <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S2</xref>.</p></sec><sec><title>Enrichment and Expansion of CMV-Specific T-Cells</title><p>CMV-specific TCPs were generated using a previously described technique (<xref rid=\"B18\" ref-type=\"bibr\">18</xref>, <xref rid=\"B36\" ref-type=\"bibr\">36</xref>, <xref rid=\"B37\" ref-type=\"bibr\">37</xref>, <xref rid=\"B42\" ref-type=\"bibr\">42</xref>). Briefly, PBMCs were stimulated for 6 h with overlapping CMV<sub>pp65/IE&#x02212;1</sub> peptide pools (JPT Peptide Technologies; 0.5 &#x003bc;g/ml each). IFN&#x003b3;-producing cells underwent positive selection using the IFN&#x003b3; Secretion Assay&#x02014;Cell Enrichment and Detection Kit according to the manufacturer\\'s instructions (Miltenyi Biotec). Enriched IFN&#x003b3;<sup>+</sup> cells were cultured for 21 days in 96- or 24-well-plates with irradiated (30 Gy using a GSR D1 [Gamma-Service Medical GmbH]) autologous feeder cells (derived from 1/5 of the capture assays\\' negative fraction) added only at d0 in complete media (VLE RPMI 1640 supplemented with penicillin (100 IU/ml) and streptomycin [all from Biochrom] and 10% fetal calf serum [FCS, PAA]), supplemented with 10 ng/ml recombinant human (rh) IL-7 and rh IL-15 (CellGenix) or 50 U/ml IL-2 in humidified incubators at 37&#x000b0;C and 5% CO<sub>2</sub>. Cells were split 1:1 when 100% confluence was reached. For Rapa-TCPs, 20 nM of Rapamycin (Rapamune, Pfizer Pharma GmbH) were added every 2nd day or upon splitting starting from day 1 (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>).</p><p>For restimulation during culture (only where indicated, <xref ref-type=\"fig\" rid=\"F1\">Figure 1</xref>, <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1</xref>), we depleted the donor\\'s PBMCs of CD3<sup>+</sup> cells using magnetically activated cell sorting (MACS) with anti-CD3 beads (Miltenyi Biotec) following the manufacturer\\'s instructions. These cells were frozen in FCS with 10% cell culture grade dimethyl-sulfoxide (Sigma-Aldrich) until restimulation at d14. Thawed autologous CD3<sup>&#x02212;</sup> PBMCs were washed twice and pulsed with overlapping CMV<sub>pp65/IE&#x02212;1</sub> peptide pools at concentrations of 2 &#x003bc;g/ml each peptide for 2 h and added at a 1:5 ratio to the T-cells in culture.</p><fig id=\"F1\" position=\"float\"><label>Figure 1</label><caption><p>Effects of Rapamycin on T-cell products: Expansion, phenotype and function. <bold>(A)</bold> Schematic overview of experiments: T-cell products (TCPs) were generated from PBMCs isolated from venous blood of healthy donors (HDs) by magnetically activated cell sorting (MACS) of T-cells producing IFN&#x003b3; in response to stimulation with CMV<sub>IE&#x02212;1/pp65</sub> peptide pools and expanded in the presence of either IL-2/IL-7 (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1</xref>) or IL-7/IL-15 without (w/o; blue) or with addition of 20 nMof Rapamycin (Rapa; red) (<bold>B&#x02013;P</bold>). Parts of the culture were re-stimulated using thawed CD3<sup>&#x02212;</sup> PBMCs loaded with CMV<sub>IE&#x02212;1/pp65</sub> peptide pools, deprived of Rapamycin or a combination of both on d14. <bold>(B)</bold> Expansion rates of IL-7/15-expanded Rapa-treated (Rapa-)TCPs (red) and untreated TCPs (blue) of <italic>n</italic> = 10 healthy donors (HDs) calculated from yield at d14 divided by the number of seeded cells at d0. We gated flow cytometric data on lymphocytes singlets living CD3<sup>+</sup> T-cells. <bold>(C)</bold> Exemplary flow cytometry plots of CD4<sup>+</sup> and CD8<sup>+</sup> populations among living CD3<sup>+</sup> T-cells in the Rapa-TCP (left plot) and untreated TCP (w/o, right plot) of one HD. <bold>(D)</bold> CD4/CD8 ratios in Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 10 HDs calculated from flow cytometry data as presented in <bold>(C)</bold>. <bold>(E)</bold> Gating strategy for CD45RA<sup>&#x02212;</sup> CCR7<sup>+</sup> central memory T-cells (T<sub>CM</sub>) among CD4<sup>+</sup> (upper panel) and CD8<sup>+</sup> (lower panel) in Rapa- (left panel) and untreated TCPs (right panel) of one exemplary HD. <bold>(F)</bold> Proportions of CD4<sup>+</sup> and CD8<sup>+</sup> T<sub>CM</sub> among Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 10 HDs determined from flow cytometric data as shown in <bold>(E)</bold> at d14. <bold>(G,H)</bold> To detect CMV-specific cytokine producers, TCPs were stimulated with CMV<sub>IE&#x02212;1/pp65</sub> peptide-loaded autologous lymphoblastic cell lines (LCLs) at a ratio of 1:10 for 6 h and Brefeldin A (BFA) was added after 1 h. <bold>(G)</bold> Representative flow cytometric plots of IFN&#x003b3;- and TNF&#x003b1;-producers in Rapa- (left panel, red) and untreated TCPs (right panel, blue) of one HD. The dark population represents unstimulated and the light population illustrates CMV<sub>IE&#x02212;1/pp65</sub>-stimulated CD4<sup>+</sup> (upper panel) and CD8<sup>+</sup> T-cells (lower panel). <bold>(H)</bold> Proportions of CMV-specific IFN&#x003b3;-producers among CD4<sup>+</sup> and CD8<sup>+</sup> T-cells in Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 10 HDs determined from flow cytometric data as shown in <bold>(G)</bold> at d14. <bold>(I&#x02013;N)</bold>: For re-stimulation on d14 of culture, thawed CD3<sup>&#x02212;</sup> autologous PBMCs were loaded with CMV<sub>IE&#x02212;1/pp65</sub> peptide pools and added at 1:5 ratio to T-cells. <bold>(I)</bold> Expansion rates of IL-7/15-expanded re-stimulated (pastel colors) or non-re-stimulated (dark colors) Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 7 HDs calculated from yield at d21 divided by the number of cells at d14. <bold>(J)</bold> CD4/CD8 ratios in Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 7 HDs calculated from flow cytometric data as presented in <bold>(C)</bold> at d21. <bold>(K</bold>,<bold>L)</bold>: Proportions of CD4<sup>+</sup>\\n<bold>(K)</bold> and CD8<sup>+</sup> T<sub>CM</sub>\\n<bold>(L)</bold> among Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 7 HDs determined from flow cytometric data as shown in <bold>(E)</bold> at d21. <bold>(M&#x02013;P)</bold> To detect CMV-specific cytokine producers, TCPs were stimulated with CMV<sub>IE&#x02212;1/pp65</sub> peptide-loaded autologous LCLs for 6 h and BFA was added after 1 h. <bold>(M&#x02013;N)</bold> Proportions of CMV-specific IFN&#x003b3;-producers among CD4<sup>+</sup>\\n<bold>(M)</bold> and CD8<sup>+</sup> T-cells <bold>(N)</bold> in Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 7 HDs determined from flow cytometric data as shown in <bold>(G)</bold> at d21. <bold>(O,P)</bold> To mimic the situation after infusion, Rapa was withdrawn and TCPs were cultivated long-term until d49. Proportions of CMV-specific IFN&#x003b3;-producers among CD4<sup>+</sup>\\n<bold>(O)</bold> and CD8<sup>+</sup> T-cells <bold>(P)</bold> in TCPs withdrawn from Rapa (red) and untreated TCPs (blue) of <italic>n</italic> = 6 HDs determined from flow cytometric data as shown in <bold>(G)</bold> at d49. For all graphs normal distribution of data points was tested with Kolmogorov-Smirnov test and paired <italic>t</italic>-test was used to determine significance in normally distributed samples or Wilcoxon\\'s matched-pairs signed rank test in not normally distributed samples, respectively. <italic>P</italic>-values below 0.05 are indicated by * and defined to be significant.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0001\"/></fig></sec><sec><title>Functional Tests, Phenotyping, Flow Cytometry, and Sorting</title><p>Expanded T-cells were analyzed for effector functions by their ability to recognize antigen-loaded target cells, which consisted of autologous lymphoblastoid B-cell lines (LCLs), transformed with B95-8 EBV and by specific production of cytokines. LCLs were generated as described previously (<xref rid=\"B43\" ref-type=\"bibr\">43</xref>).</p><p>For CMV-specific stimulation of expanded TCPs for detecting intracellular effector cytokine production, CMV<sub>pp65/IE&#x02212;1</sub> peptide pool-loaded LCLs were added to cultured T-cells at a ratio of 1:10 and incubated for 6 h. Un-pulsed LCLs served as unstimulated control. For effector cytokine detection <italic>ex vivo</italic>, PBMCs were stimulated with 1 &#x003bc;g/ml overlapping CMV<sub>pp65/IE1</sub> peptide pools <italic>ex vivo</italic> for 14 h. After 1 h, 2 &#x003bc;g/ml Brefeldin A (BFA, Sigma-Aldrich) were added to the stimulation to allow for intracellular capture of cytokines.</p><p>To induce apoptosis, 1 &#x003bc;g/ml of LEAF-purified Fas-activating antibody (EOS9.1; BioLegend) was added to cultures for 16 h. To determine survival, LIVE/DEAD&#x000ae; Fixable Blue Dead Cell Stain (Invitrogen) and Annexin V (BioLegend) were added.</p><p>For determination of killing capacity, autologous LCLs were pulsed with 2 &#x003bc;g/ml CMV<sub>pp65/IE&#x02212;1</sub> peptide pools, whereas unpulsed allogenic LCLs were used as non-target controls. Targets were labeled with 10 mM Carboxyfluorescein-diacetate-succinimidyl-ester (Sigma-Aldrich) and non-target controls with 5 mM CellTrace&#x02122; Far Red (Invitrogen). Cells were co-cultured at a T-cell/target-cell ratio of 10:1 for 14 h. Samples were analyzed using a LSR II Fortessa flow cytometer. Samples without T-cells, containing only LCLs, served as an internal control and reference for calculation of the killing capacity. For analysis, we gated on LIVE/DEAD&#x000ae; Fixable Blue Dead Cell Stain-negative cells and calculated ratios of target to non-target cells as described previously (<xref rid=\"B44\" ref-type=\"bibr\">44</xref>, <xref rid=\"B45\" ref-type=\"bibr\">45</xref>).</p><p>To define memory subsets, T-cells were stained extracellularly for surface markers CCR7 (G043H7), CD45RA (HI100), CD45RO (UCHL1), CD62L (DREG-56; eBioscience), and CD95 (DX2). Subsequently, cells were permeabilized and fixed with Foxp3/Transcription Factor Staining Buffer Set (eBioscience) and stained intracellularly for CD3 (OKT3), CD4 (SK3), and CD8 (RPA-T8), IFN&#x003b3; (4S.B3, eBioscience), TNF&#x003b1; (MAb11), and Granzyme B (GZB) (GB11, BD Pharmingen). Cells were analyzed on a LSR II Fortessa flow cytometer using FlowJo Version 10 software (Tree Star). Lymphocytes were gated based on the FSC <italic>vs</italic>. SSC profile and subsequently gated on FSC (height) <italic>vs</italic>. FSC to exclude doublets.</p><p>For evaluating of T-cell subsets on transcriptome level, T-cell subsets were sorted from PBMCs from <italic>n</italic> = 3 HDs\\' buffy coats (DRK) at d0 or derived TCPs at d18 based on the expression of CD3, CD45RA, and CCR7 by the Core Facility Flow Cytometry of the BCRT using a FACS Aria II Calliope (BD).</p><p>All antibodies were purchased from BioLegend, unless indicated otherwise.</p></sec><sec><title>Metabolic Analysis</title><p>Extracellular acidification rate (ECAR) and oxygen consumption rate (OCR) were analyzed using a Seahorse-XFe96-Analyzer following the manufacturer\\'s instructions for non-adherent cells including immobilization of cells with Cell-Tak (Corning). Assay medium consisted of Dulbecco\\'s Modified Eagle\\'s Medium D5030 (Sigma) supplemented with 3 g/l D-glucose (Roth) and 300 mg/ml L-glutamine (Gibco) and was sterile-filtered. For T-cell activation, 0.5 &#x003bc;g/ml of CMV <sub>pp65/IE&#x02212;1</sub> peptide pools were added to the microwells relying on reciprocal antigen-presentation of T-cells 0.5 h before the measurement.</p></sec><sec><title>RNA Sequencing and Bioinformatics Analysis</title><p>RNA was isolated using an All-Prep DNA/RNA Kit (Qiagen) following the manufacturer\\'s instructions. RNA samples were sent to the Deep Sequencing Core Facility in G&#x000f6;ttingen, where samples were prepared using TrueSeq Kits (Illumina) and HiSeq_4000 performing 50 million reads/sample.</p><p>Fastq-files were quality checked with FastQC (Babraham Bioinformatics) and trimmed for residual adapter sequences. Reads were aligned to the GRCh38 human genome using TopHat<sup>R</sup> (2.1.0&#x02013;Johns Hopkins University, Center for Computational Biology) and Bowtie2 (<xref rid=\"B46\" ref-type=\"bibr\">46</xref>). Counts per gene were determined as sum of all reads mapped within a gene region. Principal component (PC) analysis was performed in R (<xref rid=\"B47\" ref-type=\"bibr\">47</xref>) using the 1,000 top-variable genes within the data set. Differentially expressed genes were identified using negative binomial distributions as implemented in the DESeq2 package (<xref rid=\"B48\" ref-type=\"bibr\">48</xref>) in R. False discovery rates (FDR) were calculated to adjust <italic>p</italic>-values for multiple testing and FDR-values below 0.05 were considered as significant. Expression data for differentially expressed genes were variance-stabilized transformed and scaled prior to visualization in heat maps. RNA sequencing data are available at the GEO platform with the accession number GSE129196.</p></sec><sec><title>T-Cell Receptor Sequencing</title><p>For sequencing of T-cell receptors (TCRs) to determine the clonality of TCPs, DNA was isolated using an All-Prep DNA/RNA Kit (Qiagen) following the manufacturer\\'s instructions. TCR&#x003b2; sequencing was performed using a hsTCRb Kit (Adaptive Biotechnologies) following the manufacturer\\'s instructions and analyzed with the corresponding ImmunoSEQ-Analyzer 3.0 software. Briefly, the most variable complementary-determining region 3 (CDR3), spanning the recombination site of V-D-J recombinations of TCR &#x003b2;-chains was sequenced. Productive rearrangements were regarded as unique in-frame nucleotide sequences without stop codon, leading to a functional TCR. Productive frequency means the individual frequency of a specific productive rearrangement (clone) among all productive rearrangements. Clonality was calculated based on productive entropy normalized to the total number of productive rearrangements. Sample overlap was investigated using the Morisita index considering unique clones, individual frequencies of clones and the probability of a common origin of two samples. TCR sequencing data is accessible at the ImmuneACCESS platform <ext-link ext-link-type=\"uri\" xlink:href=\"http://adaptivebiotech.com/pub/amini-2019-frontimmunol\">http://adaptivebiotech.com/pub/amini-2019-frontimmunol</ext-link> (Adaptive Biotechnologies).</p></sec><sec><title>Statistical Analysis and Calculations</title><p>Graph Pad Prism version 7 was used for graph generation. To test for normal Gaussian distribution Kolmogorov-Smirnov test was performed. If data were normally distributed, Student\\'s paired or unpaired <italic>t</italic>-test were employed for analysis. If data were not normally distributed, Wilcoxon\\'s matched pairs test was applied to paired samples and Man-Whitney\\'s test to unpaired samples. All tests were two-tailed. Probability (p) values of &#x02264;0.05 were considered statistically significant and significance is denoted as follows: <sup>*</sup> = <italic>p</italic> &#x0003c; 0.05. Correlation analysis was assessed by Pearson\\'s correlation coefficients for normally distributed data or non-parametric Spearman\\'s rank correlation. Fold expansion expresses the manually counted cell count (Neubauer\\'s counting chamber) excluding dead cells by Trypan blue staining (Sigma-Aldrich) at the day indicated divided by the initially seeded cell amount from the positive fraction of the IFN&#x003b3; Secretion Cell Enrichment Assay.</p><p>All datasets are available upon reasonable request.</p></sec></sec><sec sec-type=\"results\" id=\"s3\"><title>Results</title><p>In order to prepare our approach for clinical translation, we deeply characterized functionality, stability and distinct molecular, metabolic and transcriptional properties of our next-generation Rapa-TCP, for which we applied mTOR inhibition by Rapamycin to enrich for CD4<sup>+</sup> T-cells and CD4<sup>+</sup>/CD8<sup>+</sup> T<sub>CM</sub> (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>). First, we addressed the question whether we can reproduce our findings and properties published for supplementation of a certain cytokine regiment, IL-2/IL-7, with a regiment commonly used for GMP applications by many groups, IL-7/IL-15, which was previously shown to support generation of T<sub>CM</sub> (<xref rid=\"B49\" ref-type=\"bibr\">49</xref>).</p><sec><title>Supplementation of IL-7/IL-15 Does Not Alter Rapamycin-Mediated Effects in TCPs</title><p>To investigate potential differences in the effects of Rapamycin administration dependent on the cytokine regiment supplemented, we expanded CMV-specific T-cells in the presence of different cytokine combinations, namely IL-7/IL-15 (<xref ref-type=\"fig\" rid=\"F1\">Figure 1A</xref>) and IL-2/IL-7 (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>) (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1A</xref>). The expansion rates of antigen-reactive T-cells were sufficient considering cell numbers used in a pilot study (<xref rid=\"B18\" ref-type=\"bibr\">18</xref>), although Rapamycin significantly reduced expansion in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figure 1B</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1B</xref>). Overall, different cytokines did not alter the beneficial effects of Rapamycin treatment (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>): Rapamycin significantly increased CD4/CD8 ratio in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1C,D</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1C</xref>) and significantly increased proportions of CD4<sup>+</sup> and CD8<sup>+</sup> T<sub>CM</sub> in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1E,F</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1D,E</xref>). Furthermore, Rapamycin increased proportions of Interferon-&#x003b3; (IFN&#x003b3;)-producing CD4<sup>+</sup> and CD8<sup>+</sup> T-cells upon exposure to CMV-specific peptides loaded onto autologous lymphoblastic cell lines (LCLs) in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1G,H</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1F,G</xref>). These data confirm the robustness of beneficial effects of mTOR inhibition using Rapamycin for TCP composition in the case of supplementing commonly used IL-7/IL-15 for expansion of TCPs.</p></sec><sec><title>Expansion Rates of Rapa-TCPs Recover Later During Culture</title><p>SOT patients often suffer from lymphopenia, which reduces the amount of PBMC, <italic>i.e</italic>., the starting material, for TCP generation and their medication can impact the functionality of T-cells (<xref rid=\"B50\" ref-type=\"bibr\">50</xref>). Thus, TCP manufacture from patient material may require longer <italic>in vitro</italic> expansion periods of up to 21 days to achieve sufficient cell numbers for successful AVTT. To assess the stability of TCPs after a longer period of expansion, we determined phenotype and functionality of TCPs after extended expansion on d21 in IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1I,N</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1H,M</xref>). Interestingly, Rapa-TCPs recovered, yet even exceeded expansion of untreated TCPs in the 3rd week of expansion (d14&#x02013;d21) (<xref ref-type=\"fig\" rid=\"F1\">Figure 1I</xref>), which was significant in IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1H</xref>). CD4/CD8 ratios remained significantly higher in Rapa-TCPs at d21 (<xref ref-type=\"fig\" rid=\"F1\">Figure 1J</xref>), but IL-2/IL-7-expanded TCPs showed significantly higher CD4/CD8 ratios than IL-7/IL-15-expanded TCPs at d21 (<xref ref-type=\"fig\" rid=\"F1\">Figure 1J</xref>\\n<italic>vs</italic>. <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1I</xref>). During expansion, T<sub>CM</sub> differentiated and the enrichment of T<sub>CM</sub> proportions upon Rapamycin-treatment lost significance in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1K,L</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1J,K</xref>).</p></sec><sec><title>Antigen Encounter Decreases Expansion Rates, but Promotes Less Differentiated Cells</title><p>We further mimicked the scenario happening once the TCPs are injected into a patient coping with CMV viremia in an <italic>in vitro</italic> model. Therefore, we modeled the situation of high antigen load by re-stimulation with CD3-depleted PBMCs pulsed with CMV-specific peptides: CMV-specific re-stimulation significantly reduced expansion rates in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figure 1I</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1H</xref>). Re-stimulation did not influence CD4/CD8 ratios in IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figure 1J</xref>) neither IL-2/IL-7-expanded Rapa-TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1I</xref>), but significantly decreased CD4/CD8 ratios in IL-7/IL-15-expanded untreated TCPs (<xref ref-type=\"fig\" rid=\"F1\">Figure 1J</xref>). Remarkably, re-stimulation significantly augmented the proportions of CD4<sup>+</sup> and CD8<sup>+</sup> T<sub>CM</sub> in both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1K,L</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1J,K</xref>). However, re-stimulation decreased the proportion of CD4<sup>+</sup> and CD8<sup>+</sup> IFN&#x003b3;-producers (<xref ref-type=\"fig\" rid=\"F1\">Figures 1M,N</xref>), which was statistically significant in IL-2/IL-7-expanded Rapa-TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1L&#x02013;M</xref>).</p></sec><sec><title>Rapamycin Preserves Superior Capacity for IFN&#x003b3; Production</title><p>Importantly, IL-2/IL-7-expanded Rapa-TCPs showed significantly higher proportions of IFN&#x003b3;-producers among CD4<sup>+</sup> T-cells at d14 and d21 (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1F,L</xref>) and CD8<sup>+</sup> T-cells at d21 (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1M</xref>) compared to untreated TCPs illustrating improved functionality. Because TCPs are deprived of Rapamycin and exposed to antigen once injected, we analyzed samples in which we withdrew Rapamycin and re-stimulated with CMV peptide-loaded CD3-depleted PBMCs on d14 (<xref ref-type=\"fig\" rid=\"F1\">Figure 1A</xref>, <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S1A</xref>). Interestingly, once treated with Rapamycin during the first 2 weeks of culture, both IL-7/IL-15- (<xref ref-type=\"fig\" rid=\"F1\">Figures 1O,P</xref>) and IL-2/IL-7-expanded TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S1N,O</xref>) comprised more CD4<sup>+</sup> and CD8<sup>+</sup> IFN&#x003b3;-producers continuously until d49 of culture.</p></sec><sec><title>Rapamycin Enhances Survival of T-Cells</title><p>Longevity is a crucial prerequisite for long-term efficacy of adoptively transferred TCPs in patients. Based on findings in B-cell lymphoma cell lines, we hypothesized that Rapamycin treatment may increase viability of T-cells (<xref rid=\"B51\" ref-type=\"bibr\">51</xref>). Thus, we analyzed overall survival of T-cells in TCPs. Strikingly, we found significantly higher proportions of living T-cells in TCPs treated with Rapamycin compared to untreated TCPs (<xref ref-type=\"fig\" rid=\"F2\">Figures 2A,B</xref>).</p><fig id=\"F2\" position=\"float\"><label>Figure 2</label><caption><p>Rapamycin promotes survival of T-cells and stabilizes their metabolism. <bold>(A)</bold> Exemplary dot-plots of flow cytometry data regarding live/dead stain and Annexin V stain (apoptosis) gated on lymphocytes singlets CD3<sup>+</sup> T-cells. Living T-cells are defined by double negative staining for Annexin V and live/dead stain in Rapa- (upper panel) and untreated TCPs (lower panel). Samples in the right panel were treated with 1 &#x003bc;g/ml activating antibody against Fas (CD95) to induce apoptosis. <bold>(B)</bold> Proportions of living T-cells in Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 8 HDs identified as shown in <bold>(A)</bold> at d21. <bold>(C)</bold> Proportions of living T-cells in Rapa- (red) and untreated TCPs (blue) of n = 8 HDs incubated with Fas-activating antibody identified as shown in <bold>(A)</bold> at d21. <bold>(D)</bold> Exemplary histograms of fluorescence intensity of Bcl-2 in CD4<sup>+</sup> (upper panel) and CD8<sup>+</sup> T-cells (lower panel) of untreated (blue) and Rapa-TCPs (red) acquired by flow cytometry. <bold>(E,F)</bold> MFIs of Bcl-2 in CD4<sup>+</sup>\\n<bold>(E)</bold> and CD8<sup>+</sup> T-cells <bold>(F)</bold> in untreated (blue) and Rapa-TCPs (red) of <italic>n</italic> = 18 HDs. <bold>(G)</bold> Oxygen consumption rare (OCR)/extracellular acidification rate (ECAR) ratio of Rapa- (red) and untreated TCPs (blue) of <italic>n</italic> = 5 HDs determined in a Seahorse assay. For stimulation (pastel colors) CMV<sub>IE&#x02212;1/pp65</sub> peptide pools were added to TCPs relying on mutual presentation of peptides by T-cells from the TCP. For all graphs normal distribution of data points was tested with Kolmogorov-Smirnov test and paired <italic>t</italic>-test was used to determine significance. <italic>P</italic>-values below 0.05 are indicated by * and defined to be significant.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0002\"/></fig><p>Immunosuppressant regiments including Tacrolimus, which are commonly used in SOT, are reported to sensitize T-cells to programmed cell death (<xref rid=\"B52\" ref-type=\"bibr\">52</xref>). Hence, we investigated the TCPs\\' sensitivity to apoptosis employing induction of the death receptor pathway by agonistic Fas-specific antibody to identify differences between untreated and Rapa-TCPs. We recorded partial resistance to Fas-induced apoptosis in Rapa-TCPs, while untreated TCPs were more sensitive to Fas-induced apoptosis (<xref ref-type=\"fig\" rid=\"F2\">Figures 2A,C</xref>). The anti-apoptotic effect of Rapamycin observed in B-cell lymphoma lines is reported to depend on upregulation of Bcl-2 on protein level (<xref rid=\"B51\" ref-type=\"bibr\">51</xref>). Hence, we assessed (<xref ref-type=\"fig\" rid=\"F2\">Figure 2D</xref>) the mean fluorescence intensity (MFI) of Bcl-2 in TCPs and found significantly higher MFIs in CD4<sup>+</sup> and CD8<sup>+</sup> T-cells of Rapa-TCPs compared to untreated TCPs (<xref ref-type=\"fig\" rid=\"F2\">Figures 2E,F</xref>). The findings regarding viability and resistance to apoptosis suggest an increased fitness of T-cells in Rapa-TCPs implying improved long-term survival <italic>in vivo</italic>.</p></sec><sec><title>Rapamycin Stabilizes T-Cell Metabolism Upon Activation</title><p>Characteristically, memory and effector T-cells are distinguished by differences in metabolic activities (<xref rid=\"B53\" ref-type=\"bibr\">53</xref>). Thus, we investigated the ratio of fatty acid oxidation to glycolysis defined by the ratio of oxygen consumption to ECAR (OCR/ECAR). In fact, we detected significantly lower OCR/ECAR ratios in Rapa-TCPs compared to untreated TCPs (<xref ref-type=\"fig\" rid=\"F2\">Figure 2G</xref>). Moreover, upon activation with CMV-specific peptides, the metabolism of Rapa-TCPs proofed to be more stable, whereas the OCR/ECAR ratio was significantly decreased in untreated TCPs (<xref ref-type=\"fig\" rid=\"F2\">Figure 2G</xref>).</p></sec><sec><title>Distinct Effects of Rapamycin on Isolated T-Cell Memory Subsets</title><p>Distinct memory T-cell subsets were reported to have defined properties and are not equally suited for long-term regeneration of T-cell immunity (<xref rid=\"B26\" ref-type=\"bibr\">26</xref>&#x02013;<xref rid=\"B33\" ref-type=\"bibr\">33</xref>). To assess the effect of Rapamycin on distinct CMV-specific memory T-cell subsets, we performed fluorescently activated cell sorting (FACS) for CCR7<sup>+</sup>CD45RA<sup>&#x02212;</sup> T<sub>CM</sub>, CCR7<sup>&#x02212;</sup>CD45RA<sup>&#x02212;</sup> T<sub>EM</sub>, and CCR7<sup>&#x02212;</sup>CD45RA<sup>+</sup> T<sub>EMRA</sub> based on their differential expression of CD45RA and CCR7 <italic>ex vivo</italic> and subsequently performed IFN&#x003b3;-secretion assays of sorted subsets to isolate CMV-specific T-cells of these particular subsets (<xref ref-type=\"fig\" rid=\"F3\">Figure 3A</xref>). The CMV-reactive T-cells of each memory T-cell subset were expanded in the presence or absence of Rapamycin and analyzed after 3 weeks of culture. Interestingly, Rapamycin significantly reduced expansion in T<sub>EM</sub>- and T<sub>EMRA</sub>-derived cultures, which was not significant in T<sub>CM</sub>-derived cultures (<xref ref-type=\"fig\" rid=\"F3\">Figure 3B</xref>), indicating preferential expansion of T<sub>CM</sub> and implying reduced sensitivity to anti-proliferative effects of Rapamycin.</p><fig id=\"F3\" position=\"float\"><label>Figure 3</label><caption><p>Influence of Rapamycin on different T-cell memory subsets. <bold>(A)</bold> Schematic experimental setup: T<sub>CM</sub>, T<sub>EM</sub>, and T<sub>EMRA</sub> were sorted out of lymphocytes singlets CD3<sup>+</sup> T-cells according to expression of CCR7 and CD45RA, CMV-reactive T-cells were isolated from each subset using an IFN&#x003b3; secretion assay and CMV-reactive T-cells from each subset were cultured with (Rapa) and without Rapamycin (w/o). Exemplary dot plots of flow cytometry data of sorted subsets of one HD and respective positive fractions of the IFN&#x003b3; secretion assay are shown. <bold>(B)</bold> Expansion rates of the indicated subsets in the presence of (red, Rapa) and absence of Rapamycin (blue, w/o) calculated from total cell numbers at d21 divided by the seeded cell number. <bold>(C,D)</bold> Proportions of CD4<sup>+</sup>\\n<bold>(C)</bold> and CD8<sup>+</sup> CD45RA<sup>&#x02212;</sup> CCR7<sup>+</sup> T<sub>CM</sub>-like cells <bold>(D)</bold> among Rapa-treated (red) and untreated (blue) cultures of indicated subsets determined from flow cytometric data at d21. <bold>(E,F)</bold> MFIs of Bcl-2 in CD4<sup>+</sup>\\n<bold>(E)</bold> and CD8<sup>+</sup> T-cells <bold>(F)</bold> in untreated (blue) and Rapa-treated cultures (red) of isolated T-cell subsets determined in flow cytometry. <bold>(G,H)</bold> To detect CMV-specific cytokine production, cultures were stimulated with CMV<sub>IE&#x02212;1/pp65</sub> peptide-loaded autologous LCLs at a ratio of 1:10 for 6 h and BFA was added after 1 h. Proportions of CMV-specific IFN&#x003b3;-producers among CD4<sup>+</sup>\\n<bold>(G)</bold> and CD8<sup>+</sup> T-cells <bold>(H)</bold> in Rapa-treated (red) and untreated (blue) cultures of isolated T-cell subsets determined from flow cytometric data. All graphs contain data from <italic>n</italic> = 6 HDs, normal distribution of data points was tested with Kolmogorov-Smirnov test and paired <italic>t</italic>-test was used to determine significance in normally distributed samples or Wilcoxon\\'s matched-pairs signed rank test in not normally distributed samples, respectively. <italic>P</italic>-values below 0.05 are indicated by * and defined to be significant.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0003\"/></fig><p>Moreover, Rapamycin prevented a significant proportion of CD4<sup>+</sup> and CD8<sup>+</sup> T<sub>CM</sub> from differentiation into late-stage memory/effector T-cells compared to control cultures (<xref ref-type=\"fig\" rid=\"F3\">Figures 3C,D</xref>). Remarkably, Rapamycin treatment even induced some cells with a T<sub>CM</sub>-like phenotype in cultures derived from T<sub>EM</sub> and T<sub>EMRA</sub> subsets suggesting some &#x0201c;rejuvenation&#x0201d; of late-stage memory cells (<xref ref-type=\"fig\" rid=\"F3\">Figures 3C,D</xref>).</p><p>Notably, Rapamycin increased the MFI of Bcl-2 in CD4<sup>+</sup> and CD8<sup>+</sup> T-cells from T<sub>CM</sub>- and T<sub>EM</sub>-, but not T<sub>EMRA</sub>-derived cultures (<xref ref-type=\"fig\" rid=\"F3\">Figures 3E,F</xref>). Interestingly, Rapamycin-treated T<sub>EM</sub>- and T<sub>EMRA</sub>-derived cultures contained significantly higher proportions of CD8<sup>+</sup> IFN&#x003b3;-producers than untreated cultures upon CMV-specific re-stimulation (<xref ref-type=\"fig\" rid=\"F3\">Figure 3H</xref>), whereas there were no significant differences in CD4<sup>+</sup> IFN&#x003b3;-producers (<xref ref-type=\"fig\" rid=\"F3\">Figure 3G</xref>). Overall, Rapamycin conferred distinct effects on different T-cell memory subsets, sustaining T<sub>CM</sub> features and counteracting differentiation into late-stage memory/effector T-cells.</p></sec><sec><title>Rapamycin-Treated T-Cell Products Have a Unique Transcriptome Resembling T<sub>CM</sub></title><p>To confirm that expansion of antigen-reactive T-cells under Rapamycin treatment &#x0201c;freezes&#x0201d; an early memory T-cell stage, we tried to extend our analysis on transcriptome level by RNA-sequencing using next-generation sequencing (NGS) of untreated and Rapa-TCPs at d21 (<xref ref-type=\"fig\" rid=\"F4\">Figure 4A</xref>). The RNA expression data revealed a total of 146 differentially expressed genes between Rapa-TCPs and untreated TCPs (<xref ref-type=\"fig\" rid=\"F4\">Figure 4A</xref>). Many of these relate to TCP performance (<xref ref-type=\"fig\" rid=\"F4\">Figure 4B</xref>, <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S1</xref>). We reviewed the literature and various databases to identify T-cell associated processes (<xref ref-type=\"fig\" rid=\"F4\">Figure 4B</xref>) and to estimate the relevance of the differentially regulated genes (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S1</xref>). With reference to previously published data, 84% of the genes identified as potentially relevant for TCP potency and longevity <italic>in vivo</italic> were regulated in a beneficial manner in Rapa-TCPs. Among these differentially expressed genes, we identified increased expression of T<sub>CM</sub> markers such as <italic>CCR7</italic> and <italic>PIM2</italic> (<xref rid=\"B54\" ref-type=\"bibr\">54</xref>), increased expression of <italic>TERT</italic>, which induces self-renewal capacity and increases the proliferative potential of human T-cells (<xref rid=\"B55\" ref-type=\"bibr\">55</xref>), and <italic>IL7R</italic>, which is reported to be a marker for persisting and protective CD8<sup>+</sup> memory T-cells (<xref rid=\"B56\" ref-type=\"bibr\">56</xref>) in Rapa-TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S1</xref>). Furthermore, our observation of sustained IFN&#x003b3; production in Rapa-TCPs is in line with the findings of increased expression of <italic>IL-13</italic> regulating IFN&#x003b3; synthesis (<xref rid=\"B57\" ref-type=\"bibr\">57</xref>), <italic>DRD2</italic> inducing IFN&#x003b3; production (<xref rid=\"B58\" ref-type=\"bibr\">58</xref>) and <italic>TNFRSF11A</italic>, which increases IFN&#x003b3; secretion upon binding its ligand (<xref rid=\"B59\" ref-type=\"bibr\">59</xref>). In addition, these data are underlined by increased expression of activation enhancing genes, including <italic>e.g., KLF7</italic> (<xref rid=\"B60\" ref-type=\"bibr\">60</xref>)<italic>, RGMB</italic> (<xref rid=\"B61\" ref-type=\"bibr\">61</xref>), and <italic>TNFRSF19</italic> (<xref rid=\"B62\" ref-type=\"bibr\">62</xref>) in Rapa-TCPs. Moreover, the fact that anti-apoptotic Bcl-2 is upregulated on protein level may be supported by increased expression of <italic>MYB</italic>, which exerts its anti-apoptotic activity <italic>via</italic> Bcl-2 (<xref rid=\"B63\" ref-type=\"bibr\">63</xref>). However, also many other genes inhibiting apoptosis, such as <italic>e.g., BEX2</italic> (<xref rid=\"B64\" ref-type=\"bibr\">64</xref>) and <italic>SIX1</italic> (<xref rid=\"B65\" ref-type=\"bibr\">65</xref>), show significantly higher expression in Rapa-TCPs compared to untreated TCPs. The metabolic data of increased glycolysis in Rapa-TCPs are in line with increased expression of <italic>EPAS1</italic> (<xref rid=\"B66\" ref-type=\"bibr\">66</xref>), however, also <italic>CHDH</italic>, a gene involved in fatty acid oxidation, (<xref rid=\"B67\" ref-type=\"bibr\">67</xref>) is higher expressed in Rapa-TCPs than in untreated TCPs. See <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S1</xref> for a complete view and annotation of the genes differentially expressed in untreated and Rapa-TCPs and their functions potentially relevant for TCP efficacy <italic>in vivo</italic>. Of note, the T<sub>CM</sub>-like cells sorted from untreated TCPs on day 18 of culture (<xref ref-type=\"fig\" rid=\"F4\">Figure 4C</xref>) clustered with Rapa-TCPs (<xref ref-type=\"fig\" rid=\"F4\">Figure 4D</xref>) regarding the differentially expressed genes identified in <xref ref-type=\"fig\" rid=\"F4\">Figure 4A</xref>. Interestingly, principle component (PC) analysis revealed some components of the transcriptome of Rapa-TCPs to be more similar to <italic>ex vivo</italic>-sorted T<sub>CM</sub> than T<sub>EM</sub> (dimension PC2; <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S2</xref>).</p><fig id=\"F4\" position=\"float\"><label>Figure 4</label><caption><p>Rapamycin-treated TCPs have a unique transcriptome resembling T<sub>CM</sub> and are clonally more diverse. RNA expression data were acquired by RNA sequencing and samples had to pass a quality control to be included in the analysis. <bold>(A)</bold> Expression heat map of differentially expressed genes between untreated and Rapa-TCPs generated from fresh blood of n = 3 HDs at d21. <bold>(B)</bold> Processes allocated to the differentially expressed genes based on the literature (see <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S1</xref> for details). <bold>(C)</bold> Exemplary dot plots of CCR7/CD45RA expression of untreated and Rapa-TCPs and the respective T<sub>CM</sub>-like cells sorted on d18 of culture. <bold>(D)</bold> Expression of differentially expressed genes between Rapa- and untreated TCPs from fresh blood of <italic>n</italic> = 3 HDs at d21, buffy coats of <italic>n</italic> = 3 HDs at d18 and T<sub>CM</sub> sorted from Rapa- and untreated TCPs generated from buffy coats of the same <italic>n</italic> = 3 HDs at d18 including clustering. Samples not included in the graph were discarded due to failure at the quality threshold. <bold>(E)</bold> Clonality of Rapa- (red) and untreated TCPs (blue) <italic>n</italic> = 3. <bold>(F)</bold> Proportions of represented sequences covered by the top 10 most represented clones in Rapa- (red) and untreated TCPs (blue), <italic>n</italic> = 3 HDs. <bold>(G)</bold> Numbers of productive rearrangements included in in Rapa- (red) and untreated TCPs (blue), <italic>n</italic> = 3 HDs. Data in <bold>(E,F)</bold> were calculated with ImmunoSEQ-Analyzer3.0 software based on TCR&#x003b2; sequencing.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0004\"/></fig></sec><sec><title>Rapamycin-Treated T-Cell Products Show Less Clonal and More Diverse TCR Repertoires</title><p>To estimate the TCR repertoire of our TCPs, we performed TCR&#x003b2; NGS. Notably, TCR&#x003b2; sequencing showed a more diverse clonal composition of Rapa-TCPs compared to untreated TCPs (<xref ref-type=\"fig\" rid=\"F4\">Figure 4E</xref>). Venn diagrams of the total numbers of clones and overlap between Rapa- and untreated TCPs are shown in <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S3A</xref> and the distribution of the top 100 clones is shown in <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S3B</xref>. Correspondingly, the top 10 clones covered around 70 and 90% of the whole TCR&#x003b2; repertoire of Rapa-TCPs and untreated TCPs, respectively (<xref ref-type=\"fig\" rid=\"F4\">Figure 4F</xref>) and Rapa-TCPs contained more different clones than untreated TCPs (<xref ref-type=\"fig\" rid=\"F4\">Figure 4G</xref>). The top 10 shared clones and their respective frequencies in Rapa- and untreated TCPs are shown in <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S3C</xref>. Comparison of unique nucleotide sequences revealed a high clonal overlap between the distinct Rapa- and untreated TCPs generated from the same donor (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S3D</xref>). In contrast, comparison of clonal repertoires between different individuals showed no overlap, confirming the specificity of the findings (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S3D</xref>).</p></sec><sec><title>Onset of Immunosuppression in Patients Does Not Influence the Starting Material for TCPs Regarding T-Cell Differentiation and CMV-Specificity</title><p>As a prerequisite for clinical translation, we aimed at confirming feasibility of Rapa-TCP generation from patient blood and therefore collected samples from end stage kidney disease (ESRD) patients before and after kidney transplantation (KTx). To investigate the influence of immunosuppression on the starting material for TCPs, 7 paired samples from ESRD patients before/after KTx were analyzed <italic>ex vivo</italic>. All KTx recipients received standard immunosuppression (characteristics in <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S2</xref>, <italic>pre/post-</italic>Tx paired samples highlighted in gray). T-cells were divided into five differentiation subsets: CCR7<sup>+</sup>CD45RA<sup>+</sup>CD95<sup>&#x02212;</sup> T<sub>N</sub> (na&#x000ef;ve T-cells), CCR7<sup>+</sup>CD45RA<sup>+</sup>CD62L<sup>+</sup>CD45RO<sup>&#x02212;</sup>CD95<sup>+</sup> T<sub>SCM</sub>, CCR7<sup>+</sup>CD45RA<sup>&#x02212;</sup> T<sub>CM</sub>, CCR7<sup>&#x02212;</sup>CD45RA<sup>&#x02212;</sup> T<sub>EM</sub>, and CCR7<sup>&#x02212;</sup>CD45RA<sup>+</sup> T<sub>EMRA</sub> (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S4A,B</xref>) revealing no substantial differences between CD4<sup>+</sup> and CD8<sup>+</sup> memory T-cell subset distributions of paired patient samples before/after KTx (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S4C,D</xref>). To assess the phenotypic and functional characteristics of CMV-specific T-cells, PBMCs were stimulated with CMV<sub>pp65/IE1</sub> peptides showing markedly higher frequencies of CMV-responsive T-cells among CD8<sup>+</sup> compared to CD4<sup>+</sup> T-cells (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S4E,F</xref>). However, frequencies of CMV-responsive T-cells were similar before and after KTx (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S4G,H</xref>) and T-cell memory subsets were comparable among CMV-responsive T-cells before and after KTx (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S4I,J</xref>).</p></sec><sec><title>Manufacturing Rapa-TCPs Is Feasible Before and After Transplantation</title><p>We assessed feasibility of TCP generation from patient material collected before and after KTx. Manufacture of untreated and Rapa-TCPs was successful with respect to yield (<xref ref-type=\"fig\" rid=\"F5\">Figure 5A</xref>), although untreated CMV-specific TCPs resulted in higher yields, which was statistically significant in TCPs generated after KTx (<xref ref-type=\"fig\" rid=\"F5\">Figure 5A</xref>). Rapa-TCPs showed beneficial, higher CD4<sup>+</sup> T-cell proportions compared to untreated TCPs, which was statistically significant in TCPs generated before KTx (<xref ref-type=\"fig\" rid=\"F5\">Figure 5B</xref>). Compared to untreated TCPs, Rapa-TCPs showed higher proportions of CD4<sup>+</sup> and CD8<sup>+</sup> T<sub>CM</sub> in KTx recipient-derived TCPs, which was significant in TCPs generated after KTx (<xref ref-type=\"fig\" rid=\"F5\">Figures 5C&#x02013;E</xref>). Upon re-stimulation with CMV-peptide-loaded autologous LCLs, we found enhanced IFN&#x003b3;-producers (<xref ref-type=\"fig\" rid=\"F5\">Figure 5F</xref>) and IFN&#x003b3;/granzyme B (GZB)-double-producers (<xref ref-type=\"fig\" rid=\"F5\">Figure 5I</xref>) among Rapamycin-treated CD8<sup>+</sup> T-cells (<xref ref-type=\"fig\" rid=\"F5\">Figures 5H,K</xref>), whereas their proportions among CD4<sup>+</sup> T-cells remained stable (<xref ref-type=\"fig\" rid=\"F5\">Figures 5G,J</xref>). We recorded comparable frequencies of IFN&#x003b3;-producers and IFN&#x003b3;/GZB-double-producers among CD4<sup>+</sup> T-cells in TCPs generated before and after KTx (<xref ref-type=\"fig\" rid=\"F5\">Figures 5G,J</xref>), whereas both increased among CD8<sup>+</sup> T-cells in TCPs generated after KTx (<xref ref-type=\"fig\" rid=\"F5\">Figures 5H,K</xref>). Remarkably, Rapamycin increased the frequency of T<sub>CM</sub> among IFN&#x003b3;-producers (<xref ref-type=\"fig\" rid=\"F5\">Figures 5L,M</xref>), which was statistically significant among CD4<sup>+</sup> IFN&#x003b3;-producers in TCPs generated after KTx (<xref ref-type=\"fig\" rid=\"F5\">Figure 5l</xref>). Regarding CMV-specific cytotoxic effects, Rapa-TCPs were as effective as untreated TCPs and TCPs generated <italic>pre</italic>- and <italic>post</italic>-KTx neither showed any differences (<xref ref-type=\"fig\" rid=\"F5\">Figure 5N</xref>).</p><fig id=\"F5\" position=\"float\"><label>Figure 5</label><caption><p>Manufacture of Rapa-TCPs is feasible before/after transplantation. <italic>N</italic> = 7 paired samples of untreated (w/o, blue) and Rapa-TCPs (red) from the same patients before (pre; pastel colors) and a few weeks after KTx (<italic>post</italic>). <bold>(A)</bold> Yield of TCPs = total cell number derived from 20 ml of patient blood on d21. <bold>(B)</bold> CD4/CD8 ratio of TCPs determined by multicolor flow cytometry on d14. <bold>(C)</bold> Exemplary dot plots of one patient\\'s untreated (right) and Rapa-TCPs (left) comparing subset distributions of CD3<sup>+</sup>CD4<sup>+</sup> (upper panel) and CD3<sup>+</sup>CD8<sup>+</sup> (lower panel) T-cells according to CCR7 and CD45RA expression on d14. <bold>(D,E)</bold> Proportions of CD45RA<sup>&#x02212;</sup> CCR7 <sup>+</sup> T<sub>CM</sub> among CD4<sup>+</sup>\\n<bold>(D)</bold> and CD8<sup>+</sup> T-cells <bold>(E)</bold> in TCPs on d14 as determined per gating strategy shown in <bold>(C)</bold>. <bold>(F)</bold> Exemplary dot plots of one patient comparing CD3<sup>+</sup>CD4<sup>+</sup> (left panel) and CD3<sup>+</sup>CD8<sup>+</sup>(right panel) IFN&#x003b3;- and TNF&#x003b1;-producers in Rapa- (red) and untreated TCPs (blue) detected by intracellular staining in multicolor flow cytometry after 6 h stimulation with autologous LCLs loaded with CMV<sub>IE&#x02212;1/pp65</sub> peptide pools (gray) or incubation with unloaded autologous LCLs as control (black) and addition of BFA after 1 h on d21. <bold>(G,H)</bold> Summary of background subtracted proportions of CD4<sup>+</sup>\\n<bold>(G)</bold> and CD8<sup>+</sup>\\n<bold>(H)</bold> CMV-reactive IFN&#x003b3;-producing T-cells in Rapa- (red) and untreated TCPs (blue) gated as illustrated in <bold>(F)</bold>. <bold>(I)</bold> Exemplary dot plots of one donor comparing CD4<sup>+</sup> (left panel) and CD8<sup>+</sup> (right panel) CMV-reactive IFN&#x003b3;- and GZB-producers in Rapa- (red) and untreated TCPs (blue) detected by intracellular staining in multicolor flow cytometry after 6 h stimulation with autologous LCLs loaded CMV<sub>IE&#x02212;1</sub> and CMV<sub>pp65</sub> peptide pools (gray), incubation with unloaded autologous LCLs as control (black) and addition of BFA after 1 h on d21. <bold>(J,K)</bold> Summary of background subtracted proportions of CD4<sup>+</sup>\\n<bold>(J)</bold> and CD8<sup>+</sup>\\n<bold>(K)</bold> CMV-reactive IFN&#x003b3;/GZB-double-producers in Rapa- (red) and untreated TCPs (blue) gated as illustrated in <bold>(I)</bold>. <bold>(L,M)</bold> Proportions of CD45RA<sup>&#x02212;</sup> CCR7 <sup>+</sup> T<sub>CM</sub> among CMV-reactive IFN&#x003b3;-producing CD4<sup>+</sup>\\n<bold>(L)</bold> and CD8<sup>+</sup> T-cells <bold>(M)</bold>. Gates were applied from gates set for global T-cell subset distribution (see <xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S4</xref>). <bold>(N)</bold> Specific killing of CMV<sub>IE&#x02212;1/pp65</sub> peptide pool loaded autologous LCLs determined by ratio with unloaded allogenic LCLs at a 1:10 ratio with TCPs after incubation overnight. All data tested for normal distribution of data points with Kolmogorov-Smirnov test; significance determined with paired <italic>t</italic>-test if normally distributed or Wilcoxon\\'s matched-pairs signed rank test for not normally distributed samples. <italic>P</italic>-values below 0.05 are indicated by * and defined to be significant.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0005\"/></fig></sec><sec><title>CMV History Affects the Composition of Starting Material From <italic>Post</italic>-KTx Patients</title><p>To determine the influence of the CMV infection status on the T-cell subset composition and function of starting material for TCP generation, 19 CMV seropositive KTx patients (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S2</xref>) and 13 CMV seropositive healthy donors (HDs) were analyzed in parallel. Patients were categorized according to their CMV reactivation status: No recorded (<italic>n</italic> = 9; 5 m/4f), history of (<italic>n</italic> = 6; 2 m/4f) or very recent CMV-DNAemia within 2 weeks before blood collection (<italic>n</italic> = 4; 4 m/0f) (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S2</xref>). The CMV reactivation status had almost no effect on the global CD4<sup>+</sup>/CD8<sup>+</sup> T-cell memory subset distribution (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S5A&#x02013;C,E,F,H&#x02013;J</xref>), except for an increase in proportions of CD8<sup>+</sup> T<sub>SCM</sub> and a decrease in CD4<sup>+</sup> T<sub>EM</sub> in the blood of patients with no record of CMV viremia compared to HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S5D,G</xref>).</p><p>We found CMV-reactive T-cells in all KTx patients and HDs, with markedly higher frequencies among CD8<sup>+</sup> vs. CD4<sup>+</sup> T-cells (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S6A,F</xref>). We did neither observe major differences in the magnitude of the CMV-response between KTx patients and HDs nor between the different groups of KTx patients (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S6A,F</xref>). Of note, proportions of below 0.2% of CMV-responsive T-cells among CD8<sup>+</sup> T-cells occurred in 38.5% of HDs and only 10.5% of patients without recorded CMV-DNAemia (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S6F</xref>).</p><p>The majority of KTx patients showed T<sub>SCM</sub> frequencies &#x0003c; 10% among CMV-reactive CD4<sup>+</sup> T-cells. However, patients with a CMV history or recent CMV-DNAemia presented with significantly elevated CMV-reactive CD4<sup>+</sup> T<sub>SCM</sub> compared to KTx patients with no recorded CMV-DNAemia and HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S6B</xref>). T<sub>CM</sub> proportions among CD4<sup>+</sup> CMV-responsive T-cells showed high inter-individual differences among the patients and HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S6C</xref>). T<sub>EM</sub> proportions among CD4<sup>+</sup> CMV-responsive T-cells were significantly lower in the cohort of KTx patients with a record of CMV-DNAemia compared to patients with no recorded CMV-DNAemia and HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S6D</xref>). The proportions of T<sub>EMRA</sub> among the CD4<sup>+</sup> CMV-responsive T-cells were below 5%, except for three patients with recent or previous CMV-DNAemia, who all received virostatic medication (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S6E</xref>, <xref ref-type=\"supplementary-material\" rid=\"SM1\">Table S2</xref>). We found no significant differences in the memory subset distribution among CMV-reactive CD8<sup>+</sup> T-cells between the different patient groups and HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S6G&#x02013;J</xref>). Notably, we could not detect CMV-responsive CD8<sup>+</sup> T<sub>CM</sub> in the majority of samples (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S6H</xref>).</p></sec><sec><title>Impact of CMV Reactivation State on Manufacture of Untreated and Rapa-TCPs</title><p>To evaluate the quality of untreated and Rapa-TCPs generated from KTx patients under maintenance immunosuppression with distinct CMV reactivation states, CMV<sub>pp65/IE1</sub>-specific T-cells were expanded with or without Rapamycin (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>). We successfully manufactured untreated and Rapa-TCPs from all KTx patients and HDs analyzed, although Rapamycin substantially reduced yields in TCPs from all patients and HDs (<xref ref-type=\"fig\" rid=\"F6\">Figure 6A</xref>). Recent CMV reactivation further significantly reduced yields of untreated and Rapa-TCPs compared to HDs and history of CMV reactivation reduced yields of Rapa-TCPs compared to HDs (<xref ref-type=\"fig\" rid=\"F6\">Figure 6A</xref>). Interestingly, there was an inverse correlation between expansion rate/yield and age in untreated and Rapa-TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S7A,B</xref>). Furthermore, the number of records with CMV-DNAemia correlated inversely with the yield of Rapa-TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S7C</xref>). We found that Rapamycin significantly increased the CD4<sup>+</sup>/CD8<sup>+</sup> T-cell ratio in TCPs of KTx patients without recorded CMV-DNAemia and HDs, which was less pronounced in TCPs of KTx patients with a record of CMV-DNAemia (<xref ref-type=\"fig\" rid=\"F6\">Figure 6B</xref>). Rapa-TCPs showed significantly higher proportions of CD8<sup>+</sup> T<sub>CM</sub> in all groups except the KTx patients with recent CMV DNAemia (<italic>n</italic> = 4) (<xref ref-type=\"fig\" rid=\"F6\">Figure 6D</xref>), while CD4<sup>+</sup> T<sub>CM</sub> were only significantly enriched in TCPs of KTx patients with no recorded CMV-DNAemia and HDs, being less pronounced in TCPs from the other groups (<xref ref-type=\"fig\" rid=\"F6\">Figure 6C</xref>).</p><fig id=\"F6\" position=\"float\"><label>Figure 6</label><caption><p>Impact of CMV history on manufacture of untreated and Rapa-TCPs. Untreated (w/o, blue) and Rapa-TCPs (red) of <italic>n</italic> = 19 patients (9 with so far no recorded CMV viremia; 4 with recent CMV viremia and 6 with a history of CMV viremia)/13 HDs. <bold>(A)</bold> Yield of TCPs = total cell number derived from 20 ml of patient blood on d21. <bold>(B)</bold> CD4/CD8 ratio of TCPs on d14. <bold>(C,D)</bold> Proportions of CD45RA<sup>&#x02212;</sup> CCR7 <sup>+</sup> T<sub>CM</sub> among CD4<sup>+</sup>\\n<bold>(C)</bold> and CD8<sup>+</sup> T-cells <bold>(D)</bold> in TCPs as determined per gating strategy shown in <xref ref-type=\"fig\" rid=\"F5\">Figure 5C</xref> on d14. <bold>(E,F)</bold> Proportions of CMV-reactive CD4<sup>+</sup>\\n<bold>(E)</bold> and CD8<sup>+</sup>\\n<bold>(F)</bold> IFN&#x003b3;-producers detected by intracellular staining in multicolor flow cytometry after 6 h stimulation with autologous LCLs loaded CMV<sub>IE&#x02212;1/pp65</sub> peptide pools at a ratio of 1:10 and addition of BFA after 1 h on d21. Gating strategy is shown in <xref ref-type=\"fig\" rid=\"F5\">Figure 5F</xref>. <bold>(G)</bold> Specific killing of CMV<sub>IE&#x02212;1/pp65</sub> peptide pool loaded autologous LCLs determined by ratio with unloaded allogenic LCLs at 1:10 ratio with TCPs after incubation overnight. All data were tested for normality with Kolmogorov-Smirnov test; significant differences for paired samples determined with paired <italic>t</italic>-test if normally distributed or Wilcoxon\\'s matched-pairs signed rank test and for unpaired samples with unpaired <italic>t</italic>-test if normally distributed or Man Whitney\\'s test. <italic>P</italic>-values below 0.05 are indicated by * and defined to be significant.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0006\"/></fig><p>Upon CMV-specific re-stimulation, we found increased frequencies of CD8<sup>+</sup> IFN&#x003b3;-producers in Rapa-TCPs, which was statistically significant in the group of KTx patients without record of CMV DNAemia (<xref ref-type=\"fig\" rid=\"F6\">Figure 6F</xref>). However, Rapa-TCPs of KTx patients with a history of CMV DNAemia contained significantly lower frequencies of IFN&#x003b3;-producers compared to Rapa-TCPs from HDs (<xref ref-type=\"fig\" rid=\"F6\">Figures 6E,F</xref>). Remarkably, the frequency of IFN&#x003b3;-producers among CD8<sup>+</sup> T-cells, but not CD4<sup>+</sup> T-cells (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8A</xref>), inversely correlated with the time from the last CMV-DNAemia in untreated, but not Rapa-TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8B</xref>). Notably, Rapa-TCPs included higher proportions of T<sub>CM</sub>-like among IFN&#x003b3;-producing CD4<sup>+</sup> and CD8<sup>+</sup> T-cells compared to the corresponding untreated TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S8C,D</xref>). This was significant in IFN&#x003b3;-producing CD4<sup>+</sup> T-cells in TCPs of KTx patients without record of CMV DNAemia and IFN&#x003b3;-producing CD8<sup>+</sup> T-cells in TCPs of KTx patients without record and history of CMV DNAemia (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figures S8C,D</xref>). The T<sub>CM</sub>-like phenotype among IFN&#x003b3;-producing CD4<sup>+</sup> T-cells was significantly more frequent in Rapa-TCPs of KTx patients compared to HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8C</xref>). IFN&#x003b3;-producing CD8<sup>+</sup> T-cells were significantly more frequent in Rapa-TCPs of KTx patients with history of CMV viremia compared to HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8D</xref>).</p><p>In order to characterize functionality, we co-cultured TCPs with CMV-antigen-loaded LCLs for 14 h and killing was analyzed. Untreated and Rapa-TCPs achieved similar target cell lysis (<xref ref-type=\"fig\" rid=\"F6\">Figure 6G</xref>). CMV-specific re-stimulation further characterized up to 65.6% of CD4<sup>+</sup> T-cells to be cytotoxic as defined by GZB/IFN&#x003b3;-double-production (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8E</xref>). Interestingly, Rapa-TCPs of KTx patients with no record of CMV viremia contained significantly more GZB/IFN&#x003b3; double producers among CD8<sup>+</sup> T-cells than their untreated counterparts (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8F</xref>). Rapa-TCPs of KTx recipients with history of CMV viremia contained significantly less GZB/IFN&#x003b3; double producers among CD8<sup>+</sup> T-cells than those of HDs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S8F</xref>).</p></sec><sec><title>Rapa-TCPs From Patients Exhibit Superior Viability After Thawing</title><p>Strikingly, every single Rapa-TCP consistently comprised higher proportions of living T-cells compared to its untreated counterpart (<xref ref-type=\"fig\" rid=\"F7\">Figure 7A</xref>). By convention, TCPs have to be frozen until GMP-compliant quality controls are accomplished and then are thawed directly before infusion into patients. This procedure is a major stress for the TCPs. We froze and thawed TCPs from ESRD/KTx patients and HDs and observed an increased frequency of living T-cells in the Rapa-TCPs, being detectable immediately and even 1 day after thawing and culture (<xref ref-type=\"fig\" rid=\"F7\">Figures 7B,C</xref>). Consistent with the findings from HDs, also Rapa-TCPs of KTx patients showed elevated MFIs of Bcl-2 compared to the untreated TCPs (<xref ref-type=\"supplementary-material\" rid=\"SM1\">Figure S9</xref>).</p><fig id=\"F7\" position=\"float\"><label>Figure 7</label><caption><p>Rapa-TCPs from patients exhibit superior viability before and after thawing. Untreated (w/o, blue) and Rapa-TCPs (red) of <italic>n</italic> = 19 patients (9 with so far no recorded CMV viremia; 4 with recent CMV viremia and 6 with a history of CMV viremia)/13 HDs. <bold>(A)</bold> Proportions of living T-cells determined by positive staining for CD3 and negative staining for live/dead stain and Annexin V. <bold>(B,C)</bold> Untreated (w/o, blue) and Rapa-TCPs (red) from <italic>n</italic> = 4 (paired <italic>pre</italic>- and <italic>post</italic>-KTx samples of patient no. 11; <italic>post</italic>-KTx TCPs of patient no. 9 and two HDs). TCPs were frozen in fetal calf serum substituted with 10 % dimethylsulfoxide and stored in liquid nitrogen. After thawing and two washing steps, proportions of living T-cells defined by positive staining for CD3 and negative staining for live/dead stain and Annexin V were determined immediately (d0, <bold>B</bold>) and after 24 h rest in complete medium in a humidified incubator at 37&#x000b0;C and 5% CO<sub>2</sub> (d1, <bold>C</bold>). All data were tested for normality with Kolmogorov-Smirnov test; significant differences for paired samples determined with paired <italic>t</italic> test if normally distributed or Wilcoxon\\'s matched-pairs signed rank test and for unpaired samples with unpaired <italic>t</italic> test if normally distributed or Man Whitney\\'s test. <italic>P</italic>-values below 0.05 are indicated by * and defined to be significant.</p></caption><graphic xlink:href=\"fimmu-10-01148-g0007\"/></fig><p>In summary, we demonstrate that CMV-specific Rapa-TCPs can be generated irrespective of the employed cytokine regiment, show better viability even after thawing, a stable metabolism, beneficial gene expression and increased clonal diversity. We likewise demonstrate the possibility to generate TCPs from patients in ESRD and <italic>post</italic>-KTx despite maintenance immunosuppression containing similar attributes as from HDs. Further, we illustrate that functional CMV-specific T-cells, the prerequisite for manufacture of CMV-specific TCPs, could be identified in all KTx patients investigated and present successful manufacture of untreated and Rapa-TCPs irrespective of the viral replication history.</p></sec></sec><sec sec-type=\"discussion\" id=\"s4\"><title>Discussion</title><p>The aim of our study was to demonstrate the benefit of the Rapamycin treatment during manufacture of TCPs with improved properties. We demonstrate stability of Rapa-TCPs irrespective of cytokine combinations administered during expansion, sustained IFN&#x003b3; production despite withdrawal of Rapamycin and re-stimulation with viral antigen. Furthermore, we found superior viability and partial resistance to death receptor-induced apoptosis, stable metabolism upon activation, favorable gene expression pattern, and enhanced clonal diversity of Rapa- compared to untreated TCPs. Moreover, we show preferential expansion of T<sub>CM</sub> in the presence of Rapamycin and partial conversion of other T-cell memory subsets to T<sub>CM</sub>-like cells. We demonstrate the feasibility of manufacturing autologous anti-CMV Rapa-TCPs from blood of ESRD patients and KTx recipients with distinct CMV reactivation history. We confirm increased CD4/CD8 ratios and T<sub>CM</sub> proportions in Rapa- compared to untreated TCPs (<xref rid=\"B37\" ref-type=\"bibr\">37</xref>), which are associated with long-term clinical efficacy of adoptively transferred TCPs (<xref rid=\"B29\" ref-type=\"bibr\">29</xref>, <xref rid=\"B30\" ref-type=\"bibr\">30</xref>, <xref rid=\"B40\" ref-type=\"bibr\">40</xref>, <xref rid=\"B41\" ref-type=\"bibr\">41</xref>). In addition, our results imply that generation of CMV-specific TCPs prior to transplantation is possible, however, not favorable. Moreover, Rapa-TCPs from patients contained more viable cells after freezing/thawing compared to untreated TCPs.</p><p>Although the cytokine combination IL-7/IL-15 was reported to preferentially promote T<sub>CM</sub> (<xref rid=\"B49\" ref-type=\"bibr\">49</xref>), we did not reveal substantial differences compared to cultures expanded with IL-2/IL7, suggesting these combinations are equally suited for TCP expansion and do not alter cell intrinsic mTOR-dependent signaling programs. Antigenic re-challenge and concomitant elimination of antigen-loaded APCs decreased T-cell expansion and resulted in T-cell culture contraction. These cultures predominantly contained long-lived memory T-cells. In line with this, among CMV-specific CD4<sup>+</sup> T-cells, we recorded significant decreases in T<sub>EM</sub> and significant increases in T<sub>SCM</sub> in KTx recipients with a record of CMV DNAemia.</p><p>Rapa-TCPs consistently contained more living T-cells than untreated TCPs, even after freezing/thawing. Hence, Rapa-TCPs may entail increased fitness following infusion, as TCPs are frozen until accomplishment of all quality controls in a clinical setting respecting GMP/GCP. As shown previously for B-cell lymphoma lines (<xref rid=\"B51\" ref-type=\"bibr\">51</xref>), Rapamycin increased Bcl-2 on protein level and moreover led to partial resistance to Fas-induced apoptosis. We did not identify altered gene expression of <italic>BCL-2</italic> on mRNA level, however, expression of <italic>MYB</italic>, exerting anti-apoptotic effects <italic>via</italic> Bcl-2 (<xref rid=\"B63\" ref-type=\"bibr\">63</xref>), was upregulated in Rapa-TCPs. Furthermore, the decreased sensitivity to apoptosis of Rapa-TCPs may also be conferred by additional proteins and pathways as we found many other anti-apoptotic genes, such as <italic>BEX2</italic> (<xref rid=\"B64\" ref-type=\"bibr\">64</xref>) and <italic>SIX1</italic> (<xref rid=\"B65\" ref-type=\"bibr\">65</xref>), to be overexpressed in Rapa-TCPs compared to untreated TCPs.</p><p>We recorded higher OCR/ECAR ratios in untreated TCPs, although inhibition of glycolysis is associated with a long-lived memory phenotype (<xref rid=\"B68\" ref-type=\"bibr\">68</xref>) and memory T-cells are reported to preferentially perform fatty acid oxidation (<xref rid=\"B53\" ref-type=\"bibr\">53</xref>). Increased glycolysis in Rapa-TCPs was supported by RNA sequencing data showing increased expression of <italic>EPAS1</italic> (<xref rid=\"B66\" ref-type=\"bibr\">66</xref>), however, also <italic>CHDH</italic>, a gene involved in fatty acid oxidation (<xref rid=\"B67\" ref-type=\"bibr\">67</xref>), showed increased expression in Rapa-TCPs. In fact, glycolysis is reported to allow immediate effector function (<xref rid=\"B69\" ref-type=\"bibr\">69</xref>), which is in line with the Rapa-TCPs\\' enhanced capacity for IFN&#x003b3; production and increased expression of <italic>MAP3K21</italic> allowing a rapid switch from a rested to an activated state (<xref rid=\"B70\" ref-type=\"bibr\">70</xref>). Furthermore, the metabolism of Rapa-TCPs remained more stable upon CMV-specific activation. Memory cell self-renewal might occur at a comparable number to that of effector T-cell generation in Rapa-TCPs, whereas in untreated TCPs, the balance might be extremely skewed toward effector T-cells upon activation, leading to this significant decrease in OCR/ECAR ratio and lack of long-lived memory T-cells.</p><p>Intriguingly, 84% of genes differentially expressed between untreated and Rapa-TCPs, whose function we could allocate to impact TCPs, were regulated toward promotion of an effective and long-lived product. Our data imply a T<sub>CM</sub>-like transcriptome of long-lived poly-functional memory T-cells for Rapa-TCPs (<xref rid=\"B54\" ref-type=\"bibr\">54</xref>&#x02013;<xref rid=\"B56\" ref-type=\"bibr\">56</xref>, <xref rid=\"B66\" ref-type=\"bibr\">66</xref>, <xref rid=\"B71\" ref-type=\"bibr\">71</xref>&#x02013;<xref rid=\"B79\" ref-type=\"bibr\">79</xref>). The fact that Rapa-TCPs show increased clonal diversity may occur due to survival of low frequency clones. Especially, the preferential expansion of T<sub>CM</sub> in Rapa-TCPs may contribute to increased clonal diversity, as this subset was shown to have a higher clonal diversity compared to further differentiated memory T-cells (<xref rid=\"B80\" ref-type=\"bibr\">80</xref>). In fact, this may also be the underlying mechanism, why we have more CD4<sup>+</sup> T-cells in the Rapa-TCP, because, evident from our <italic>ex vivo</italic> data, proportions of CMV-specific T<sub>CM</sub> are much higher among CD4<sup>+</sup> compared to CD8<sup>+</sup> T-cells.</p><p>When CMV-specific T-cells from different memory T-cell subsets were cultured individually, we revealed preferential expansion of T<sub>CM</sub> with preserved T<sub>CM</sub> phenotype, partially protected from differentiation in the presence of Rapamycin, but partially also conversion of T-cells from other memory T-cell subsets into T<sub>CM</sub>-like cells (<xref rid=\"B81\" ref-type=\"bibr\">81</xref>). Interestingly, patient 12 lacked CMV-specific early CD8<sup>+</sup> memory T-cells and his CMV-specific T-cells consisted to 90.6% of T<sub>EMRA</sub>. However, the respective Rapa<italic>-</italic>TCP included a strikingly high proportion of CD8<sup>+</sup> T<sub>CM</sub> suggesting reprogramming of T<sub>EM</sub>/T<sub>EMRA</sub> to T<sub>CM</sub>.</p><p>We did not record major differences in the characteristics of the TCPs irrespective of whether the TCPs were generated <italic>post</italic>-KTx or <italic>pre</italic>-KTx. Nonetheless, CD8<sup>+</sup> T-cells comprised higher proportions of T<sub>CM</sub> and cytokine producers in TCPs generated <italic>pre</italic>-KTx. Hence, there is no benefit to generate anti-CMV TCPs prior to KTx.</p><p>Investigating <italic>ex vivo-</italic>T-cell responses to CMV-specific stimuli, we found higher frequencies of CD8<sup>+</sup> than of CD4<sup>+</sup> CMV-specific T-cells as described previously (<xref rid=\"B82\" ref-type=\"bibr\">82</xref>). Frequencies of below 0.2% of CMV-reactive CD8<sup>+</sup> T-cells were found in 2/19 patients compared to 6/13 HDs, which matches findings suggesting an increase of CD8<sup>+</sup> CMV-reactive T-cells <italic>post</italic>-Tx (<xref rid=\"B83\" ref-type=\"bibr\">83</xref>). Interestingly, low frequencies of CMV-reactive CD8<sup>+</sup> T-cells did not cause low TCP yields, underlining the feasibility of TCP generation. Intriguingly, a high proportion of CD8<sup>+</sup> T<sub>EMRA</sub> among CMV-responsive T-cells did neither cause low expansion/yield nor high CD4/CD8 ratio in the TCP, which stresses the applicability even to patients with a high degree of terminal T-cell differentiation.</p><p>The dosage for successful CMV-specific AVTT in the SOT setting is undetermined. Case reports suggest numbers between 30 and 245 million T-cells (<xref rid=\"B19\" ref-type=\"bibr\">19</xref>&#x02013;<xref rid=\"B21\" ref-type=\"bibr\">21</xref>, <xref rid=\"B36\" ref-type=\"bibr\">36</xref>). Importantly, our protocol for TCP generation achieves medians of IFN&#x003b3;-producers of 20 and 50% in CD4<sup>+</sup> and CD8<sup>+</sup> T-cells, respectively, and median killing rates of 85%, while these frequencies are much lower in other published approaches reporting a maximum of 8% of CMV-reactive IFN&#x003b3;-producers in the TCP and specific lysis of &#x02264;50% at higher T-cell/target ratios (<xref rid=\"B19\" ref-type=\"bibr\">19</xref>, <xref rid=\"B20\" ref-type=\"bibr\">20</xref>). Comparable values are only achieved in a recent study (<xref rid=\"B21\" ref-type=\"bibr\">21</xref>). Based on the superiority in function and phenotype of Rapa-TCPs, we assume that also smaller T-cell numbers would be efficient for long-term control of CMV in KTx patients. The patients in this study whose TCPs yielded &#x0003c; 5 million T-cells from 20 ml of peripheral blood, included the two patients with the lowest lymphocyte counts (Patients 6/10), suggesting an amendment of the amount of blood collected to the lymphocyte count. Patient 6, with no recorded CMV DNAemia, was diagnosed with acute rejection and treated with ATG before blood collection. ATG administration may be an indication to collect blood for TCP generation preventively, given the patients\\' risk of developing CMV disease (<xref rid=\"B84\" ref-type=\"bibr\">84</xref>, <xref rid=\"B85\" ref-type=\"bibr\">85</xref>). Indeed, transient CMV DNAemia was recorded in this patient 54 days after blood collection for the study. Patient 10 had a record of recent CMV viremia, was seronegative <italic>pre</italic>-KTx, received a kidney from a seropositive donor and had extensive CMV-associated complications. Notably, he was the oldest patient included and correlation analysis revealed a negative correlation between age and yield of Rapa-TCPs. We also recorded a low yield of the Rapa-TCP of Patient 19, who had a history of CMV-DNAemia and was receiving Acyclovir-treatment at the time of blood collection, which was reported to diminish IFN&#x003b3;-production in response to CMV<sub>pp65</sub> peptides (<xref rid=\"B50\" ref-type=\"bibr\">50</xref>). In cases as described above, we suggest to first generate untreated TCPs for patients with acute CMV disease to diminish viral load (<xref rid=\"B36\" ref-type=\"bibr\">36</xref>) and then successively generate and infuse Rapa-TCPs for long-term control of the virus. The fact that we recorded significantly lower yields in patients with a recent CMV DNAemia compared to HDs also suggests to follow the proposed approach in these patients and motivates to investigate more than the four patients that we were able to recruit for this study. In fact, we also found a negative correlation of the number of records with viremia and the yield of Rapa-TCPs for the ten patients with detected CMV DNAemia. However, this analysis may be biased by a more thorough screening of problematic patients, as the tests for CMV DNAemia were not of equal frequency in all patients. Moreover, patient 12, for whom we could also only generate a Rapa-TCP with a yield of around 5 million cells, suffered from chronic hepatitis. These, of course limited data, indicate that problems may occur during manufacture in the case of different chronic infections. It has to be thoroughly overthought whether it is possible to begin with more blood as starting material in these patients with numerous reactivating infections or use first the conventional approach and then generate a Rapa-TCP for long-term protection.</p><p>Recently, Smith and colleagues published a study about the application of a comparable autologous CMV-specific TCP in SOT patients (<xref rid=\"B21\" ref-type=\"bibr\">21</xref>). However, their production process varies in many points, as they do not select for CMV-specific T-cells starting their culture with PBMCs, use G-Rex reactors instead of classical well-plates for expansion and have a different cytokine supplementation strategy using IL-21 and IL-2. They infuse multiple doses of TCPs at up to 6 different points in time. Of note, they demonstrate safety and clinical improvement in the majority of patients and could decrease or stop antiviral medication in many patients. Interestingly, compared to the time of infusion, they see an increase in viral load in 9 of 13 patients after infusion of TCPs (<xref rid=\"B21\" ref-type=\"bibr\">21</xref>). In 5 of the 9 monitored patients CD8<sup>+</sup> CMV-specific T-cells were reduced by the end of monitoring (max. day 300) and in three of the cases this correlated with an increase in viral load. One of the patients died of CMV disease (<xref rid=\"B21\" ref-type=\"bibr\">21</xref>). These data demonstrate, that there is still room for optimization in the long-term outcome of SOT patients treated with autologous CMV-specific TCPs. Probably, Rapa-TCPs could improve long-term efficacy, however, the actual clinical performance of our TCP has to be demonstrated. Our preclinical data imply a long-lived TCP with beneficial properties maybe even allowing for a single infusion.</p><p>A variety of other putative strategies are reported to rejuvenate T-cells with beneficial characteristics for AVTT. These include among others interference with different signaling pathways (<xref rid=\"B86\" ref-type=\"bibr\">86</xref>&#x02013;<xref rid=\"B90\" ref-type=\"bibr\">90</xref>), use of different cytokine supplementation strategies (<xref rid=\"B91\" ref-type=\"bibr\">91</xref>, <xref rid=\"B92\" ref-type=\"bibr\">92</xref>), employment of certain microRNAs (<xref rid=\"B93\" ref-type=\"bibr\">93</xref>), modulation of metabolism (<xref rid=\"B53\" ref-type=\"bibr\">53</xref>, <xref rid=\"B68\" ref-type=\"bibr\">68</xref>), inhibition of ion channels (<xref rid=\"B94\" ref-type=\"bibr\">94</xref>), and promotion of autophagy (<xref rid=\"B95\" ref-type=\"bibr\">95</xref>). However, most of these are far from being practicable under GMP conditions for contemporary application to a clinical setting. This is also the case for genetic manipulation of T-cells for optimization of AVTT. Proposed strategies for genetic engineering include induction of resistance to immunosuppressive medication (<xref rid=\"B96\" ref-type=\"bibr\">96</xref>, <xref rid=\"B97\" ref-type=\"bibr\">97</xref>), introduction of suicide genes as safety switch (<xref rid=\"B49\" ref-type=\"bibr\">49</xref>, <xref rid=\"B98\" ref-type=\"bibr\">98</xref>) and knock out of anti-inflammatory signaling components such as PD-1 and LAG-3 (<xref rid=\"B99\" ref-type=\"bibr\">99</xref>, <xref rid=\"B100\" ref-type=\"bibr\">100</xref>). All these suggestions have to be adapted to realistic GMP-feasible conditions and then may be valuable upgrades for even more sophisticated AVTT approaches. In contrast to other approaches, whose translation is less progressed at the moment, our minimally manipulative next-generation anti-CMV AVTT may help many transplanted patients whose endogenous immune system is not capable of defying the virus. Furthermore, the beneficial properties of Rapa-TCPs may also be transferred to other approaches using antigen-specific T-cells, e.g., other viruses or cancer immunotherapy with known antigens (<xref rid=\"B42\" ref-type=\"bibr\">42</xref>).</p><p>In summary, our study revealed favorable phenotypic and functional properties of Rapa-TCPs as well as their applicability to a variety of ESRD/KTx patient samples. Ultimately, we seek for clinical confirmation of functionality and efficacy of Rapa<italic>-</italic>TCPs in a clinical proof-of-concept trial.</p></sec><sec id=\"s5\"><title>Ethics Statement</title><p>The Charit&#x000e9; Ethics Committee (IRB) approved the study protocol and all blood donors provided written informed consent.</p></sec><sec id=\"s6\"><title>Author Contributions</title><p>LA, H-DV, PR, and MS-H conceptualized and designed the study. PR and NO provided patient samples and data. LA, TV, DW, and AJ performed experiments supervised by MS-H. LA acquired, analyzed, and interpreted data. LA and MS-H composed figures and manuscript. KJ performed bioinformatics analyses, created the respective graphs and made RNA seq data available at the GEO platform. TV, DW, SL-K, KJ, H-DV, and PR critically revised and all authors approved the final version of the manuscript.</p><sec><title>Conflict of Interest Statement</title><p>H-DV, PR, and MS-H own a patent on the manufacture of Rapa-TCPs. The remaining authors declare that the research was conducted in the absence of any commercial or financial relationships that could be construed as a potential conflict of interest.</p></sec></sec></body><back><ack><p>We acknowledge the assistance of the BCRT Flow Cytometry Core Facility, Dr. D. Kunkel and Dr. Sarah Meier. We thank the Institute for Experimental Endocrinology for providing their Seahorse analyzer and especially Dr. Nancy Schanze for her kind support with the handling. We are thankful to Marten J&#x000e4;ger for the performance of the quality control of RNA sequencing data. We express our deep gratitude to Anett Sefrin (study nurse) for organization of blood collection from patients as well as all patients and HDs for their willingness to donate blood for this study. and to Sarina Richter for the performance of DNA/RNA extractions and preparation of TCR sequencing. We thank Dr. Ulrike Kr&#x000fc;ger for her help with the sequencing devices, Dr. J&#x000f6;rn Schuldes (Labor Berlin GmbH) for providing the miSeq device and Gundula Leschik for operational assistance with the sequencing device.</p></ack><fn-group><fn fn-type=\"financial-disclosure\"><p><bold>Funding.</bold> The study was generously supported in parts by the Deutsche Forschungsgemeinschaft (DFG-SFB-TR36-project A3), the German Federal Ministry of Education and Research (Berlin-Brandenburg Center for Regenerative Therapies grant). The funders had no role in study design, data collection and analysis, decision to publish or preparation of the manuscript.</p></fn></fn-group><sec sec-type=\"supplementary-material\" id=\"s7\"><title>Supplementary Material</title><p>The Supplementary Material for this article can be found online at: <ext-link ext-link-type=\"uri\" xlink:href=\"https://www.frontiersin.org/articles/10.3389/fimmu.2019.01148/full#supplementary-material\">https://www.frontiersin.org/articles/10.3389/fimmu.2019.01148/full#supplementary-material</ext-link></p><supplementary-material content-type=\"local-data\" id=\"SM1\"><media xlink:href=\"Presentation_1.pdf\"><caption><p>Click here for additional data file.</p></caption></media></supplementary-material></sec><ref-list><title>References</title><ref id=\"B1\"><label>1.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Harvala</surname><given-names>H</given-names></name><name><surname>Stewart</surname><given-names>C</given-names></name><name><surname>Muller</surname><given-names>K</given-names></name><name><surname>Burns</surname><given-names>S</given-names></name><name><surname>Marson</surname><given-names>L</given-names></name><name><surname>MacGilchrist</surname><given-names>A</given-names></name><etal/></person-group>. <article-title>High risk of cytomegalovirus infection following solid organ transplantation despite prophylactic therapy</article-title>. <source>J Med Virol.</source> (<year>2013</year>) <volume>85</volume>:<fpage>893</fpage>&#x02013;<lpage>8</lpage>. <pub-id pub-id-type=\"doi\">10.1002/jmv.23539</pub-id><pub-id pub-id-type=\"pmid\">23508914</pub-id></mixed-citation></ref><ref id=\"B2\"><label>2.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tong</surname><given-names>CY</given-names></name><name><surname>Bakran</surname><given-names>A</given-names></name><name><surname>Peiris</surname><given-names>JS</given-names></name><name><surname>Muir</surname><given-names>P</given-names></name><name><surname>Herrington</surname><given-names>CS</given-names></name></person-group>. <article-title>The association of viral infection and chronic allograft nephropathy with graft dysfunction after renal transplantation</article-title>. <source>Transplantation.</source> (<year>2002</year>) <volume>74</volume>:<fpage>576</fpage>&#x02013;<lpage>8</lpage>. <pub-id pub-id-type=\"doi\">10.1097/00007890-200208270-00026</pub-id><pub-id pub-id-type=\"pmid\">12352923</pub-id></mixed-citation></ref><ref id=\"B3\"><label>3.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Bonatti</surname><given-names>H</given-names></name><name><surname>Sifri</surname><given-names>CD</given-names></name><name><surname>Larcher</surname><given-names>C</given-names></name><name><surname>Schneeberger</surname><given-names>S</given-names></name><name><surname>Kotton</surname><given-names>C</given-names></name><name><surname>Geltner</surname><given-names>C</given-names></name></person-group>. <article-title>Use of cidofovir for cytomegalovirus disease refractory to ganciclovir in solid organ recipients</article-title>. <source>Surg Infect</source>. (<year>2017</year>) <volume>18</volume>:<fpage>128</fpage>&#x02013;<lpage>36</lpage>. <pub-id pub-id-type=\"doi\">10.1089/sur.2015.266</pub-id><pub-id pub-id-type=\"pmid\">27849440</pub-id></mixed-citation></ref><ref id=\"B4\"><label>4.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Grossi</surname><given-names>PA</given-names></name><name><surname>Costa</surname><given-names>AN</given-names></name><name><surname>Fehily</surname><given-names>D</given-names></name><name><surname>Blumberg</surname><given-names>EA</given-names></name><name><surname>Kuehnert</surname><given-names>MJ</given-names></name><name><surname>Fishman</surname><given-names>JA</given-names></name><etal/></person-group>. <article-title>Infections and organ transplantation: new challenges for prevention and treatment&#x02013;a colloquium</article-title>. <source>Transplantation.</source> (<year>2012</year>) <volume>93</volume>:<fpage>S4</fpage>&#x02013;<lpage>39</lpage>. <pub-id pub-id-type=\"doi\">10.1097/TP.0b013e3182481347</pub-id><pub-id pub-id-type=\"pmid\">22374265</pub-id></mixed-citation></ref><ref id=\"B5\"><label>5.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kumar</surname><given-names>D</given-names></name><name><surname>Chernenko</surname><given-names>S</given-names></name><name><surname>Moussa</surname><given-names>G</given-names></name><name><surname>Cobos</surname><given-names>I</given-names></name><name><surname>Manuel</surname><given-names>O</given-names></name><name><surname>Preiksaitis</surname><given-names>J</given-names></name><etal/></person-group>. <article-title>Cell-mediated immunity to predict cytomegalovirus disease in high-risk solid organ transplant recipients</article-title>. <source>Am J Transplant.</source> (<year>2009</year>) <volume>9</volume>:<fpage>1214</fpage>&#x02013;<lpage>22</lpage>. <pub-id pub-id-type=\"doi\">10.1111/j.1600-6143.2009.02618.x</pub-id><pub-id pub-id-type=\"pmid\">19422346</pub-id></mixed-citation></ref><ref id=\"B6\"><label>6.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Schachtner</surname><given-names>T</given-names></name><name><surname>Stein</surname><given-names>M</given-names></name><name><surname>Reinke</surname><given-names>P</given-names></name></person-group>. <article-title>CMV-specific T cell monitoring offers superior risk stratification of CMV-seronegative kidney transplant recipients of a CMV-seropositive donor</article-title>. <source>Transplantation.</source> (<year>2017</year>) <volume>101</volume>:<fpage>e315</fpage>&#x02013;<lpage>25</lpage>. <pub-id pub-id-type=\"doi\">10.1097/TP.0000000000001825</pub-id><pub-id pub-id-type=\"pmid\">28594749</pub-id></mixed-citation></ref><ref id=\"B7\"><label>7.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Bunde</surname><given-names>T</given-names></name><name><surname>Kirchner</surname><given-names>A</given-names></name><name><surname>Hoffmeister</surname><given-names>B</given-names></name><name><surname>Habedank</surname><given-names>D</given-names></name><name><surname>Hetzer</surname><given-names>R</given-names></name><name><surname>Cherepnev</surname><given-names>G</given-names></name><etal/></person-group>. <article-title>Protection from cytomegalovirus after transplantation is correlated with immediate early 1-specific CD8 T cells</article-title>. <source>J Exp Med.</source> (<year>2005</year>) <volume>201</volume>:<fpage>1031</fpage>. <pub-id pub-id-type=\"doi\">10.1084/jem.20042384</pub-id><pub-id pub-id-type=\"pmid\">15795239</pub-id></mixed-citation></ref><ref id=\"B8\"><label>8.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Tormo</surname><given-names>N</given-names></name><name><surname>Solano</surname><given-names>C</given-names></name><name><surname>Benet</surname><given-names>I</given-names></name><name><surname>Nieto</surname><given-names>J</given-names></name><name><surname>de la Camara</surname><given-names>R</given-names></name><name><surname>Lopez</surname><given-names>J</given-names></name><etal/></person-group>. <article-title>Reconstitution of CMV pp65 and IE-1-specific IFN-gamma CD8(+) and CD4(+) T-cell responses affording protection from CMV DNAemia following allogeneic hematopoietic SCT</article-title>. <source>Bone Marrow Transplant.</source> (<year>2011</year>) <volume>46</volume>:<fpage>1437</fpage>&#x02013;<lpage>43</lpage>. <pub-id pub-id-type=\"doi\">10.1038/bmt.2010.330</pub-id><pub-id pub-id-type=\"pmid\">21243030</pub-id></mixed-citation></ref><ref id=\"B9\"><label>9.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gratama</surname><given-names>JW</given-names></name><name><surname>Brooimans</surname><given-names>RA</given-names></name><name><surname>van der Holt</surname><given-names>B</given-names></name><name><surname>Sintnicolaas</surname><given-names>K</given-names></name><name><surname>van Doornum</surname><given-names>G</given-names></name><name><surname>Niesters</surname><given-names>HG</given-names></name><etal/></person-group>. <article-title>Monitoring cytomegalovirus IE-1 and pp65-specific CD4+ and CD8+ T-cell responses after allogeneic stem cell transplantation may identify patients at risk for recurrent CMV reactivations</article-title>. <source>Cytometry B Clin Cytom.</source> (<year>2008</year>) <volume>74</volume>:<fpage>211</fpage>&#x02013;<lpage>20</lpage>. <pub-id pub-id-type=\"doi\">10.1002/cyto.b.20420</pub-id><pub-id pub-id-type=\"pmid\">18454493</pub-id></mixed-citation></ref><ref id=\"B10\"><label>10.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Guerreiro</surname><given-names>M</given-names></name><name><surname>Na</surname><given-names>KI</given-names></name><name><surname>Letsch</surname><given-names>A</given-names></name><name><surname>Haase</surname><given-names>D</given-names></name><name><surname>Bauer</surname><given-names>S</given-names></name><name><surname>Meisel</surname><given-names>C</given-names></name><etal/></person-group>. <article-title>Human peripheral blood and bone marrow Epstein-Barr virus-specific T-cell repertoire in latent infection reveals distinct memory T-cell subsets</article-title>. <source>Eur J Immunol.</source> (<year>2010</year>) <volume>40</volume>:<fpage>1566</fpage>&#x02013;<lpage>76</lpage>. <pub-id pub-id-type=\"doi\">10.1002/eji.200940000</pub-id><pub-id pub-id-type=\"pmid\">20232341</pub-id></mixed-citation></ref><ref id=\"B11\"><label>11.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Schachtner</surname><given-names>T</given-names></name><name><surname>Stein</surname><given-names>M</given-names></name><name><surname>Babel</surname><given-names>N</given-names></name><name><surname>Reinke</surname><given-names>P</given-names></name></person-group>. <article-title>The loss of BKV-specific immunity from pretransplantation to posttransplantation identifies kidney transplant recipients at increased risk of BKV replication</article-title>. <source>Am J Transplant.</source> (<year>2015</year>) <volume>15</volume>:<fpage>2159</fpage>&#x02013;<lpage>69</lpage>. <pub-id pub-id-type=\"doi\">10.1111/ajt.13252</pub-id><pub-id pub-id-type=\"pmid\">25808077</pub-id></mixed-citation></ref><ref id=\"B12\"><label>12.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Riddell</surname><given-names>SR</given-names></name><name><surname>Watanabe</surname><given-names>KS</given-names></name><name><surname>Goodrich</surname><given-names>JM</given-names></name><name><surname>Li</surname><given-names>CR</given-names></name><name><surname>Agha</surname><given-names>ME</given-names></name><name><surname>Greenberg</surname><given-names>PD</given-names></name></person-group>. <article-title>Restoration of viral immunity in immunodeficient humans by the adoptive transfer of T cell clones</article-title>. <source>Science.</source> (<year>1992</year>) <volume>257</volume>:<fpage>238</fpage>&#x02013;<lpage>41</lpage>. <pub-id pub-id-type=\"doi\">10.1126/science.1352912</pub-id><pub-id pub-id-type=\"pmid\">1352912</pub-id></mixed-citation></ref><ref id=\"B13\"><label>13.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hill</surname><given-names>GR</given-names></name><name><surname>Tey</surname><given-names>SK</given-names></name><name><surname>Beagley</surname><given-names>L</given-names></name><name><surname>Crough</surname><given-names>T</given-names></name><name><surname>Morton</surname><given-names>JA</given-names></name><name><surname>Clouston</surname><given-names>AD</given-names></name><etal/></person-group>. <article-title>Successful immunotherapy of HCMV disease using virus-specific T cells expanded from an allogeneic stem cell transplant recipient</article-title>. <source>Am J Transplant.</source> (<year>2010</year>) <volume>10</volume>:<fpage>173</fpage>&#x02013;<lpage>9</lpage>. <pub-id pub-id-type=\"doi\">10.1111/j.1600-6143.2009.02872.x</pub-id><pub-id pub-id-type=\"pmid\">19919660</pub-id></mixed-citation></ref><ref id=\"B14\"><label>14.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Clancy</surname><given-names>LE</given-names></name><name><surname>Blyth</surname><given-names>E</given-names></name><name><surname>Simms</surname><given-names>RM</given-names></name><name><surname>Micklethwaite</surname><given-names>KP</given-names></name><name><surname>Ma</surname><given-names>CK</given-names></name><name><surname>Burgess</surname><given-names>JS</given-names></name><etal/></person-group>. <article-title>Cytomegalovirus-specific cytotoxic T lymphocytes can be efficiently expanded from granulocyte colony-stimulating factor-mobilized hemopoietic progenitor cell products ex vivo and safely transferred to stem cell transplantation recipients to facilitate immune reconstitution</article-title>. <source>Biol Blood Marrow Transplant.</source> (<year>2013</year>) <volume>19</volume>:<fpage>725</fpage>&#x02013;<lpage>34</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.bbmt.2013.01.021</pub-id><pub-id pub-id-type=\"pmid\">23380344</pub-id></mixed-citation></ref><ref id=\"B15\"><label>15.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Stemberger</surname><given-names>C</given-names></name><name><surname>Graef</surname><given-names>P</given-names></name><name><surname>Odendahl</surname><given-names>M</given-names></name><name><surname>Albrecht</surname><given-names>J</given-names></name><name><surname>Dossinger</surname><given-names>G</given-names></name><name><surname>Anderl</surname><given-names>F</given-names></name><etal/></person-group>. <article-title>Lowest numbers of primary CD8(+) T cells can reconstitute protective immunity upon adoptive immunotherapy</article-title>. <source>Blood.</source> (<year>2014</year>) <volume>124</volume>:<fpage>628</fpage>&#x02013;<lpage>37</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2013-12-547349</pub-id><pub-id pub-id-type=\"pmid\">24855206</pub-id></mixed-citation></ref><ref id=\"B16\"><label>16.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Stuehler</surname><given-names>C</given-names></name><name><surname>Stussi</surname><given-names>G</given-names></name><name><surname>Halter</surname><given-names>J</given-names></name><name><surname>Nowakowska</surname><given-names>J</given-names></name><name><surname>Schibli</surname><given-names>A</given-names></name><name><surname>Battegay</surname><given-names>M</given-names></name><etal/></person-group>. <article-title>Combination therapy for multidrug-resistant cytomegalovirus disease</article-title>. <source>Transpl Infect Dis.</source> (<year>2015</year>) <volume>17</volume>:<fpage>751</fpage>&#x02013;<lpage>5</lpage>. <pub-id pub-id-type=\"doi\">10.1111/tid.12435</pub-id><pub-id pub-id-type=\"pmid\">26432076</pub-id></mixed-citation></ref><ref id=\"B17\"><label>17.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Neuenhahn</surname><given-names>M</given-names></name><name><surname>Albrecht</surname><given-names>J</given-names></name><name><surname>Odendahl</surname><given-names>M</given-names></name><name><surname>Schlott</surname><given-names>F</given-names></name><name><surname>Dossinger</surname><given-names>G</given-names></name><name><surname>Schiemann</surname><given-names>M</given-names></name><etal/></person-group>. <article-title>Transfer of minimally manipulated CMV-specific T cells from stem cell or third-party donors to treat CMV infection after allo-HSC</article-title>. <source>Leukemia.</source> (<year>2017</year>) <volume>31</volume>:<fpage>2161</fpage>&#x02013;<lpage>71</lpage>. <pub-id pub-id-type=\"doi\">10.1038/leu.2017.16</pub-id><pub-id pub-id-type=\"pmid\">28090089</pub-id></mixed-citation></ref><ref id=\"B18\"><label>18.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Brestrich</surname><given-names>G</given-names></name><name><surname>Zwinger</surname><given-names>S</given-names></name><name><surname>Fischer</surname><given-names>A</given-names></name><name><surname>Schmuck</surname><given-names>M</given-names></name><name><surname>Rohmhild</surname><given-names>A</given-names></name><name><surname>Hammer</surname><given-names>MH</given-names></name><etal/></person-group>. <article-title>Adoptive T-cell therapy of a lung transplanted patient with severe CMV disease and resistance to antiviral therapy</article-title>. <source>Am J Transplant.</source> (<year>2009</year>) <volume>9</volume>:<fpage>1679</fpage>&#x02013;<lpage>84</lpage>. <pub-id pub-id-type=\"doi\">10.1111/j.1600-6143.2009.02672.x</pub-id><pub-id pub-id-type=\"pmid\">19459791</pub-id></mixed-citation></ref><ref id=\"B19\"><label>19.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Macesic</surname><given-names>N</given-names></name><name><surname>Langsford</surname><given-names>D</given-names></name><name><surname>Nicholls</surname><given-names>K</given-names></name><name><surname>Hughes</surname><given-names>P</given-names></name><name><surname>Gottlieb</surname><given-names>DJ</given-names></name><name><surname>Clancy</surname><given-names>L</given-names></name><etal/></person-group>. <article-title>Adoptive T cell immunotherapy for treatment of ganciclovir-resistant cytomegalovirus disease in a renal transplant recipient</article-title>. <source>Am J Transplant.</source> (<year>2015</year>) <volume>15</volume>:<fpage>827</fpage>&#x02013;<lpage>32</lpage>. <pub-id pub-id-type=\"doi\">10.1111/ajt.13023</pub-id><pub-id pub-id-type=\"pmid\">25648555</pub-id></mixed-citation></ref><ref id=\"B20\"><label>20.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Holmes-Liew</surname><given-names>CL</given-names></name><name><surname>Holmes</surname><given-names>M</given-names></name><name><surname>Beagley</surname><given-names>L</given-names></name><name><surname>Hopkins</surname><given-names>P</given-names></name><name><surname>Chambers</surname><given-names>D</given-names></name><name><surname>Smith</surname><given-names>C</given-names></name><etal/></person-group>. <article-title>Adoptive T-cell immunotherapy for ganciclovir-resistant CMV disease after lung transplantation</article-title>. <source>Clin Transl Immunology.</source> (<year>2015</year>) <volume>4</volume>:<fpage>e35</fpage>. <pub-id pub-id-type=\"doi\">10.1038/cti.2015.5</pub-id><pub-id pub-id-type=\"pmid\">25859390</pub-id></mixed-citation></ref><ref id=\"B21\"><label>21.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Smith</surname><given-names>C</given-names></name><name><surname>Beagley</surname><given-names>L</given-names></name><name><surname>Rehan</surname><given-names>S</given-names></name><name><surname>Neller</surname><given-names>MA</given-names></name><name><surname>Crooks</surname><given-names>P</given-names></name><name><surname>Solomon</surname><given-names>M</given-names></name><etal/></person-group>. <article-title>Autologous adoptive T-cell therapy for recurrent or drug-resistant cytomegalovirus complications in solid organ transplant patients: a single-arm open-label phase I clinical trial</article-title>. <source>Clin Infect Dis</source>. (<year>2018</year>). <pub-id pub-id-type=\"doi\">10.1093/cid/ciy549</pub-id><pub-id pub-id-type=\"pmid\">29982441</pub-id></mixed-citation></ref><ref id=\"B22\"><label>22.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Khanna</surname><given-names>R</given-names></name><name><surname>Bell</surname><given-names>S</given-names></name><name><surname>Sherritt</surname><given-names>M</given-names></name><name><surname>Galbraith</surname><given-names>A</given-names></name><name><surname>Burrows</surname><given-names>SR</given-names></name><name><surname>Rafter</surname><given-names>L</given-names></name><etal/></person-group>. <article-title>Activation and adoptive transfer of Epstein-Barr virus-specific cytotoxic T cells in solid organ transplant patients with posttransplant lymphoproliferative disease</article-title>. <source>Proc Natl Acad Sci USA.</source> (<year>1999</year>) <volume>96</volume>:<fpage>10391</fpage>&#x02013;<lpage>6</lpage>. <pub-id pub-id-type=\"doi\">10.1073/pnas.96.18.10391</pub-id><pub-id pub-id-type=\"pmid\">10468618</pub-id></mixed-citation></ref><ref id=\"B23\"><label>23.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Savoldo</surname><given-names>B</given-names></name><name><surname>Huls</surname><given-names>MH</given-names></name><name><surname>Liu</surname><given-names>Z</given-names></name><name><surname>Okamura</surname><given-names>T</given-names></name><name><surname>Volk</surname><given-names>HD</given-names></name><name><surname>Reinke</surname><given-names>P</given-names></name><etal/></person-group>. <article-title>Autologous Epstein-Barr virus (EBV)-specific cytotoxic T cells for the treatment of persistent active EBV infection</article-title>. <source>Blood.</source> (<year>2002</year>) <volume>100</volume>:<fpage>4059</fpage>&#x02013;<lpage>66</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2002-01-0039</pub-id><pub-id pub-id-type=\"pmid\">12393655</pub-id></mixed-citation></ref><ref id=\"B24\"><label>24.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Haque</surname><given-names>T</given-names></name><name><surname>Wilkie</surname><given-names>GM</given-names></name><name><surname>Jones</surname><given-names>MM</given-names></name><name><surname>Higgins</surname><given-names>CD</given-names></name><name><surname>Urquhart</surname><given-names>G</given-names></name><name><surname>Wingate</surname><given-names>P</given-names></name><etal/></person-group>. <article-title>Allogeneic cytotoxic T-cell therapy for EBV-positive posttransplantation lymphoproliferative disease: results of a phase 2 multicenter clinical trial</article-title>. <source>Blood.</source> (<year>2007</year>) <volume>110</volume>:<fpage>1123</fpage>&#x02013;<lpage>31</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2006-12-063008</pub-id><pub-id pub-id-type=\"pmid\">17468341</pub-id></mixed-citation></ref><ref id=\"B25\"><label>25.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Savoldo</surname><given-names>B</given-names></name><name><surname>Goss</surname><given-names>JA</given-names></name><name><surname>Hammer</surname><given-names>MM</given-names></name><name><surname>Zhang</surname><given-names>L</given-names></name><name><surname>Lopez</surname><given-names>T</given-names></name><name><surname>Gee</surname><given-names>AP</given-names></name><etal/></person-group>. <article-title>Treatment of solid organ transplant recipients with autologous Epstein Barr virus-specific cytotoxic T lymphocytes (CTLs)</article-title>. <source>Blood.</source> (<year>2006</year>) <volume>108</volume>:<fpage>2942</fpage>&#x02013;<lpage>9</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2006-05-021782</pub-id><pub-id pub-id-type=\"pmid\">16835376</pub-id></mixed-citation></ref><ref id=\"B26\"><label>26.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Sallusto</surname><given-names>F</given-names></name><name><surname>Lenig</surname><given-names>D</given-names></name><name><surname>Forster</surname><given-names>R</given-names></name><name><surname>Lipp</surname><given-names>M</given-names></name><name><surname>Lanzavecchia</surname><given-names>A</given-names></name></person-group>. <article-title>Two subsets of memory T lymphocytes with distinct homing potentials and effector functions</article-title>. <source>Nature.</source> (<year>1999</year>) <volume>401</volume>:<fpage>708</fpage>&#x02013;<lpage>12</lpage>. <pub-id pub-id-type=\"doi\">10.1038/44385</pub-id><pub-id pub-id-type=\"pmid\">10537110</pub-id></mixed-citation></ref><ref id=\"B27\"><label>27.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Geginat</surname><given-names>J</given-names></name><name><surname>Lanzavecchia</surname><given-names>A</given-names></name><name><surname>Sallusto</surname><given-names>F</given-names></name></person-group>\\n<article-title>Proliferation and differentiation potential of human CD8+ memory T-cell subsets in response to antigen or homeostatic cytokines</article-title>. <source>Blood.</source> (<year>2003</year>) <volume>101</volume>:<fpage>4260</fpage>&#x02013;<lpage>6</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2002-11-3577</pub-id><pub-id pub-id-type=\"pmid\">12576317</pub-id></mixed-citation></ref><ref id=\"B28\"><label>28.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Berger</surname><given-names>C</given-names></name><name><surname>Jensen</surname><given-names>MC</given-names></name><name><surname>Lansdorp</surname><given-names>PM</given-names></name><name><surname>Gough</surname><given-names>M</given-names></name><name><surname>Elliott</surname><given-names>C</given-names></name><name><surname>Riddell</surname><given-names>SR</given-names></name></person-group>. <article-title>Adoptive transfer of effector CD8+ T cells derived from central memory cells establishes persistent T cell memory in primates</article-title>. <source>J Clin Invest.</source> (<year>2008</year>) <volume>118</volume>:<fpage>294</fpage>&#x02013;<lpage>305</lpage>. <pub-id pub-id-type=\"doi\">10.1172/JCI32103</pub-id><pub-id pub-id-type=\"pmid\">18060041</pub-id></mixed-citation></ref><ref id=\"B29\"><label>29.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Louis</surname><given-names>CU</given-names></name><name><surname>Savoldo</surname><given-names>B</given-names></name><name><surname>Dotti</surname><given-names>G</given-names></name><name><surname>Pule</surname><given-names>M</given-names></name><name><surname>Yvon</surname><given-names>E</given-names></name><name><surname>Myers</surname><given-names>GD</given-names></name><etal/></person-group>. <article-title>Antitumor activity and long-term fate of chimeric antigen receptor-positive T cells in patients with neuroblastoma</article-title>. <source>Blood.</source> (<year>2011</year>) <volume>118</volume>:<fpage>6050</fpage>&#x02013;<lpage>6</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2011-05-354449</pub-id><pub-id pub-id-type=\"pmid\">21984804</pub-id></mixed-citation></ref><ref id=\"B30\"><label>30.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wang</surname><given-names>CS</given-names></name><name><surname>Shah</surname><given-names>SA</given-names></name><name><surname>Chiu</surname><given-names>Y</given-names></name><name><surname>Paria</surname><given-names>BC</given-names></name><name><surname>Aghamolla</surname><given-names>T</given-names></name><name><surname>Alvarez-Downing</surname><given-names>MM</given-names></name><etal/></person-group>. <article-title>The stoichiometric production of IL-2 and IFN-gamma mRNA defines memory T cells that can self-renew after adoptive transfer in humans</article-title>. <source>Sci Transl Med.</source> (<year>2012</year>) <volume>4</volume>:<fpage>149ra120</fpage>. <pub-id pub-id-type=\"doi\">10.1126/scitranslmed.3004306</pub-id><pub-id pub-id-type=\"pmid\">22932225</pub-id></mixed-citation></ref><ref id=\"B31\"><label>31.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Graef</surname><given-names>P</given-names></name><name><surname>Buchholz</surname><given-names>VR</given-names></name><name><surname>Stemberger</surname><given-names>C</given-names></name><name><surname>Flossdorf</surname><given-names>M</given-names></name><name><surname>Henkel</surname><given-names>L</given-names></name><name><surname>Schiemann</surname><given-names>M</given-names></name><etal/></person-group>. <article-title>Serial transfer of single-cell-derived immunocompetence reveals stemness of CD8(+) central memory T cells</article-title>. <source>Immunity.</source> (<year>2014</year>) <volume>41</volume>:<fpage>116</fpage>&#x02013;<lpage>26</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.immuni.2014.05.018</pub-id><pub-id pub-id-type=\"pmid\">25035956</pub-id></mixed-citation></ref><ref id=\"B32\"><label>32.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Oliveira</surname><given-names>G</given-names></name><name><surname>Ruggiero</surname><given-names>E</given-names></name><name><surname>Stanghellini</surname><given-names>MT</given-names></name><name><surname>Cieri</surname><given-names>N</given-names></name><name><surname>D\\'Agostino</surname><given-names>M</given-names></name><name><surname>Fronza</surname><given-names>R</given-names></name><etal/></person-group>. <article-title>Tracking genetically engineered lymphocytes long-term reveals the dynamics of T cell immunological memory</article-title>. <source>Sci Transl Med.</source> (<year>2015</year>) <volume>7</volume>:<fpage>317ra198</fpage>. <pub-id pub-id-type=\"doi\">10.1126/scitranslmed.aac8265</pub-id><pub-id pub-id-type=\"pmid\">26659572</pub-id></mixed-citation></ref><ref id=\"B33\"><label>33.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kueberuwa</surname><given-names>G</given-names></name><name><surname>Gornall</surname><given-names>H</given-names></name><name><surname>Alcantar-Orozco</surname><given-names>EM</given-names></name><name><surname>Bouvier</surname><given-names>D</given-names></name><name><surname>Kapacee</surname><given-names>ZA</given-names></name><name><surname>Hawkins</surname><given-names>R</given-names></name><etal/></person-group>. <article-title>CCR7+ selected gene-modified T cells maintain a central memory phenotype and display enhanced persistence in peripheral blood <italic>in vivo</italic></article-title>. <source>J Immunother Cancer.</source> (<year>2017</year>) <volume>5</volume>:<fpage>14</fpage>. <pub-id pub-id-type=\"doi\">10.1186/s40425-017-0216-7</pub-id><pub-id pub-id-type=\"pmid\">28239467</pub-id></mixed-citation></ref><ref id=\"B34\"><label>34.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Sallusto</surname><given-names>F</given-names></name><name><surname>Geginat</surname><given-names>J</given-names></name><name><surname>Lanzavecchia</surname><given-names>A</given-names></name></person-group>. <article-title>Central memory and effector memory T cell subsets: function, generation, and maintenance</article-title>. <source>Annu Rev Immunol.</source> (<year>2004</year>) <volume>22</volume>:<fpage>745</fpage>&#x02013;<lpage>63</lpage>. <pub-id pub-id-type=\"doi\">10.1146/annurev.immunol.22.012703.104702</pub-id><pub-id pub-id-type=\"pmid\">15032595</pub-id></mixed-citation></ref><ref id=\"B35\"><label>35.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Liu</surname><given-names>J</given-names></name><name><surname>Chang</surname><given-names>YJ</given-names></name><name><surname>Yan</surname><given-names>CH</given-names></name><name><surname>Xu</surname><given-names>LP</given-names></name><name><surname>Jiang</surname><given-names>ZF</given-names></name><name><surname>Zhang</surname><given-names>XH</given-names></name><etal/></person-group>. <article-title>Poor CMV-specific CD8+ T central memory subset recovery at early stage post-HSCT associates with refractory and recurrent CMV reactivation</article-title>. <source>J Infect.</source> (<year>2016</year>) <volume>73</volume>:<fpage>261</fpage>&#x02013;<lpage>70</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.jinf.2016.04.033</pub-id><pub-id pub-id-type=\"pmid\">27311748</pub-id></mixed-citation></ref><ref id=\"B36\"><label>36.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Brestrich</surname><given-names>G</given-names></name><name><surname>Zwinger</surname><given-names>S</given-names></name><name><surname>Roemhild</surname><given-names>A</given-names></name><name><surname>Noutsias</surname><given-names>M</given-names></name><name><surname>Rohde</surname><given-names>M</given-names></name><name><surname>Keeren</surname><given-names>K</given-names></name><etal/></person-group>. <article-title>Generation of HCMV-specific T-cell lines from seropositive solid-organ-transplant recipients for adoptive T-cell therapy</article-title>. <source>J Immunother.</source> (<year>2009</year>) <volume>32</volume>:<fpage>932</fpage>&#x02013;<lpage>40</lpage>. <pub-id pub-id-type=\"doi\">10.1097/CJI.0b013e3181b88fda</pub-id><pub-id pub-id-type=\"pmid\">19816188</pub-id></mixed-citation></ref><ref id=\"B37\"><label>37.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Schmueck</surname><given-names>M</given-names></name><name><surname>Fischer</surname><given-names>AM</given-names></name><name><surname>Hammoud</surname><given-names>B</given-names></name><name><surname>Brestrich</surname><given-names>G</given-names></name><name><surname>Fuehrer</surname><given-names>H</given-names></name><name><surname>Luu</surname><given-names>SH</given-names></name><etal/></person-group>. <article-title>Preferential expansion of human virus-specific multifunctional central memory T cells by partial targeting of the IL-2 receptor signaling pathway: the key role of CD4+ T cells</article-title>. <source>J Immunol.</source> (<year>2012</year>) <volume>188</volume>:<fpage>5189</fpage>&#x02013;<lpage>98</lpage>. <pub-id pub-id-type=\"doi\">10.4049/jimmunol.1103763</pub-id><pub-id pub-id-type=\"pmid\">22516956</pub-id></mixed-citation></ref><ref id=\"B38\"><label>38.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Araki</surname><given-names>K</given-names></name><name><surname>Turner</surname><given-names>AP</given-names></name><name><surname>Shaffer</surname><given-names>VO</given-names></name><name><surname>Gangappa</surname><given-names>S</given-names></name><name><surname>Keller</surname><given-names>SA</given-names></name><name><surname>Bachmann</surname><given-names>MF</given-names></name><etal/></person-group>. <article-title>mTOR regulates memory CD8 T-cell differentiation</article-title>. <source>Nature.</source> (<year>2009</year>) <volume>460</volume>:<fpage>108</fpage>&#x02013;<lpage>12</lpage>. <pub-id pub-id-type=\"doi\">10.1038/nature08155</pub-id><pub-id pub-id-type=\"pmid\">19543266</pub-id></mixed-citation></ref><ref id=\"B39\"><label>39.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Pepper</surname><given-names>M</given-names></name><name><surname>Pagan</surname><given-names>AJ</given-names></name><name><surname>Igyarto</surname><given-names>BZ</given-names></name><name><surname>Taylor</surname><given-names>JJ</given-names></name><name><surname>Jenkins</surname><given-names>MK</given-names></name></person-group>. <article-title>Opposing signals from the Bcl6 transcription factor and the interleukin-2 receptor generate T helper 1 central and effector memory cells</article-title>. <source>Immunity</source>. (<year>2011</year>) <volume>35</volume>:<fpage>583</fpage>&#x02013;<lpage>95</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.immuni.2011.09.009</pub-id><pub-id pub-id-type=\"pmid\">22018468</pub-id></mixed-citation></ref><ref id=\"B40\"><label>40.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Walter</surname><given-names>EA</given-names></name><name><surname>Greenberg</surname><given-names>PD</given-names></name><name><surname>Gilbert</surname><given-names>MJ</given-names></name><name><surname>Finch</surname><given-names>RJ</given-names></name><name><surname>Watanabe</surname><given-names>KS</given-names></name><name><surname>Thomas</surname><given-names>ED</given-names></name></person-group>\\n<article-title>Riddell SR, Reconstitution of cellular immunity against cytomegalovirus in recipients of allogeneic bone marrow by transfer of T-cell clones from the donor</article-title>. <source>N Engl J Med.</source> (<year>1995</year>) <volume>333</volume>:<fpage>1038</fpage>&#x02013;<lpage>44</lpage>. <pub-id pub-id-type=\"doi\">10.1056/NEJM199510193331603</pub-id><pub-id pub-id-type=\"pmid\">7675046</pub-id></mixed-citation></ref><ref id=\"B41\"><label>41.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wang</surname><given-names>X</given-names></name><name><surname>Popplewell</surname><given-names>LL</given-names></name><name><surname>Wagner</surname><given-names>JR</given-names></name><name><surname>Naranjo</surname><given-names>A</given-names></name><name><surname>Blanchard</surname><given-names>MS</given-names></name><name><surname>Mott</surname><given-names>MR</given-names></name><etal/></person-group>. <article-title>Phase 1 studies of central memory-derived CD19 CAR T-cell therapy following autologous HSCT in patients with B-cell NHL</article-title>. <source>Blood.</source> (<year>2016</year>) <volume>127</volume>:<fpage>2980</fpage>&#x02013;<lpage>90</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2015-12-686725</pub-id><pub-id pub-id-type=\"pmid\">27118452</pub-id></mixed-citation></ref><ref id=\"B42\"><label>42.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hammer</surname><given-names>MH</given-names></name><name><surname>Meyer</surname><given-names>S</given-names></name><name><surname>Brestrich</surname><given-names>G</given-names></name><name><surname>Moosmann</surname><given-names>A</given-names></name><name><surname>Kern</surname><given-names>F</given-names></name><name><surname>Tesfa</surname><given-names>L</given-names></name><etal/></person-group>. <article-title>HLA type-independent generation of antigen-specific T cells for adoptive immunotherapy</article-title>. <source>Eur J Immunol.</source> (<year>2005</year>) <volume>35</volume>:<fpage>2250</fpage>&#x02013;<lpage>8</lpage>. <pub-id pub-id-type=\"doi\">10.1002/eji.200526230</pub-id><pub-id pub-id-type=\"pmid\">15915543</pub-id></mixed-citation></ref><ref id=\"B43\"><label>43.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Moosmann Khan</surname><given-names>N</given-names></name><name><surname>Cobbold</surname><given-names>M</given-names></name><name><surname>Zentz</surname><given-names>C</given-names></name><name><surname>Delecluse</surname><given-names>HJ</given-names></name><name><surname>Hollweck</surname><given-names>G</given-names></name><name><surname>Hislop</surname><given-names>AD</given-names></name><etal/></person-group>. <article-title>B cells immortalized by a mini-Epstein-Barr virus encoding a foreign antigen efficiently reactivate specific cytotoxic T cells</article-title>. <source>Blood.</source> (<year>2002</year>) <volume>100</volume>:<fpage>1755</fpage>&#x02013;<lpage>64</lpage>.<pub-id pub-id-type=\"pmid\">12176897</pub-id></mixed-citation></ref><ref id=\"B44\"><label>44.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hermans</surname><given-names>IF</given-names></name><name><surname>Silk</surname><given-names>JD</given-names></name><name><surname>Yang</surname><given-names>J</given-names></name><name><surname>Palmowski</surname><given-names>MJ</given-names></name><name><surname>Gileadi</surname><given-names>U</given-names></name><name><surname>McCarthy</surname><given-names>C</given-names></name><etal/></person-group>. <article-title>The VITAL assay: a versatile fluorometric technique for assessing CTL- and NKT-mediated cytotoxicity against multiple targets in vitro and <italic>in vivo</italic></article-title>. <source>J Immunol Methods.</source> (<year>2004</year>) <volume>285</volume>:<fpage>25</fpage>&#x02013;<lpage>40</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.jim.2003.10.017</pub-id><pub-id pub-id-type=\"pmid\">14871532</pub-id></mixed-citation></ref><ref id=\"B45\"><label>45.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Hammoud</surname><given-names>B</given-names></name><name><surname>Schmueck</surname><given-names>M</given-names></name><name><surname>Fischer</surname><given-names>AM</given-names></name><name><surname>Fuehrer</surname><given-names>H</given-names></name><name><surname>Park</surname><given-names>SJ</given-names></name><name><surname>Akyuez</surname><given-names>L</given-names></name><etal/></person-group>\\n<article-title>HCMV-specific T-cell therapy: do not forget supply of help</article-title>. <source>J Immunother.</source> (<year>2013</year>) <volume>36</volume>:<fpage>93</fpage>&#x02013;<lpage>101</lpage>. <pub-id pub-id-type=\"doi\">10.1097/CJI.0b013e31827b87cc</pub-id><pub-id pub-id-type=\"pmid\">23377662</pub-id></mixed-citation></ref><ref id=\"B46\"><label>46.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Langmead</surname><given-names>B</given-names></name><name><surname>Salzberg</surname><given-names>SL</given-names></name></person-group>\\n<article-title>Fast gapped-read alignment with Bowtie 2</article-title>. <source>Nat Methods.</source> (<year>2012</year>) <volume>9</volume>:<fpage>357</fpage>&#x02013;<lpage>9</lpage>. <pub-id pub-id-type=\"doi\">10.1038/nmeth.1923</pub-id><pub-id pub-id-type=\"pmid\">22388286</pub-id></mixed-citation></ref><ref id=\"B47\"><label>47.</label><mixed-citation publication-type=\"book\"><person-group person-group-type=\"author\"><name><surname>Team</surname><given-names>RC</given-names></name></person-group>\\n<source>R: A Language and Environment For Statistical Computing.</source>\\n<publisher-name>R Foundation for Statistical Computing</publisher-name>, <publisher-loc>Vienna, Austria</publisher-loc> (<year>2018</year>).</mixed-citation></ref><ref id=\"B48\"><label>48.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Love</surname><given-names>MF</given-names></name><name><surname>Huber</surname><given-names>W</given-names></name><name><surname>Anders</surname><given-names>S</given-names></name></person-group>. <article-title>Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2</article-title>. <source>Genome Biol.</source> (<year>2014</year>) <volume>15</volume>:<fpage>550</fpage>. <pub-id pub-id-type=\"doi\">10.1186/s13059-014-0550-8</pub-id><pub-id pub-id-type=\"pmid\">25516281</pub-id></mixed-citation></ref><ref id=\"B49\"><label>49.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kaneko</surname><given-names>S</given-names></name><name><surname>Mastaglio</surname><given-names>S</given-names></name><name><surname>Bondanza</surname><given-names>A</given-names></name><name><surname>Ponzoni</surname><given-names>M</given-names></name><name><surname>Sanvito</surname><given-names>F</given-names></name><name><surname>Aldrighetti</surname><given-names>L</given-names></name><etal/></person-group>. <article-title>IL-7 and IL-15 allow the generation of suicide gene-modified alloreactive self-renewing central memory human T lymphocytes</article-title>. <source>Blood.</source> (<year>2009</year>) <volume>113</volume>:<fpage>1006</fpage>&#x02013;<lpage>15</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2008-05-156059</pub-id><pub-id pub-id-type=\"pmid\">18978209</pub-id></mixed-citation></ref><ref id=\"B50\"><label>50.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Pachnio</surname><given-names>A</given-names></name><name><surname>Begum</surname><given-names>J</given-names></name><name><surname>Fox</surname><given-names>A</given-names></name><name><surname>Moss</surname><given-names>P</given-names></name></person-group>. <article-title>Acyclovir therapy reduces the CD4+ T cell response against the immunodominant pp65 protein from cytomegalovirus in immune competent individuals</article-title>. <source>PLoS ONE</source>. (<year>2015</year>) <volume>10</volume>:<fpage>e0125287</fpage>. <pub-id pub-id-type=\"doi\">10.1371/journal.pone.0125287</pub-id><pub-id pub-id-type=\"pmid\">25923913</pub-id></mixed-citation></ref><ref id=\"B51\"><label>51.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Calastretti</surname><given-names>A</given-names></name><name><surname>Rancati</surname><given-names>F</given-names></name><name><surname>Ceriani</surname><given-names>MC</given-names></name><name><surname>Asnaghi</surname><given-names>L</given-names></name><name><surname>Canti</surname><given-names>G</given-names></name><name><surname>Nicolin</surname><given-names>A</given-names></name></person-group>. <article-title>Rapamycin increases the cellular concentration of the BCL-2 protein and exerts an anti-apoptotic effect</article-title>. <source>Eur J Cancer</source>. (<year>2001</year>) <volume>37</volume>:<fpage>2121</fpage>&#x02013;<lpage>8</lpage>. <pub-id pub-id-type=\"doi\">10.1016/S0959-8049(01)00256-8</pub-id><pub-id pub-id-type=\"pmid\">11597394</pub-id></mixed-citation></ref><ref id=\"B52\"><label>52.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Choi</surname><given-names>SJ</given-names></name><name><surname>You</surname><given-names>HS</given-names></name><name><surname>Chung</surname><given-names>SY</given-names></name></person-group>. <article-title>Tacrolimus-Induced Apoptotic Signal Transduction Pathway</article-title>. <source>Transplant Proc</source>. (<year>2008</year>) <volume>40</volume>:<fpage>2734</fpage>&#x02013;<lpage>6</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.transproceed.2008.08.028</pub-id><pub-id pub-id-type=\"pmid\">18929848</pub-id></mixed-citation></ref><ref id=\"B53\"><label>53.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Pearce</surname><given-names>EL</given-names></name><name><surname>Walsh</surname><given-names>MC</given-names></name><name><surname>Cejas</surname><given-names>PJ</given-names></name><name><surname>Harms</surname><given-names>GM</given-names></name><name><surname>Shen</surname><given-names>H</given-names></name><name><surname>Wang</surname><given-names>LS</given-names></name><etal/></person-group>. <article-title>Enhancing CD8 T-cell memory by modulating fatty acid metabolism</article-title>. <source>Nature</source>. (<year>2009</year>) <volume>460</volume>:<fpage>103</fpage>&#x02013;<lpage>7</lpage>. <pub-id pub-id-type=\"doi\">10.1038/nature08097</pub-id><pub-id pub-id-type=\"pmid\">19494812</pub-id></mixed-citation></ref><ref id=\"B54\"><label>54.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Riou</surname><given-names>C</given-names></name><name><surname>Yassine-Diab</surname><given-names>B</given-names></name><name><surname>Van grevenynghe</surname><given-names>J</given-names></name><name><surname>Somogyi</surname><given-names>R</given-names></name><name><surname>Greller</surname><given-names>LD</given-names></name><name><surname>Gagnon</surname><given-names>D</given-names></name><etal/></person-group>. <article-title>Convergence of TCR and cytokine signaling leads to FOXO3a phosphorylation and drives the survival of CD4+ central memory T cells</article-title>. <source>J Exp Med.</source> (<year>2007</year>) <volume>204</volume>:<fpage>79</fpage>&#x02013;<lpage>91</lpage>. <pub-id pub-id-type=\"doi\">10.1084/jem.20061681</pub-id><pub-id pub-id-type=\"pmid\">17190839</pub-id></mixed-citation></ref><ref id=\"B55\"><label>55.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Rufer</surname><given-names>N</given-names></name><name><surname>Migliaccio</surname><given-names>M</given-names></name><name><surname>Antonchuk</surname><given-names>J</given-names></name><name><surname>Humphries</surname><given-names>RK</given-names></name><name><surname>Roosnek</surname><given-names>E</given-names></name><name><surname>Lansdorp</surname><given-names>PM</given-names></name></person-group>. <article-title>Transfer of the human telomerase reverse transcriptase (TERT) gene into T lymphocytes results in extension of replicative potential</article-title>. <source>Blood</source>. (<year>2001</year>) <volume>98</volume>:<fpage>597</fpage>&#x02013;<lpage>603</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood.V98.3.597</pub-id><pub-id pub-id-type=\"pmid\">11468156</pub-id></mixed-citation></ref><ref id=\"B56\"><label>56.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kaech</surname><given-names>SM</given-names></name><name><surname>Tan</surname><given-names>JT</given-names></name><name><surname>Wherry</surname><given-names>EJ</given-names></name><name><surname>Konieczny</surname><given-names>BT</given-names></name><name><surname>Surh</surname><given-names>CD</given-names></name><name><surname>Ahmed</surname><given-names>R</given-names></name></person-group>. <article-title>Selective expression of the interleukin 7 receptor identifies effector CD8 T cells that give rise to long-lived memory cells</article-title>. <source>Nat Immunol</source>. (<year>2003</year>) <volume>4</volume>:<fpage>1191</fpage>. <pub-id pub-id-type=\"doi\">10.1038/ni1009</pub-id><pub-id pub-id-type=\"pmid\">14625547</pub-id></mixed-citation></ref><ref id=\"B57\"><label>57.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Minty</surname><given-names>A</given-names></name><name><surname>Chalon</surname><given-names>P</given-names></name><name><surname>Derocq</surname><given-names>JM</given-names></name><name><surname>Dumont</surname><given-names>X</given-names></name><name><surname>Guillemot</surname><given-names>JC</given-names></name><name><surname>Kaghad</surname><given-names>M</given-names></name><etal/></person-group>\\n<article-title>lnterleukin-13 is a new human lymphokine regulating inflammatory and immune responses</article-title>. <source>Nature</source>. (<year>1993</year>) <volume>362</volume>:<fpage>248</fpage>\\n<pub-id pub-id-type=\"doi\">10.1038/362248a0</pub-id><pub-id pub-id-type=\"pmid\">8096327</pub-id></mixed-citation></ref><ref id=\"B58\"><label>58.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Chen</surname><given-names>ML</given-names></name><name><surname>Tsai</surname><given-names>TC</given-names></name><name><surname>Lin</surname><given-names>YY</given-names></name><name><surname>Tsai</surname><given-names>YM</given-names></name><name><surname>Wang</surname><given-names>LK</given-names></name><name><surname>Lee</surname><given-names>MC</given-names></name><etal/></person-group>. <article-title>Antipsychotic drugs suppress the AKT/NF-&#x003ba;B pathway and regulate the differentiation of T-cell subsets</article-title>. <source>Immunol Lett</source>. (<year>2011</year>) <volume>140</volume>:<fpage>81</fpage>&#x02013;<lpage>91</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.imlet.2011.06.011</pub-id><pub-id pub-id-type=\"pmid\">21763349</pub-id></mixed-citation></ref><ref id=\"B59\"><label>59.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Chen</surname><given-names>N-J</given-names></name><name><surname>Huang</surname><given-names>M-W</given-names></name><name><surname>Hsieh</surname><given-names>S-L</given-names></name></person-group>. <article-title>Enhanced secretion of IFN-&#x003b3; by activated Th1 cells occurs via reverse signaling through TNF-related activation-induced cytokine</article-title>. <source>J Immunol</source>. (<year>2001</year>) <volume>166</volume>:<fpage>270</fpage>. <pub-id pub-id-type=\"doi\">10.4049/jimmunol.166.1.270</pub-id><pub-id pub-id-type=\"pmid\">11123302</pub-id></mixed-citation></ref><ref id=\"B60\"><label>60.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Best</surname><given-names>JA</given-names></name><name><surname>Blair</surname><given-names>DA</given-names></name><name><surname>Knell</surname><given-names>J</given-names></name><name><surname>Yang</surname><given-names>E</given-names></name><name><surname>Mayya</surname><given-names>V</given-names></name><name><surname>Doedens</surname><given-names>A</given-names></name><etal/></person-group>. <article-title>Transcriptional insights into the CD8+ T cell response to infection and memory T cell formation</article-title>. <source>Nat Immunol</source>. (<year>2013</year>) <volume>14</volume>:<fpage>404</fpage>. <pub-id pub-id-type=\"doi\">10.1038/ni.2536</pub-id><pub-id pub-id-type=\"pmid\">23396170</pub-id></mixed-citation></ref><ref id=\"B61\"><label>61.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Xiao</surname><given-names>Y</given-names></name><name><surname>Yu</surname><given-names>S</given-names></name><name><surname>Zhu</surname><given-names>B</given-names></name><name><surname>Bedoret</surname><given-names>D</given-names></name><name><surname>Bu</surname><given-names>X</given-names></name><name><surname>Francisco</surname><given-names>LM</given-names></name><etal/></person-group>. <article-title>RGMb is a novel binding partner for PD-L2 and its engagement with PD-L2 promotes respiratory tolerance</article-title>. <source>J Exp Med</source>. (<year>2014</year>) <volume>211</volume>:<fpage>943</fpage>&#x02013;<lpage>59</lpage>. <pub-id pub-id-type=\"doi\">10.1084/jem.20130790</pub-id><pub-id pub-id-type=\"pmid\">24752301</pub-id></mixed-citation></ref><ref id=\"B62\"><label>62.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gaur</surname><given-names>U</given-names></name><name><surname>Aggarwal</surname><given-names>BB</given-names></name></person-group>. <article-title>Regulation of proliferation, survival and apoptosis by members of the TNF superfamily</article-title>. <source>Biochem Pharmacol</source>. (<year>2003</year>) <volume>66</volume>:<fpage>1403</fpage>&#x02013;<lpage>8</lpage>. <pub-id pub-id-type=\"doi\">10.1016/S0006-2952(03)00490-8</pub-id><pub-id pub-id-type=\"pmid\">14555214</pub-id></mixed-citation></ref><ref id=\"B63\"><label>63.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Chen</surname><given-names>Z</given-names></name><name><surname>Stelekati</surname><given-names>E</given-names></name><name><surname>Kurachi</surname><given-names>M</given-names></name><name><surname>Yu</surname><given-names>S</given-names></name><name><surname>Cai</surname><given-names>Z</given-names></name><name><surname>Manne</surname><given-names>S</given-names></name><etal/></person-group>. <article-title>miR-150 regulates memory CD8 T cell differentiation via c-Myb</article-title>. <source>Cell Rep</source>. (<year>2017</year>) <volume>20</volume>:<fpage>2584</fpage>&#x02013;<lpage>97</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.celrep.2017.08.060</pub-id><pub-id pub-id-type=\"pmid\">28903040</pub-id></mixed-citation></ref><ref id=\"B64\"><label>64.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Naderi</surname><given-names>A</given-names></name><name><surname>Liu</surname><given-names>J</given-names></name><name><surname>Bennett</surname><given-names>IC</given-names></name></person-group>. <article-title>BEX2 regulates mitochondrial apoptosis and G1 cell cycle in breast cancer</article-title>. <source>Int J Cancer.</source> (<year>2009</year>) <volume>126</volume>:<fpage>1596</fpage>&#x02013;<lpage>610</lpage>. <pub-id pub-id-type=\"doi\">10.1002/ijc.24866</pub-id><pub-id pub-id-type=\"pmid\">19711341</pub-id></mixed-citation></ref><ref id=\"B65\"><label>65.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Behbakht</surname><given-names>K</given-names></name><name><surname>Qamar</surname><given-names>L</given-names></name><name><surname>Aldridge</surname><given-names>CS</given-names></name><name><surname>Coletta</surname><given-names>RD</given-names></name><name><surname>Davidson</surname><given-names>SA</given-names></name><name><surname>Thorburn</surname><given-names>A</given-names></name><etal/></person-group>. <article-title>Six1 overexpression in ovarian carcinoma causes resistance to TRAIL-mediated apoptosis and is associated with poor survival</article-title>. <source>Cancer Res</source>. (<year>2007</year>) <volume>67</volume>:<fpage>3036</fpage>&#x02013;<lpage>42</lpage>. <pub-id pub-id-type=\"doi\">10.1158/0008-5472.CAN-06-3755</pub-id><pub-id pub-id-type=\"pmid\">17409410</pub-id></mixed-citation></ref><ref id=\"B66\"><label>66.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Doedens</surname><given-names>AL</given-names></name><name><surname>Phan</surname><given-names>AT</given-names></name><name><surname>Stradner</surname><given-names>MH</given-names></name><name><surname>Fujimoto</surname><given-names>JK</given-names></name><name><surname>Nguyen</surname><given-names>JV</given-names></name><name><surname>Yang</surname><given-names>E</given-names></name><etal/></person-group>. <article-title>Hypoxia-inducible factors enhance the effector responses of CD8+ T cells to persistent antigen</article-title>. <source>Nat Immunol</source>. (<year>2013</year>) <volume>14</volume>:<fpage>1173</fpage>&#x02013;<lpage>82</lpage>. <pub-id pub-id-type=\"doi\">10.1038/ni.2714</pub-id><pub-id pub-id-type=\"pmid\">24076634</pub-id></mixed-citation></ref><ref id=\"B67\"><label>67.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Vafai</surname><given-names>SB</given-names></name><name><surname>Mootha</surname><given-names>VK</given-names></name></person-group>. <article-title>Mitochondrial disorders as windows into an ancient organelle</article-title>. <source>Nature</source>. (<year>2012</year>) <volume>491</volume>:<fpage>374</fpage>&#x02013;<lpage>83</lpage>. <pub-id pub-id-type=\"doi\">10.1038/nature11707</pub-id><pub-id pub-id-type=\"pmid\">23151580</pub-id></mixed-citation></ref><ref id=\"B68\"><label>68.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Sukumar</surname><given-names>M</given-names></name><name><surname>Liu</surname><given-names>J</given-names></name><name><surname>Ji</surname><given-names>Y</given-names></name><name><surname>Subramanian</surname><given-names>M</given-names></name><name><surname>Crompton</surname><given-names>JG</given-names></name><name><surname>Yu</surname><given-names>Z</given-names></name><etal/></person-group>. <article-title>Inhibiting glycolytic metabolism enhances CD8+ T cell memory and antitumor function</article-title>. <source>J Clin Invest.</source> (<year>2013</year>) <volume>123</volume>:<fpage>4479</fpage>&#x02013;<lpage>88</lpage>. <pub-id pub-id-type=\"doi\">10.1172/JCI69589</pub-id><pub-id pub-id-type=\"pmid\">24091329</pub-id></mixed-citation></ref><ref id=\"B69\"><label>69.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gubser</surname><given-names>PM</given-names></name><name><surname>Bantug</surname><given-names>GR</given-names></name><name><surname>Razik</surname><given-names>L</given-names></name><name><surname>Fischer</surname><given-names>M</given-names></name><name><surname>Dimeloe</surname><given-names>S</given-names></name><name><surname>Hoenger</surname><given-names>G</given-names></name><etal/></person-group>. <article-title>Rapid effector function of memory CD8+ T cells requires an immediate-early glycolytic switch</article-title>. <source>Nat Immunol.</source> (<year>2013</year>) <volume>14</volume>:<fpage>1064</fpage>&#x02013;<lpage>72</lpage>. <pub-id pub-id-type=\"doi\">10.1038/ni.2687</pub-id><pub-id pub-id-type=\"pmid\">23955661</pub-id></mixed-citation></ref><ref id=\"B70\"><label>70.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Araki</surname><given-names>Y</given-names></name><name><surname>Wang</surname><given-names>Z</given-names></name><name><surname>Zang</surname><given-names>C</given-names></name><name><surname>Wood</surname><given-names>WH</given-names></name><name><surname>Schones</surname><given-names>D</given-names></name><name><surname>Cui</surname><given-names>K</given-names></name><etal/></person-group>. <article-title>Genome-wide analysis of histone methylation reveals chromatin state-based regulation of gene transcription and function of memory CD8+ T cells</article-title>. <source>Immunity.</source> (<year>2009</year>) <volume>30</volume>:<fpage>912</fpage>&#x02013;<lpage>25</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.immuni.2009.05.006</pub-id><pub-id pub-id-type=\"pmid\">19523850</pub-id></mixed-citation></ref><ref id=\"B71\"><label>71.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Douglas</surname><given-names>RS</given-names></name><name><surname>Gianoukakis</surname><given-names>AG</given-names></name><name><surname>Kamat</surname><given-names>S</given-names></name><name><surname>Smith</surname><given-names>TJ</given-names></name></person-group>. <article-title>Aberrant expression of the insulin-like growth factor-1 receptor by T cells from patients with Graves\\' Disease may carry functional consequences for disease pathogenesis</article-title>. <source>J Immunol</source>. (<year>2007</year>) <volume>178</volume>:<fpage>3281</fpage>. <pub-id pub-id-type=\"doi\">10.4049/jimmunol.178.5.3281</pub-id><pub-id pub-id-type=\"pmid\">17312178</pub-id></mixed-citation></ref><ref id=\"B72\"><label>72.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wherry</surname><given-names>EJ</given-names></name><name><surname>Ha</surname><given-names>S-J</given-names></name><name><surname>Kaech</surname><given-names>SM</given-names></name><name><surname>Haining</surname><given-names>WN</given-names></name><name><surname>Sarkar</surname><given-names>S</given-names></name><name><surname>Kalia</surname><given-names>V</given-names></name><etal/></person-group>. <article-title>Molecular signature of CD8+ T cell exhaustion during chronic viral infection</article-title>. <source>Immunity</source>. (<year>2007</year>) <volume>27</volume>:<fpage>670</fpage>&#x02013;<lpage>84</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.immuni.2007.09.006</pub-id><pub-id pub-id-type=\"pmid\">17950003</pub-id></mixed-citation></ref><ref id=\"B73\"><label>73.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Li</surname><given-names>G</given-names></name><name><surname>Weyand</surname><given-names>CM</given-names></name><name><surname>Goronzy</surname><given-names>JJ</given-names></name></person-group>. <article-title>Epigenetic mechanisms of age-dependent KIR2DL4 expression in T cells</article-title>. <source>J Leuk Biol</source>. (<year>2008</year>) <volume>84</volume>:<fpage>824</fpage>&#x02013;<lpage>34</lpage>. <pub-id pub-id-type=\"doi\">10.1189/jlb.0807583</pub-id><pub-id pub-id-type=\"pmid\">18586981</pub-id></mixed-citation></ref><ref id=\"B74\"><label>74.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gattinoni</surname><given-names>L</given-names></name><name><surname>Klebanoff</surname><given-names>CA</given-names></name><name><surname>Restifo</surname><given-names>NP</given-names></name></person-group>. <article-title>Pharmacologic induction of CD8+ T cell memory: better living through chemistry</article-title>. <source>Sci Transl Med.</source> (<year>2009</year>) <volume>1</volume>:<fpage>11ps12</fpage>. <pub-id pub-id-type=\"doi\">10.1126/scitranslmed.3000302</pub-id><pub-id pub-id-type=\"pmid\">20371454</pub-id></mixed-citation></ref><ref id=\"B75\"><label>75.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kondo</surname><given-names>T</given-names></name><name><surname>Takiguchi</surname><given-names>M</given-names></name></person-group>. <article-title>Human memory CCR4+CD8+ T cell subset has the ability to produce multiple cytokines</article-title>. <source>Int Immunol</source>. (<year>2009</year>) <volume>21</volume>:<fpage>523</fpage>&#x02013;<lpage>32</lpage>. <pub-id pub-id-type=\"doi\">10.1093/intimm/dxp019</pub-id><pub-id pub-id-type=\"pmid\">19261691</pub-id></mixed-citation></ref><ref id=\"B76\"><label>76.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Lazuardi</surname><given-names>L</given-names></name><name><surname>Herndler-Brandstetter</surname><given-names>D</given-names></name><name><surname>Brunner</surname><given-names>S</given-names></name><name><surname>Laschober</surname><given-names>GT</given-names></name><name><surname>Lepperdinger</surname><given-names>G</given-names></name><name><surname>Grubeck-Loebenstein</surname><given-names>B</given-names></name></person-group>\\n<article-title>Microarray analysis reveals similarity between CD8+CD28&#x02013; T cells from young and elderly persons, but not of CD8+CD28+ T cells</article-title>. <source>Biogerontology</source>. (<year>2009</year>) <volume>10</volume>:<fpage>191</fpage>&#x02013;<lpage>202</lpage>. <pub-id pub-id-type=\"doi\">10.1007/s10522-008-9167-1</pub-id><pub-id pub-id-type=\"pmid\">18751903</pub-id></mixed-citation></ref><ref id=\"B77\"><label>77.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Mohandas</surname></name><name><surname>Zola</surname><given-names>H</given-names></name><name><surname>Barry</surname><given-names>S</given-names></name><name><surname>Krumbiegel</surname><given-names>D</given-names></name></person-group>\\n<article-title>Peptidase inhibitor 16 identifies a unique subset of memory T helper cells with hyperproliferative and proinflammatory properties (IRC8P.477)</article-title>. <source>J Immunol</source>. (<year>2014</year>) <volume>192</volume>:<fpage>190</fpage>.5.</mixed-citation></ref><ref id=\"B78\"><label>78.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Komori</surname><given-names>HK</given-names></name><name><surname>Hart</surname><given-names>T</given-names></name><name><surname>LaMere</surname><given-names>SA</given-names></name><name><surname>Chew</surname><given-names>PV</given-names></name><name><surname>Salomon</surname><given-names>DR</given-names></name></person-group>. <article-title>Defining CD4 T cell memory by the epigenetic landscape of CpG DNA methylation</article-title>. <source>J Immunol.</source> (<year>2015</year>) <volume>194</volume>:<fpage>1565</fpage>&#x02013;<lpage>79</lpage>. <pub-id pub-id-type=\"doi\">10.4049/jimmunol.1401162</pub-id><pub-id pub-id-type=\"pmid\">25576597</pub-id></mixed-citation></ref><ref id=\"B79\"><label>79.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Wang</surname><given-names>G</given-names></name><name><surname>Zhao</surname><given-names>N</given-names></name><name><surname>Berkhout</surname><given-names>B</given-names></name><name><surname>Das</surname><given-names>AT</given-names></name></person-group>. <article-title>A combinatorial CRISPR-Cas9 attack on HIV-1 DNA extinguishes all infectious provirus in infected T cell cultures</article-title>. <source>Cell Rep.</source> (<year>2016</year>) <volume>17</volume>:<fpage>2819</fpage>&#x02013;<lpage>26</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.celrep.2016.11.057</pub-id><pub-id pub-id-type=\"pmid\">27974196</pub-id></mixed-citation></ref><ref id=\"B80\"><label>80.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Qi</surname><given-names>Q</given-names></name><name><surname>Liu</surname><given-names>Y</given-names></name><name><surname>Cheng</surname><given-names>Y</given-names></name><name><surname>Glanville</surname><given-names>J</given-names></name><name><surname>Zhang</surname><given-names>D</given-names></name><name><surname>Lee</surname><given-names>JY</given-names></name><etal/></person-group>. <article-title>Diversity and clonal selection in the human T-cell repertoire</article-title>. <source>Proc Natl Acad Sci USA</source>. (<year>2014</year>) <volume>111</volume>:<fpage>13139</fpage>&#x02013;<lpage>44</lpage>. <pub-id pub-id-type=\"doi\">10.1073/pnas.1409155111</pub-id><pub-id pub-id-type=\"pmid\">25157137</pub-id></mixed-citation></ref><ref id=\"B81\"><label>81.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Youngblood</surname><given-names>B</given-names></name><name><surname>Hale</surname><given-names>JS</given-names></name><name><surname>Kissick</surname><given-names>HT</given-names></name><name><surname>Ahn</surname><given-names>E</given-names></name><name><surname>Xu</surname><given-names>X</given-names></name><name><surname>Wieland</surname><given-names>A</given-names></name><etal/></person-group>. <article-title>Effector CD8 T cells dedifferentiate into long-lived memory cells</article-title>. <source>Nature.</source> (<year>2017</year>) <volume>552</volume>:<fpage>404</fpage>&#x02013;<lpage>9</lpage>. <pub-id pub-id-type=\"doi\">10.1038/nature25144</pub-id><pub-id pub-id-type=\"pmid\">29236683</pub-id></mixed-citation></ref><ref id=\"B82\"><label>82.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kern</surname><given-names>F</given-names></name><name><surname>Bunde</surname><given-names>T</given-names></name><name><surname>Faulhaber</surname><given-names>N</given-names></name><name><surname>Kiecker</surname><given-names>F</given-names></name><name><surname>Khatamzas</surname><given-names>E</given-names></name><name><surname>Rudawski</surname><given-names>IM</given-names></name><etal/></person-group>. <article-title>Cytomegalovirus (CMV) phosphoprotein 65 makes a large contribution to shaping the T cell repertoire in CMV-exposed individuals</article-title>. <source>J Infect Dis.</source> (<year>2002</year>) <volume>185</volume>:<fpage>1709</fpage>&#x02013;<lpage>16</lpage>. <pub-id pub-id-type=\"doi\">10.1086/340637</pub-id><pub-id pub-id-type=\"pmid\">12085315</pub-id></mixed-citation></ref><ref id=\"B83\"><label>83.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Higdon</surname><given-names>LE</given-names></name><name><surname>Trofe-Clark</surname><given-names>J</given-names></name><name><surname>Liu</surname><given-names>S</given-names></name><name><surname>Margulies</surname><given-names>KB</given-names></name><name><surname>Sahoo</surname><given-names>MK</given-names></name><name><surname>Blumberg</surname><given-names>E</given-names></name><etal/></person-group>. <article-title>Cytomegalovirus responsive CD8+ T cells expand after solid organ transplantation in the absence of CMV disease</article-title>. <source>Am J Transplant.</source> (<year>2017</year>) <volume>17</volume>:<fpage>2045</fpage>&#x02013;<lpage>54</lpage>. <pub-id pub-id-type=\"doi\">10.1111/ajt.14227</pub-id><pub-id pub-id-type=\"pmid\">28199780</pub-id></mixed-citation></ref><ref id=\"B84\"><label>84.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Humar</surname><given-names>A</given-names></name><name><surname>Snydman</surname><given-names>D</given-names></name></person-group>\\n<article-title>AST Infectious Diseases Community of Practice. Cytomegalovirus in solid organ transplant recipients</article-title>. <source>Am J Transplant.</source> (<year>2009</year>) <volume>9</volume> (<issue>Suppl. 4</issue>):<fpage>S78</fpage>&#x02013;<lpage>86</lpage>. <pub-id pub-id-type=\"doi\">10.1111/j.1600-6143.2009.02897.x</pub-id><pub-id pub-id-type=\"pmid\">20070700</pub-id></mixed-citation></ref><ref id=\"B85\"><label>85.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Taherimahmoudi</surname><given-names>M</given-names></name><name><surname>Ahmadi</surname><given-names>H</given-names></name><name><surname>Baradaran</surname><given-names>N</given-names></name><name><surname>Montaser-Kouhsari</surname><given-names>L</given-names></name><name><surname>Salem</surname><given-names>S</given-names></name><name><surname>Mehrsai</surname><given-names>A</given-names></name><etal/></person-group>. <article-title>Cytomegalovirus infection and disease following renal transplantation: preliminary report of incidence and potential risk factors</article-title>. <source>Transplant Proc.</source> (<year>2009</year>) <volume>41</volume>:<fpage>2841</fpage>&#x02013;<lpage>4</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.transproceed.2009.07.027</pub-id><pub-id pub-id-type=\"pmid\">19765452</pub-id></mixed-citation></ref><ref id=\"B86\"><label>86.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Gattinoni</surname><given-names>L</given-names></name><name><surname>Zhong</surname><given-names>XS</given-names></name><name><surname>Palmer</surname><given-names>DC</given-names></name><name><surname>Ji</surname><given-names>Y</given-names></name><name><surname>Hinrichs</surname><given-names>CS</given-names></name><name><surname>Yu</surname><given-names>Z</given-names></name><etal/></person-group>. <article-title>Wnt signaling arrests effector T cell differentiation and generates CD8+ memory stem cells</article-title>. <source>Nat Med.</source> (<year>2009</year>) <volume>15</volume>:<fpage>808</fpage>&#x02013;<lpage>13</lpage>. <pub-id pub-id-type=\"doi\">10.1038/nm.1982</pub-id><pub-id pub-id-type=\"pmid\">19525962</pub-id></mixed-citation></ref><ref id=\"B87\"><label>87.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kondo</surname><given-names>T</given-names></name><name><surname>Morita</surname><given-names>R</given-names></name><name><surname>Okuzono</surname><given-names>Y</given-names></name><name><surname>Nakatsukasa</surname><given-names>H</given-names></name><name><surname>Sekiya</surname><given-names>T</given-names></name><name><surname>Chikuma</surname><given-names>S</given-names></name><etal/></person-group>. <article-title>Notch-mediated conversion of activated T cells into stem cell memory-like T cells for adoptive immunotherapy</article-title>. <source>Nat Commun.</source> (<year>2017</year>) <volume>8</volume>:<fpage>15338</fpage>. <pub-id pub-id-type=\"doi\">10.1038/ncomms15338</pub-id><pub-id pub-id-type=\"pmid\">28530241</pub-id></mixed-citation></ref><ref id=\"B88\"><label>88.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Abu Eid</surname><given-names>R</given-names></name><name><surname>Friedman</surname><given-names>KM</given-names></name><name><surname>Mkrtichyan</surname><given-names>M</given-names></name><name><surname>Walens</surname><given-names>A</given-names></name><name><surname>King</surname><given-names>W</given-names></name><name><surname>Janik</surname><given-names>J</given-names></name><etal/></person-group>. <article-title>Akt1 and&#x02212;2 inhibition diminishes terminal differentiation and enhances central memory CD8+ T-cell proliferation and survival</article-title>. <source>Oncoimmunology</source>. (<year>2015</year>) <volume>4</volume>:<fpage>e1005448</fpage>. <pub-id pub-id-type=\"doi\">10.1080/2162402X.2015.1005448</pub-id><pub-id pub-id-type=\"pmid\">26155399</pub-id></mixed-citation></ref><ref id=\"B89\"><label>89.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Abu Eid</surname><given-names>R</given-names></name><name><surname>Ahmad</surname><given-names>S</given-names></name><name><surname>Lin</surname><given-names>Y</given-names></name><name><surname>Webb</surname><given-names>M</given-names></name><name><surname>Berrong</surname><given-names>Z</given-names></name><name><surname>Shrimali</surname><given-names>R</given-names></name><etal/></person-group>. <article-title>Enhanced therapeutic efficacy and memory of tumor-specific CD8 T cells by <italic>ex vivo</italic> PI3K-delta inhibition</article-title>. <source>Cancer Res.</source> (<year>2017</year>) <volume>77</volume>:<fpage>4135</fpage>&#x02013;<lpage>45</lpage>. <pub-id pub-id-type=\"doi\">10.1158/0008-5472.CAN-16-1925</pub-id><pub-id pub-id-type=\"pmid\">28615225</pub-id></mixed-citation></ref><ref id=\"B90\"><label>90.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kagoya</surname><given-names>Y</given-names></name><name><surname>Nakatsugawa</surname><given-names>M</given-names></name><name><surname>Yamashita</surname><given-names>Y</given-names></name><name><surname>Ochi</surname><given-names>T</given-names></name><name><surname>Guo</surname><given-names>T</given-names></name><name><surname>Anczurowski</surname><given-names>M</given-names></name><etal/></person-group>. <article-title>BET bromodomain inhibition enhances T cell persistence and function in adoptive immunotherapy models</article-title>. <source>J Clin Invest.</source> (<year>2016</year>) <volume>126</volume>:<fpage>3479</fpage>&#x02013;<lpage>94</lpage>. <pub-id pub-id-type=\"doi\">10.1172/JCI86437</pub-id><pub-id pub-id-type=\"pmid\">27548527</pub-id></mixed-citation></ref><ref id=\"B91\"><label>91.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Orio</surname><given-names>J</given-names></name><name><surname>Carli</surname><given-names>C</given-names></name><name><surname>Janelle</surname><given-names>V</given-names></name><name><surname>Giroux</surname><given-names>M</given-names></name><name><surname>Taillefer</surname><given-names>J</given-names></name><name><surname>Goupil</surname><given-names>M</given-names></name><etal/></person-group>. <article-title>Early exposure to interleukin-21 limits rapidly generated anti-Epstein-Barr virus T-cell line differentiation</article-title>. <source>Cytotherapy</source>. (<year>2015</year>) <volume>17</volume>:<fpage>496</fpage>&#x02013;<lpage>508</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.jcyt.2014.12.009</pub-id><pub-id pub-id-type=\"pmid\">25661862</pub-id></mixed-citation></ref><ref id=\"B92\"><label>92.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Kaartinen</surname><given-names>T</given-names></name><name><surname>Luostarinen</surname><given-names>A</given-names></name><name><surname>Maliniemi</surname><given-names>P</given-names></name><name><surname>Keto</surname><given-names>J</given-names></name><name><surname>Arvas</surname><given-names>M</given-names></name><name><surname>Belt</surname><given-names>H</given-names></name><etal/></person-group>\\n<article-title>Low interleukin-2 concentration favors generation of early memory T cells over effector phenotypes during chimeric antigen receptor T-cell expansion</article-title>. <source>Cytotherapy</source>. (<year>2017</year>) <volume>19</volume>:<fpage>689</fpage>&#x02013;<lpage>702</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.jcyt.2017.03.067</pub-id><pub-id pub-id-type=\"pmid\">28411126</pub-id></mixed-citation></ref><ref id=\"B93\"><label>93.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ji</surname><given-names>Y</given-names></name><name><surname>Hocker</surname><given-names>JD</given-names></name><name><surname>Gattinoni</surname><given-names>L</given-names></name></person-group>. <article-title>Enhancing adoptive T cell immunotherapy with microRNA therapeutics</article-title>. <source>Semin Immunol</source>. (<year>2016</year>) <volume>28</volume>:<fpage>45</fpage>&#x02013;<lpage>53</lpage>. <pub-id pub-id-type=\"doi\">10.1016/j.smim.2015.11.006</pub-id><pub-id pub-id-type=\"pmid\">26710685</pub-id></mixed-citation></ref><ref id=\"B94\"><label>94.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Singh</surname><given-names>DK</given-names></name><name><surname>Dwivedi</surname><given-names>VP</given-names></name><name><surname>Ranganathan</surname><given-names>A</given-names></name><name><surname>Bishai</surname><given-names>WR</given-names></name><name><surname>Van Kaer</surname><given-names>L</given-names></name><name><surname>Das</surname><given-names>G</given-names></name></person-group>. <article-title>Blockade of the Kv1.3 K+ channel enhances BCG vaccine efficacy by expanding central memory T lymphocytes</article-title>. <source>J Infect Dis.</source> (<year>2016</year>) <volume>214</volume>:<fpage>1456</fpage>&#x02013;<lpage>64</lpage>. <pub-id pub-id-type=\"doi\">10.1093/infdis/jiw395</pub-id><pub-id pub-id-type=\"pmid\">27571906</pub-id></mixed-citation></ref><ref id=\"B95\"><label>95.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Puleston</surname><given-names>DJ</given-names></name><name><surname>Zhang</surname><given-names>H</given-names></name><name><surname>Powell</surname><given-names>TJ</given-names></name><name><surname>Lipina</surname><given-names>E</given-names></name><name><surname>Sims</surname><given-names>S</given-names></name><name><surname>Panse</surname><given-names>I</given-names></name><etal/></person-group>. <article-title>Autophagy is a critical regulator of memory CD8(+) T cell formation</article-title>. <source>Elife.</source> (<year>2014</year>) <volume>3</volume>:<fpage>e03706</fpage>. 3. <pub-id pub-id-type=\"doi\">10.7554/eLife.03706</pub-id><pub-id pub-id-type=\"pmid\">25385531</pub-id></mixed-citation></ref><ref id=\"B96\"><label>96.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Ricciardelli</surname><given-names>I</given-names></name><name><surname>Brewin</surname><given-names>J</given-names></name><name><surname>Lugthart</surname><given-names>G</given-names></name><name><surname>Albon</surname><given-names>SJ</given-names></name><name><surname>Pule</surname><given-names>M</given-names></name><name><surname>Amrolia</surname><given-names>PJ</given-names></name></person-group>. <article-title>Rapid generation of EBV-specific cytotoxic T lymphocytes resistant to calcineurin inhibitors for adoptive immunotherapy</article-title>. <source>Am J Transplant</source>. (<year>2013</year>) <volume>13</volume>:<fpage>3244</fpage>&#x02013;<lpage>52</lpage>. <pub-id pub-id-type=\"doi\">10.1111/ajt.12475</pub-id><pub-id pub-id-type=\"pmid\">24266973</pub-id></mixed-citation></ref><ref id=\"B97\"><label>97.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Menger</surname><given-names>L</given-names></name><name><surname>Gouble</surname><given-names>A</given-names></name><name><surname>Marzolini</surname><given-names>MA</given-names></name><name><surname>Pachnio</surname><given-names>A</given-names></name><name><surname>Bergerhoff</surname><given-names>K</given-names></name><name><surname>Henry</surname><given-names>JY</given-names></name><etal/></person-group>. <article-title>TALEN-mediated genetic inactivation of the glucocorticoid receptor in cytomegalovirus-specific T cells</article-title>. <source>Blood</source>. (<year>2015</year>) <volume>126</volume>:<fpage>2781</fpage>&#x02013;<lpage>9</lpage>. <pub-id pub-id-type=\"doi\">10.1182/blood-2015-08-664755</pub-id><pub-id pub-id-type=\"pmid\">26508783</pub-id></mixed-citation></ref><ref id=\"B98\"><label>98.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Zhan</surname><given-names>H</given-names></name><name><surname>Gilmour</surname><given-names>K</given-names></name><name><surname>Chan</surname><given-names>L</given-names></name><name><surname>Farzaneh</surname><given-names>F</given-names></name><name><surname>McNicol</surname><given-names>AM</given-names></name><name><surname>Xu</surname><given-names>JH</given-names></name><etal/></person-group>. <article-title>Production and first-in-man use of T cells engineered to express a HSVTK-CD34 sort-suicide gene</article-title>. <source>PLoS ONE</source>. (<year>2013</year>) <volume>8</volume>:<fpage>e77106</fpage>. <pub-id pub-id-type=\"doi\">10.1371/journal.pone.0077106</pub-id><pub-id pub-id-type=\"pmid\">24204746</pub-id></mixed-citation></ref><ref id=\"B99\"><label>99.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Rupp</surname><given-names>LJ</given-names></name><name><surname>Schumann</surname><given-names>K</given-names></name><name><surname>Roybal</surname><given-names>KT</given-names></name><name><surname>Gate</surname><given-names>RE</given-names></name><name><surname>Ye</surname><given-names>CJ</given-names></name><name><surname>Lim</surname><given-names>WA</given-names></name><etal/></person-group>. <article-title>CRISPR/Cas9-mediated PD-1 disruption enhances anti-tumor efficacy of human chimeric antigen receptor T cells</article-title>. <source>Sci Rep</source>. (<year>2017</year>) <volume>7</volume>:<fpage>737</fpage>. <pub-id pub-id-type=\"doi\">10.1038/s41598-017-00462-8</pub-id><pub-id pub-id-type=\"pmid\">28389661</pub-id></mixed-citation></ref><ref id=\"B100\"><label>100.</label><mixed-citation publication-type=\"journal\"><person-group person-group-type=\"author\"><name><surname>Zhang</surname><given-names>Y</given-names></name><name><surname>Zhang</surname><given-names>X</given-names></name><name><surname>Cheng</surname><given-names>C</given-names></name><name><surname>Mu</surname><given-names>W</given-names></name><name><surname>Liu</surname><given-names>X</given-names></name><name><surname>Li</surname><given-names>N</given-names></name><etal/></person-group>. <article-title>CRISPR-Cas9 mediated LAG-3 disruption in CAR-T cells</article-title>. <source>Front Med.</source> (<year>2017</year>) <volume>11</volume>:<fpage>554</fpage>&#x02013;<lpage>62</lpage>. <pub-id pub-id-type=\"doi\">10.1007/s11684-017-0543-6</pub-id><pub-id pub-id-type=\"pmid\">28625015</pub-id></mixed-citation></ref></ref-list></back></article>'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tar = tarfile.open(osp.join(oadir, 'comm_use.C-H.xml.tar.gz'), \"r:gz\")\n",
    "text = tar.extractfile(tar.getmember('Front_Immunol/PMC6546853.nxml')).read().decode('utf-8', errors='ignore')\n",
    "text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run src/integration.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft = parse_nxml(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1 entries, 0 to 0\n",
      "Data columns (total 11 columns):\n",
      "abstract          1 non-null object\n",
      "body              1 non-null object\n",
      "date_accepted     1 non-null datetime64[ns]\n",
      "date_pub          1 non-null datetime64[ns]\n",
      "date_received     1 non-null datetime64[ns]\n",
      "id_doi            1 non-null object\n",
      "id_pmc            1 non-null object\n",
      "id_pmid           1 non-null object\n",
      "journal_ids       1 non-null object\n",
      "journal_titles    1 non-null object\n",
      "title             1 non-null object\n",
      "dtypes: datetime64[ns](3), object(8)\n",
      "memory usage: 168.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "dft.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    Front. Immunol.||Front Immunol\n",
       "Name: journal_ids, dtype: object"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft['journal_ids']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"IntroductionSevere viral infections have a major impact on the clinical course of immunocompromised patients. Despite availability of powerful antiviral medication, cytomegalovirus (CMV) still accounts for significant morbidity and mortality in solid organ transplant (SOT) recipients. CMV can trigger direct and indirect morbidities such as chronic allograft rejection or in the case of kidney transplantation (KTx) chronic nephropathy. Therapeutic control of CMV may be hampered by the development of anti-viral drug resistance. Moreover, after discontinuation of anti-viral prophylaxis, late-onset CMV disease frequently occurs and overall mortality is significantly higher in CMV-infected compared to uninfected KTx patients. Of note, T-cell-mediated anti-CMV immunity was reported to be predictive for the development of late-onset disease  and anti-CMVIE−1-specific CD8+ T-cell responses stratify risk of CMV disease in heart and lung transplant as well as KTx patients. In addition, the magnitude of the CMVpp65- and CMVIE−1-specific T-cell responses turned out to be protective against complications with CMV in hematopoietic stem cell transplantation (HSCT). Consequently, regeneration of the endogenous T-cell response against these antigens, as aspired by AVTT, may prevent and reduce virus-associated morbidities/mortality in the SOT setting. Other viruses with impact on SOT outcomes are Epstein-Barr-virus and BK-virus, for which less efficient or no antiviral drugs are available. T-cells play a key role in protection from severe viral infections. Thus, adoptive T-cell therapy (AVTT) is a potent novel treatment strategy to tackle fatal viral complications in immunosuppressed transplant patients. Mechanisms of success or failure of new AVTT approaches need to be thoroughly understood and specific characteristics of patient cohorts have to be considered for successful translation of AVTT.For clinical application of AVTT, ex vivo enrichment and expansion of virus-specific T-cells under GMP conditions are crucial and thus various protocols have been developed for CMV-specific AVTT after HSCT. However, the success of these approaches is limited in SOT patients due to the T-cell products (TCPs) being derived from patients instead of healthy HSCT donors, the lack of lymphodepletive preconditioning and the need for concomitant immunosuppression. Nevertheless, we and other groups demonstrated not only safety of AVTT, but also significant reduction of viral load and control of clinical symptoms of CMV disease in SOT recipients under maintenance immunosuppression in proof-of-concept studies. These observations are in line with positive results of AVTT for treatment of patients with EBV-related post-transplant lymphoproliferative disease. Yet, long-term efficacy failed in some patients, who experienced recurrence of CMV or EBV load and symptoms. To adapt AVTT to combat these clinical challenges, it is crucial to consider the respective patient cohort and the TCPs' characteristics. Specifically, our aim was to increase longevity of transferred T-cells to improve sustainability of clinical efficacy of AVTT in SOT patients. Failure of long-term control of CMV/EBV infections may be due to limited persistence of adoptively transferred T-cells in vivo, which might occur due to the late differentiation state implying limited longevity of infused T-cells. Therefore, advancing the quality of adoptively transferred TCPs with defined compositions by the enrichment for distinct T-cell memory subsets may improve therapeutic outcome. In particular, central-memory T-cells (TCM; CCR7+ CD62L+ CD45RO+ CD45RA−) and memory-stem T-cells (TSCM; CCR7+ CD62L+ CD45RO− CD45RA+ CD95+) have high proliferative potential, self-renewal capacity and are reported to show superior engraftment, persistence, and survival compared to more differentiated memory T-cells. Conversely, late-differentiated short-lived effector-memory T-cells (TEM; CCR7− CD62L− CD45RA−CD45RO+) and terminally-differentiated effector T-cells (TEMRA; CCR7− CD62L− CD45RA+CD45RO−) exert immediate effector function, but fail to establish long-lasting protective memory, because of poor proliferative potential and limited survival following antigenic rechallenge. Remarkably, these observations match clinical data demonstrating T-cell reconstitution after HSCT and prevention of CMV disease related to TCM proportions in peripheral blood. Direct sorting strategies to isolate only CMV-specific long-lived T-cells are barely feasible under GMP conditions and would yield very small cell numbers likely not sufficient for successful AVTT in immunosuppressed SOT recipients. To ensure applicability in a clinical setting, we recently optimized our GMP-conform manufacturing process for autologous virus-specific TCPs and succeeded in attenuating T-cell differentiation by treatment with low doses of Rapamycin (inhibits the mechanistic-target-of-rapamycin-complex-1: mTOR-C1, favorable results with 20 nM) during expansion cultures. This next-generation antiviral TCP comprises enriched proportions of early-differentiated TCM being superior for AVTT. Furthermore, next-generation Rapamycin-treated (Rapa-)TCPs contain higher proportions of CD4+ T-cells  reported to improve clinical efficacy.Detailed knowledge regarding the characteristics of Rapa-TCPs is a prerequisite for realization of clinical translation. Thus, we closely investigated the molecular properties of this Rapa-TCP regarding dependence on cytokine supplementation regiments during in vitro expansion, long-term stability, survival/sensitivity to apoptosis, metabolism, transcriptome, clonal composition, the role of the different memory T-cell subsets and applicability to SOT patient samples. Our data reveal a beneficial early differentiated phenotype, profound function, elevated clonal diversity, and superior survival of Rapa-TCPs compared to first-generation TCPs, which is further underlined and confirmed by a distinct gene expression signature revealed by mRNA sequencing.We used in vitro models to mimic the situation of TCPs once injected into a patient coping with CMV disease, i.e., massive antigen exposure and withdrawal of Rapamycin. Here, we observed a preserved capacity for CMV-specific production of effector cytokines. Moreover, we tested manufacture of CMV-specific TCPs from material of patients with chronic end-stage renal disease (ESRD) before and after KTx to investigate the impact of chronic immunosuppression, showing no benefits of TCP generation before KTx. We further investigated the influence of CMV-reactivation history after KTx on the differentiation of virus-reactive memory T-cells and the resulting composition of untreated and Rapa-TCPs, implying feasibility of TCP generation from all groups investigated. This next-generation AVTT approach may also be applied to other viral specificities, such as EBV and BKV or even cancer-specific T-cells. Prospectively, implementation of next-generation AVTT may allow for reduction or complete ablation of toxic anti-viral medication and minimize the risk for virus-associated complications in the SOT setting.MethodsPatients' and Healthy Donors' Blood SamplesVenous blood samples were collected from 19 healthy donors (HDs) (10 m/9 f; 25–81 years) and 19 KTx patients (11 m/8 f; 34–78 years; ) of the Kidney Transplant Ambulance, Charité Virchow Klinikum, Berlin. We worked with buffy coats from 3 of the 19 different HDs to have sufficient cells for different cell sorting steps. Peripheral blood mononuclear cells (PBMC) were isolated by Biocoll Separating Solution density gradient centrifugation (Biochrom). The Charité Ethics Committee (IRB) approved the study protocol and all blood donors provided written informed consent. Detailed characteristics of each patient are presented in.Enrichment and Expansion of CMV-Specific T-CellsCMV-specific TCPs were generated using a previously described technique. Briefly, PBMCs were stimulated for 6 h with overlapping CMVpp65/IE−1 peptide pools (JPT Peptide Technologies; 0.5 μg/ml each). IFNγ-producing cells underwent positive selection using the IFNγ Secretion Assay—Cell Enrichment and Detection Kit according to the manufacturer's instructions (Miltenyi Biotec). Enriched IFNγ+ cells were cultured for 21 days in 96- or 24-well-plates with irradiated (30 Gy using a GSR D1 [Gamma-Service Medical GmbH]) autologous feeder cells (derived from 1/5 of the capture assays' negative fraction) added only at d0 in complete media (VLE RPMI 1640 supplemented with penicillin (100 IU/ml) and streptomycin [all from Biochrom] and 10% fetal calf serum [FCS, PAA]), supplemented with 10 ng/ml recombinant human (rh) IL-7 and rh IL-15 (CellGenix) or 50 U/ml IL-2 in humidified incubators at 37°C and 5% CO2. Cells were split 1:1 when 100% confluence was reached. For Rapa-TCPs, 20 nM of Rapamycin (Rapamune, Pfizer Pharma GmbH) were added every 2nd day or upon splitting starting from day 1.For restimulation during culture (only where indicated, , ), we depleted the donor's PBMCs of CD3+ cells using magnetically activated cell sorting (MACS) with anti-CD3 beads (Miltenyi Biotec) following the manufacturer's instructions. These cells were frozen in FCS with 10% cell culture grade dimethyl-sulfoxide (Sigma-Aldrich) until restimulation at d14. Thawed autologous CD3− PBMCs were washed twice and pulsed with overlapping CMVpp65/IE−1 peptide pools at concentrations of 2 μg/ml each peptide for 2 h and added at a 1:5 ratio to the T-cells in culture.Figure 1Effects of Rapamycin on T-cell products: Expansion, phenotype and function. (A) Schematic overview of experiments: T-cell products (TCPs) were generated from PBMCs isolated from venous blood of healthy donors (HDs) by magnetically activated cell sorting (MACS) of T-cells producing IFNγ in response to stimulation with CMVIE−1/pp65 peptide pools and expanded in the presence of either IL-2/IL-7  or IL-7/IL-15 without (w/o; blue) or with addition of 20 nMof Rapamycin (Rapa; red) (B–P). Parts of the culture were re-stimulated using thawed CD3− PBMCs loaded with CMVIE−1/pp65 peptide pools, deprived of Rapamycin or a combination of both on d14. (B) Expansion rates of IL-7/15-expanded Rapa-treated (Rapa-)TCPs (red) and untreated TCPs (blue) of n = 10 healthy donors (HDs) calculated from yield at d14 divided by the number of seeded cells at d0. We gated flow cytometric data on lymphocytes singlets living CD3+ T-cells. (C) Exemplary flow cytometry plots of CD4+ and CD8+ populations among living CD3+ T-cells in the Rapa-TCP (left plot) and untreated TCP (w/o, right plot) of one HD. (D) CD4/CD8 ratios in Rapa- (red) and untreated TCPs (blue) of n = 10 HDs calculated from flow cytometry data as presented in (C). (E) Gating strategy for CD45RA− CCR7+ central memory T-cells (TCM) among CD4+ (upper panel) and CD8+ (lower panel) in Rapa- (left panel) and untreated TCPs (right panel) of one exemplary HD. (F) Proportions of CD4+ and CD8+ TCM among Rapa- (red) and untreated TCPs (blue) of n = 10 HDs determined from flow cytometric data as shown in (E) at d14. (G,H) To detect CMV-specific cytokine producers, TCPs were stimulated with CMVIE−1/pp65 peptide-loaded autologous lymphoblastic cell lines (LCLs) at a ratio of 1:10 for 6 h and Brefeldin A (BFA) was added after 1 h. (G) Representative flow cytometric plots of IFNγ- and TNFα-producers in Rapa- (left panel, red) and untreated TCPs (right panel, blue) of one HD. The dark population represents unstimulated and the light population illustrates CMVIE−1/pp65-stimulated CD4+ (upper panel) and CD8+ T-cells (lower panel). (H) Proportions of CMV-specific IFNγ-producers among CD4+ and CD8+ T-cells in Rapa- (red) and untreated TCPs (blue) of n = 10 HDs determined from flow cytometric data as shown in (G) at d14. (I–N): For re-stimulation on d14 of culture, thawed CD3− autologous PBMCs were loaded with CMVIE−1/pp65 peptide pools and added at 1:5 ratio to T-cells. (I) Expansion rates of IL-7/15-expanded re-stimulated (pastel colors) or non-re-stimulated (dark colors) Rapa- (red) and untreated TCPs (blue) of n = 7 HDs calculated from yield at d21 divided by the number of cells at d14. (J) CD4/CD8 ratios in Rapa- (red) and untreated TCPs (blue) of n = 7 HDs calculated from flow cytometric data as presented in (C) at d21. (K,L): Proportions of CD4+\\n(L) among Rapa- (red) and untreated TCPs (blue) of n = 7 HDs determined from flow cytometric data as shown in (E) at d21. (M–P) To detect CMV-specific cytokine producers, TCPs were stimulated with CMVIE−1/pp65 peptide-loaded autologous LCLs for 6 h and BFA was added after 1 h. (M–N) Proportions of CMV-specific IFNγ-producers among CD4+\\n(M) and CD8+ T-cells (N) in Rapa- (red) and untreated TCPs (blue) of n = 7 HDs determined from flow cytometric data as shown in (G) at d21. (O,P) To mimic the situation after infusion, Rapa was withdrawn and TCPs were cultivated long-term until d49. Proportions of CMV-specific IFNγ-producers among CD4+\\n(O) and CD8+ T-cells (P) in TCPs withdrawn from Rapa (red) and untreated TCPs (blue) of n = 6 HDs determined from flow cytometric data as shown in (G) at d49. For all graphs normal distribution of data points was tested with Kolmogorov-Smirnov test and paired t-test was used to determine significance in normally distributed samples or Wilcoxon's matched-pairs signed rank test in not normally distributed samples, respectively. P-values below 0.05 are indicated by * and defined to be significant.Functional Tests, Phenotyping, Flow Cytometry, and SortingExpanded T-cells were analyzed for effector functions by their ability to recognize antigen-loaded target cells, which consisted of autologous lymphoblastoid B-cell lines (LCLs), transformed with B95-8 EBV and by specific production of cytokines. LCLs were generated as described previously.For CMV-specific stimulation of expanded TCPs for detecting intracellular effector cytokine production, CMVpp65/IE−1 peptide pool-loaded LCLs were added to cultured T-cells at a ratio of 1:10 and incubated for 6 h. Un-pulsed LCLs served as unstimulated control. For effector cytokine detection ex vivo, PBMCs were stimulated with 1 μg/ml overlapping CMVpp65/IE1 peptide pools ex vivo for 14 h. After 1 h, 2 μg/ml Brefeldin A (BFA, Sigma-Aldrich) were added to the stimulation to allow for intracellular capture of cytokines.To induce apoptosis, 1 μg/ml of LEAF-purified Fas-activating antibody (EOS9.1; BioLegend) was added to cultures for 16 h. To determine survival, LIVE/DEAD® Fixable Blue Dead Cell Stain (Invitrogen) and Annexin V (BioLegend) were added.For determination of killing capacity, autologous LCLs were pulsed with 2 μg/ml CMVpp65/IE−1 peptide pools, whereas unpulsed allogenic LCLs were used as non-target controls. Targets were labeled with 10 mM Carboxyfluorescein-diacetate-succinimidyl-ester (Sigma-Aldrich) and non-target controls with 5 mM CellTrace™ Far Red (Invitrogen). Cells were co-cultured at a T-cell/target-cell ratio of 10:1 for 14 h. Samples were analyzed using a LSR II Fortessa flow cytometer. Samples without T-cells, containing only LCLs, served as an internal control and reference for calculation of the killing capacity. For analysis, we gated on LIVE/DEAD® Fixable Blue Dead Cell Stain-negative cells and calculated ratios of target to non-target cells as described previously.To define memory subsets, T-cells were stained extracellularly for surface markers CCR7 (G043H7), CD45RA (HI100), CD45RO (UCHL1), CD62L (DREG-56; eBioscience), and CD95 (DX2). Subsequently, cells were permeabilized and fixed with Foxp3/Transcription Factor Staining Buffer Set (eBioscience) and stained intracellularly for CD3 (OKT3), CD4 (SK3), and CD8 (RPA-T8), IFNγ (4S.B3, eBioscience), TNFα (MAb11), and Granzyme B (GZB) (GB11, BD Pharmingen). Cells were analyzed on a LSR II Fortessa flow cytometer using FlowJo Version 10 software (Tree Star). Lymphocytes were gated based on the FSC vs. SSC profile and subsequently gated on FSC (height) vs. FSC to exclude doublets.For evaluating of T-cell subsets on transcriptome level, T-cell subsets were sorted from PBMCs from n = 3 HDs' buffy coats (DRK) at d0 or derived TCPs at d18 based on the expression of CD3, CD45RA, and CCR7 by the Core Facility Flow Cytometry of the BCRT using a FACS Aria II Calliope (BD).All antibodies were purchased from BioLegend, unless indicated otherwise.Metabolic AnalysisExtracellular acidification rate (ECAR) and oxygen consumption rate (OCR) were analyzed using a Seahorse-XFe96-Analyzer following the manufacturer's instructions for non-adherent cells including immobilization of cells with Cell-Tak (Corning). Assay medium consisted of Dulbecco's Modified Eagle's Medium D5030 (Sigma) supplemented with 3 g/l D-glucose (Roth) and 300 mg/ml L-glutamine (Gibco) and was sterile-filtered. For T-cell activation, 0.5 μg/ml of CMV pp65/IE−1 peptide pools were added to the microwells relying on reciprocal antigen-presentation of T-cells 0.5 h before the measurement.RNA Sequencing and Bioinformatics AnalysisRNA was isolated using an All-Prep DNA/RNA Kit (Qiagen) following the manufacturer's instructions. RNA samples were sent to the Deep Sequencing Core Facility in Göttingen, where samples were prepared using TrueSeq Kits (Illumina) and HiSeq_4000 performing 50 million reads/sample.Fastq-files were quality checked with FastQC (Babraham Bioinformatics) and trimmed for residual adapter sequences. Reads were aligned to the GRCh38 human genome using TopHatR (2.1.0–Johns Hopkins University, Center for Computational Biology) and Bowtie2. Counts per gene were determined as sum of all reads mapped within a gene region. Principal component (PC) analysis was performed in R  using the 1,000 top-variable genes within the data set. Differentially expressed genes were identified using negative binomial distributions as implemented in the DESeq2 package  in R. False discovery rates (FDR) were calculated to adjust p-values for multiple testing and FDR-values below 0.05 were considered as significant. Expression data for differentially expressed genes were variance-stabilized transformed and scaled prior to visualization in heat maps. RNA sequencing data are available at the GEO platform with the accession number GSE129196.T-Cell Receptor SequencingFor sequencing of T-cell receptors (TCRs) to determine the clonality of TCPs, DNA was isolated using an All-Prep DNA/RNA Kit (Qiagen) following the manufacturer's instructions. TCRβ sequencing was performed using a hsTCRb Kit (Adaptive Biotechnologies) following the manufacturer's instructions and analyzed with the corresponding ImmunoSEQ-Analyzer 3.0 software. Briefly, the most variable complementary-determining region 3 (CDR3), spanning the recombination site of V-D-J recombinations of TCR β-chains was sequenced. Productive rearrangements were regarded as unique in-frame nucleotide sequences without stop codon, leading to a functional TCR. Productive frequency means the individual frequency of a specific productive rearrangement (clone) among all productive rearrangements. Clonality was calculated based on productive entropy normalized to the total number of productive rearrangements. Sample overlap was investigated using the Morisita index considering unique clones, individual frequencies of clones and the probability of a common origin of two samples. TCR sequencing data is accessible at the ImmuneACCESS platform http://adaptivebiotech.com/pub/amini-2019-frontimmunol (Adaptive Biotechnologies).Statistical Analysis and CalculationsGraph Pad Prism version 7 was used for graph generation. To test for normal Gaussian distribution Kolmogorov-Smirnov test was performed. If data were normally distributed, Student's paired or unpaired t-test were employed for analysis. If data were not normally distributed, Wilcoxon's matched pairs test was applied to paired samples and Man-Whitney's test to unpaired samples. All tests were two-tailed. Probability (p) values of ≤0.05 were considered statistically significant and significance is denoted as follows: * = p < 0.05. Correlation analysis was assessed by Pearson's correlation coefficients for normally distributed data or non-parametric Spearman's rank correlation. Fold expansion expresses the manually counted cell count (Neubauer's counting chamber) excluding dead cells by Trypan blue staining (Sigma-Aldrich) at the day indicated divided by the initially seeded cell amount from the positive fraction of the IFNγ Secretion Cell Enrichment Assay.All datasets are available upon reasonable request.ResultsIn order to prepare our approach for clinical translation, we deeply characterized functionality, stability and distinct molecular, metabolic and transcriptional properties of our next-generation Rapa-TCP, for which we applied mTOR inhibition by Rapamycin to enrich for CD4+ T-cells and CD4+/CD8+ TCM. First, we addressed the question whether we can reproduce our findings and properties published for supplementation of a certain cytokine regiment, IL-2/IL-7, with a regiment commonly used for GMP applications by many groups, IL-7/IL-15, which was previously shown to support generation of TCM.Supplementation of IL-7/IL-15 Does Not Alter Rapamycin-Mediated Effects in TCPsTo investigate potential differences in the effects of Rapamycin administration dependent on the cytokine regiment supplemented, we expanded CMV-specific T-cells in the presence of different cytokine combinations, namely IL-7/IL-15  and IL-2/IL-7. The expansion rates of antigen-reactive T-cells were sufficient considering cell numbers used in a pilot study , although Rapamycin significantly reduced expansion in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs. Overall, different cytokines did not alter the beneficial effects of Rapamycin treatment : Rapamycin significantly increased CD4/CD8 ratio in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs  and significantly increased proportions of CD4+ and CD8+ TCM in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs. Furthermore, Rapamycin increased proportions of Interferon-γ (IFNγ)-producing CD4+ and CD8+ T-cells upon exposure to CMV-specific peptides loaded onto autologous lymphoblastic cell lines (LCLs) in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs. These data confirm the robustness of beneficial effects of mTOR inhibition using Rapamycin for TCP composition in the case of supplementing commonly used IL-7/IL-15 for expansion of TCPs.Expansion Rates of Rapa-TCPs Recover Later During CultureSOT patients often suffer from lymphopenia, which reduces the amount of PBMC, i.e., the starting material, for TCP generation and their medication can impact the functionality of T-cells. Thus, TCP manufacture from patient material may require longer in vitro expansion periods of up to 21 days to achieve sufficient cell numbers for successful AVTT. To assess the stability of TCPs after a longer period of expansion, we determined phenotype and functionality of TCPs after extended expansion on d21 in IL-7/IL-15-  and IL-2/IL-7-expanded TCPs. Interestingly, Rapa-TCPs recovered, yet even exceeded expansion of untreated TCPs in the 3rd week of expansion (d14–d21) , which was significant in IL-2/IL-7-expanded TCPs. CD4/CD8 ratios remained significantly higher in Rapa-TCPs at d21 , but IL-2/IL-7-expanded TCPs showed significantly higher CD4/CD8 ratios than IL-7/IL-15-expanded TCPs at d21 (\\nvs. ). During expansion, TCM differentiated and the enrichment of TCM proportions upon Rapamycin-treatment lost significance in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs.Antigen Encounter Decreases Expansion Rates, but Promotes Less Differentiated CellsWe further mimicked the scenario happening once the TCPs are injected into a patient coping with CMV viremia in an in vitro model. Therefore, we modeled the situation of high antigen load by re-stimulation with CD3-depleted PBMCs pulsed with CMV-specific peptides: CMV-specific re-stimulation significantly reduced expansion rates in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs. Re-stimulation did not influence CD4/CD8 ratios in IL-7/IL-15-  neither IL-2/IL-7-expanded Rapa-TCPs , but significantly decreased CD4/CD8 ratios in IL-7/IL-15-expanded untreated TCPs. Remarkably, re-stimulation significantly augmented the proportions of CD4+ and CD8+ TCM in both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs. However, re-stimulation decreased the proportion of CD4+ and CD8+ IFNγ-producers , which was statistically significant in IL-2/IL-7-expanded Rapa-TCPs.Rapamycin Preserves Superior Capacity for IFNγ ProductionImportantly, IL-2/IL-7-expanded Rapa-TCPs showed significantly higher proportions of IFNγ-producers among CD4+ T-cells at d14 and d21  and CD8+ T-cells at d21  compared to untreated TCPs illustrating improved functionality. Because TCPs are deprived of Rapamycin and exposed to antigen once injected, we analyzed samples in which we withdrew Rapamycin and re-stimulated with CMV peptide-loaded CD3-depleted PBMCs on d14. Interestingly, once treated with Rapamycin during the first 2 weeks of culture, both IL-7/IL-15-  and IL-2/IL-7-expanded TCPs  comprised more CD4+ and CD8+ IFNγ-producers continuously until d49 of culture.Rapamycin Enhances Survival of T-CellsLongevity is a crucial prerequisite for long-term efficacy of adoptively transferred TCPs in patients. Based on findings in B-cell lymphoma cell lines, we hypothesized that Rapamycin treatment may increase viability of T-cells. Thus, we analyzed overall survival of T-cells in TCPs. Strikingly, we found significantly higher proportions of living T-cells in TCPs treated with Rapamycin compared to untreated TCPs.Figure 2Rapamycin promotes survival of T-cells and stabilizes their metabolism. (A) Exemplary dot-plots of flow cytometry data regarding live/dead stain and Annexin V stain (apoptosis) gated on lymphocytes singlets CD3+ T-cells. Living T-cells are defined by double negative staining for Annexin V and live/dead stain in Rapa- (upper panel) and untreated TCPs (lower panel). Samples in the right panel were treated with 1 μg/ml activating antibody against Fas (CD95) to induce apoptosis. (B) Proportions of living T-cells in Rapa- (red) and untreated TCPs (blue) of n = 8 HDs identified as shown in (A) at d21. (C) Proportions of living T-cells in Rapa- (red) and untreated TCPs (blue) of n = 8 HDs incubated with Fas-activating antibody identified as shown in (A) at d21. (D) Exemplary histograms of fluorescence intensity of Bcl-2 in CD4+ (upper panel) and CD8+ T-cells (lower panel) of untreated (blue) and Rapa-TCPs (red) acquired by flow cytometry. (E,F) MFIs of Bcl-2 in CD4+\\n(E) and CD8+ T-cells (F) in untreated (blue) and Rapa-TCPs (red) of n = 18 HDs. (G) Oxygen consumption rare (OCR)/extracellular acidification rate (ECAR) ratio of Rapa- (red) and untreated TCPs (blue) of n = 5 HDs determined in a Seahorse assay. For stimulation (pastel colors) CMVIE−1/pp65 peptide pools were added to TCPs relying on mutual presentation of peptides by T-cells from the TCP. For all graphs normal distribution of data points was tested with Kolmogorov-Smirnov test and paired t-test was used to determine significance. P-values below 0.05 are indicated by * and defined to be significant.Immunosuppressant regiments including Tacrolimus, which are commonly used in SOT, are reported to sensitize T-cells to programmed cell death. Hence, we investigated the TCPs' sensitivity to apoptosis employing induction of the death receptor pathway by agonistic Fas-specific antibody to identify differences between untreated and Rapa-TCPs. We recorded partial resistance to Fas-induced apoptosis in Rapa-TCPs, while untreated TCPs were more sensitive to Fas-induced apoptosis. The anti-apoptotic effect of Rapamycin observed in B-cell lymphoma lines is reported to depend on upregulation of Bcl-2 on protein level. Hence, we assessed  the mean fluorescence intensity (MFI) of Bcl-2 in TCPs and found significantly higher MFIs in CD4+ and CD8+ T-cells of Rapa-TCPs compared to untreated TCPs. The findings regarding viability and resistance to apoptosis suggest an increased fitness of T-cells in Rapa-TCPs implying improved long-term survival in vivo.Rapamycin Stabilizes T-Cell Metabolism Upon ActivationCharacteristically, memory and effector T-cells are distinguished by differences in metabolic activities. Thus, we investigated the ratio of fatty acid oxidation to glycolysis defined by the ratio of oxygen consumption to ECAR (OCR/ECAR). In fact, we detected significantly lower OCR/ECAR ratios in Rapa-TCPs compared to untreated TCPs. Moreover, upon activation with CMV-specific peptides, the metabolism of Rapa-TCPs proofed to be more stable, whereas the OCR/ECAR ratio was significantly decreased in untreated TCPs.Distinct Effects of Rapamycin on Isolated T-Cell Memory SubsetsDistinct memory T-cell subsets were reported to have defined properties and are not equally suited for long-term regeneration of T-cell immunity. To assess the effect of Rapamycin on distinct CMV-specific memory T-cell subsets, we performed fluorescently activated cell sorting (FACS) for CCR7+CD45RA− TCM, CCR7−CD45RA− TEM, and CCR7−CD45RA+ TEMRA based on their differential expression of CD45RA and CCR7 ex vivo and subsequently performed IFNγ-secretion assays of sorted subsets to isolate CMV-specific T-cells of these particular subsets. The CMV-reactive T-cells of each memory T-cell subset were expanded in the presence or absence of Rapamycin and analyzed after 3 weeks of culture. Interestingly, Rapamycin significantly reduced expansion in TEM- and TEMRA-derived cultures, which was not significant in TCM-derived cultures , indicating preferential expansion of TCM and implying reduced sensitivity to anti-proliferative effects of Rapamycin.Figure 3Influence of Rapamycin on different T-cell memory subsets. (A) Schematic experimental setup: TCM, TEM, and TEMRA were sorted out of lymphocytes singlets CD3+ T-cells according to expression of CCR7 and CD45RA, CMV-reactive T-cells were isolated from each subset using an IFNγ secretion assay and CMV-reactive T-cells from each subset were cultured with (Rapa) and without Rapamycin (w/o). Exemplary dot plots of flow cytometry data of sorted subsets of one HD and respective positive fractions of the IFNγ secretion assay are shown. (B) Expansion rates of the indicated subsets in the presence of (red, Rapa) and absence of Rapamycin (blue, w/o) calculated from total cell numbers at d21 divided by the seeded cell number. (C,D) Proportions of CD4+\\n(C) and CD8+ CD45RA− CCR7+ TCM-like cells (D) among Rapa-treated (red) and untreated (blue) cultures of indicated subsets determined from flow cytometric data at d21. (E,F) MFIs of Bcl-2 in CD4+\\n(E) and CD8+ T-cells (F) in untreated (blue) and Rapa-treated cultures (red) of isolated T-cell subsets determined in flow cytometry. (G,H) To detect CMV-specific cytokine production, cultures were stimulated with CMVIE−1/pp65 peptide-loaded autologous LCLs at a ratio of 1:10 for 6 h and BFA was added after 1 h. Proportions of CMV-specific IFNγ-producers among CD4+\\n(G) and CD8+ T-cells (H) in Rapa-treated (red) and untreated (blue) cultures of isolated T-cell subsets determined from flow cytometric data. All graphs contain data from n = 6 HDs, normal distribution of data points was tested with Kolmogorov-Smirnov test and paired t-test was used to determine significance in normally distributed samples or Wilcoxon's matched-pairs signed rank test in not normally distributed samples, respectively. P-values below 0.05 are indicated by * and defined to be significant.Moreover, Rapamycin prevented a significant proportion of CD4+ and CD8+ TCM from differentiation into late-stage memory/effector T-cells compared to control cultures. Remarkably, Rapamycin treatment even induced some cells with a TCM-like phenotype in cultures derived from TEM and TEMRA subsets suggesting some “rejuvenation” of late-stage memory cells.Notably, Rapamycin increased the MFI of Bcl-2 in CD4+ and CD8+ T-cells from TCM- and TEM-, but not TEMRA-derived cultures. Interestingly, Rapamycin-treated TEM- and TEMRA-derived cultures contained significantly higher proportions of CD8+ IFNγ-producers than untreated cultures upon CMV-specific re-stimulation , whereas there were no significant differences in CD4+ IFNγ-producers. Overall, Rapamycin conferred distinct effects on different T-cell memory subsets, sustaining TCM features and counteracting differentiation into late-stage memory/effector T-cells.Rapamycin-Treated T-Cell Products Have a Unique Transcriptome Resembling TCMTo confirm that expansion of antigen-reactive T-cells under Rapamycin treatment “freezes” an early memory T-cell stage, we tried to extend our analysis on transcriptome level by RNA-sequencing using next-generation sequencing (NGS) of untreated and Rapa-TCPs at d21. The RNA expression data revealed a total of 146 differentially expressed genes between Rapa-TCPs and untreated TCPs. Many of these relate to TCP performance. We reviewed the literature and various databases to identify T-cell associated processes  and to estimate the relevance of the differentially regulated genes. With reference to previously published data, 84% of the genes identified as potentially relevant for TCP potency and longevity in vivo were regulated in a beneficial manner in Rapa-TCPs. Among these differentially expressed genes, we identified increased expression of TCM markers such as CCR7 and PIM2 , increased expression of TERT, which induces self-renewal capacity and increases the proliferative potential of human T-cells , and IL7R, which is reported to be a marker for persisting and protective CD8+ memory T-cells  in Rapa-TCPs. Furthermore, our observation of sustained IFNγ production in Rapa-TCPs is in line with the findings of increased expression of IL-13 regulating IFNγ synthesis , DRD2 inducing IFNγ production  and TNFRSF11A, which increases IFNγ secretion upon binding its ligand. In addition, these data are underlined by increased expression of activation enhancing genes, including e.g., KLF7 , RGMB , and TNFRSF19  in Rapa-TCPs. Moreover, the fact that anti-apoptotic Bcl-2 is upregulated on protein level may be supported by increased expression of MYB, which exerts its anti-apoptotic activity via Bcl-2. However, also many other genes inhibiting apoptosis, such as e.g., BEX2  and SIX1 , show significantly higher expression in Rapa-TCPs compared to untreated TCPs. The metabolic data of increased glycolysis in Rapa-TCPs are in line with increased expression of EPAS1 , however, also CHDH, a gene involved in fatty acid oxidation,  is higher expressed in Rapa-TCPs than in untreated TCPs. See  for a complete view and annotation of the genes differentially expressed in untreated and Rapa-TCPs and their functions potentially relevant for TCP efficacy in vivo. Of note, the TCM-like cells sorted from untreated TCPs on day 18 of culture  clustered with Rapa-TCPs  regarding the differentially expressed genes identified in. Interestingly, principle component (PC) analysis revealed some components of the transcriptome of Rapa-TCPs to be more similar to ex vivo-sorted TCM than TEM (dimension PC2; ).Figure 4Rapamycin-treated TCPs have a unique transcriptome resembling TCM and are clonally more diverse. RNA expression data were acquired by RNA sequencing and samples had to pass a quality control to be included in the analysis. (A) Expression heat map of differentially expressed genes between untreated and Rapa-TCPs generated from fresh blood of n = 3 HDs at d21. (B) Processes allocated to the differentially expressed genes based on the literature (see  for details). (C) Exemplary dot plots of CCR7/CD45RA expression of untreated and Rapa-TCPs and the respective TCM-like cells sorted on d18 of culture. (D) Expression of differentially expressed genes between Rapa- and untreated TCPs from fresh blood of n = 3 HDs at d21, buffy coats of n = 3 HDs at d18 and TCM sorted from Rapa- and untreated TCPs generated from buffy coats of the same n = 3 HDs at d18 including clustering. Samples not included in the graph were discarded due to failure at the quality threshold. (E) Clonality of Rapa- (red) and untreated TCPs (blue) n = 3. (F) Proportions of represented sequences covered by the top 10 most represented clones in Rapa- (red) and untreated TCPs (blue), n = 3 HDs. (G) Numbers of productive rearrangements included in in Rapa- (red) and untreated TCPs (blue), n = 3 HDs. Data in (E,F) were calculated with ImmunoSEQ-Analyzer3.0 software based on TCRβ sequencing.Rapamycin-Treated T-Cell Products Show Less Clonal and More Diverse TCR RepertoiresTo estimate the TCR repertoire of our TCPs, we performed TCRβ NGS. Notably, TCRβ sequencing showed a more diverse clonal composition of Rapa-TCPs compared to untreated TCPs. Venn diagrams of the total numbers of clones and overlap between Rapa- and untreated TCPs are shown in  and the distribution of the top 100 clones is shown in. Correspondingly, the top 10 clones covered around 70 and 90% of the whole TCRβ repertoire of Rapa-TCPs and untreated TCPs, respectively  and Rapa-TCPs contained more different clones than untreated TCPs. The top 10 shared clones and their respective frequencies in Rapa- and untreated TCPs are shown in. Comparison of unique nucleotide sequences revealed a high clonal overlap between the distinct Rapa- and untreated TCPs generated from the same donor. In contrast, comparison of clonal repertoires between different individuals showed no overlap, confirming the specificity of the findings.Onset of Immunosuppression in Patients Does Not Influence the Starting Material for TCPs Regarding T-Cell Differentiation and CMV-SpecificityAs a prerequisite for clinical translation, we aimed at confirming feasibility of Rapa-TCP generation from patient blood and therefore collected samples from end stage kidney disease (ESRD) patients before and after kidney transplantation (KTx). To investigate the influence of immunosuppression on the starting material for TCPs, 7 paired samples from ESRD patients before/after KTx were analyzed ex vivo. All KTx recipients received standard immunosuppression (characteristics in , pre/post-Tx paired samples highlighted in gray). T-cells were divided into five differentiation subsets: CCR7+CD45RA+CD95− TN (naïve T-cells), CCR7+CD45RA+CD62L+CD45RO−CD95+ TSCM, CCR7+CD45RA− TCM, CCR7−CD45RA− TEM, and CCR7−CD45RA+ TEMRA  revealing no substantial differences between CD4+ and CD8+ memory T-cell subset distributions of paired patient samples before/after KTx. To assess the phenotypic and functional characteristics of CMV-specific T-cells, PBMCs were stimulated with CMVpp65/IE1 peptides showing markedly higher frequencies of CMV-responsive T-cells among CD8+ compared to CD4+ T-cells. However, frequencies of CMV-responsive T-cells were similar before and after KTx  and T-cell memory subsets were comparable among CMV-responsive T-cells before and after KTx.Manufacturing Rapa-TCPs Is Feasible Before and After TransplantationWe assessed feasibility of TCP generation from patient material collected before and after KTx. Manufacture of untreated and Rapa-TCPs was successful with respect to yield , although untreated CMV-specific TCPs resulted in higher yields, which was statistically significant in TCPs generated after KTx. Rapa-TCPs showed beneficial, higher CD4+ T-cell proportions compared to untreated TCPs, which was statistically significant in TCPs generated before KTx. Compared to untreated TCPs, Rapa-TCPs showed higher proportions of CD4+ and CD8+ TCM in KTx recipient-derived TCPs, which was significant in TCPs generated after KTx. Upon re-stimulation with CMV-peptide-loaded autologous LCLs, we found enhanced IFNγ-producers  and IFNγ/granzyme B (GZB)-double-producers  among Rapamycin-treated CD8+ T-cells , whereas their proportions among CD4+ T-cells remained stable. We recorded comparable frequencies of IFNγ-producers and IFNγ/GZB-double-producers among CD4+ T-cells in TCPs generated before and after KTx , whereas both increased among CD8+ T-cells in TCPs generated after KTx. Remarkably, Rapamycin increased the frequency of TCM among IFNγ-producers , which was statistically significant among CD4+ IFNγ-producers in TCPs generated after KTx. Regarding CMV-specific cytotoxic effects, Rapa-TCPs were as effective as untreated TCPs and TCPs generated pre- and post-KTx neither showed any differences.Figure 5Manufacture of Rapa-TCPs is feasible before/after transplantation. N = 7 paired samples of untreated (w/o, blue) and Rapa-TCPs (red) from the same patients before (pre; pastel colors) and a few weeks after KTx (post). (A) Yield of TCPs = total cell number derived from 20 ml of patient blood on d21. (B) CD4/CD8 ratio of TCPs determined by multicolor flow cytometry on d14. (C) Exemplary dot plots of one patient's untreated (right) and Rapa-TCPs (left) comparing subset distributions of CD3+CD4+ (upper panel) and CD3+CD8+ (lower panel) T-cells according to CCR7 and CD45RA expression on d14. (D,E) Proportions of CD45RA− CCR7 + TCM among CD4+\\n(D) and CD8+ T-cells (E) in TCPs on d14 as determined per gating strategy shown in (C). (F) Exemplary dot plots of one patient comparing CD3+CD4+ (left panel) and CD3+CD8+(right panel) IFNγ- and TNFα-producers in Rapa- (red) and untreated TCPs (blue) detected by intracellular staining in multicolor flow cytometry after 6 h stimulation with autologous LCLs loaded with CMVIE−1/pp65 peptide pools (gray) or incubation with unloaded autologous LCLs as control (black) and addition of BFA after 1 h on d21. (G,H) Summary of background subtracted proportions of CD4+\\n(H) CMV-reactive IFNγ-producing T-cells in Rapa- (red) and untreated TCPs (blue) gated as illustrated in (F). (I) Exemplary dot plots of one donor comparing CD4+ (left panel) and CD8+ (right panel) CMV-reactive IFNγ- and GZB-producers in Rapa- (red) and untreated TCPs (blue) detected by intracellular staining in multicolor flow cytometry after 6 h stimulation with autologous LCLs loaded CMVIE−1 and CMVpp65 peptide pools (gray), incubation with unloaded autologous LCLs as control (black) and addition of BFA after 1 h on d21. (J,K) Summary of background subtracted proportions of CD4+\\n(K) CMV-reactive IFNγ/GZB-double-producers in Rapa- (red) and untreated TCPs (blue) gated as illustrated in (I). (L,M) Proportions of CD45RA− CCR7 + TCM among CMV-reactive IFNγ-producing CD4+\\n(L) and CD8+ T-cells (M). Gates were applied from gates set for global T-cell subset distribution (see ). (N) Specific killing of CMVIE−1/pp65 peptide pool loaded autologous LCLs determined by ratio with unloaded allogenic LCLs at a 1:10 ratio with TCPs after incubation overnight. All data tested for normal distribution of data points with Kolmogorov-Smirnov test; significance determined with paired t-test if normally distributed or Wilcoxon's matched-pairs signed rank test for not normally distributed samples. P-values below 0.05 are indicated by * and defined to be significant.CMV History Affects the Composition of Starting Material From Post-KTx PatientsTo determine the influence of the CMV infection status on the T-cell subset composition and function of starting material for TCP generation, 19 CMV seropositive KTx patients  and 13 CMV seropositive healthy donors (HDs) were analyzed in parallel. Patients were categorized according to their CMV reactivation status: No recorded (n = 9; 5 m/4f), history of (n = 6; 2 m/4f) or very recent CMV-DNAemia within 2 weeks before blood collection (n = 4; 4 m/0f). The CMV reactivation status had almost no effect on the global CD4+/CD8+ T-cell memory subset distribution , except for an increase in proportions of CD8+ TSCM and a decrease in CD4+ TEM in the blood of patients with no record of CMV viremia compared to HDs.We found CMV-reactive T-cells in all KTx patients and HDs, with markedly higher frequencies among CD8+ vs. CD4+ T-cells. We did neither observe major differences in the magnitude of the CMV-response between KTx patients and HDs nor between the different groups of KTx patients. Of note, proportions of below 0.2% of CMV-responsive T-cells among CD8+ T-cells occurred in 38.5% of HDs and only 10.5% of patients without recorded CMV-DNAemia.The majority of KTx patients showed TSCM frequencies < 10% among CMV-reactive CD4+ T-cells. However, patients with a CMV history or recent CMV-DNAemia presented with significantly elevated CMV-reactive CD4+ TSCM compared to KTx patients with no recorded CMV-DNAemia and HDs. TCM proportions among CD4+ CMV-responsive T-cells showed high inter-individual differences among the patients and HDs. TEM proportions among CD4+ CMV-responsive T-cells were significantly lower in the cohort of KTx patients with a record of CMV-DNAemia compared to patients with no recorded CMV-DNAemia and HDs. The proportions of TEMRA among the CD4+ CMV-responsive T-cells were below 5%, except for three patients with recent or previous CMV-DNAemia, who all received virostatic medication. We found no significant differences in the memory subset distribution among CMV-reactive CD8+ T-cells between the different patient groups and HDs. Notably, we could not detect CMV-responsive CD8+ TCM in the majority of samples.Impact of CMV Reactivation State on Manufacture of Untreated and Rapa-TCPsTo evaluate the quality of untreated and Rapa-TCPs generated from KTx patients under maintenance immunosuppression with distinct CMV reactivation states, CMVpp65/IE1-specific T-cells were expanded with or without Rapamycin. We successfully manufactured untreated and Rapa-TCPs from all KTx patients and HDs analyzed, although Rapamycin substantially reduced yields in TCPs from all patients and HDs. Recent CMV reactivation further significantly reduced yields of untreated and Rapa-TCPs compared to HDs and history of CMV reactivation reduced yields of Rapa-TCPs compared to HDs. Interestingly, there was an inverse correlation between expansion rate/yield and age in untreated and Rapa-TCPs. Furthermore, the number of records with CMV-DNAemia correlated inversely with the yield of Rapa-TCPs. We found that Rapamycin significantly increased the CD4+/CD8+ T-cell ratio in TCPs of KTx patients without recorded CMV-DNAemia and HDs, which was less pronounced in TCPs of KTx patients with a record of CMV-DNAemia. Rapa-TCPs showed significantly higher proportions of CD8+ TCM in all groups except the KTx patients with recent CMV DNAemia (n = 4) , while CD4+ TCM were only significantly enriched in TCPs of KTx patients with no recorded CMV-DNAemia and HDs, being less pronounced in TCPs from the other groups.Figure 6Impact of CMV history on manufacture of untreated and Rapa-TCPs. Untreated (w/o, blue) and Rapa-TCPs (red) of n = 19 patients (9 with so far no recorded CMV viremia; 4 with recent CMV viremia and 6 with a history of CMV viremia)/13 HDs. (A) Yield of TCPs = total cell number derived from 20 ml of patient blood on d21. (B) CD4/CD8 ratio of TCPs on d14. (C,D) Proportions of CD45RA− CCR7 + TCM among CD4+\\n(C) and CD8+ T-cells (D) in TCPs as determined per gating strategy shown in  on d14. (E,F) Proportions of CMV-reactive CD4+\\n(F) IFNγ-producers detected by intracellular staining in multicolor flow cytometry after 6 h stimulation with autologous LCLs loaded CMVIE−1/pp65 peptide pools at a ratio of 1:10 and addition of BFA after 1 h on d21. Gating strategy is shown in. (G) Specific killing of CMVIE−1/pp65 peptide pool loaded autologous LCLs determined by ratio with unloaded allogenic LCLs at 1:10 ratio with TCPs after incubation overnight. All data were tested for normality with Kolmogorov-Smirnov test; significant differences for paired samples determined with paired t-test if normally distributed or Wilcoxon's matched-pairs signed rank test and for unpaired samples with unpaired t-test if normally distributed or Man Whitney's test. P-values below 0.05 are indicated by * and defined to be significant.Upon CMV-specific re-stimulation, we found increased frequencies of CD8+ IFNγ-producers in Rapa-TCPs, which was statistically significant in the group of KTx patients without record of CMV DNAemia. However, Rapa-TCPs of KTx patients with a history of CMV DNAemia contained significantly lower frequencies of IFNγ-producers compared to Rapa-TCPs from HDs. Remarkably, the frequency of IFNγ-producers among CD8+ T-cells, but not CD4+ T-cells , inversely correlated with the time from the last CMV-DNAemia in untreated, but not Rapa-TCPs. Notably, Rapa-TCPs included higher proportions of TCM-like among IFNγ-producing CD4+ and CD8+ T-cells compared to the corresponding untreated TCPs. This was significant in IFNγ-producing CD4+ T-cells in TCPs of KTx patients without record of CMV DNAemia and IFNγ-producing CD8+ T-cells in TCPs of KTx patients without record and history of CMV DNAemia. The TCM-like phenotype among IFNγ-producing CD4+ T-cells was significantly more frequent in Rapa-TCPs of KTx patients compared to HDs. IFNγ-producing CD8+ T-cells were significantly more frequent in Rapa-TCPs of KTx patients with history of CMV viremia compared to HDs.In order to characterize functionality, we co-cultured TCPs with CMV-antigen-loaded LCLs for 14 h and killing was analyzed. Untreated and Rapa-TCPs achieved similar target cell lysis. CMV-specific re-stimulation further characterized up to 65.6% of CD4+ T-cells to be cytotoxic as defined by GZB/IFNγ-double-production. Interestingly, Rapa-TCPs of KTx patients with no record of CMV viremia contained significantly more GZB/IFNγ double producers among CD8+ T-cells than their untreated counterparts. Rapa-TCPs of KTx recipients with history of CMV viremia contained significantly less GZB/IFNγ double producers among CD8+ T-cells than those of HDs.Rapa-TCPs From Patients Exhibit Superior Viability After ThawingStrikingly, every single Rapa-TCP consistently comprised higher proportions of living T-cells compared to its untreated counterpart. By convention, TCPs have to be frozen until GMP-compliant quality controls are accomplished and then are thawed directly before infusion into patients. This procedure is a major stress for the TCPs. We froze and thawed TCPs from ESRD/KTx patients and HDs and observed an increased frequency of living T-cells in the Rapa-TCPs, being detectable immediately and even 1 day after thawing and culture. Consistent with the findings from HDs, also Rapa-TCPs of KTx patients showed elevated MFIs of Bcl-2 compared to the untreated TCPs.Figure 7Rapa-TCPs from patients exhibit superior viability before and after thawing. Untreated (w/o, blue) and Rapa-TCPs (red) of n = 19 patients (9 with so far no recorded CMV viremia; 4 with recent CMV viremia and 6 with a history of CMV viremia)/13 HDs. (A) Proportions of living T-cells determined by positive staining for CD3 and negative staining for live/dead stain and Annexin V. (B,C) Untreated (w/o, blue) and Rapa-TCPs (red) from n = 4 (paired pre- and post-KTx samples of patient no. 11; post-KTx TCPs of patient no. 9 and two HDs). TCPs were frozen in fetal calf serum substituted with 10 % dimethylsulfoxide and stored in liquid nitrogen. After thawing and two washing steps, proportions of living T-cells defined by positive staining for CD3 and negative staining for live/dead stain and Annexin V were determined immediately (d0, B) and after 24 h rest in complete medium in a humidified incubator at 37°C and 5% CO2 (d1, C). All data were tested for normality with Kolmogorov-Smirnov test; significant differences for paired samples determined with paired t test if normally distributed or Wilcoxon's matched-pairs signed rank test and for unpaired samples with unpaired t test if normally distributed or Man Whitney's test. P-values below 0.05 are indicated by * and defined to be significant.In summary, we demonstrate that CMV-specific Rapa-TCPs can be generated irrespective of the employed cytokine regiment, show better viability even after thawing, a stable metabolism, beneficial gene expression and increased clonal diversity. We likewise demonstrate the possibility to generate TCPs from patients in ESRD and post-KTx despite maintenance immunosuppression containing similar attributes as from HDs. Further, we illustrate that functional CMV-specific T-cells, the prerequisite for manufacture of CMV-specific TCPs, could be identified in all KTx patients investigated and present successful manufacture of untreated and Rapa-TCPs irrespective of the viral replication history.DiscussionThe aim of our study was to demonstrate the benefit of the Rapamycin treatment during manufacture of TCPs with improved properties. We demonstrate stability of Rapa-TCPs irrespective of cytokine combinations administered during expansion, sustained IFNγ production despite withdrawal of Rapamycin and re-stimulation with viral antigen. Furthermore, we found superior viability and partial resistance to death receptor-induced apoptosis, stable metabolism upon activation, favorable gene expression pattern, and enhanced clonal diversity of Rapa- compared to untreated TCPs. Moreover, we show preferential expansion of TCM in the presence of Rapamycin and partial conversion of other T-cell memory subsets to TCM-like cells. We demonstrate the feasibility of manufacturing autologous anti-CMV Rapa-TCPs from blood of ESRD patients and KTx recipients with distinct CMV reactivation history. We confirm increased CD4/CD8 ratios and TCM proportions in Rapa- compared to untreated TCPs , which are associated with long-term clinical efficacy of adoptively transferred TCPs. In addition, our results imply that generation of CMV-specific TCPs prior to transplantation is possible, however, not favorable. Moreover, Rapa-TCPs from patients contained more viable cells after freezing/thawing compared to untreated TCPs.Although the cytokine combination IL-7/IL-15 was reported to preferentially promote TCM , we did not reveal substantial differences compared to cultures expanded with IL-2/IL7, suggesting these combinations are equally suited for TCP expansion and do not alter cell intrinsic mTOR-dependent signaling programs. Antigenic re-challenge and concomitant elimination of antigen-loaded APCs decreased T-cell expansion and resulted in T-cell culture contraction. These cultures predominantly contained long-lived memory T-cells. In line with this, among CMV-specific CD4+ T-cells, we recorded significant decreases in TEM and significant increases in TSCM in KTx recipients with a record of CMV DNAemia.Rapa-TCPs consistently contained more living T-cells than untreated TCPs, even after freezing/thawing. Hence, Rapa-TCPs may entail increased fitness following infusion, as TCPs are frozen until accomplishment of all quality controls in a clinical setting respecting GMP/GCP. As shown previously for B-cell lymphoma lines , Rapamycin increased Bcl-2 on protein level and moreover led to partial resistance to Fas-induced apoptosis. We did not identify altered gene expression of BCL-2 on mRNA level, however, expression of MYB, exerting anti-apoptotic effects via Bcl-2 , was upregulated in Rapa-TCPs. Furthermore, the decreased sensitivity to apoptosis of Rapa-TCPs may also be conferred by additional proteins and pathways as we found many other anti-apoptotic genes, such as BEX2  and SIX1 , to be overexpressed in Rapa-TCPs compared to untreated TCPs.We recorded higher OCR/ECAR ratios in untreated TCPs, although inhibition of glycolysis is associated with a long-lived memory phenotype  and memory T-cells are reported to preferentially perform fatty acid oxidation. Increased glycolysis in Rapa-TCPs was supported by RNA sequencing data showing increased expression of EPAS1 , however, also CHDH, a gene involved in fatty acid oxidation , showed increased expression in Rapa-TCPs. In fact, glycolysis is reported to allow immediate effector function , which is in line with the Rapa-TCPs' enhanced capacity for IFNγ production and increased expression of MAP3K21 allowing a rapid switch from a rested to an activated state. Furthermore, the metabolism of Rapa-TCPs remained more stable upon CMV-specific activation. Memory cell self-renewal might occur at a comparable number to that of effector T-cell generation in Rapa-TCPs, whereas in untreated TCPs, the balance might be extremely skewed toward effector T-cells upon activation, leading to this significant decrease in OCR/ECAR ratio and lack of long-lived memory T-cells.Intriguingly, 84% of genes differentially expressed between untreated and Rapa-TCPs, whose function we could allocate to impact TCPs, were regulated toward promotion of an effective and long-lived product. Our data imply a TCM-like transcriptome of long-lived poly-functional memory T-cells for Rapa-TCPs. The fact that Rapa-TCPs show increased clonal diversity may occur due to survival of low frequency clones. Especially, the preferential expansion of TCM in Rapa-TCPs may contribute to increased clonal diversity, as this subset was shown to have a higher clonal diversity compared to further differentiated memory T-cells. In fact, this may also be the underlying mechanism, why we have more CD4+ T-cells in the Rapa-TCP, because, evident from our ex vivo data, proportions of CMV-specific TCM are much higher among CD4+ compared to CD8+ T-cells.When CMV-specific T-cells from different memory T-cell subsets were cultured individually, we revealed preferential expansion of TCM with preserved TCM phenotype, partially protected from differentiation in the presence of Rapamycin, but partially also conversion of T-cells from other memory T-cell subsets into TCM-like cells. Interestingly, patient 12 lacked CMV-specific early CD8+ memory T-cells and his CMV-specific T-cells consisted to 90.6% of TEMRA. However, the respective Rapa-TCP included a strikingly high proportion of CD8+ TCM suggesting reprogramming of TEM/TEMRA to TCM.We did not record major differences in the characteristics of the TCPs irrespective of whether the TCPs were generated post-KTx or pre-KTx. Nonetheless, CD8+ T-cells comprised higher proportions of TCM and cytokine producers in TCPs generated pre-KTx. Hence, there is no benefit to generate anti-CMV TCPs prior to KTx.Investigating ex vivo-T-cell responses to CMV-specific stimuli, we found higher frequencies of CD8+ than of CD4+ CMV-specific T-cells as described previously. Frequencies of below 0.2% of CMV-reactive CD8+ T-cells were found in 2/19 patients compared to 6/13 HDs, which matches findings suggesting an increase of CD8+ CMV-reactive T-cells post-Tx. Interestingly, low frequencies of CMV-reactive CD8+ T-cells did not cause low TCP yields, underlining the feasibility of TCP generation. Intriguingly, a high proportion of CD8+ TEMRA among CMV-responsive T-cells did neither cause low expansion/yield nor high CD4/CD8 ratio in the TCP, which stresses the applicability even to patients with a high degree of terminal T-cell differentiation.The dosage for successful CMV-specific AVTT in the SOT setting is undetermined. Case reports suggest numbers between 30 and 245 million T-cells. Importantly, our protocol for TCP generation achieves medians of IFNγ-producers of 20 and 50% in CD4+ and CD8+ T-cells, respectively, and median killing rates of 85%, while these frequencies are much lower in other published approaches reporting a maximum of 8% of CMV-reactive IFNγ-producers in the TCP and specific lysis of ≤50% at higher T-cell/target ratios. Comparable values are only achieved in a recent study. Based on the superiority in function and phenotype of Rapa-TCPs, we assume that also smaller T-cell numbers would be efficient for long-term control of CMV in KTx patients. The patients in this study whose TCPs yielded < 5 million T-cells from 20 ml of peripheral blood, included the two patients with the lowest lymphocyte counts (Patients 6/10), suggesting an amendment of the amount of blood collected to the lymphocyte count. Patient 6, with no recorded CMV DNAemia, was diagnosed with acute rejection and treated with ATG before blood collection. ATG administration may be an indication to collect blood for TCP generation preventively, given the patients' risk of developing CMV disease. Indeed, transient CMV DNAemia was recorded in this patient 54 days after blood collection for the study. Patient 10 had a record of recent CMV viremia, was seronegative pre-KTx, received a kidney from a seropositive donor and had extensive CMV-associated complications. Notably, he was the oldest patient included and correlation analysis revealed a negative correlation between age and yield of Rapa-TCPs. We also recorded a low yield of the Rapa-TCP of Patient 19, who had a history of CMV-DNAemia and was receiving Acyclovir-treatment at the time of blood collection, which was reported to diminish IFNγ-production in response to CMVpp65 peptides. In cases as described above, we suggest to first generate untreated TCPs for patients with acute CMV disease to diminish viral load  and then successively generate and infuse Rapa-TCPs for long-term control of the virus. The fact that we recorded significantly lower yields in patients with a recent CMV DNAemia compared to HDs also suggests to follow the proposed approach in these patients and motivates to investigate more than the four patients that we were able to recruit for this study. In fact, we also found a negative correlation of the number of records with viremia and the yield of Rapa-TCPs for the ten patients with detected CMV DNAemia. However, this analysis may be biased by a more thorough screening of problematic patients, as the tests for CMV DNAemia were not of equal frequency in all patients. Moreover, patient 12, for whom we could also only generate a Rapa-TCP with a yield of around 5 million cells, suffered from chronic hepatitis. These, of course limited data, indicate that problems may occur during manufacture in the case of different chronic infections. It has to be thoroughly overthought whether it is possible to begin with more blood as starting material in these patients with numerous reactivating infections or use first the conventional approach and then generate a Rapa-TCP for long-term protection.Recently, Smith and colleagues published a study about the application of a comparable autologous CMV-specific TCP in SOT patients. However, their production process varies in many points, as they do not select for CMV-specific T-cells starting their culture with PBMCs, use G-Rex reactors instead of classical well-plates for expansion and have a different cytokine supplementation strategy using IL-21 and IL-2. They infuse multiple doses of TCPs at up to 6 different points in time. Of note, they demonstrate safety and clinical improvement in the majority of patients and could decrease or stop antiviral medication in many patients. Interestingly, compared to the time of infusion, they see an increase in viral load in 9 of 13 patients after infusion of TCPs. In 5 of the 9 monitored patients CD8+ CMV-specific T-cells were reduced by the end of monitoring (max. day 300) and in three of the cases this correlated with an increase in viral load. One of the patients died of CMV disease. These data demonstrate, that there is still room for optimization in the long-term outcome of SOT patients treated with autologous CMV-specific TCPs. Probably, Rapa-TCPs could improve long-term efficacy, however, the actual clinical performance of our TCP has to be demonstrated. Our preclinical data imply a long-lived TCP with beneficial properties maybe even allowing for a single infusion.A variety of other putative strategies are reported to rejuvenate T-cells with beneficial characteristics for AVTT. These include among others interference with different signaling pathways , use of different cytokine supplementation strategies , employment of certain microRNAs , modulation of metabolism , inhibition of ion channels , and promotion of autophagy. However, most of these are far from being practicable under GMP conditions for contemporary application to a clinical setting. This is also the case for genetic manipulation of T-cells for optimization of AVTT. Proposed strategies for genetic engineering include induction of resistance to immunosuppressive medication , introduction of suicide genes as safety switch  and knock out of anti-inflammatory signaling components such as PD-1 and LAG-3. All these suggestions have to be adapted to realistic GMP-feasible conditions and then may be valuable upgrades for even more sophisticated AVTT approaches. In contrast to other approaches, whose translation is less progressed at the moment, our minimally manipulative next-generation anti-CMV AVTT may help many transplanted patients whose endogenous immune system is not capable of defying the virus. Furthermore, the beneficial properties of Rapa-TCPs may also be transferred to other approaches using antigen-specific T-cells, e.g., other viruses or cancer immunotherapy with known antigens.In summary, our study revealed favorable phenotypic and functional properties of Rapa-TCPs as well as their applicability to a variety of ESRD/KTx patient samples. Ultimately, we seek for clinical confirmation of functionality and efficacy of Rapa-TCPs in a clinical proof-of-concept trial.Ethics StatementThe Charité Ethics Committee (IRB) approved the study protocol and all blood donors provided written informed consent.Author ContributionsLA, H-DV, PR, and MS-H conceptualized and designed the study. PR and NO provided patient samples and data. LA, TV, DW, and AJ performed experiments supervised by MS-H. LA acquired, analyzed, and interpreted data. LA and MS-H composed figures and manuscript. KJ performed bioinformatics analyses, created the respective graphs and made RNA seq data available at the GEO platform. TV, DW, SL-K, KJ, H-DV, and PR critically revised and all authors approved the final version of the manuscript.Conflict of Interest StatementH-DV, PR, and MS-H own a patent on the manufacture of Rapa-TCPs. The remaining authors declare that the research was conducted in the absence of any commercial or financial relationships that could be construed as a potential conflict of interest.\""
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft['body'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>date</th>\n",
       "      <th>has_text</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>xml</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Viral infections have a major impact on morbid...</td>\n",
       "      <td>2019-01-08</td>\n",
       "      <td>True</td>\n",
       "      <td>6546853</td>\n",
       "      <td>Comprehensive Characterization of a Next-Gener...</td>\n",
       "      <td>&lt;article article-type=\"research-article\" xmlns...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            abstract       date  has_text  \\\n",
       "0  Viral infections have a major impact on morbid... 2019-01-08      True   \n",
       "\n",
       "        id                                              title  \\\n",
       "0  6546853  Comprehensive Characterization of a Next-Gener...   \n",
       "\n",
       "                                                 xml  \n",
       "0  <article article-type=\"research-article\" xmlns...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import itertools\n",
    "# stream = row_stream(doc_stream(file_meta_stream(archives)))\n",
    "# rows = list(itertools.islice(stream, 5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tar = tarfile.open(osp.join(oadir, 'non_comm_use.0-9A-B.txt.tar.gz'), \"r:gz\")\n",
    "#/lab/data/pmc_oa/bulk/files/comm_use.0-9A-B.txt.tar.gz, file BMC_Cancer/PMC5481910.txt\n",
    "#tar = tarfile.open(osp.join(oadir, 'comm_use.0-9A-B.txt.tar.gz'), \"r:gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# f = tar.getmember('BMC_Cancer/PMC5481910.txt')\n",
    "# f.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# arr = tar.extractfile(f).read()\n",
    "# len(arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#arr.decode('utf-8', errors='strict')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tar = tarfile.open(osp.join(oadir, 'non_comm_use.0-9A-B.txt.tar.gz'), \"r:gz\")\n",
    "# files = tar.getmembers()\n",
    "# pd.Series([f.name for f in files]).sample(35).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tar.extractfile([f for f in files if f.name == 'Am_J_Respir_Crit_Care_Med/PMC6444650.txt'][0]).read()\n",
    "#tar.extractfile([f for f in files if f.name == 'Bioinformatics/PMC6084620.txt'][0]).read().decode('utf-8')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
