{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata Integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os.path as osp\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "%run env.py\n",
    "%run src/lib.py\n",
    "%run src/meta.py\n",
    "%matplotlib inline\n",
    "MIN_SYM_LEN = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hashlib\n",
    "ID_TYP_CK = 'CK'\n",
    "ID_TYP_TF = 'TF'\n",
    "ID_TYP_CT = 'CT'\n",
    "\n",
    "def get_ids(df, typ):\n",
    "    ids = [':'.join([r['src'], str(r['spid']), r['sym'], r['lbl']]) for _, r in df.iterrows()]\n",
    "    return [typ + hashlib.md5(v.encode('utf-8')).hexdigest()[:16].upper() for v in ids]\n",
    "\n",
    "def merge(dfs, typ):\n",
    "    cols = ['id', 'src', 'sym', 'lbl', 'spid', 'extid']\n",
    "    dfs = [df.copy() for df in dfs]\n",
    "    for df in dfs:\n",
    "        if 'id' in df:\n",
    "            raise ValueError('ID should not already be assigned')\n",
    "        df['id'] = get_ids(df, typ)\n",
    "    df = pd.concat([df[cols] for df in dfs])\n",
    "    return df\n",
    "\n",
    "def add_preferred_id(g):\n",
    "    \"\"\"Add preferred id for synonym record for a given source, label group\"\"\"\n",
    "    pid = g[g['sym'] == g['lbl']]\n",
    "    assert len(pid) == 1, \\\n",
    "        'Found {} records for src={}, lbl={}'\\\n",
    "        .format(len(pid), g['src'].iloc[0], g['lbl'].iloc[0])\n",
    "    pid = pid.iloc[0]['id']\n",
    "    return g.assign(prefid=pid)\n",
    "\n",
    "def add_preferred_ids(df):\n",
    "    return df.groupby(['src', 'lbl'], group_keys=False).apply(add_preferred_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cytokines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load from book table (Cameron and Kelvin)\n",
    "\n",
    "From: https://www.ncbi.nlm.nih.gov/books/NBK6294/table/A13506/?report=objectonly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>synonym</th>\n",
       "      <th>amino</th>\n",
       "      <th>chromosome_acids</th>\n",
       "      <th>mwt</th>\n",
       "      <th>receptor_form</th>\n",
       "      <th>receptor_location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL-1α</td>\n",
       "      <td>hematopoietin-1</td>\n",
       "      <td>271</td>\n",
       "      <td>2q14</td>\n",
       "      <td>30606</td>\n",
       "      <td>CD121a, CDw121b</td>\n",
       "      <td>2q12, 2q12-q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL-1β</td>\n",
       "      <td>catabolin</td>\n",
       "      <td>269</td>\n",
       "      <td>2q14</td>\n",
       "      <td>20747</td>\n",
       "      <td>CD121a, CDw121b</td>\n",
       "      <td>2q12,2q12-q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL-1RA</td>\n",
       "      <td>IL-1 receptor antagonist</td>\n",
       "      <td>177</td>\n",
       "      <td>2q14.2</td>\n",
       "      <td>20055</td>\n",
       "      <td>CD121a</td>\n",
       "      <td>2q12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL-18</td>\n",
       "      <td>interferon-γ inducing factor</td>\n",
       "      <td>193</td>\n",
       "      <td>11q22.2-q22.3</td>\n",
       "      <td>22326</td>\n",
       "      <td>IL-18Rα, β</td>\n",
       "      <td>2q12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL-2</td>\n",
       "      <td>T cell growth factor</td>\n",
       "      <td>153</td>\n",
       "      <td>4q26-q27</td>\n",
       "      <td>17628</td>\n",
       "      <td>CD25, 122,132</td>\n",
       "      <td>10p15-p14, 22q13.1, Xq13.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name                       synonym amino chromosome_acids    mwt  \\\n",
       "0   IL-1α               hematopoietin-1   271             2q14  30606   \n",
       "1   IL-1β                     catabolin   269             2q14  20747   \n",
       "2  IL-1RA      IL-1 receptor antagonist   177           2q14.2  20055   \n",
       "3   IL-18  interferon-γ inducing factor   193    11q22.2-q22.3  22326   \n",
       "4    IL-2          T cell growth factor   153         4q26-q27  17628   \n",
       "\n",
       "     receptor_form           receptor_location  \n",
       "0  CD121a, CDw121b              2q12, 2q12-q22  \n",
       "1  CD121a, CDw121b               2q12,2q12-q22  \n",
       "2           CD121a                        2q12  \n",
       "3       IL-18Rα, β                        2q12  \n",
       "4    CD25, 122,132  10p15-p14, 22q13.1, Xq13.1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRC = 'cameron'\n",
    "dfr = pd.read_csv(osp.join(META_DATA_DIR, 'raw', 'cytokines.cameron.csv'), sep='\\t')\n",
    "dfr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL-1α</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL-1β</td>\n",
       "      <td>IL-1β</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL-1RA</td>\n",
       "      <td>IL-1RA</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL-18</td>\n",
       "      <td>IL-18</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL-2</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sym     lbl  spid      src\n",
       "0   IL-1α   IL-1α     1  cameron\n",
       "1   IL-1β   IL-1β     1  cameron\n",
       "2  IL-1RA  IL-1RA     1  cameron\n",
       "3   IL-18   IL-18     1  cameron\n",
       "4    IL-2    IL-2     1  cameron"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = (\n",
    "    pd.concat([\n",
    "        dfr[['name', 'name']].set_axis(['sym', 'lbl'], axis=1, inplace=False),\n",
    "        dfr.set_index('name')['synonym'].str.split(',', expand=True).stack().str.strip()\n",
    "            .reset_index().set_axis(['lbl', 'i', 'sym'], axis=1, inplace=False)\n",
    "            .filter(items=['sym', 'lbl'])\n",
    "    ])\n",
    "    .assign(spid=SPECIES_HUMAN_ID, src=SRC)\n",
    ")\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "      <th>extid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL-1α</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL-1β</td>\n",
       "      <td>IL-1β</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL-1RA</td>\n",
       "      <td>IL-1RA</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL-18</td>\n",
       "      <td>IL-18</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL-2</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>1</td>\n",
       "      <td>cameron</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sym     lbl  spid      src extid\n",
       "0   IL-1α   IL-1α     1  cameron  None\n",
       "1   IL-1β   IL-1β     1  cameron  None\n",
       "2  IL-1RA  IL-1RA     1  cameron  None\n",
       "3   IL-18   IL-18     1  cameron  None\n",
       "4    IL-2    IL-2     1  cameron  None"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def transform(df, mask, fn):\n",
    "    df = df.loc[mask.values].copy()\n",
    "    df['sym'] = df['sym'].apply(fn)\n",
    "    return df\n",
    "df = dft.append(transform(dft, dft['sym'].str.match(r'IL-[^ ]+$|TNF-|IFN-|TGF-'), lambda v: v.replace('-', '')))\n",
    "df['sym'] = df['sym'].str.strip()\n",
    "df['lbl'] = df['lbl'].str.strip()\n",
    "df['extid'] = None\n",
    "df = df[df['sym'].str.len() > 0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['IL-1α', 'IL-1β', 'IL-1RA', 'IL-18', 'IL-2', 'IL-4', 'IL-7',\n",
       "       'IL-9', 'IL-13', 'IL-15', 'IL-3', 'IL-5', 'GM-CSF', 'IL-6',\n",
       "       'IL-11', 'G-CSF', 'IL-12', 'LIF', 'OSM', 'IL-10', 'IL-20', 'IL-14',\n",
       "       'IL-16', 'IL-17', 'IFN-α', 'IFN-β', 'IFN-γ', 'CD154', 'LT-β',\n",
       "       'TNF-α', 'TNF-β', '4-1BBL', 'APRIL', 'CD70', 'CD153', 'CD178',\n",
       "       'GITRL', 'LIGHT', 'OX40L', 'TALL-1', 'TRAIL', 'TWEAK', 'TRANCE',\n",
       "       'TGF-β', 'TGF-β2', 'TGF-β3', 'hematopoietin-1', 'catabolin',\n",
       "       'IL-1 receptor antagonist', 'interferon-γ inducing factor',\n",
       "       'T cell growth factor', 'BSF-1', 'T cell growth factor P40',\n",
       "       'P600', 'multipotential CSF', 'MCGF', 'BCDF-1', 'CSF-2', 'IFN-β2',\n",
       "       'BSF-2', 'AGIF', 'CSF-3', 'NK cell stimulatory factor',\n",
       "       'leukemia inhibitory factor', 'oncostatin M', 'CSIF', 'HMW-BCGF',\n",
       "       'LCF', 'CTLA-8', 'CD40L', 'TRAP', 'cachectin', 'LT-α', 'TALL-2',\n",
       "       'CD27L', 'CD30L', 'FasL', 'Apo2L', 'Apo3L', 'OPGL', 'TGF-β1',\n",
       "       'IL1α', 'IL1β', 'IL1RA', 'IL18', 'IL2', 'IL4', 'IL7', 'IL9',\n",
       "       'IL13', 'IL15', 'IL3', 'IL5', 'IL6', 'IL11', 'IL12', 'IL10',\n",
       "       'IL20', 'IL14', 'IL16', 'IL17', 'IFNα', 'IFNβ', 'IFNγ', 'TNFα',\n",
       "       'TNFβ', 'TGFβ', 'TGFβ2', 'TGFβ3', 'IFNβ2', 'TGFβ1'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['sym'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cameron = df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load from mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original result info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 568 entries, 0 to 567\n",
      "Data columns (total 8 columns):\n",
      "_id             568 non-null object\n",
      "_score          568 non-null float64\n",
      "alias           531 non-null object\n",
      "ensembl         30 non-null object\n",
      "ensembl.gene    522 non-null object\n",
      "name            568 non-null object\n",
      "symbol          568 non-null object\n",
      "taxid           568 non-null int64\n",
      "dtypes: float64(1), int64(1), object(6)\n",
      "memory usage: 35.6+ KB\n",
      "None\n",
      "Score histogram:\n",
      "Original result info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 255 entries, 0 to 254\n",
      "Data columns (total 8 columns):\n",
      "_id             255 non-null object\n",
      "_score          255 non-null float64\n",
      "alias           230 non-null object\n",
      "ensembl         20 non-null object\n",
      "ensembl.gene    225 non-null object\n",
      "name            255 non-null object\n",
      "symbol          255 non-null object\n",
      "taxid           255 non-null int64\n",
      "dtypes: float64(1), int64(1), object(6)\n",
      "memory usage: 16.0+ KB\n",
      "None\n",
      "Score histogram:\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAEHpJREFUeJzt3X+MZWV9x/H3R5ZKw1oXip1sl22HplsbLBV1gjT6x6BpRSBdTAwBqaLSrn9Agsk27eo/YA0JTYpa09ZkFSKm4kqUFiLUlm65pfyBuotU+VHiVobKZoVYEbkYaAa+/WMOesFl753fd555v5LJnPPc55z7zHdnPvPsc889k6pCktSul632ACRJy8ugl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDVuw2oPAOCEE06oycnJeR/31FNPceyxxy79gBpjnUZjnYazRqNZqTrt37//B1X1qmH9xiLoJycn2bdv37yP6/V6TE9PL/2AGmOdRmOdhrNGo1mpOiV5eJR+Lt1IUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjxuKdsYsxueuWkfrNXHX2Mo9EksaTM3pJatzQoE+yNcntSe5Pcl+Sy7r2K5IcTHJP93HWwDEfSnIgyYNJ3racX4Ak6chGWbqZBXZW1d1JXgHsT3Jb99jHq+qvBjsnORk4H3gN8KvAvyb5rap6dikHLkkazdAZfVUdqqq7u+0ngQeALUc4ZDuwp6qeqaqHgAPAaUsxWEnS/M1rjT7JJPA64Gtd06VJvpXk2iTHdW1bgO8NHPYIR/7FIElaRqmq0TomG4F/B66sqhuTTAA/AAr4KLC5qt6f5G+Au6rq77vjrgH+qaq+9KLz7QB2AExMTLxhz5498x58v9/noSdGWxE6Zcsr533+VvT7fTZu3Ljawxh71mk4azSalarTGWecsb+qpob1G+nyyiRHA18GPl9VNwJU1aMDj38a+Eq3exDYOnD4iV3bC1TVbmA3wNTUVC3kJv29Xo+r73xqpL4zF87//K3wj0WMxjoNZ41GM251GuWqmwDXAA9U1ccG2jcPdHsHcG+3fTNwfpKXJzkJ2AZ8femGLEmaj1Fm9G8C3g18O8k9XduHgQuSnMrc0s0M8AGAqrovyQ3A/cxdsXOJV9xI0uoZGvRVdSeQwzx06xGOuRK4chHjkiQtEd8ZK0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWrc0KBPsjXJ7UnuT3Jfksu69uOT3JbkO93n47r2JPlkkgNJvpXk9cv9RUiSXtooM/pZYGdVnQycDlyS5GRgF7C3qrYBe7t9gLcD27qPHcCnlnzUkqSRDQ36qjpUVXd3208CDwBbgO3AdV2364Bzu+3twOdqzl3ApiSbl3zkkqSRpKpG75xMAncAvwP8T1Vt6toDPF5Vm5J8Bbiqqu7sHtsL/HlV7XvRuXYwN+NnYmLiDXv27Jn34Pv9Pg898exIfU/Z8sp5n78V/X6fjRs3rvYwxp51Gs4ajWal6nTGGWfsr6qpYf02jHrCJBuBLwMfrKofz2X7nKqqJKP/xpg7ZjewG2Bqaqqmp6fnczgAvV6Pq+98aqS+MxfO//yt6PV6LKS+6411Gs4ajWbc6jTSVTdJjmYu5D9fVTd2zY8+vyTTfX6saz8IbB04/MSuTZK0Cka56ibANcADVfWxgYduBi7qti8Cbhpof0939c3pwBNVdWgJxyxJmodRlm7eBLwb+HaSe7q2DwNXATckuRh4GDive+xW4CzgAPAT4H1LOmJJ0rwMDfruRdW8xMNvPUz/Ai5Z5LgkSUvEd8ZKUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktS4oUGf5NokjyW5d6DtiiQHk9zTfZw18NiHkhxI8mCSty3XwCVJoxllRv9Z4MzDtH+8qk7tPm4FSHIycD7wmu6Yv0ty1FINVpI0f0ODvqruAH444vm2A3uq6pmqegg4AJy2iPFJkhZpwyKOvTTJe4B9wM6qehzYAtw10OeRru3nJNkB7ACYmJig1+vNewD9fp+dpzw7Ut+FnL8V/X5/XX/9o7JOw1mj0YxbnRYa9J8CPgpU9/lq4P3zOUFV7QZ2A0xNTdX09PS8B9Hr9bj6zqdG6jtz4fzP34per8dC6rveWKfhrNFoxq1OC7rqpqoerapnq+o54NP8bHnmILB1oOuJXZskaZUsKOiTbB7YfQfw/BU5NwPnJ3l5kpOAbcDXFzdESdJiDF26SfIFYBo4IckjwOXAdJJTmVu6mQE+AFBV9yW5AbgfmAUuqarRFtElSctiaNBX1QWHab7mCP2vBK5czKAkSUvHd8ZKUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjFnOb4jVlctctQ/vMXHX2CoxEklaWM3pJapxBL0mNM+glqXFNrNHPHPOuoX0mn75+BUYiSePHGb0kNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1bmjQJ7k2yWNJ7h1oOz7JbUm+030+rmtPkk8mOZDkW0lev5yDlyQNN8qM/rPAmS9q2wXsraptwN5uH+DtwLbuYwfwqaUZpiRpoYYGfVXdAfzwRc3bgeu67euAcwfaP1dz7gI2Jdm8VIOVJM3fQtfoJ6rqULf9fWCi294CfG+g3yNdmyRplWxY7AmqqpLUfI9LsoO55R0mJibo9Xrzfu5+v8/OU56l97KPDO2787nZoX0WMoa1oN/vN/u1LZVzzjmHquKWW25Z7aGMNb+XRjNudVpo0D+aZHNVHeqWZh7r2g8CWwf6ndi1/Zyq2g3sBpiamqrp6el5D6LX63H1nU8xc8zlQ/u+9+nrh/aZuXD+Y1gLer0eC6nverJhwwZmZ2et0xB+L41m3Oq00KWbm4GLuu2LgJsG2t/TXX1zOvDEwBKPJGkVDJ3RJ/kCMA2ckOQR4HLgKuCGJBcDDwPndd1vBc4CDgA/Ad63DGOWJM3D0KCvqgte4qG3HqZvAZcsdlCSpKXjO2MlqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWrchsUcnGQGeBJ4FpitqqkkxwNfBCaBGeC8qnp8ccOUJC3UUszoz6iqU6tqqtvfBeytqm3A3m5fkrRKlmPpZjtwXbd9HXDuMjyHJGlEiw36Av4lyf4kO7q2iao61G1/H5hY5HNIkhZhUWv0wJur6mCSXwFuS/Jfgw9WVSWpwx3Y/WLYATAxMUGv15v3k/f7fXae8iy9l31kaN+dz80O7bOQMawF/X6/2a9tqczOzlJV1mkIv5dGM251StVhc3j+J0quAPrAnwDTVXUoyWagV1WvPtKxU1NTtW/fvnk/Z6/X471ffYqZY941tO/k09cP7TNz1dnzHsNa0Ov1mJ6eXu1hjLVNmzYxOztLv99f7aGMNb+XRrNSdUqyf+D10Ze04KWbJMcmecXz28AfAPcCNwMXdd0uAm5a6HNIkhZvMUs3E8A/JHn+PNdX1VeTfAO4IcnFwMPAeYsfpiRpoRYc9FX1XeC1h2n/X+CtixmUJGnp+M5YSWrcYq+6acrkrltG6tfqi7aS2mTQS+vJFa8coc8Tyz8OrSiXbiSpcQa9JDXOoJekxhn0ktQ4g16SGudVN5K0DMbpcm1n9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc7r6KW1YAXvOvlS1397e+61yxm9JDVu3czoZ4551xEfn3z6+hUaiSStLGf0ktQ4g16SGmfQS1LjDHpJapxBL0mNWzdX3UgaH+N0r/b1wBm9JDXOGf0CjDIbOeJMZJ7vcnT2Iy3eon9u1zCDviEv9Y2885RZ3jvwWKvfzOvd4f79/bcWGPQ/Neyds+C7ZyWtTQb9MjnSfxNnjlnBgUha9wx6abWN8pqNtAjLFvRJzgT+GjgK+ExVXbVcz7VeeaM2SaNYlqBPchTwt8DvA48A30hyc1XdvxzP16SBWZ5LPVq3VvA+/C1brhn9acCBqvouQJI9wHag+aAf5UVdSVpJyxX0W4DvDew/ArxxmZ5L87SerydeCOv1IkvwmsLMMeO1tPjTCdoVR+i0hv/nkKpa+pMm7wTOrKo/7vbfDbyxqi4d6LMD2NHtvhp4cAFPdQLwg0UOdz2wTqOxTsNZo9GsVJ1+vapeNazTcs3oDwJbB/ZP7Np+qqp2A7sX8yRJ9lXV1GLOsR5Yp9FYp+Gs0WjGrU7Lda+bbwDbkpyU5BeA84Gbl+m5JElHsCwz+qqaTXIp8M/MXV55bVXdtxzPJUk6smW7jr6qbgVuXa7zdxa19LOOWKfRWKfhrNFoxqpOy/JirCRpfHg/eklq3JoN+iRnJnkwyYEku1Z7POMiybVJHkty70Db8UluS/Kd7vNxqznG1ZZka5Lbk9yf5L4kl3Xt1mlAkmOSfD3Jf3Z1+kjXflKSr3U/e1/sLrhY15IcleSbSb7S7Y9VjdZk0A/cYuHtwMnABUlOXt1RjY3PAme+qG0XsLeqtgF7u/31bBbYWVUnA6cDl3TfP9bphZ4B3lJVrwVOBc5Mcjrwl8DHq+o3gceBi1dxjOPiMuCBgf2xqtGaDHoGbrFQVf8HPH+LhXWvqu4Afvii5u3Add32dcC5KzqoMVNVh6rq7m77SeZ+QLdgnV6g5vS73aO7jwLeAnypa1/3dUpyInA28JluP4xZjdZq0B/uFgtbVmksa8FEVR3qtr8PTKzmYMZJkkngdcDXsE4/p1uSuAd4DLgN+G/gR1U123XxZw8+AfwZ8Fy3/8uMWY3WatBrgWruMisvtQKSbAS+DHywqn48+Jh1mlNVz1bVqcy9u/004LdXeUhjJck5wGNVtX+1x3Ika/UPjwy9xYJe4NEkm6vqUJLNzM3O1rUkRzMX8p+vqhu7Zuv0EqrqR0luB34P2JRkQzdjXe8/e28C/jDJWcAxwC8x93c4xqpGa3VG7y0W5udm4KJu+yLgplUcy6rr1lCvAR6oqo8NPGSdBiR5VZJN3fYvMvf3JR4Abgfe2XVb13Wqqg9V1YlVNclcDv1bVV3ImNVozb5hqvsN+gl+douFK1d5SGMhyReAaebunvcocDnwj8ANwK8BDwPnVdWLX7BdN5K8GfgP4Nv8bF31w8yt01unTpLfZe6FxKOYmxTeUFV/keQ3mLsA4njgm8AfVdUzqzfS8ZBkGvjTqjpn3Gq0ZoNekjSatbp0I0kakUEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1Lj/h/y+uzLzYVwWgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "SRC = 'mygene'\n",
    "df = pd.concat([\n",
    "    mygene_query('interleukin'),\n",
    "    mygene_query('chemokine')\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 172 entries, 0 to 76\n",
      "Data columns (total 7 columns):\n",
      "extid           172 non-null object\n",
      "alias           164 non-null object\n",
      "ensembl         10 non-null object\n",
      "ensembl.gene    157 non-null object\n",
      "name            172 non-null object\n",
      "symbol          172 non-null object\n",
      "taxid           172 non-null int64\n",
      "dtypes: int64(1), object(6)\n",
      "memory usage: 10.8+ KB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 102 entries, 0 to 76\n",
      "Data columns (total 7 columns):\n",
      "extid           102 non-null object\n",
      "alias           98 non-null object\n",
      "ensembl         10 non-null object\n",
      "ensembl.gene    87 non-null object\n",
      "name            102 non-null object\n",
      "symbol          102 non-null object\n",
      "taxid           102 non-null int64\n",
      "dtypes: int64(1), object(6)\n",
      "memory usage: 6.4+ KB\n"
     ]
    }
   ],
   "source": [
    "df = df[(~df['name'].str.contains('receptor'))]\n",
    "df = df[df['name'].str.contains('interleukin') | df['name'].str.contains('chemokine')]\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 747 entries, 0 to 775\n",
      "Data columns (total 5 columns):\n",
      "sym      747 non-null object\n",
      "lbl      747 non-null object\n",
      "extid    747 non-null object\n",
      "spid     747 non-null int64\n",
      "src      747 non-null object\n",
      "dtypes: int64(1), object(4)\n",
      "memory usage: 35.0+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>extid</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL-17F</td>\n",
       "      <td>IL-17F</td>\n",
       "      <td>112744</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>interleukin 17F</td>\n",
       "      <td>IL-17F</td>\n",
       "      <td>112744</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL17F</td>\n",
       "      <td>IL-17F</td>\n",
       "      <td>112744</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CANDF6</td>\n",
       "      <td>IL-17F</td>\n",
       "      <td>112744</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ML-1</td>\n",
       "      <td>IL-17F</td>\n",
       "      <td>112744</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sym     lbl   extid  spid     src\n",
       "0           IL-17F  IL-17F  112744     1  mygene\n",
       "1  interleukin 17F  IL-17F  112744     1  mygene\n",
       "2            IL17F  IL-17F  112744     1  mygene\n",
       "3           CANDF6  IL-17F  112744     1  mygene\n",
       "5             ML-1  IL-17F  112744     1  mygene"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def label_fn(v):\n",
    "    # Prefer hyphonated labels (e.g. IL-2) for compatibility with other sources\n",
    "    if not v.startswith('IL'):\n",
    "        return v\n",
    "    if 'IL-' in v:\n",
    "        return v\n",
    "    return 'IL-' + v[2:]\n",
    "df = mygene_prep(df, label_fn=label_fn)\n",
    "df = df.assign(spid=SPECIES_HUMAN_ID, src=SRC)\n",
    "df = df.drop_duplicates(subset=['sym', 'lbl'])\n",
    "print(df.info())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#np.sort(df['lbl'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['IL-17F', 'interleukin 17F', 'IL17F', 'CANDF6', 'ML-1', 'ML1',\n",
       "       'IL-34', 'interleukin 34', 'IL34', 'C16orf77', 'IL-17C',\n",
       "       'interleukin 17C', 'IL17C', 'CX2', 'IL-17B', 'interleukin 17B',\n",
       "       'IL17B', 'IL-20', 'NIRF', 'ZCYTO7', 'IL-9', 'interleukin 9', 'IL9',\n",
       "       'HP40', 'P40', 'IL-15', 'interleukin 15', 'IL15', 'IL-7',\n",
       "       'interleukin 7', 'IL7', 'IL-18', 'interleukin 18', 'IL18', 'IGIF',\n",
       "       'IL-1g', 'IL1F4', 'IL-12A', 'interleukin 12A', 'IL12A', 'CLMF',\n",
       "       'NFSK', 'NKSF1', 'P35', 'IL-31', 'interleukin 31', 'IL31', 'IL-24',\n",
       "       'interleukin 24', 'IL24', 'C49A', 'FISP', 'IL10B', 'MDA7', 'MOB5',\n",
       "       'ST16', 'IL-2', 'interleukin 2', 'IL2', 'TCGF', 'lymphokine',\n",
       "       'IL-32', 'interleukin 32', 'IL32', 'IL-32alpha', 'IL-32beta',\n",
       "       'IL-32delta', 'IL-32gamma', 'NK4', 'TAIF', 'TAIFa', 'TAIFb',\n",
       "       'TAIFc', 'TAIFd', 'IL-22', 'interleukin 22', 'IL22', 'IL-21',\n",
       "       'IL-D110', 'IL-TIF', 'ILTIF', 'TIFIL-23', 'TIFa', 'zcyto18',\n",
       "       'IL-26', 'interleukin 26', 'IL26', 'AK155', 'IL-3',\n",
       "       'interleukin 3', 'IL3', 'MCGF', 'MULTI-CSF', 'IL-17A',\n",
       "       'interleukin 17A', 'IL17A', 'CTLA-8', 'CTLA8', 'IL-17', 'IL17',\n",
       "       'IL-25', 'interleukin 25', 'IL25', 'IL17E', 'IL-17D',\n",
       "       'interleukin 17D', 'IL17D', 'IL-33', 'interleukin 33', 'IL33',\n",
       "       'C9orf26', 'DVS27', 'IL1F11', 'NF-HEV', 'NFEHEV', 'interleukin 20',\n",
       "       'IL20', 'IL10D', 'ZCYTO10', 'IL-27', 'interleukin 27', 'IL27',\n",
       "       'IL-27A', 'IL27A', 'IL27p28', 'IL30', 'p28', 'IL-16',\n",
       "       'interleukin 16', 'IL16', 'LCF', 'NIL16', 'PRIL16', 'prIL-16',\n",
       "       'IL-13', 'interleukin 13', 'IL13', 'P600', 'IL-12B',\n",
       "       'interleukin 12B', 'IL12B', 'CLMF2', 'IMD28', 'IMD29', 'NKSF',\n",
       "       'NKSF2', 'IL-5', 'interleukin 5', 'IL5', 'EDF', 'TRF', 'IL-6',\n",
       "       'interleukin 6', 'IL6', 'BSF-2', 'BSF2', 'CDF', 'HGF', 'HSF',\n",
       "       'IFN-beta-2', 'IFNB2', 'IL-10', 'interleukin 10', 'IL10', 'CSIF',\n",
       "       'GVHDS', 'IL10A', 'TGIF', 'IL-11', 'interleukin 11', 'IL11',\n",
       "       'AGIF', 'IL-19', 'interleukin 19', 'IL19', 'IL-10C', 'MDA1',\n",
       "       'NG.1', 'ZMDA1', 'IL-37', 'interleukin 37', 'IL37', 'FIL1',\n",
       "       'FIL1(ZETA)', 'FIL1Z', 'IL-1F7', 'IL-1H', 'IL-1H4', 'IL-1RP1',\n",
       "       'IL1F7', 'IL1H4', 'IL1RP1', 'IL-4', 'interleukin 4', 'IL4',\n",
       "       'BCGF-1', 'BCGF1', 'BSF-1', 'BSF1', 'interleukin 21', 'IL21',\n",
       "       'CVID11', 'Za11', 'IL-1A', 'interleukin 1 alpha', 'IL1A', 'IL1',\n",
       "       'IL1-ALPHA', 'IL1F1', 'IL-36A', 'interleukin 36 alpha', 'IL36A',\n",
       "       'FIL1(EPSILON)', 'FIL1E', 'IL-1F6', 'IL1(EPSILON)', 'IL1F6',\n",
       "       'IL-36G', 'interleukin 36 gamma', 'IL36G', 'IL-1F9', 'IL-1H1',\n",
       "       'IL-1RP2', 'IL1E', 'IL1F9', 'IL1H1', 'IL1RP2', 'IL-1B',\n",
       "       'interleukin 1 beta', 'IL1B', 'IL-1', 'IL1-BETA', 'IL1F2',\n",
       "       'IL-36B', 'interleukin 36 beta', 'IL36B', 'FIL1-(ETA)', 'FIL1H',\n",
       "       'FILI-(ETA)', 'IL-1F8', 'IL-1H2', 'IL1-ETA', 'IL1F8', 'IL1H2',\n",
       "       'IL-4I1', 'interleukin 4 induced 1', 'IL4I1', 'FIG1', 'LAAO',\n",
       "       'LAO', 'IL-6ST', 'interleukin 6 signal transducer', 'IL6ST',\n",
       "       'CD130', 'CDW130', 'GP130', 'IL-6RB', 'IL-23A',\n",
       "       'interleukin 23 subunit alpha', 'IL23A', 'IL-23', 'IL23P19', 'P19',\n",
       "       'SGRF', 'IL-18BP', 'interleukin 18 binding protein', 'IL18BP',\n",
       "       'IL18BPa', 'IL-F3', 'interleukin enhancer binding factor 3',\n",
       "       'ILF3', 'CBTF', 'DRBF', 'DRBP76', 'MMP4', 'MPHOSPH4', 'MPP4',\n",
       "       'MPP4110', 'NF-AT-90', 'NF110', 'NF110b', 'NF90', 'NF90a', 'NF90b',\n",
       "       'NF90c', 'NF90ctv', 'NFAR', 'NFAR-1', 'NFAR-2', 'NFAR110', 'NFAR2',\n",
       "       'NFAR90', 'TCP110', 'TCP80', 'IL-1F10',\n",
       "       'interleukin 1 family member 10', 'IL1F10', 'FIL1-theta', 'FKSG75',\n",
       "       'IL-1HY2', 'IL-38', 'IL1-theta', 'IL1HY2', 'NFIL3',\n",
       "       'nuclear factor, interleukin 3 regulated', 'E4BP4', 'IL3BP1',\n",
       "       'NF-IL3A', 'NFIL3A', 'IL-F2',\n",
       "       'interleukin enhancer binding factor 2', 'ILF2', 'NF45', 'PRO3063',\n",
       "       'LIF', 'LIF interleukin 6 family cytokine', 'DIA', 'HILDA',\n",
       "       'MLPLI', 'LOC107983991',\n",
       "       'nuclear factor interleukin-3-regulated protein-like', 'CKLF',\n",
       "       'chemokine like factor', 'C32', 'CKLF1', 'CKLF2', 'CKLF3', 'CKLF4',\n",
       "       'HSPC224', 'UCK-1', 'FP248',\n",
       "       'Putative chemokine-related protein FP248', 'CCL23',\n",
       "       'C-C motif chemokine ligand 23', 'CK-BETA-8', 'CKb8', 'Ckb-8',\n",
       "       'Ckb-8-1', 'MIP-3', 'MIP3', 'MPIF-1', 'SCYA23', 'hmrp-2a', 'TAFA1',\n",
       "       'TAFA chemokine like family member 1', 'FAM19A1', 'TAFA-1',\n",
       "       'TAFA2', 'TAFA chemokine like family member 2', 'FAM19A2',\n",
       "       'TAFA-2', 'CCL24', 'C-C motif chemokine ligand 24', 'Ckb-6',\n",
       "       'MPIF-2', 'MPIF2', 'SCYA24', 'CCL5',\n",
       "       'C-C motif chemokine ligand 5', 'D17S136E', 'RANTES', 'SCYA5',\n",
       "       'SIS-delta', 'SISd', 'TCP228', 'eoCP', 'CCL14',\n",
       "       'C-C motif chemokine ligand 14', 'CC-1', 'CC-3', 'CKB1', 'HCC-1',\n",
       "       'HCC-1(1-74)', 'HCC-1/HCC-3', 'HCC-3', 'MCIF', 'NCC-2', 'NCC2',\n",
       "       'SCYA14', 'SCYL2', 'SY14', 'CCL11',\n",
       "       'C-C motif chemokine ligand 11', 'SCYA11', 'CCL2',\n",
       "       'C-C motif chemokine ligand 2', 'GDCF-2', 'HC11', 'HSMCR30',\n",
       "       'MCAF', 'MCP-1', 'MCP1', 'SCYA2', 'SMC-CF', 'CCL25',\n",
       "       'C-C motif chemokine ligand 25', 'Ckb15', 'SCYA25', 'TECK',\n",
       "       'CCL13', 'C-C motif chemokine ligand 13', 'CKb10', 'MCP-4',\n",
       "       'NCC-1', 'NCC1', 'SCYA13', 'SCYL1', 'CCL27',\n",
       "       'C-C motif chemokine ligand 27', 'ALP', 'CTACK', 'CTAK', 'ESKINE',\n",
       "       'ILC', 'PESKY', 'SCYA27', 'XCL1', 'X-C motif chemokine ligand 1',\n",
       "       'ATAC', 'LPTN', 'LTN', 'SCM-1', 'SCM-1a', 'SCM1', 'SCM1A', 'SCYC1',\n",
       "       'XCL2', 'X-C motif chemokine ligand 2', 'SCM-1b', 'SCM1B', 'SCYC2',\n",
       "       'CCL28', 'C-C motif chemokine ligand 28', 'CCK1', 'MEC', 'SCYA28',\n",
       "       'CCL20', 'C-C motif chemokine ligand 20', 'CKb4', 'Exodus', 'LARC',\n",
       "       'MIP-3-alpha', 'MIP-3a', 'MIP3A', 'SCYA20', 'ST38', 'CCL26',\n",
       "       'C-C motif chemokine ligand 26', 'IMAC', 'MIP-4a', 'MIP-4alpha',\n",
       "       'SCYA26', 'TSC-1', 'CCL15', 'C-C motif chemokine ligand 15',\n",
       "       'HCC-2', 'HMRP-2B', 'LKN-1', 'LKN1', 'MIP-1 delta', 'MIP-1D',\n",
       "       'MIP-5', 'MRP-2B', 'NCC-3', 'NCC3', 'SCYA15', 'SCYL3', 'SY15',\n",
       "       'CCL16', 'C-C motif chemokine ligand 16', 'CKb12', 'HCC-4',\n",
       "       'ILINCK', 'LCC-1', 'LEC', 'LMC', 'Mtn-1', 'NCC-4', 'NCC4',\n",
       "       'SCYA16', 'SCYL4', 'TAFA3', 'TAFA chemokine like family member 3',\n",
       "       'FAM19A3', 'TAFA-3', 'TAFA4',\n",
       "       'TAFA chemokine like family member 4', 'FAM19A4', 'TAFA-4',\n",
       "       'CCL18', 'C-C motif chemokine ligand 18', 'AMAC-1', 'AMAC1',\n",
       "       'CKb7', 'DC-CK1', 'DCCK1', 'MIP-4', 'PARC', 'SCYA18', 'CCL17',\n",
       "       'C-C motif chemokine ligand 17', 'A-152E5.3', 'ABCD-2', 'SCYA17',\n",
       "       'TARC', 'CCL22', 'C-C motif chemokine ligand 22', 'A-152E5.1',\n",
       "       'ABCD-1', 'DC/B-CK', 'MDC', 'SCYA22', 'STCP-1', 'TAFA5',\n",
       "       'TAFA chemokine like family member 5', 'FAM19A5', 'QLLK5208',\n",
       "       'TAFA-5', 'UNQ5208', 'CCL3', 'C-C motif chemokine ligand 3',\n",
       "       'G0S19-1', 'LD78ALPHA', 'MIP-1-alpha', 'MIP1A', 'SCYA3', 'CCL19',\n",
       "       'C-C motif chemokine ligand 19', 'CKb11', 'ELC', 'MIP-3b', 'MIP3B',\n",
       "       'SCYA19', 'CCL7', 'C-C motif chemokine ligand 7', 'FIC', 'MARC',\n",
       "       'MCP-3', 'MCP3', 'NC28', 'SCYA6', 'SCYA7', 'CCL8',\n",
       "       'C-C motif chemokine ligand 8', 'HC14', 'MCP-2', 'MCP2', 'SCYA10',\n",
       "       'SCYA8', 'CCL4', 'C-C motif chemokine ligand 4', 'ACT2', 'AT744.1',\n",
       "       'G-26', 'HC21', 'LAG-1', 'LAG1', 'MIP-1-beta', 'MIP1B', 'MIP1B1',\n",
       "       'SCYA4', 'CCL1', 'C-C motif chemokine ligand 1', 'I-309', 'P500',\n",
       "       'SCYA1', 'SISe', 'TCA3', 'CCL21', 'C-C motif chemokine ligand 21',\n",
       "       '6Ckine', 'CKb9', 'ECL', 'SCYA21', 'SLC', 'TCA4', 'CXCL13',\n",
       "       'C-X-C motif chemokine ligand 13', 'ANGIE', 'ANGIE2', 'BCA-1',\n",
       "       'BCA1', 'BLC', 'BLR1L', 'SCYB13', 'CXCL8',\n",
       "       'C-X-C motif chemokine ligand 8', 'GCP-1', 'GCP1', 'IL8', 'LECT',\n",
       "       'LUCT', 'LYNAP', 'MDNCF', 'MONAP', 'NAF', 'NAP-1', 'NAP1', 'CXCL1',\n",
       "       'C-X-C motif chemokine ligand 1', 'FSP', 'GRO1', 'GROa', 'MGSA',\n",
       "       'MGSA-a', 'NAP-3', 'SCYB1', 'CXCL9',\n",
       "       'C-X-C motif chemokine ligand 9', 'CMK', 'Humig', 'MIG', 'SCYB9',\n",
       "       'crg-10', 'CXCL10', 'C-X-C motif chemokine ligand 10', 'C7',\n",
       "       'IFI10', 'INP10', 'IP-10', 'SCYB10', 'crg-2', 'gIP-10', 'mob-1',\n",
       "       'CXCL11', 'C-X-C motif chemokine ligand 11', 'H174', 'I-TAC',\n",
       "       'IP-9', 'IP9', 'SCYB11', 'SCYB9B', 'b-R1', 'CXCL6',\n",
       "       'C-X-C motif chemokine ligand 6', 'CKA-3', 'GCP-2', 'GCP2',\n",
       "       'SCYB6', 'CXCL3', 'C-X-C motif chemokine ligand 3', 'CINC-2b',\n",
       "       'GRO3', 'GROg', 'MIP-2b', 'MIP2B', 'SCYB3', 'CXCL12',\n",
       "       'C-X-C motif chemokine ligand 12', 'IRH', 'PBSF', 'SCYB12', 'SDF1',\n",
       "       'TLSF', 'TPAR1', 'CXCL17', 'C-X-C motif chemokine ligand 17',\n",
       "       'DMC', 'Dcip1', 'UNQ473', 'VCC-1', 'VCC1', 'CXCL2',\n",
       "       'C-X-C motif chemokine ligand 2', 'CINC-2a', 'GRO2', 'GROb',\n",
       "       'MGSA-b', 'MIP-2a', 'MIP2', 'MIP2A', 'SCYB2', 'CXCL5',\n",
       "       'C-X-C motif chemokine ligand 5', 'ENA-78', 'SCYB5', 'CXCL14',\n",
       "       'C-X-C motif chemokine ligand 14', 'BMAC', 'BRAK', 'KEC', 'KS1',\n",
       "       'MIP-2g', 'MIP2G', 'NJAC', 'SCYB14', 'CXCL16',\n",
       "       'C-X-C motif chemokine ligand 16', 'CXCLG16', 'SR-PSOX', 'SRPSOX',\n",
       "       'CX3CL1', 'C-X3-C motif chemokine ligand 1', 'ABCD-3', 'C3Xkine',\n",
       "       'CXC3', 'CXC3C', 'NTN', 'NTT', 'SCYD1', 'fractalkine',\n",
       "       'neurotactin', 'CCL4L2', 'C-C motif chemokine ligand 4 like 2',\n",
       "       'AT744.2', 'CCL4L', 'SCYA4L', 'SCYQ4L2', 'CCL4L1',\n",
       "       'C-C motif chemokine ligand 4 like 1', 'SCYA4L1', 'SCYA4L2',\n",
       "       'CCL3L3', 'C-C motif chemokine ligand 3 like 3', '464.2',\n",
       "       'D17S1718', 'G0S19-2', 'LD78', 'LD78BETA', 'SCYA3L', 'SCYA3L1',\n",
       "       'CCL3L1', 'C-C motif chemokine ligand 3 like 1', 'LD78-beta(1-70)',\n",
       "       'MIP1AP', 'UMLILO',\n",
       "       'upstream master lncRNA of the inflammatory chemokine locus'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for (and remove) any excessively vague symbols\n",
    "df['sym'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>extid</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>lymphokine</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>3558</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sym   lbl extid  spid     src\n",
       "71  lymphokine  IL-2  3558     1  mygene"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['sym'].str.match('lymphokine')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(746, 747)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = df[~df['sym'].str.match('lymphokine')]\n",
    "len(dfs), len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mygene = dfs.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>extid</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>IL-2</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>3558</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>interleukin 2</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>3558</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>IL2</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>3558</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>TCGF</td>\n",
       "      <td>IL-2</td>\n",
       "      <td>3558</td>\n",
       "      <td>1</td>\n",
       "      <td>mygene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              sym   lbl extid  spid     src\n",
       "66           IL-2  IL-2  3558     1  mygene\n",
       "67  interleukin 2  IL-2  3558     1  mygene\n",
       "68            IL2  IL-2  3558     1  mygene\n",
       "70           TCGF  IL-2  3558     1  mygene"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mygene[df_mygene['lbl'] == 'IL-2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load Manual Entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "      <th>extid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL-23</td>\n",
       "      <td>IL-23</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL23</td>\n",
       "      <td>IL-23</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL-35</td>\n",
       "      <td>IL-35</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL35</td>\n",
       "      <td>IL-35</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL-32</td>\n",
       "      <td>IL-32</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sym    lbl  spid     src extid\n",
       "0  IL-23  IL-23     1  manual  None\n",
       "1   IL23  IL-23     1  manual  None\n",
       "2  IL-35  IL-35     1  manual  None\n",
       "3   IL35  IL-35     1  manual  None\n",
       "4  IL-32  IL-32     1  manual  None"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRC = 'manual'\n",
    "df = pd.read_csv(osp.join(META_DATA_DIR, 'raw', 'cytokines.manual.csv'), sep=',')\n",
    "df = df.assign(src=SRC, extid=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_manual = df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate Programmtic Transformations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>letter</th>\n",
       "      <th>name</th>\n",
       "      <th>trans</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>α</td>\n",
       "      <td>alpha</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>β</td>\n",
       "      <td>beta</td>\n",
       "      <td>b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>γ</td>\n",
       "      <td>gamma</td>\n",
       "      <td>g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>δ</td>\n",
       "      <td>delta</td>\n",
       "      <td>d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>κ</td>\n",
       "      <td>kappa</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  letter   name trans\n",
       "0      α  alpha     a\n",
       "1      β   beta     b\n",
       "2      γ  gamma     g\n",
       "3      δ  delta     d\n",
       "9      κ  kappa     k"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "greeks = pd.DataFrame(get_greek_alphabet(), columns=['letter', 'name', 'trans'])\n",
    "greeks = greeks[greeks['name'].isin(['alpha', 'beta', 'delta', 'gamma', 'kappa'])]\n",
    "greeks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to add greek symbol variations\n",
    "def expand_greeks(sym):\n",
    "    syms = [sym]\n",
    "    for r in greeks.itertuples():\n",
    "        if r.letter in sym:\n",
    "            syms.append(sym.replace(r.letter, r.name)) # IFN-γ -> IFN-gamma\n",
    "            syms.append(sym.replace(r.letter, r.trans)) # IFN-γ -> IFN-g\n",
    "            break\n",
    "        if r.name in sym:\n",
    "            syms.append(sym.replace(r.name, r.letter)) # IFN-gamma -> IFN-γ\n",
    "            syms.append(sym.replace(r.name, r.trans)) # IFN-gamma -> IFN-g\n",
    "            break\n",
    "    return syms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>src</th>\n",
       "      <th>extid</th>\n",
       "      <th>spid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL-1α</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>transform</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL-1alpha</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>transform</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL-1a</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>transform</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL1α</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>transform</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL1alpha</td>\n",
       "      <td>IL-1α</td>\n",
       "      <td>transform</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         sym    lbl        src extid  spid\n",
       "0      IL-1α  IL-1α  transform  None     1\n",
       "1  IL-1alpha  IL-1α  transform  None     1\n",
       "2      IL-1a  IL-1α  transform  None     1\n",
       "3       IL1α  IL-1α  transform  None     1\n",
       "4   IL1alpha  IL-1α  transform  None     1"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRC = 'transform'\n",
    "\n",
    "# Add symbols to resolve references like \"interleukin (IL)-4\" or \"interferon (IFN)-γ\"\n",
    "# as well as ensure that every IL-\\d or IL\\d+ is present\n",
    "\n",
    "df = pd.concat([df_cameron, df_mygene, df_manual], sort=True).drop_duplicates()\n",
    "dfr = []\n",
    "for prefix, name in [\n",
    "    ('IL', 'interleukin'), \n",
    "    ('IFN', 'interferon'),\n",
    "    ('TNF', 'tumor necrosis factor'),\n",
    "    ('TGF', 'transforming growth factor'),\n",
    "    ('LT', 'lymphotoxin')\n",
    "]:\n",
    "    dfp = df[df['sym'].str.startswith(prefix)]\n",
    "    # Get unique IL numbers, IFN types (alpha, beta, gamma), etc.\n",
    "    suffixes = dfp['lbl'].str.replace(prefix + '[-]?', '').unique()\n",
    "    for suffix in suffixes:\n",
    "        lbl = prefix + '-' + suffix\n",
    "        syms = [\n",
    "            prefix + '-' + suffix, # IL-2\n",
    "            prefix + suffix, # IL2\n",
    "            '('+ prefix + ')' + suffix, # (IL)2\n",
    "            '('+ prefix + ') ' + suffix, # (IL) 2\n",
    "            '('+ prefix + ')-' + suffix, # (IL)-2\n",
    "            name + ' ' + suffix, # interleukin 2\n",
    "            name + '-' + suffix, # interleukin-2\n",
    "            name + ' (' + prefix + ')-' + suffix, # interleukin (IL)-2\n",
    "        ]\n",
    "        # Add variations for greek letters\n",
    "        syms = [v for sym in syms for v in expand_greeks(sym)]\n",
    "        for sym in syms:\n",
    "            dfr.append((sym, lbl))\n",
    "df = pd.DataFrame(dfr, columns=['sym', 'lbl'])\n",
    "df = df.assign(src=SRC, extid=None, spid=SPECIES_HUMAN_ID)\n",
    "df_transform = df.copy()\n",
    "df_transform.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_transform[df_transform['lbl'].str.contains('TNF')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 1953 entries, 31 to 791\n",
      "Data columns (total 7 columns):\n",
      "id        1953 non-null object\n",
      "src       1953 non-null object\n",
      "sym       1953 non-null object\n",
      "lbl       1953 non-null object\n",
      "spid      1953 non-null int64\n",
      "extid     746 non-null object\n",
      "prefid    1953 non-null object\n",
      "dtypes: int64(1), object(6)\n",
      "memory usage: 122.1+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>src</th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>extid</th>\n",
       "      <th>prefid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CK66480A09B13AD79D</td>\n",
       "      <td>cameron</td>\n",
       "      <td>4-1BBL</td>\n",
       "      <td>4-1BBL</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK66480A09B13AD79D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CK3C8B23133684A426</td>\n",
       "      <td>cameron</td>\n",
       "      <td>APRIL</td>\n",
       "      <td>APRIL</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK3C8B23133684A426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CKA1A67F50E3346A8F</td>\n",
       "      <td>cameron</td>\n",
       "      <td>TALL-2</td>\n",
       "      <td>APRIL</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK3C8B23133684A426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CKB4A58D7EF45C45C0</td>\n",
       "      <td>cameron</td>\n",
       "      <td>CD153</td>\n",
       "      <td>CD153</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CKB4A58D7EF45C45C0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CKDB38E36EFC121A94</td>\n",
       "      <td>cameron</td>\n",
       "      <td>CD30L</td>\n",
       "      <td>CD153</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CKB4A58D7EF45C45C0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    id      src     sym     lbl  spid extid  \\\n",
       "31  CK66480A09B13AD79D  cameron  4-1BBL  4-1BBL     1  None   \n",
       "32  CK3C8B23133684A426  cameron   APRIL   APRIL     1  None   \n",
       "35  CKA1A67F50E3346A8F  cameron  TALL-2   APRIL     1  None   \n",
       "34  CKB4A58D7EF45C45C0  cameron   CD153   CD153     1  None   \n",
       "37  CKDB38E36EFC121A94  cameron   CD30L   CD153     1  None   \n",
       "\n",
       "                prefid  \n",
       "31  CK66480A09B13AD79D  \n",
       "32  CK3C8B23133684A426  \n",
       "35  CK3C8B23133684A426  \n",
       "34  CKB4A58D7EF45C45C0  \n",
       "37  CKB4A58D7EF45C45C0  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = merge([df_cameron, df_mygene, df_manual, df_transform], ID_TYP_CK).drop_duplicates()\n",
    "df = add_preferred_ids(df)\n",
    "print(df.info())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 1 records with excessively short symbols: ['C7']\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 1952 entries, 31 to 791\n",
      "Data columns (total 7 columns):\n",
      "id        1952 non-null object\n",
      "src       1952 non-null object\n",
      "sym       1952 non-null object\n",
      "lbl       1952 non-null object\n",
      "spid      1952 non-null int64\n",
      "extid     745 non-null object\n",
      "prefid    1952 non-null object\n",
      "dtypes: int64(1), object(6)\n",
      "memory usage: 122.0+ KB\n"
     ]
    }
   ],
   "source": [
    "rm_sym = df[df['sym'].str.len() < MIN_SYM_LEN]['sym'].unique()\n",
    "print('Removing {} records with excessively short symbols: {}'.format(len(rm_sym), rm_sym))\n",
    "df = df[~df['sym'].isin(rm_sym)]\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['4-1BB', '4-1BBL', 'APRIL', 'CCL1', 'CCL11', 'CCL13', 'CCL14',\n",
       "       'CCL15', 'CCL16', 'CCL17', 'CCL18', 'CCL19', 'CCL2', 'CCL20',\n",
       "       'CCL21', 'CCL22', 'CCL23', 'CCL24', 'CCL25', 'CCL26', 'CCL27',\n",
       "       'CCL28', 'CCL3', 'CCL3L1', 'CCL3L3', 'CCL4', 'CCL4L1', 'CCL4L2',\n",
       "       'CCL5', 'CCL6', 'CCL7', 'CCL8', 'CCL9', 'CD153', 'CD154', 'CD178',\n",
       "       'CD258', 'CD40LG', 'CD70', 'CKLF', 'CSF1', 'CSF2', 'CSF3',\n",
       "       'CX3CL1', 'CXCL1', 'CXCL10', 'CXCL11', 'CXCL12', 'CXCL13',\n",
       "       'CXCL14', 'CXCL16', 'CXCL17', 'CXCL2', 'CXCL3', 'CXCL4', 'CXCL5',\n",
       "       'CXCL6', 'CXCL7', 'CXCL8', 'CXCL9', 'EBI3', 'EGF', 'EPO', 'FGF1',\n",
       "       'FGF2', 'FP248', 'G-CSF', 'GDF15', 'GITRL', 'GM-CSF', 'IFN-1',\n",
       "       'IFN-IL-28A', 'IFN-IL-28B', 'IFN-IL-6', 'IFN-α', 'IFN-α1',\n",
       "       'IFN-α2', 'IFN-α6', 'IFN-β', 'IFN-β1', 'IFN-γ', 'IL-1', 'IL-10',\n",
       "       'IL-11', 'IL-12', 'IL-12A', 'IL-12B', 'IL-12α', 'IL-13', 'IL-14',\n",
       "       'IL-15', 'IL-16', 'IL-17', 'IL-17A', 'IL-17B', 'IL-17C', 'IL-17D',\n",
       "       'IL-17F', 'IL-18', 'IL-18BP', 'IL-19', 'IL-1A', 'IL-1B', 'IL-1F10',\n",
       "       'IL-1RA', 'IL-1α', 'IL-1β', 'IL-2', 'IL-20', 'IL-21', 'IL-22',\n",
       "       'IL-23', 'IL-23A', 'IL-24', 'IL-25', 'IL-26', 'IL-27', 'IL-28',\n",
       "       'IL-28A', 'IL-28B', 'IL-3', 'IL-31', 'IL-32', 'IL-33', 'IL-34',\n",
       "       'IL-35', 'IL-36A', 'IL-36B', 'IL-36G', 'IL-37', 'IL-4', 'IL-4I1',\n",
       "       'IL-5', 'IL-6', 'IL-6ST', 'IL-7', 'IL-9', 'IL-CCL16', 'IL-CCL27',\n",
       "       'IL-CXCL8', 'IL-F2', 'IL-F3', 'IL-NF3', 'LIF', 'LIGHT',\n",
       "       'LOC107983991', 'LT-TNF-β', 'LT-XCL1', 'LT-β', 'MIF', 'NFIL3',\n",
       "       'OSM', 'OX40L', 'SPP1', 'TAFA1', 'TAFA2', 'TAFA3', 'TAFA4',\n",
       "       'TAFA5', 'TALL-1', 'TGF-β', 'TGF-β2', 'TGF-β3', 'TNF-4-1BB',\n",
       "       'TNF-APRIL', 'TNF-CD153', 'TNF-CD258', 'TNF-OX40L', 'TNF-SF13B',\n",
       "       'TNF-VEGI', 'TNF-α', 'TNF-β', 'TNFSF13B', 'TRAIL', 'TRANCE',\n",
       "       'TWEAK', 'UMLILO', 'VEGI', 'XCL1', 'XCL2'], dtype=object)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sort(df['lbl'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>src</th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>extid</th>\n",
       "      <th>prefid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CK128EC732B281BF60</td>\n",
       "      <td>cameron</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK128EC732B281BF60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CK8074B95EB01901B1</td>\n",
       "      <td>cameron</td>\n",
       "      <td>IFNγ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK128EC732B281BF60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "      <td>transform</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>CKB14BD5778A003DBA</td>\n",
       "      <td>transform</td>\n",
       "      <td>IFN-gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>CKAB0A4DC198C5F1ED</td>\n",
       "      <td>transform</td>\n",
       "      <td>IFN-g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>CKF88B95F709297E33</td>\n",
       "      <td>transform</td>\n",
       "      <td>IFNγ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>CK28F65925FFAAE009</td>\n",
       "      <td>transform</td>\n",
       "      <td>IFNgamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>CK0D10BE8E50A659D4</td>\n",
       "      <td>transform</td>\n",
       "      <td>IFNg</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>CKAC6AB0DEF89E5113</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN)γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>CKF5A331D97EF0549F</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN)gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>CK9C1DCC534F34890F</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN)g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>CK8FAB99D623C11005</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN) γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>CKE0B9163A8DAECB0E</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN) gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>CK27080E286065B3C6</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN) g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>CKDC084D8D066CBDAD</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN)-γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>CK64A840435F9DBA7B</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN)-gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>CKCFA041B4E878C9C0</td>\n",
       "      <td>transform</td>\n",
       "      <td>(IFN)-g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>CK30FCB87957F6228C</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>CK47D24BFC95CEBC4D</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>CKC35DFD6C81BFC5CA</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610</th>\n",
       "      <td>CKB1327E1355BFD9DD</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon-γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>CK8E78E2514321441A</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon-gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>CK320F7A36DB4839C9</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon-g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>CKD58AE8DCDCFEC12B</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon (IFN)-γ</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>CK39DF3C229ADCB13F</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon (IFN)-gamma</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>CK8EB605E5401B4F2F</td>\n",
       "      <td>transform</td>\n",
       "      <td>interferon (IFN)-g</td>\n",
       "      <td>IFN-γ</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>CK1B71668FDDECE3CF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id        src                     sym    lbl  spid extid  \\\n",
       "26   CK128EC732B281BF60    cameron                   IFN-γ  IFN-γ     1  None   \n",
       "26   CK8074B95EB01901B1    cameron                    IFNγ  IFN-γ     1  None   \n",
       "592  CK1B71668FDDECE3CF  transform                   IFN-γ  IFN-γ     1  None   \n",
       "593  CKB14BD5778A003DBA  transform               IFN-gamma  IFN-γ     1  None   \n",
       "594  CKAB0A4DC198C5F1ED  transform                   IFN-g  IFN-γ     1  None   \n",
       "595  CKF88B95F709297E33  transform                    IFNγ  IFN-γ     1  None   \n",
       "596  CK28F65925FFAAE009  transform                IFNgamma  IFN-γ     1  None   \n",
       "597  CK0D10BE8E50A659D4  transform                    IFNg  IFN-γ     1  None   \n",
       "598  CKAC6AB0DEF89E5113  transform                  (IFN)γ  IFN-γ     1  None   \n",
       "599  CKF5A331D97EF0549F  transform              (IFN)gamma  IFN-γ     1  None   \n",
       "600  CK9C1DCC534F34890F  transform                  (IFN)g  IFN-γ     1  None   \n",
       "601  CK8FAB99D623C11005  transform                 (IFN) γ  IFN-γ     1  None   \n",
       "602  CKE0B9163A8DAECB0E  transform             (IFN) gamma  IFN-γ     1  None   \n",
       "603  CK27080E286065B3C6  transform                 (IFN) g  IFN-γ     1  None   \n",
       "604  CKDC084D8D066CBDAD  transform                 (IFN)-γ  IFN-γ     1  None   \n",
       "605  CK64A840435F9DBA7B  transform             (IFN)-gamma  IFN-γ     1  None   \n",
       "606  CKCFA041B4E878C9C0  transform                 (IFN)-g  IFN-γ     1  None   \n",
       "607  CK30FCB87957F6228C  transform            interferon γ  IFN-γ     1  None   \n",
       "608  CK47D24BFC95CEBC4D  transform        interferon gamma  IFN-γ     1  None   \n",
       "609  CKC35DFD6C81BFC5CA  transform            interferon g  IFN-γ     1  None   \n",
       "610  CKB1327E1355BFD9DD  transform            interferon-γ  IFN-γ     1  None   \n",
       "611  CK8E78E2514321441A  transform        interferon-gamma  IFN-γ     1  None   \n",
       "612  CK320F7A36DB4839C9  transform            interferon-g  IFN-γ     1  None   \n",
       "613  CKD58AE8DCDCFEC12B  transform      interferon (IFN)-γ  IFN-γ     1  None   \n",
       "614  CK39DF3C229ADCB13F  transform  interferon (IFN)-gamma  IFN-γ     1  None   \n",
       "615  CK8EB605E5401B4F2F  transform      interferon (IFN)-g  IFN-γ     1  None   \n",
       "\n",
       "                 prefid  \n",
       "26   CK128EC732B281BF60  \n",
       "26   CK128EC732B281BF60  \n",
       "592  CK1B71668FDDECE3CF  \n",
       "593  CK1B71668FDDECE3CF  \n",
       "594  CK1B71668FDDECE3CF  \n",
       "595  CK1B71668FDDECE3CF  \n",
       "596  CK1B71668FDDECE3CF  \n",
       "597  CK1B71668FDDECE3CF  \n",
       "598  CK1B71668FDDECE3CF  \n",
       "599  CK1B71668FDDECE3CF  \n",
       "600  CK1B71668FDDECE3CF  \n",
       "601  CK1B71668FDDECE3CF  \n",
       "602  CK1B71668FDDECE3CF  \n",
       "603  CK1B71668FDDECE3CF  \n",
       "604  CK1B71668FDDECE3CF  \n",
       "605  CK1B71668FDDECE3CF  \n",
       "606  CK1B71668FDDECE3CF  \n",
       "607  CK1B71668FDDECE3CF  \n",
       "608  CK1B71668FDDECE3CF  \n",
       "609  CK1B71668FDDECE3CF  \n",
       "610  CK1B71668FDDECE3CF  \n",
       "611  CK1B71668FDDECE3CF  \n",
       "612  CK1B71668FDDECE3CF  \n",
       "613  CK1B71668FDDECE3CF  \n",
       "614  CK1B71668FDDECE3CF  \n",
       "615  CK1B71668FDDECE3CF  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['lbl'] == 'IFN-γ']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/eczech/repos/hammer/t-cell-relation-extraction/pm_subtype_protein_relations/data/meta/cytokines.csv'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = osp.join(META_DATA_DIR, CYTOKINES + '.csv')\n",
    "df.to_csv(path, index=False)\n",
    "path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcription Factors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load from Lambert, et al. 2016"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 1639 entries, 0 to 2763\n",
      "Data columns (total 5 columns):\n",
      "extid    1639 non-null object\n",
      "lbl      1639 non-null object\n",
      "sym      1639 non-null object\n",
      "spid     1639 non-null int64\n",
      "src      1639 non-null object\n",
      "dtypes: int64(1), object(4)\n",
      "memory usage: 76.8+ KB\n"
     ]
    }
   ],
   "source": [
    "SRC = 'lambert'\n",
    "# Load export of tab \"Table S1. Related to Figure 1B-Table 1.csv\" from Document S1 in\n",
    "# https://doi.org/10.1016%2Fj.cell.2018.01.029 (containing *human* transcription factors)\n",
    "df = pd.read_csv(osp.join(META_DATA_DIR, 'raw', 'transcription_factors.lambert.csv'), header=[0,1])\n",
    "df = df[df['Is TF?'].iloc[:,0] == 'Yes']\n",
    "df = df[[('Gene Information', 'ID'), ('Unnamed: 1_level_0', 'Name')]]\n",
    "df.columns = ['extid', 'lbl']\n",
    "df = df.assign(sym=df['lbl'].values, spid=SPECIES_HUMAN_ID, src=SRC)\n",
    "df['spid'] = SPECIES_HUMAN_ID\n",
    "# Ensure number of records from spreadsheet export matches https://en.wikipedia.org/wiki/List_of_human_transcription_factors\n",
    "assert len(df) == 1639, \\\n",
    "    'Data frame does not have expected row count 1639 '\\\n",
    "    '(count should match https://en.wikipedia.org/wiki/List_of_human_transcription_factors)'\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>extid</th>\n",
       "      <th>lbl</th>\n",
       "      <th>sym</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>1</td>\n",
       "      <td>lambert</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000008196</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>1</td>\n",
       "      <td>lambert</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000087510</td>\n",
       "      <td>TFAP2C</td>\n",
       "      <td>TFAP2C</td>\n",
       "      <td>1</td>\n",
       "      <td>lambert</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000008197</td>\n",
       "      <td>TFAP2D</td>\n",
       "      <td>TFAP2D</td>\n",
       "      <td>1</td>\n",
       "      <td>lambert</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000116819</td>\n",
       "      <td>TFAP2E</td>\n",
       "      <td>TFAP2E</td>\n",
       "      <td>1</td>\n",
       "      <td>lambert</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             extid     lbl     sym  spid      src\n",
       "0  ENSG00000137203  TFAP2A  TFAP2A     1  lambert\n",
       "1  ENSG00000008196  TFAP2B  TFAP2B     1  lambert\n",
       "2  ENSG00000087510  TFAP2C  TFAP2C     1  lambert\n",
       "3  ENSG00000008197  TFAP2D  TFAP2D     1  lambert\n",
       "4  ENSG00000116819  TFAP2E  TFAP2E     1  lambert"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-1639...done.\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 1639 entries, ENSG00000137203 to ENSG00000168152\n",
      "Data columns (total 10 columns):\n",
      "_id             1636 non-null object\n",
      "_score          1636 non-null float64\n",
      "alias           1309 non-null object\n",
      "ensembl         77 non-null object\n",
      "ensembl.gene    1559 non-null object\n",
      "name            1634 non-null object\n",
      "notfound        3 non-null object\n",
      "symbol          1636 non-null object\n",
      "taxid           1636 non-null float64\n",
      "extid           1639 non-null object\n",
      "dtypes: float64(2), object(8)\n",
      "memory usage: 140.9+ KB\n"
     ]
    }
   ],
   "source": [
    "# Use MyGene to get aliases for genes above (by ensembl id, not symbol/name)\n",
    "dfa = mg.getgenes(\n",
    "    ids=df['extid'].unique(),\n",
    "    scopes=[\"symbol\", \"retired\", \"name\", \"alias\"],\n",
    "    fields='symbol,name,taxid,ensembl.gene,alias', \n",
    "    as_dataframe=True\n",
    ")\n",
    "dfa['extid'] = dfa.index\n",
    "dfa.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1001-1639...done.\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 7365 entries, 0 to 7375\n",
      "Data columns (total 5 columns):\n",
      "sym      7365 non-null object\n",
      "lbl      7365 non-null object\n",
      "extid    7365 non-null object\n",
      "spid     7365 non-null int64\n",
      "src      7365 non-null object\n",
      "dtypes: int64(1), object(4)\n",
      "memory usage: 345.2+ KB\n"
     ]
    }
   ],
   "source": [
    "# Use MyGene to get aliases for genes above (by ensembl id, not symbol/name)\n",
    "dfa = mg.getgenes(\n",
    "    ids=df['extid'].unique(),\n",
    "    scopes=[\"symbol\", \"retired\", \"name\", \"alias\"],\n",
    "    fields='symbol,name,taxid,ensembl.gene,alias', \n",
    "    as_dataframe=True\n",
    ")\n",
    "# The query term is set as the index in results so use that to get the ensembl id\n",
    "# rather than the `ensenbl.gene` field since this is inexplicably empty sometimes\n",
    "# even when the query term as an ensembl id and all other fields are valid (e.g.\n",
    "# ENSG00000232040 --> gives symbol ZBED9 and valid aliases but empty gene id field)\n",
    "dfa['extid'] = dfa.index\n",
    "dfa = mygene_prep(dfa)\n",
    "dfa = dfa.assign(spid=SPECIES_HUMAN_ID, src=SRC)\n",
    "dfa = dfa[dfa['lbl'].notnull() & dfa['sym'].notnull()]\n",
    "assert dfa['extid'].notnull().all()\n",
    "dfa.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 7373 entries, 0 to 7375\n",
      "Data columns (total 5 columns):\n",
      "extid    7373 non-null object\n",
      "lbl      7373 non-null object\n",
      "sym      7373 non-null object\n",
      "spid     7373 non-null int64\n",
      "src      7373 non-null object\n",
      "dtypes: int64(1), object(4)\n",
      "memory usage: 345.6+ KB\n"
     ]
    }
   ],
   "source": [
    "dfm = pd.concat([df, dfa[df.columns]])\n",
    "# As the mygene results may contain aliases matching the names\n",
    "# in the lambert spreadsheet and the `src` is the same, duplicates\n",
    "# should be removed \n",
    "dfm = dfm.drop_duplicates(subset=['sym', 'lbl', 'spid'])\n",
    "dfm.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lambert = dfm.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_lambert[df_lambert['sym'].str.lower().str.contains('t-bet')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load Manual Entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "      <th>extid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sym, lbl, spid, src, extid]\n",
       "Index: []"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRC = 'manual'\n",
    "df = pd.read_csv(osp.join(META_DATA_DIR, 'raw', 'transcription_factors.manual.csv'), sep=',')\n",
    "df = df.assign(src=SRC, extid=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_manual = df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge and Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 7373 entries, 1070 to 6151\n",
      "Data columns (total 7 columns):\n",
      "id        7373 non-null object\n",
      "src       7373 non-null object\n",
      "sym       7373 non-null object\n",
      "lbl       7373 non-null object\n",
      "spid      7373 non-null object\n",
      "extid     7373 non-null object\n",
      "prefid    7373 non-null object\n",
      "dtypes: object(7)\n",
      "memory usage: 460.8+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>src</th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>extid</th>\n",
       "      <th>prefid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1070</th>\n",
       "      <td>TFC491EFF7A86A1701</td>\n",
       "      <td>lambert</td>\n",
       "      <td>AC008770.2</td>\n",
       "      <td>AC008770.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>TFC491EFF7A86A1701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>TF366ACF42CE4CCEBB</td>\n",
       "      <td>lambert</td>\n",
       "      <td>AC008770.3</td>\n",
       "      <td>AC008770.3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000267179</td>\n",
       "      <td>TF366ACF42CE4CCEBB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>TF9837F24D5ADF39CB</td>\n",
       "      <td>lambert</td>\n",
       "      <td>AC023509.3</td>\n",
       "      <td>AC023509.3</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000267281</td>\n",
       "      <td>TF9837F24D5ADF39CB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>TFA91BACA6A8187757</td>\n",
       "      <td>lambert</td>\n",
       "      <td>AC092835.1</td>\n",
       "      <td>AC092835.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000233757</td>\n",
       "      <td>TFA91BACA6A8187757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>TF25E2A8B125012DB6</td>\n",
       "      <td>lambert</td>\n",
       "      <td>AC138696.1</td>\n",
       "      <td>AC138696.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000264668</td>\n",
       "      <td>TF25E2A8B125012DB6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id      src         sym         lbl spid  \\\n",
       "1070  TFC491EFF7A86A1701  lambert  AC008770.2  AC008770.2    1   \n",
       "232   TF366ACF42CE4CCEBB  lambert  AC008770.3  AC008770.3    1   \n",
       "160   TF9837F24D5ADF39CB  lambert  AC023509.3  AC023509.3    1   \n",
       "233   TFA91BACA6A8187757  lambert  AC092835.1  AC092835.1    1   \n",
       "234   TF25E2A8B125012DB6  lambert  AC138696.1  AC138696.1    1   \n",
       "\n",
       "                extid              prefid  \n",
       "1070  ENSG00000267179  TFC491EFF7A86A1701  \n",
       "232   ENSG00000267179  TF366ACF42CE4CCEBB  \n",
       "160   ENSG00000267281  TF9837F24D5ADF39CB  \n",
       "233   ENSG00000233757  TFA91BACA6A8187757  \n",
       "234   ENSG00000264668  TF25E2A8B125012DB6  "
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = merge([df_lambert, df_manual], ID_TYP_TF).drop_duplicates()\n",
    "df = add_preferred_ids(df)\n",
    "print(df.info())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 25 records with excessively short symbols: ['AR' 'KD' 'D9' 'ZF' 'ER' 'LD' 'H6' 'PL' 'K8' 'LF' 'RS' 'MI' 'L8' 'GR'\n",
      " 'MR' 'AN' 'PR' 'RX' 'SS' 'OF' 'T' 'AA' 'Y8' 'RN' 'NP']\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 7346 entries, 1070 to 6151\n",
      "Data columns (total 7 columns):\n",
      "id        7346 non-null object\n",
      "src       7346 non-null object\n",
      "sym       7346 non-null object\n",
      "lbl       7346 non-null object\n",
      "spid      7346 non-null object\n",
      "extid     7346 non-null object\n",
      "prefid    7346 non-null object\n",
      "dtypes: object(7)\n",
      "memory usage: 459.1+ KB\n"
     ]
    }
   ],
   "source": [
    "rm_sym = df[df['sym'].str.len() < MIN_SYM_LEN]['sym'].unique()\n",
    "print('Removing {} records with excessively short symbols: {}'.format(len(rm_sym), rm_sym))\n",
    "df = df[~df['sym'].isin(rm_sym)]\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>src</th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>extid</th>\n",
       "      <th>prefid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>TFD6425765B0AF5303</td>\n",
       "      <td>lambert</td>\n",
       "      <td>NGFI-A</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000120738</td>\n",
       "      <td>TF9C933C7F3601C428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1216</th>\n",
       "      <td>TF6AD173F0E42A506C</td>\n",
       "      <td>lambert</td>\n",
       "      <td>NGFI-C</td>\n",
       "      <td>EGR4</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000135625</td>\n",
       "      <td>TF193163E659BF0520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1217</th>\n",
       "      <td>TFA77682C38E2A0D4F</td>\n",
       "      <td>lambert</td>\n",
       "      <td>NGFIC</td>\n",
       "      <td>EGR4</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000135625</td>\n",
       "      <td>TF193163E659BF0520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>TF1FEE043B45D9654E</td>\n",
       "      <td>lambert</td>\n",
       "      <td>GFI1</td>\n",
       "      <td>GFI1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000162676</td>\n",
       "      <td>TF1FEE043B45D9654E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1240</th>\n",
       "      <td>TFE156A51136C217BC</td>\n",
       "      <td>lambert</td>\n",
       "      <td>GFI-1</td>\n",
       "      <td>GFI1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000162676</td>\n",
       "      <td>TF1FEE043B45D9654E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1241</th>\n",
       "      <td>TFD513B64FD7B6A5DF</td>\n",
       "      <td>lambert</td>\n",
       "      <td>GFI1A</td>\n",
       "      <td>GFI1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000162676</td>\n",
       "      <td>TF1FEE043B45D9654E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>TF06386BF18D80F1C5</td>\n",
       "      <td>lambert</td>\n",
       "      <td>GFI1B</td>\n",
       "      <td>GFI1B</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000165702</td>\n",
       "      <td>TF06386BF18D80F1C5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6377</th>\n",
       "      <td>TF03C9CAC32D70E497</td>\n",
       "      <td>lambert</td>\n",
       "      <td>NGFIB</td>\n",
       "      <td>NR4A1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000123358</td>\n",
       "      <td>TF139F6699C1A7EEBB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id      src     sym    lbl spid            extid  \\\n",
       "1198  TFD6425765B0AF5303  lambert  NGFI-A   EGR1    1  ENSG00000120738   \n",
       "1216  TF6AD173F0E42A506C  lambert  NGFI-C   EGR4    1  ENSG00000135625   \n",
       "1217  TFA77682C38E2A0D4F  lambert   NGFIC   EGR4    1  ENSG00000135625   \n",
       "265   TF1FEE043B45D9654E  lambert    GFI1   GFI1    1  ENSG00000162676   \n",
       "1240  TFE156A51136C217BC  lambert   GFI-1   GFI1    1  ENSG00000162676   \n",
       "1241  TFD513B64FD7B6A5DF  lambert   GFI1A   GFI1    1  ENSG00000162676   \n",
       "266   TF06386BF18D80F1C5  lambert   GFI1B  GFI1B    1  ENSG00000165702   \n",
       "6377  TF03C9CAC32D70E497  lambert   NGFIB  NR4A1    1  ENSG00000123358   \n",
       "\n",
       "                  prefid  \n",
       "1198  TF9C933C7F3601C428  \n",
       "1216  TF193163E659BF0520  \n",
       "1217  TF193163E659BF0520  \n",
       "265   TF1FEE043B45D9654E  \n",
       "1240  TF1FEE043B45D9654E  \n",
       "1241  TF1FEE043B45D9654E  \n",
       "266   TF06386BF18D80F1C5  \n",
       "6377  TF139F6699C1A7EEBB  "
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['sym'].str.lower().str.contains('gfi')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/eczech/repos/hammer/t-cell-relation-extraction/pm_subtype_protein_relations/data/meta/transcription_factors.csv'"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = osp.join(META_DATA_DIR, TRANSCRIPTION_FACTORS + '.csv')\n",
    "df.to_csv(path, index=False)\n",
    "path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRC = 'manual'\n",
    "cell_types = [\n",
    "    ('TMEM', ['T-mem', 'memory-T', 'mem-T']),\n",
    "    ('TN', ['T-naïve', 'naïve-T']),\n",
    "    ('TEMRA', ['T-emra', 'emra-T']),\n",
    "    ('TCM', ['T-cm', 'cm-T', 'central-memory-T']),\n",
    "    ('TEFF', ['T-eff', 'T-effector', 'eff-T', 'effector-T']),\n",
    "    ('TEM', ['T-em', 'em-T', 'effector-memory-T', 'effector-memory-T']),\n",
    "    ('IEL', ['IELs', 'Intraepithelial-lymphocyte', 'Intraepithelial-lymphocytes']),\n",
    "    ('DETC', ['DETCs', 'Dendritic-epidermal-T']),\n",
    "    ('Thymocyte', ['thymocyte', 'thymocytes']),\n",
    "    ('Th', ['T-Helper', 'Helper-T']),\n",
    "    ('Tc', ['T-Cytotoxic', 'Cytotoxic-T', 'Cytolitic-T', 'T-Cytolitic']),\n",
    "    ('ThP', ['Th-Precursor', 'Precursor-Th']),\n",
    "    ('Trm', ['T-rm', 'T-resident-memory', 'resident-memory-T']),\n",
    "    ('Tscm', ['T-SCM', 'T-SC', 'stem-memory-T', 'T-stem-memory', 'memory-stem-T', 'T-memory-stem']),\n",
    "    ('NKT', ['natural-killer-T', 'T-natural-killer', 'INKT', 'TRNKT', 'CNKT', 'NK/T']),\n",
    "    ('MAIT', ['Mucosal-associated invariant T']),\n",
    "    ('Treg', ['T-reg', 'T-regs', 'T-regulatory', 'regulatory-T']),\n",
    "    ('Tsupp', ['Tsuppressor', 'Ts/c', 'Ts/Tc','T-suppressor/cytotoxic'], 'Treg'),\n",
    "    ('Treg1', ['T-reg1', 'T-reg-1', 'Treg/Th1', 'Treg-Th1', 'Tr-1'], 'Treg'),\n",
    "    ('Treg17', ['T-reg17', 'T-reg-17', 'Treg/Th17', 'Treg-Th17', 'Tr-17'], 'Treg'),\n",
    "    ('Tfh', ['T-FH', 'T-follicular-helper', 'follicular-helper-T', 'fhT'], 'Th'),\n",
    "    ('Tfreg', ['Tf-reg', 'T-follicular-regulatory', 'follicular-regulatory-T', 'follicular-Treg', 'fTreg'], 'Treg'),\n",
    "    ('iTreg', [\n",
    "        'i-Treg', 'induced-Treg', 'inducible-Treg', 'induced-T-reg', 'inducible-T-reg',\n",
    "        'induced T-regulatory', 'induced-T-regulatory', 'induced regulatory-T', 'induced-regulatory-T',\n",
    "        'inducible T-regulatory', 'inducible-T-regulatory', 'inducible regulatory-T', 'inducible-regulatory-T',\n",
    "        ], 'Treg'),\n",
    "    ('nTreg', [\n",
    "        'n-Treg', 'natural-Treg', 'natural-T-reg', \n",
    "        'natural T-regulatory', 'natural-T-regulatory', 'natural regulatory-T', 'natural-regulatory-T'\n",
    "        ], 'Treg'),\n",
    "    ('pTreg', [\n",
    "        'p-Treg', 'peripheral-Treg', 'peripheral-T-reg', \n",
    "        'peripheral T-regulatory', 'peripheral-T-regulatory', 'peripheral regulatory-T', 'peripheral-regulatory-T',\n",
    "        'peripherally-induced-Treg'\n",
    "        ], 'Treg'),\n",
    "    ('γδT', ['Gamma-Delta-T', 'T-Gamma-Delta', 'Gamma/Delta-T', 'γ/δ-T', 'γδ-T', 'gd-T', 'Tgd', 'g/dT', 'Tg/d']),\n",
    "    ('γδT-Vδ1', ['VDelta1-T', 'Vδ1-T'], 'γδT'),\n",
    "    ('γδT-Vδ2', ['VDelta2-T', 'Vδ2-T'], 'γδT'),\n",
    "    ('γδT-Vγ1', ['VGamma1-T', 'Vγ1-T'], 'γδT'),\n",
    "    ('γδT-Vγ4', ['VGamma4-T', 'Vγ4-T'], 'γδT'),\n",
    "    ('γδT-Vγ9', ['VGamma9-T', 'Vγ9-T'], 'γδT'),\n",
    "    ('γδT-Vγ9Vδ2', ['T-VGamma9-Delta2', 'VGamma9-Delta2-T', 'Vγ9Vδ2T'], 'γδT'),\n",
    "    ('γδT-17', ['γδ17T', 'T-Gamma-Delta17', 'Gamma-Delta17-T'], 'γδT'),\n",
    "    ('γδT-TCS1', ['TCS1'], 'γδT')\n",
    "] + [\n",
    "    ('Tc' + str(i), ['Tc-' + str(i), 'T-cytolitic-' + str(i), 'T cytolitic-' + str(i), 'T-cytotoxic-' + str(i), 'T cytotoxic-' + str(i)], 'Tc') \n",
    "    for i in [0, 1, 2, 3, 9, 17, 22]\n",
    "] + [\n",
    "    ('Th' + str(i), ['Th-' + str(i), 'T-helper-' + str(i), 'T helper-' + str(i)], 'Th') \n",
    "    for i in [0, 1, 2, 3, 9, 17, 22]\n",
    "] + [\n",
    "    ('Tfh{}like'.format(i), ['Tfh-{}-like'.format(i), 'Tfh-{}like'.format(i), 'Tfh{}-like'.format(i)], 'Tfh') \n",
    "    for i in [0, 1, 2, 3, 9, 17, 22]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "      <th>extid</th>\n",
       "      <th>id</th>\n",
       "      <th>prefid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>DETC</td>\n",
       "      <td>DETC</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT231343DE336492F1</td>\n",
       "      <td>CT231343DE336492F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>DETCs</td>\n",
       "      <td>DETC</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT4CF1E0FFE26A395F</td>\n",
       "      <td>CT231343DE336492F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Dendritic-epidermal-T</td>\n",
       "      <td>DETC</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CTAD8B4B3EE6ECB6BB</td>\n",
       "      <td>CT231343DE336492F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Dendritic epidermal T</td>\n",
       "      <td>DETC</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT728A6AA916C71387</td>\n",
       "      <td>CT231343DE336492F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>DendriticepidermalT</td>\n",
       "      <td>DETC</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CTE89D7A3B2E8FAACE</td>\n",
       "      <td>CT231343DE336492F1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       sym   lbl  spid     src extid                  id  \\\n",
       "133                   DETC  DETC     1  manual  None  CT231343DE336492F1   \n",
       "134                  DETCs  DETC     1  manual  None  CT4CF1E0FFE26A395F   \n",
       "140  Dendritic-epidermal-T  DETC     1  manual  None  CTAD8B4B3EE6ECB6BB   \n",
       "141  Dendritic epidermal T  DETC     1  manual  None  CT728A6AA916C71387   \n",
       "142    DendriticepidermalT  DETC     1  manual  None  CTE89D7A3B2E8FAACE   \n",
       "\n",
       "                 prefid  \n",
       "133  CT231343DE336492F1  \n",
       "134  CT231343DE336492F1  \n",
       "140  CT231343DE336492F1  \n",
       "141  CT231343DE336492F1  \n",
       "142  CT231343DE336492F1  "
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unidecode\n",
    "df = []\n",
    "for r in cell_types:\n",
    "    assert len(r) in [2, 3]\n",
    "    lbl, alias = r[:2]\n",
    "    parent = None if len(r) == 2 else r[2]\n",
    "    df.append((lbl, lbl))\n",
    "    for a1 in alias:\n",
    "        for a2 in [a1, unidecode.unidecode(a1)]:\n",
    "            df.append((a2, lbl))\n",
    "            df.append((a2.replace('-', ' '), lbl))\n",
    "            df.append((a2.replace('-', ''), lbl))\n",
    "df = pd.DataFrame(df, columns=['sym', 'lbl']).drop_duplicates()\n",
    "df = df.assign(spid=SPECIES_HUMAN_ID, src=SRC, extid=None)\n",
    "df['id'] = get_ids(df, ID_TYP_CT)\n",
    "df = df.drop_duplicates()\n",
    "df = add_preferred_ids(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sym</th>\n",
       "      <th>lbl</th>\n",
       "      <th>spid</th>\n",
       "      <th>src</th>\n",
       "      <th>extid</th>\n",
       "      <th>id</th>\n",
       "      <th>prefid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>T-follicular-helper</td>\n",
       "      <td>Tfh</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT521FD23411A69F7B</td>\n",
       "      <td>CTE9F6070561C95355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>T follicular helper</td>\n",
       "      <td>Tfh</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT5C03730892B0B7E0</td>\n",
       "      <td>CTE9F6070561C95355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>Tfollicularhelper</td>\n",
       "      <td>Tfh</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT10F75C923DC5D960</td>\n",
       "      <td>CTE9F6070561C95355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>follicular-helper-T</td>\n",
       "      <td>Tfh</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT8A54A7929F3074FF</td>\n",
       "      <td>CTE9F6070561C95355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>follicular helper T</td>\n",
       "      <td>Tfh</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT48DDD25612FEC28B</td>\n",
       "      <td>CTE9F6070561C95355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>follicularhelperT</td>\n",
       "      <td>Tfh</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT92FD47DCB488F926</td>\n",
       "      <td>CTE9F6070561C95355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>T-Helper</td>\n",
       "      <td>Th</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT58AC91B99EE259CE</td>\n",
       "      <td>CTB7DFCED683E6F0A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>T Helper</td>\n",
       "      <td>Th</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CTCC4CF8EAB76DFC68</td>\n",
       "      <td>CTB7DFCED683E6F0A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>THelper</td>\n",
       "      <td>Th</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT0107866BBCD97E04</td>\n",
       "      <td>CTB7DFCED683E6F0A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>Helper-T</td>\n",
       "      <td>Th</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT068AEEF38840649E</td>\n",
       "      <td>CTB7DFCED683E6F0A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>Helper T</td>\n",
       "      <td>Th</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT7580C71A46994EC5</td>\n",
       "      <td>CTB7DFCED683E6F0A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>HelperT</td>\n",
       "      <td>Th</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CTAF0873D2974B1075</td>\n",
       "      <td>CTB7DFCED683E6F0A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1038</th>\n",
       "      <td>T-helper-0</td>\n",
       "      <td>Th0</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CTBD1BDE561333FA0A</td>\n",
       "      <td>CTBDDE0043F47EDCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1039</th>\n",
       "      <td>T helper 0</td>\n",
       "      <td>Th0</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT550BD2F6E7C9F4A6</td>\n",
       "      <td>CTBDDE0043F47EDCC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1040</th>\n",
       "      <td>Thelper0</td>\n",
       "      <td>Th0</td>\n",
       "      <td>1</td>\n",
       "      <td>manual</td>\n",
       "      <td>None</td>\n",
       "      <td>CT2B7FD02233168156</td>\n",
       "      <td>CTBDDE0043F47EDCC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      sym  lbl  spid     src extid                  id  \\\n",
       "429   T-follicular-helper  Tfh     1  manual  None  CT521FD23411A69F7B   \n",
       "430   T follicular helper  Tfh     1  manual  None  CT5C03730892B0B7E0   \n",
       "431     Tfollicularhelper  Tfh     1  manual  None  CT10F75C923DC5D960   \n",
       "435   follicular-helper-T  Tfh     1  manual  None  CT8A54A7929F3074FF   \n",
       "436   follicular helper T  Tfh     1  manual  None  CT48DDD25612FEC28B   \n",
       "437     follicularhelperT  Tfh     1  manual  None  CT92FD47DCB488F926   \n",
       "160              T-Helper   Th     1  manual  None  CT58AC91B99EE259CE   \n",
       "161              T Helper   Th     1  manual  None  CTCC4CF8EAB76DFC68   \n",
       "162               THelper   Th     1  manual  None  CT0107866BBCD97E04   \n",
       "166              Helper-T   Th     1  manual  None  CT068AEEF38840649E   \n",
       "167              Helper T   Th     1  manual  None  CT7580C71A46994EC5   \n",
       "168               HelperT   Th     1  manual  None  CTAF0873D2974B1075   \n",
       "1038           T-helper-0  Th0     1  manual  None  CTBD1BDE561333FA0A   \n",
       "1039           T helper 0  Th0     1  manual  None  CT550BD2F6E7C9F4A6   \n",
       "1040             Thelper0  Th0     1  manual  None  CT2B7FD02233168156   \n",
       "\n",
       "                  prefid  \n",
       "429   CTE9F6070561C95355  \n",
       "430   CTE9F6070561C95355  \n",
       "431   CTE9F6070561C95355  \n",
       "435   CTE9F6070561C95355  \n",
       "436   CTE9F6070561C95355  \n",
       "437   CTE9F6070561C95355  \n",
       "160   CTB7DFCED683E6F0A9  \n",
       "161   CTB7DFCED683E6F0A9  \n",
       "162   CTB7DFCED683E6F0A9  \n",
       "166   CTB7DFCED683E6F0A9  \n",
       "167   CTB7DFCED683E6F0A9  \n",
       "168   CTB7DFCED683E6F0A9  \n",
       "1038  CTBDDE0043F47EDCC1  \n",
       "1039  CTBDDE0043F47EDCC1  \n",
       "1040  CTBDDE0043F47EDCC1  "
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['sym'].str.lower().str.match('.*help.*')].head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/eczech/repos/hammer/t-cell-relation-extraction/pm_subtype_protein_relations/data/meta/cell_types.csv'"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = osp.join(META_DATA_DIR, CELL_TYPES + '.csv')\n",
    "df.to_csv(path, index=False)\n",
    "path"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
